{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "kernel": "SoS"
   },
   "source": [
    "# Association analysis for GTEx V8 data\n",
    "\n",
    "We use this pipeline to perform association analysis for GTEx V8 data, including fine-mapping and prediction applications."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "kernel": "SoS"
   },
   "source": [
    "## Authors\n",
    "\n",
    "Gao (fine-mapping) and Fabio (prediction) at UChicago."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "kernel": "SoS"
   },
   "source": [
    "## Pipeline interface"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "kernel": "Bash"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "usage: sos run GTEx_V8_Association.ipynb\n",
      "               [workflow_name | -t targets] [options] [workflow_options]\n",
      "  workflow_name:        Single or combined workflows defined in this script\n",
      "  targets:              One or more targets to generate\n",
      "  options:              Single-hyphen sos parameters (see \"sos run -h\" for details)\n",
      "  workflow_options:     Double-hyphen workflow-specific parameters\n",
      "\n",
      "Workflows:\n",
      "  mnm\n",
      "  fqtl\n",
      "\n",
      "Global Workflow Options:\n",
      "  --output-dir output (as path)\n",
      "  --data-dir /project2/compbio/GTEx_eQTL/cis_eqtl_analysis_ready (as path)\n",
      "  --gene-id-file  path(f\"{data_dir}/multitissue_genes_list.txt\")\n",
      "\n",
      "  --missing-rate-cutoff 0.05 (as float)\n",
      "  --maf-cutoff 0.01 (as float)\n",
      "  --mash-model /project2/compbio/GTEx_eQTL/mashr_flashr_workflow_output/FastQTLSumStats.mash.FL_PC3.mash_model_est_v.rds (as path)\n",
      "  --prop-test 0.0 (as float)\n",
      "                        If test_prop is set > 0 then prediction will also be\n",
      "                        performed and evaluated\n",
      "\n",
      "Sections\n",
      "  mnm_0, fqtl_0:\n",
      "  mnm_1:\n",
      "    Workflow Options:\n",
      "      --max-mixtures 20 (as int)\n",
      "                        maximum number of mixture components to use\n",
      "      --max-effects 10 (as int)\n",
      "                        maximum number of single effects to capture\n",
      "  fqtl_1:\n"
     ]
    }
   ],
   "source": [
    "sos run GTEx_V8_Association.ipynb -h"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "kernel": "Bash"
   },
   "source": [
    "A test command is (on a local computer):\n",
    "\n",
    "```\n",
    "cat /project2/compbio/GTEx_eQTL/cis_eqtl_analysis_ready/multitissue_genes_list.txt | grep -v \"#\" | head -1 > test.txt\n",
    "sos run GTEx_V8_Association.ipynb mnm \\\n",
    "    --gene-id-file test.txt \\\n",
    "    --prop-test 0.2 \\\n",
    "    --output-dir ./output\n",
    "```\n",
    "\n",
    "To submit jobs, for say, 1000 genes from the list,\n",
    "\n",
    "```\n",
    "cat /project2/compbio/GTEx_eQTL/cis_eqtl_analysis_ready/multitissue_genes_list.txt | grep -v \"#\" | head -1000 > 1000_genes.txt\n",
    "sos run GTEx_V8_Association.ipynb mnm \\\n",
    "    --gene-id-file 1000_genes.txt \\\n",
    "    --prop-test 0.2 \\\n",
    "    --output-dir /project2/compbio/GTEx_eQTL/cis_eqtl_mnm_v8 \\\n",
    "    -c midway2.yml -q midway2\n",
    "```\n",
    "\n",
    "To run `fqtl` pipeline that uses a different job template (configured as a different queue),\n",
    "\n",
    "```\n",
    "cat /project2/compbio/GTEx_eQTL/cis_eqtl_analysis_ready/multitissue_genes_list.txt | grep -v \"#\" | head -1000 > 1000_genes.txt\n",
    "sos run GTEx_V8_Association.ipynb fqtl \\\n",
    "    --gene-id-file 1000_genes.txt \\\n",
    "    --prop-test 0.2 \\\n",
    "    --output-dir /project2/compbio/MnM-TWAS/Output/cis_eqtl_fqtl_v8 \\\n",
    "    -c midway2.yml -q midway2_rmkl\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "kernel": "SoS"
   },
   "outputs": [],
   "source": [
    "[global]\n",
    "parameter: output_dir = path('./output')\n",
    "parameter: data_dir = path('/project2/compbio/GTEx_eQTL/cis_eqtl_analysis_ready')\n",
    "parameter: gene_id_file = path(f\"{data_dir}/multitissue_genes_list.txt\")\n",
    "parameter: missing_rate_cutoff = 0.05\n",
    "parameter: maf_cutoff = 0.01\n",
    "parameter: mash_model = path('/project2/compbio/GTEx_eQTL/mashr_flashr_workflow_output/FastQTLSumStats.mash.FL_PC3.mash_model_est_v.rds')\n",
    "# If test_prop is set > 0 then prediction will also be performed and evaluated\n",
    "parameter: prop_test = 0.0\n",
    "# specify seed\n",
    "parameter: seed = 1\n",
    "genes = [x.strip() for x in open(gene_id_file).readlines() if not x.startswith('#')]\n",
    "fail_if(len(genes) == 0, msg = 'No gene to analyze!')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "kernel": "SoS"
   },
   "source": [
    "## Utilities\n",
    "\n",
    "Here are some predefined functions from Fabio to use elsewhere in the workflow. In the spirit of minimal changes made I'm saving this as a utility file for other workflows to use as necessary. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "kernel": "SoS"
   },
   "outputs": [],
   "source": [
    "[mnm_0,fqtl_0]\n",
    "report: expand = \"${ }\", output = '.sos/utils.R'\n",
    "\n",
    "    ###Function needed for fqtl prediction\n",
    "    predict_fqtl <- function(coeffs, newx){\n",
    "        for(i in 1:ncol(coeffs)){\n",
    "            if(i==1){res <- newx %*% coeffs[, i]} else if(i>1){res <- cbind(res, newx %*% coeffs[, i])}\n",
    "        }\n",
    "        return(res)\n",
    "    }\n",
    "\n",
    "    ###Functions to compute MAF and missing genotype rate\n",
    "    compute_maf <- function(geno){\n",
    "      f <- mean(geno,na.rm = TRUE)/2\n",
    "      return(min(f, 1-f))\n",
    "    }\n",
    "\n",
    "    compute_missing <- function(geno){\n",
    "      miss <- sum(is.na(geno))/length(geno)\n",
    "      return(miss)\n",
    "    }\n",
    "    \n",
    "    mean_impute <- function(geno){\n",
    "      f <- apply(geno, 2, function(x) mean(x,na.rm = TRUE))\n",
    "      for (i in 1:length(f)) geno[,i][which(is.na(geno[,i]))] <- f[i]\n",
    "      return(geno)\n",
    "    }\n",
    "\n",
    "    is_zero_variance <- function(x) {\n",
    "      if (length(unique(x))==1) return(T)\n",
    "      else return(F)\n",
    "    }\n",
    "\n",
    "    ###Function to sample test set individuals\n",
    "    create_test_set <- function(Y, test_prop){\n",
    "        Ntot <- nrow(Y)\n",
    "        test_ind <- sample(x=c(1:Ntot), size=round(Ntot*test_prop), replace=F)\n",
    "\n",
    "        return(test_ind)\n",
    "    }\n",
    "\n",
    "    ###Function to calculate the covariance matrix of Y via diag\n",
    "    compute_cov_diag <- function(Y, miss=NULL){\n",
    "        if(is.null(miss)){\n",
    "            covar <- diag(apply(Y, 2, var, na.rm=T))\n",
    "        } else {\n",
    "            covar <- diag(apply(Y[-miss, ], 2, var, na.rm=T))\n",
    "        }\n",
    "\n",
    "        return(covar)\n",
    "    }\n",
    "\n",
    "    ###Function to calculate the covariance matrix of Y via flash\n",
    "    compute_cov_flash <- function(Y, miss=NULL){\n",
    "        if(is.null(miss)){\n",
    "            fl <- flashier::flash(Y, var.type = 2, prior.family = c(flashier::prior.normal(), flashier::prior.normal.scale.mix()), backfit = TRUE, verbose.lvl=0)\n",
    "        } else {\n",
    "            fl <- flashier::flash(Y[-miss, ], var.type = 2, prior.family = c(flashier::prior.normal(), flashier::prior.normal.scale.mix()), backfit = TRUE, verbose.lvl=0)\n",
    "        }  \n",
    "    \n",
    "        if(fl$n.factors==0){\n",
    "          covar <- diag(fl$residuals.sd^2)\n",
    "        } else {\n",
    "          fsd <- sapply(fl$fitted.g[[1]], '[[', \"sd\")\n",
    "          covar <- diag(fl$residuals.sd^2) + crossprod(t(fl$flash.fit$EF[[2]]) * fsd)\n",
    "        }\n",
    "\n",
    "        return(covar)\n",
    "    }\n",
    "\n",
    "    ### mashr null correlation code\n",
    "    estimate_null_correlation_simple = function(bhat, sbhat, z_thresh=2, est_cor = TRUE){\n",
    "      z = bhat/sbhat\n",
    "      z[which(is.nan(z) | is.infinite(z))] = 0\n",
    "      max_absz = apply(abs(z),1, max)\n",
    "      nullish = which(max_absz < z_thresh)\n",
    "      if(length(nullish)<ncol(z)){\n",
    "        print(\"not enough null data to estimate null correlation; return diag instead\")\n",
    "        return(diag(ncol(z)))\n",
    "      }\n",
    "      nullish_z = z[nullish,]\n",
    "      Vhat = cov(nullish_z)\n",
    "      if(est_cor){\n",
    "        Vhat = cov2cor(Vhat)\n",
    "      }\n",
    "      return(Vhat)\n",
    "    }\n",
    "\n",
    "    ###Functions to do prediction\n",
    "    fqtl_fit_pred <- function(X, Y, miss, K=1){\n",
    "        ###Scale X and Y\n",
    "        X_scaled <- scale(X, center=T, scale=T)\n",
    "        Y_scaled <- scale(Y, center=T, scale=T)\n",
    "\n",
    "        ###Fit fqtl\n",
    "        opt.gtex <- list(vbiter = 5000, gammax = 1e4, tol = 1e-8,\n",
    "                         rate = 1e-2, decay = -1e-2,\n",
    "                         pi.ub = -1/2, pi.lb = -2, tau = -4, do.hyper = TRUE,\n",
    "                         jitter = 0.1, svd.init = TRUE, out.residual = FALSE,\n",
    "                         print.interv = 100, k = K, mf.right.nn = FALSE)\n",
    "\n",
    "        out <- fqtl::fit.fqtl(Y_scaled[-miss, ], X_scaled[-miss, ], factored=TRUE, nthread = 1, options = opt.gtex)\n",
    "        tis_effect <- out$mean.right$theta\n",
    "        snp_effect <- out$mean.left$theta\n",
    "        combined_effect <- tcrossprod(snp_effect, tis_effect)\n",
    "\n",
    "        ###Predict phenotypes in the test set\n",
    "        pred_scaled <- predict_fqtl(combined_effect, X_scaled[miss, ])\n",
    "        pred <- pred_scaled * attr(Y_scaled, 'scaled:scale')[col(pred_scaled)] + attr(Y_scaled, 'scaled:center')[col(pred_scaled)]\n",
    "\n",
    "        return(list(fit=out, ymiss=Y[miss, ], ymiss_pred=pred))\n",
    "    }\n",
    "\n",
    "    ### Filter X matrix\n",
    "    filter_X <- function(X, missing_rate_thresh, maf_thresh) {\n",
    "        rm_col <- which(apply(X, 2, compute_missing) > missing_rate_thresh)\n",
    "        if (length(rm_col)) X <- X[, -rm_col]\n",
    "        rm_col <- which(apply(X, 2, compute_maf) < maf_thresh)\n",
    "        if (length(rm_col)) X <- X[, -rm_col]\n",
    "        rm_col <- which(apply(X, 2, is_zero_variance))\n",
    "        if (length(rm_col)) X <- X[, -rm_col]\n",
    "        return(mean_impute(X))\n",
    "    }\n",
    "\n",
    "    TISSUE_LIST = c('Adipose_Subcutaneous',\n",
    "    'Adipose_Visceral_Omentum',\n",
    "    'Adrenal_Gland',\n",
    "    'Artery_Aorta',\n",
    "    'Artery_Coronary',\n",
    "    'Artery_Tibial',\n",
    "    'Brain_Amygdala',\n",
    "    'Brain_Anterior_cingulate_cortex_BA24',\n",
    "    'Brain_Caudate_basal_ganglia',\n",
    "    'Brain_Cerebellar_Hemisphere',\n",
    "    'Brain_Cerebellum',\n",
    "    'Brain_Cortex',\n",
    "    'Brain_Frontal_Cortex_BA9',\n",
    "    'Brain_Hippocampus',\n",
    "    'Brain_Hypothalamus',\n",
    "    'Brain_Nucleus_accumbens_basal_ganglia',\n",
    "    'Brain_Putamen_basal_ganglia',\n",
    "    'Brain_Spinal_cord_cervical_c-1',\n",
    "    'Brain_Substantia_nigra',\n",
    "    'Breast_Mammary_Tissue',\n",
    "    'Cells_Cultured_fibroblasts',\n",
    "    'Cells_EBV-transformed_lymphocytes',\n",
    "    'Colon_Sigmoid',\n",
    "    'Colon_Transverse',\n",
    "    'Esophagus_Gastroesophageal_Junction',\n",
    "    'Esophagus_Mucosa',\n",
    "    'Esophagus_Muscularis',\n",
    "    'Heart_Atrial_Appendage',\n",
    "    'Heart_Left_Ventricle',\n",
    "    'Kidney_Cortex',\n",
    "    'Liver',\n",
    "    'Lung',\n",
    "    'Minor_Salivary_Gland',\n",
    "    'Muscle_Skeletal',\n",
    "    'Nerve_Tibial',\n",
    "    'Ovary',\n",
    "    'Pancreas',\n",
    "    'Pituitary',\n",
    "    'Prostate',\n",
    "    'Skin_Not_Sun_Exposed_Suprapubic',\n",
    "    'Skin_Sun_Exposed_Lower_leg',\n",
    "    'Small_Intestine_Terminal_Ileum',\n",
    "    'Spleen',\n",
    "    'Stomach',\n",
    "    'Testis',\n",
    "    'Thyroid',\n",
    "    'Uterus',\n",
    "    'Vagina',\n",
    "    'Whole_Blood')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "kernel": "SoS"
   },
   "source": [
    "## M&M analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "kernel": "SoS"
   },
   "outputs": [],
   "source": [
    "# Fit M&M\n",
    "[mnm_1]\n",
    "depends: R_library('mvsusieR'), R_library('flashier'), R_library('susieR'), mash_model\n",
    "# maximum number of mixture components to use\n",
    "parameter: max_mixtures = 20\n",
    "# maximum number of single effects to capture\n",
    "parameter: max_effects = 10\n",
    "# covariance matrix computed by \"flash\", \"diag\" or \"znull\" method\n",
    "parameter: cov_method = 'flash'\n",
    "input: [f'{data_dir}/{gene}.Multi_Tissues.rds' for gene in genes], group_by = 1\n",
    "output: f'{output_dir}/{_input:bn}.mnm_{cov_method}.rds'\n",
    "# each job uses 10 nodes, each node 4 cores in parallel each core using 2G memory; and jobs are created in batches of 40.\n",
    "task: trunk_workers = [4] * 10, trunk_size = 40, walltime = '30m', mem = '8G', cores = 1, tags = f'{step_name}_{_output:bn}'\n",
    "\n",
    "R: expand = '${ }', input = '.sos/utils.R'\n",
    "    set.seed(${seed})\n",
    "    attach(readRDS(${_input:r}))\n",
    "    # setup data\n",
    "    X <- filter_X(X, ${missing_rate_cutoff}, ${maf_cutoff})\n",
    "    prop_test <- ${prop_test}\n",
    "    \n",
    "    # Get original effects results\n",
    "    # to keep for a comparison, not to be used in M&M\n",
    "    univariate_res = lapply(1:ncol(y_res), function(i) susieR:::univariate_regression(X,y_res[,i]))\n",
    "    bhat = do.call(cbind, lapply(1:ncol(y_res), function(i) univariate_res[[i]]$betahat))\n",
    "    shat = do.call(cbind, lapply(1:ncol(y_res), function(i) univariate_res[[i]]$sebetahat))\n",
    "    # Get mash model\n",
    "    mash <- readRDS(${mash_model:r})\n",
    "    res = list()\n",
    "    # For fine-mapping\n",
    "    if (\"${cov_method}\" == 'znull') {\n",
    "      vhat = estimate_null_correlation_simple(bhat, shat)\n",
    "      s_y = apply(y_res, 2, function(x) sd(x, na.rm=T))\n",
    "      cov_Y = diag(s_y) %*% vhat %*% diag(s_y)\n",
    "    } else {\n",
    "      cov_Y <- compute_cov_${cov_method}(y_res)\n",
    "    }\n",
    "    if (mash$fitted_g$usepointmass) {\n",
    "      prior_weights = mash$fitted_g$pi[-1]\n",
    "      null_weight = mash$fitted_g$pi[1]\n",
    "    } else {\n",
    "      prior_weights = mash$fitted_g$pi\n",
    "      null_weight = 0\n",
    "    }\n",
    "    include_cond = match(colnames(y_res), TISSUE_LIST)\n",
    "    m_init <- mvsusieR:::MashInitializer$new(Ulist=mash$fitted_g$Ulist, grid=mash$fitted_g$grid, prior_weights=prior_weights, null_weight=null_weight, top_mixtures=${max_mixtures}, include_conditions=include_cond)\n",
    "    res$fitted <- mvsusieR::mvsusie(X, y_res, L=${max_effects}, prior_variance=m_init, residual_variance=cov_Y, compute_objective=F, estimate_residual_variance=F, estimate_prior_variance=T, estimate_prior_method=\"simple\")\n",
    "    res$fitted$bhat = bhat\n",
    "    res$fitted$shat = shat\n",
    "    # For prediction\n",
    "    if (prop_test>0) {\n",
    "        miss <- create_test_set(y_res, prop_test)\n",
    "        cov_Y <- compute_cov_${cov_method}(y_res, miss)\n",
    "        m <- mvsusieR::mvsusie(X[-miss, ], y_res[-miss, ], L=${max_effects}, prior_variance=m_init, residual_variance=cov_Y, compute_objective=F, estimate_residual_variance=F, estimate_prior_variance=T, estimate_prior_method=\"simple\")\n",
    "        res$ymiss_pred <- mvsusieR::predict.mvsusieR(m, X[miss, ])\n",
    "        res$ymiss <- y_res[miss, ]\n",
    "    }\n",
    "    saveRDS(res, ${_output:r})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "kernel": "SoS"
   },
   "outputs": [],
   "source": [
    "# Make some M&M plots\n",
    "[mnm_2]\n",
    "depends: executable('convert')\n",
    "output: f'{_input:n}.png'\n",
    "# each job uses 10 nodes, each node 4 cores in parallel each core using 2G memory; and jobs are created in batches of 40.\n",
    "task: trunk_workers = [4] * 10, trunk_size = 40, walltime = '5m', mem = '2G', cores = 1, tags = f'{step_name}_{_output:bn}'\n",
    "R: expand = '${ }'\n",
    "    m = readRDS(${_input:r})$fitted\n",
    "    pdf(\"${_output:n}.dot.pdf\", width=12, height=6)\n",
    "    susieR::susie_plot(m,y='PIP', main = 'Cross-condition Posterior Inclusion Probability', xlab = 'SNP positions', add_legend = T)\n",
    "    dev.off()\n",
    "    p = mvsusieR::mvsusie_plot(m)\n",
    "    pdf(\"${_output:n}.bubble.pdf\", width=20, height=15)\n",
    "    print(p$plot)\n",
    "    dev.off()\n",
    "    p = mvsusieR::mvsusie_plot(m, original_sumstat=T)\n",
    "    pdf(\"${_output:nn}.z_scores.pdf\", width=20, height=15)\n",
    "    print(p$plot)\n",
    "    dev.off()\n",
    "\n",
    "bash: expand = True\n",
    "  convert -append {_output:n}.dot.pdf {_output:n}.bubble.pdf {_output:nn}.z_scores.pdf {_output}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "kernel": "SoS"
   },
   "source": [
    "## fqtl analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "kernel": "SoS"
   },
   "outputs": [],
   "source": [
    "[fqtl_1]\n",
    "depends: R_library('fqtl')\n",
    "input: [f'{data_dir}/{gene}.Multi_Tissues.rds' for gene in genes], group_by = 1\n",
    "output: f'{output_dir}/{_input:bn}.fqtl.rds'\n",
    "# each job uses 10 nodes, each node 4 cores in parallel each core using 2G memory; and jobs are created in batches of 40.\n",
    "task: trunk_workers = 4, trunk_size = 8, walltime = '150m', mem = '8G', cores = 1, tags = f'{step_name}_{_output:bn}'\n",
    "\n",
    "R: expand = '${ }', input = '.sos/utils.R'\n",
    "    set.seed(${seed})\n",
    "    attach(readRDS(${_input:r}))\n",
    "    # setup data\n",
    "    X <- filter_X(X, ${missing_rate_cutoff}, ${maf_cutoff})\n",
    "    prop_test <- ${prop_test}\n",
    "    #\n",
    "    miss <- create_test_set(y_res, prop_test)\n",
    "    K <- min((nrow(y_res)-length(miss)), ncol(y_res))\n",
    "    res <- fqtl_fit_pred(X, y_res, miss, K=K)\n",
    "    saveRDS(res, ${_output:r})"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "SoS",
   "language": "sos",
   "name": "sos"
  },
  "language_info": {
   "codemirror_mode": "sos",
   "file_extension": ".sos",
   "mimetype": "text/x-sos",
   "name": "sos",
   "nbconvert_exporter": "sos_notebook.converter.SoS_Exporter",
   "pygments_lexer": "sos"
  },
  "sos": {
   "kernels": [
    [
     "Bash",
     "bash",
     "Bash",
     "#E6EEFF",
     ""
    ],
    [
     "R",
     "ir",
     "R",
     "#DCDCDA",
     ""
    ],
    [
     "SoS",
     "sos",
     "",
     "",
     "sos"
    ]
   ],
   "version": "0.21.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
