<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta charset="utf-8">
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link rel="stylesheet" type="text/css" href="../css/jt.css">
<link rel="stylesheet" type="text/css" href="../css/toc2.css">
<link href="../site_libs/jqueryui-1.11.4/jquery-ui.css">
<link rel="stylesheet" href="../site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<link rel="stylesheet" href="../site_libs/font-awesome-4.5.0/css/font-awesome.min.css" rel="stylesheet" />
<script type="text/javascript" src="https://ajax.googleapis.com/ajax/libs/jquery/1.9.1/jquery.min.js"></script>
<script type="text/javascript" src="https://ajax.googleapis.com/ajax/libs/jqueryui/1.9.1/jquery-ui.min.js"></script>
<script src="../site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="../site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="../site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<link rel="stylesheet"
      href="../site_libs/highlight/textmate.css"
      type="text/css" />
<script src="../site_libs/highlight/highlight.js"></script>
<script type="text/javascript">
if (window.hljs && document.readyState && document.readyState === "complete") {
   window.setTimeout(function() {
      hljs.initHighlighting();
   }, 0);
}
</script>
<script src="../js/toc2.js"></script>
<script src="../js/docs.js"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.1/MathJax.js?config=TeX-AMS_HTML"></script>
<script>
    MathJax.Hub.Config({
        extensions: ["tex2jax.js"],
        jax: ["input/TeX", "output/HTML-CSS"],
        tex2jax: {
        inlineMath: [ ['$','$'], ["\\(","\\)"] ],
        displayMath: [ ['$$','$$'], ["\\[","\\]"] ],
        processEscapes: true
        },
        "HTML-CSS": {
            preferredFont: "TeX",
            availableFonts: ["STIX","TeX"],
            styles: {
                scale: 110,
                ".MathJax_Display": {
                    "font-size": "110%",
                }
            }
        }
    });
</script>
<script>
$( document ).ready(function(){
            var cfg={'threshold':3,     // depth of toc (number of levels)
             'number_sections': false,
             'toc_cell': false,          // useless here
             'toc_window_display': true, // display the toc window
             "toc_section_display": "block", // display toc contents in the window
             'sideBar':true,       // sidebar or floating window
             'navigate_menu':false       // navigation menu (only in liveNotebook -- do not change)
            }
            var st={};                  // some variables used in the script
            st.rendering_toc_cell = false;
            st.config_loaded = false;
            st.extension_initialized=false;
            st.nbcontainer_marginleft = $('#notebook-container').css('margin-left')
            st.nbcontainer_marginright = $('#notebook-container').css('margin-right')
            st.nbcontainer_width = $('#notebook-container').css('width')
            st.oldTocHeight = undefined
            st.cell_toc = undefined;
            st.toc_index=0;
            // fire the main function with these parameters
            table_of_contents(cfg, st);
            var file=writeupDict[$("h1:first").attr("id")];
            $("#toc-level0 a").css("color","#126dce");
            $('a[href="#'+$("h1:first").attr("id")+'"]').hide()
            var docs=writeupArray;
            var pos=writeupArray.indexOf(file);
            for (var a=pos;a>=0;a--){
                  var name=docs[a]
                  $('<li><a href="'+name+'.html"><font color="#073642"><b>'+name.replace(/_/g," ")+'</b></font></a></li>').insertBefore("#toc-level0 li:eq(0)");
            }
            $('a[href="'+file+'.html'+'"]').css("color","#126dce");
            for (var a=pos+1;a<docs.length;a++){
                  var name=docs[a]
                  $(".toc #toc-level0").append('<li><a href="'+name+'.html"><font color="#073642"><b>'+name.replace(/_/g," ")+'</b></font></a></li>');
            }
            $("#toc-header").hide();
    });
</script>
<script>
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');
  // mark it active
  menuAnchor.parent().addClass('active');
  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');
});
</script>
<div class="container-fluid main-container">
<!-- tabsets -->
<script src="../site_libs/navigation-1.1/tabsets.js"></script>
<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});
</script>
<title>m&m ash</title>
<style type = "text/css">
body {
  font-family: "Droid Sans";
  padding-top: 66px;
  padding-bottom: 40px;
}
</style>
</head>
<body>
<div tabindex="-1" id="notebook" class="border-box-sizing">
<div class="container" id="notebook-container">
<!-- code folding -->
<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="../index.html">m&m ash</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">Overview</a>
</li>
<li>
  <a href="../analysis.html">Analysis</a>
</li>
<li>
  <a href="../pipeline.html">Pipeline</a>
</li>
<li>
  <a href="../writeup.html">Writeup</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
    <a href="http://github.com/gaow/mvarbvs">
    <span class="fa fa-github"></span>
    </a>
    </li>
    </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->
<div class="cell border-box-sizing text_cell rendered">
<div class="prompt input_prompt">
</div>
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h1 id="GTEx-V7-eQTL-data-analysis-procedure">GTEx V7 eQTL data analysis procedure<a class="anchor-link" href="#GTEx-V7-eQTL-data-analysis-procedure">&#182;</a></h1><p>See <a href="https://gtexportal.org/home/documentationPage">this page</a> for analysis outline of other releases.</p>
<p><strong>last committed:</strong> Thu Jul 6 23:35:45 2017 (revision 161, <a href="http://github.com/gaow/mvarbvs/commit/8f994eaed346b3790718e19a23624dbdb2e71449">8f994ea</a>)</p>
</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered">
<div class="prompt input_prompt">
</div>
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h2 id="Preprocessing">Preprocessing<a class="anchor-link" href="#Preprocessing">&#182;</a></h2>
</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered">
<div class="prompt input_prompt">
</div>
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h3 id="What-we-have">What we have<a class="anchor-link" href="#What-we-have">&#182;</a></h3><ul>
<li>Genotype data, filted and phased, yet not imputed</li>
<li>RNA-Seq data of gene expression for all sample tissues, available both the counts and RPKM</li>
<li>Sample phenotype / covariates</li>
</ul>
<h3 id="What-we-need-to-do">What we need to do<a class="anchor-link" href="#What-we-need-to-do">&#182;</a></h3><h4 id="Genotype-data">Genotype data<a class="anchor-link" href="#Genotype-data">&#182;</a></h4><ul>
<li>We need to impute missing genotypes. We use UMichigen Imputation Panel <code>minimac3</code> + HRC reference data</li>
<li>Do some PLINK QC, including<ul>
<li>Keep variants with at least 10 samples having the minor allele (more stringent that 1% MAF filter)</li>
</ul>
</li>
<li>Genotype samples have to match RNA-Seq samples by ID<ul>
<li>Do it to RNA-Seq data when creating HDF5 matrices</li>
</ul>
</li>
</ul>
<h4 id="RNA-Seq-data">RNA-Seq data<a class="anchor-link" href="#RNA-Seq-data">&#182;</a></h4><ul>
<li>Genes should have at least 10 samples with RPKM &gt; 0.1 and read counts &gt; 6</li>
<li>Use quantile normalization</li>
</ul>
<h4 id="Covariates">Covariates<a class="anchor-link" href="#Covariates">&#182;</a></h4><ul>
<li>3 PCs, gender, genotyping platform, and PEER factors</li>
<li>PEER factors are generated using the top 10000 expressed genes per tissue, after normalization.<ul>
<li>In V6P guideline it seems to vaguely imply using all genes from all tissues. But it seems PEER cannot handle data of that size (too slow to process).</li>
<li>In <a href="https://storage.googleapis.com/gtex-public-data/Portal_Analysis_Methods_v6_08182016.pdf">V6 guideline</a> it suggests using top 10K expressed genes per tissue. Since we will remove these covariates via conventional multiple regression anyways it is Ok to correct for it this way.</li>
</ul>
</li>
<li>Number of PEER depends on sample size N. <ul>
<li>$N < 150$, use 15 PEERs, $150 \le N < 250$, use 30 PEERs, $N \ge 250$ use 35 PEERs</li>
</ul>
</li>
<li>Regress out these factors separately for each tissue, save the residual to HDF5 as the Y for later analysis</li>
</ul>
<h4 id="FastQTL-analysis">FastQTL analysis<a class="anchor-link" href="#FastQTL-analysis">&#182;</a></h4><p>Adapted from <a href="https://github.com/francois-a/fastqtl">Broad FastQTL wrapper tool</a></p>
</div>
</div>
</div>
<hr>
&copy 2016-2017 Gao Wang et al at Stephens Lab, University of Chicago
</div>
</div>
</body>
</html>
