<!DOCTYPE html>
<html>
<head><meta charset="utf-8" />
<title>Pipelines</title><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.1.10/require.min.js"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/2.0.3/jquery.min.js"></script>








<!-- Custom stylesheet, it must be in the same directory as the html file -->
<link rel="stylesheet" href="custom.css">

<!-- Loading mathjax macro -->
<!-- Load mathjax -->
    <script src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS_HTML"></script>
    <!-- MathJax configuration -->
    <script type="text/x-mathjax-config">
    MathJax.Hub.Config({
        tex2jax: {
            inlineMath: [ ['$','$'], ["\\(","\\)"] ],
            displayMath: [ ['$$','$$'], ["\\[","\\]"] ],
            processEscapes: true,
            processEnvironments: true
        },
        // Center justify equations in code and markdown cells. Elsewhere
        // we use CSS to left justify single line equations in code cells.
        displayAlign: 'center',
        "HTML-CSS": {
            styles: {'.MathJax_Display': {"margin": 0}},
            linebreaks: { automatic: true }
        }
    });
    </script>
    <!-- End of mathjax configuration --></head>

 <link rel="stylesheet" href="https://code.jquery.com/ui/1.11.4/themes/smoothness/jquery-ui.css">

<script type="text/javascript" src="https://ajax.googleapis.com/ajax/libs/jquery/1.9.1/jquery.min.js"></script>
<script type="text/javascript" src="https://ajax.googleapis.com/ajax/libs/jqueryui/1.9.1/jquery-ui.min.js"></script>

<style>  /* defined here in case the main.css below cannot be loaded */
.lev1 {margin-left: 80px}
.lev2 {margin-left: 100px}
.lev3 {margin-left: 120px}
.lev4 {margin-left: 140px}
.lev5 {margin-left: 160px}
.lev6 {margin-left: 180px}
</style>

<link rel="stylesheet" type="text/css" href="../../css/jt.css">
<link rel="stylesheet" type="text/css" href="../../css/toc2.css">

<script src="../../js/toc2.js"></script>
<script src="../../js/docs.js"></script>

<script>
    MathJax.Hub.Config({
        extensions: ["tex2jax.js"],
        jax: ["input/TeX", "output/HTML-CSS"],
        tex2jax: {
        inlineMath: [ ['$','$'], ["\\(","\\)"] ],
        displayMath: [ ['$$','$$'], ["\\[","\\]"] ],
        processEscapes: true
        },
        "HTML-CSS": {
            preferredFont: "TeX",
            availableFonts: ["STIX","TeX"],
            styles: {
                scale: 110,
                ".MathJax_Display": {
                    "font-size": "110%",
                }
            }
        }
    });
</script>

<script>
$( document ).ready(function(){

            var cfg={'threshold':3,     // depth of toc (number of levels)
             'number_sections': false, 
             'toc_cell': false,          // useless here
             'toc_window_display': true, // display the toc window
             "toc_section_display": "block", // display toc contents in the window
             'sideBar':true,       // sidebar or floating window
             'navigate_menu':false       // navigation menu (only in liveNotebook -- do not change)
            }

            var st={};                  // some variables used in the script
            st.rendering_toc_cell = false;
            st.config_loaded = false;
            st.extension_initialized=false;
            st.nbcontainer_marginleft = $('#notebook-container').css('margin-left')
            st.nbcontainer_marginright = $('#notebook-container').css('margin-right')
            st.nbcontainer_width = $('#notebook-container').css('width')
            st.oldTocHeight = undefined
            st.cell_toc = undefined;
            st.toc_index=0;

            // fire the main function with these parameters


            table_of_contents(cfg,st);

            var file=notebookDict[$("h1:first").attr("id")];
            $("#toc-level0 a").css("color","#126dce");
            $('a[href="#'+$("h1:first").attr("id")+'"]').hide()
            var docs=documentation;
            var pos=documentation.indexOf(file);
        
            for (var a=pos;a>=0;a--){
                  var name=docs[a]
                  $('<li><a href="'+name+'.html">'+name.replace(/_/g," ")+'</a></li>').insertBefore("#toc-level0 li:eq(0)");
            }
            $('a[href="'+file+'.html'+'"]').css("color","#126dce");


            $('<li id="indexHome"><a href="../../documentation.html"><b>Analysis Home<b></a></li>').insertBefore("#toc-level0 li:eq(0)");
            for (var a=pos+1;a<docs.length;a++){
                  var name=docs[a]
                  $(".toc #toc-level0").append('<li><a href="'+name+'.html">'+name.replace(/_/g," ")+'</a></li>');
            }
            $("#toc-header").hide();

    });
</script>
<body>
  <div tabindex="-1" id="notebook" class="border-box-sizing">
    <div class="container" id="notebook-container">

<div class="cell border-box-sizing text_cell rendered">
<div class="prompt input_prompt">
</div>
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h1 id="GTEx-pipeline-execution-interface">GTEx pipeline execution interface<a class="anchor-link" href="#GTEx-pipeline-execution-interface">&#182;</a></h1>
</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered">
<div class="prompt input_prompt">
</div>
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h2 id="Preparation-&amp;-preprocessing">Preparation &amp; preprocessing<a class="anchor-link" href="#Preparation-&amp;-preprocessing">&#182;</a></h2><p>See <a href="https://gaow.github.io/mvarbvs/doc/writeup/GTEx7_Analysis_Plan.html#Preprocessing">this page</a> and <a href="https://gaow.github.io/mvarbvs/doc/writeup/Meetings.html#Project-meeting-20170518">this meeting note</a> for details.</p>

</div>
</div>
</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[2]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-sos"><pre><span></span><span class="o">%</span><span class="n">sossave</span> <span class="n">prep</span><span class="o">.</span><span class="n">sos</span> <span class="o">-</span><span class="n">f</span> <span class="o">-</span><span class="n">x</span>
<span class="c1">#!/usr/bin/env sos-runner</span>
<span class="c1">#fileformat=SOS1.0</span>

<span class="c1"># Usage:</span>
<span class="c1"># ./prep.sos download</span>
<span class="c1"># ./prep.sos </span>

<span class="o">%</span><span class="n">include</span> <span class="n">ResourceManagement</span> <span class="k">as</span> <span class="n">RM</span>
<span class="o">%</span><span class="n">include</span> <span class="n">Misc</span> <span class="k">as</span> <span class="n">MC</span>
<span class="o">%</span><span class="n">include</span> <span class="n">DataWrestling</span> <span class="k">as</span> <span class="n">DW</span>

<span class="p">[</span><span class="k">global</span><span class="p">]</span>

<span class="c1">#</span>
<span class="c1"># Auxiliary steps</span>
<span class="c1">#</span>

<span class="p">[</span><span class="kp">download</span><span class="p">]</span>
<span class="c1"># Resource preparation</span>
<span class="kp">sos_run</span><span class="p">(</span><span class="s1">&#39;RM.plink&#39;</span><span class="p">,</span> <span class="kp">workdir</span> <span class="o">=</span> <span class="n">CONFIG</span><span class="p">[</span><span class="s1">&#39;wd&#39;</span><span class="p">])</span>
<span class="kp">sos_run</span><span class="p">(</span><span class="s1">&#39;RM.minimac3&#39;</span><span class="p">,</span> <span class="kp">workdir</span> <span class="o">=</span> <span class="n">CONFIG</span><span class="p">[</span><span class="s1">&#39;wd&#39;</span><span class="p">])</span>
<span class="kp">sos_run</span><span class="p">(</span><span class="s1">&#39;RM.vcftools&#39;</span><span class="p">,</span> <span class="kp">workdir</span> <span class="o">=</span> <span class="n">CONFIG</span><span class="p">[</span><span class="s1">&#39;wd&#39;</span><span class="p">])</span>
<span class="kp">sos_run</span><span class="p">(</span><span class="s1">&#39;RM.peer&#39;</span><span class="p">,</span> <span class="kp">workdir</span> <span class="o">=</span> <span class="n">CONFIG</span><span class="p">[</span><span class="s1">&#39;wd&#39;</span><span class="p">])</span>
<span class="kp">sos_run</span><span class="p">(</span><span class="s1">&#39;RM.king&#39;</span><span class="p">,</span> <span class="kp">workdir</span> <span class="o">=</span> <span class="n">CONFIG</span><span class="p">[</span><span class="s1">&#39;wd&#39;</span><span class="p">])</span>


<span class="c1">#</span>
<span class="c1"># Workhorse</span>
<span class="c1">#</span>

<span class="p">[</span><span class="n">data_summary</span><span class="p">]</span>
<span class="kn">input:</span> <span class="n">CONFIG</span><span class="p">[</span><span class="s1">&#39;genotype&#39;</span><span class="p">]</span>
<span class="kp">sos_run</span><span class="p">(</span><span class="s2">&quot;MC.genotype_stats&quot;</span><span class="p">,</span> <span class="kp">workdir</span> <span class="o">=</span> <span class="n">CONFIG</span><span class="p">[</span><span class="s1">&#39;wd&#39;</span><span class="p">])</span>

<span class="p">[</span><span class="n">genotype_preprocessing</span><span class="p">]</span>
<span class="kn">input:</span> <span class="n">CONFIG</span><span class="p">[</span><span class="s1">&#39;genotype&#39;</span><span class="p">]</span>
<span class="kp">sos_run</span><span class="p">(</span><span class="s2">&quot;DW.vcf_by_chrom&quot;</span><span class="p">,</span> <span class="kp">workdir</span> <span class="o">=</span> <span class="n">CONFIG</span><span class="p">[</span><span class="s1">&#39;wd&#39;</span><span class="p">])</span>

<span class="p">[</span><span class="n">rna_preprocessing</span><span class="p">]</span>
<span class="kn">input:</span> <span class="n">CONFIG</span><span class="p">[</span><span class="s1">&#39;rna_rpkm&#39;</span><span class="p">],</span> <span class="n">CONFIG</span><span class="p">[</span><span class="s1">&#39;rna_cnts&#39;</span><span class="p">],</span> <span class="n">CONFIG</span><span class="p">[</span><span class="s1">&#39;genotype&#39;</span><span class="p">],</span> <span class="n">CONFIG</span><span class="p">[</span><span class="s1">&#39;sample_attr&#39;</span><span class="p">]</span> 
<span class="kp">sos_run</span><span class="p">(</span><span class="s2">&quot;MC.rnaseq&quot;</span><span class="p">,</span> <span class="kp">workdir</span> <span class="o">=</span>  <span class="n">CONFIG</span><span class="p">[</span><span class="s1">&#39;wd&#39;</span><span class="p">])</span>

<span class="p">[</span><span class="n">pca_plot_broad</span><span class="p">]</span>
<span class="kn">input:</span> <span class="s1">&#39;/tmp/GTExPCA.ped&#39;</span>
<span class="kp">sos_run</span><span class="p">(</span><span class="s2">&quot;MC.global_ancestry:2&quot;</span><span class="p">,</span> <span class="kp">workdir</span> <span class="o">=</span> <span class="s1">&#39;/tmp&#39;</span><span class="p">)</span>

<span class="p">[</span><span class="n">genotype_pca_broad</span><span class="p">]</span>
<span class="kn">parameter:</span> <span class="n">project_name</span> <span class="o">=</span> <span class="s2">&quot;GTEx7.dbGaP&quot;</span>
<span class="kn">input:</span> <span class="n">CONFIG</span><span class="p">[</span><span class="s1">&#39;genotype&#39;</span><span class="p">]</span>
<span class="kp">sos_run</span><span class="p">(</span><span class="s2">&quot;DW.vcf_by_chrom+DW.broad_to_plink+MC.LD_pruning+MC.global_ancestry&quot;</span><span class="p">,</span> 
        <span class="kp">workdir</span> <span class="o">=</span> <span class="n">CONFIG</span><span class="p">[</span><span class="s1">&#39;wd&#39;</span><span class="p">],</span>
        <span class="n">project_name</span> <span class="o">=</span> <span class="n">project_name</span><span class="p">)</span>

<span class="p">[</span><span class="n">genotype_pca_umich</span><span class="p">]</span>
<span class="kn">parameter:</span> <span class="n">project_name</span> <span class="o">=</span> <span class="s2">&quot;GTEx7.Imputed&quot;</span>
<span class="kn">input:</span> <span class="n">CONFIG</span><span class="p">[</span><span class="s1">&#39;imputed_genotype&#39;</span><span class="p">]</span>
<span class="kp">sos_run</span><span class="p">(</span><span class="s2">&quot;DW.umich_to_plink+MC.LD_pruning+MC.global_ancestry&quot;</span><span class="p">,</span> 
        <span class="kp">workdir</span> <span class="o">=</span> <span class="n">CONFIG</span><span class="p">[</span><span class="s1">&#39;wd&#39;</span><span class="p">],</span>
        <span class="n">project_name</span> <span class="o">=</span> <span class="n">project_name</span><span class="p">)</span>

<span class="p">[</span><span class="n">genotype_pca_umich_filtered</span><span class="p">]</span>
<span class="c1"># Filtered imputation data removing imputed sites</span>
<span class="kn">input:</span> <span class="s2">&quot;{}/GTEx7.Imputed.genotyped.filtered.bed&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">CONFIG</span><span class="p">[</span><span class="s1">&#39;wd&#39;</span><span class="p">])</span>
<span class="kp">sos_run</span><span class="p">(</span><span class="s2">&quot;MC.LD_pruning+MC.global_ancestry&quot;</span><span class="p">,</span> 
        <span class="kp">workdir</span> <span class="o">=</span> <span class="n">CONFIG</span><span class="p">[</span><span class="s1">&#39;wd&#39;</span><span class="p">])</span>

<span class="p">[</span><span class="n">gene_annotation</span><span class="p">:</span> <span class="kp">provides</span> <span class="o">=</span> <span class="s2">&quot;${CONFIG[&#39;rna_cnts&#39;]!n}.annotation&quot;</span><span class="p">]</span>
<span class="kn">input:</span> <span class="s2">&quot;${CONFIG[&#39;rna_cnts&#39;]}&quot;</span>
<span class="kn">output:</span> <span class="s2">&quot;${CONFIG[&#39;rna_cnts&#39;]!n}.annotation&quot;</span>
<span class="kp">sos_run</span><span class="p">(</span><span class="s2">&quot;MC.ensembl_annotation&quot;</span><span class="p">,</span> <span class="kp">workdir</span> <span class="o">=</span> <span class="n">CONFIG</span><span class="p">[</span><span class="s1">&#39;wd&#39;</span><span class="p">])</span>

<span class="p">[</span><span class="n">genotype_formatting</span><span class="p">]</span>
<span class="kn">parameter:</span> <span class="n">original_variants</span> <span class="o">=</span> <span class="s2">&quot;{}/GTEx7.dbGaP.bed&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">CONFIG</span><span class="p">[</span><span class="s1">&#39;wd&#39;</span><span class="p">])</span>
<span class="kn">parameter:</span> <span class="n">gene_annotation</span> <span class="o">=</span> <span class="s2">&quot;${CONFIG[&#39;rna_cnts&#39;]!n}.annotation&quot;</span>
<span class="kn">depends:</span> <span class="n">original_variants</span>
<span class="kn">input:</span> <span class="s2">&quot;{}/GTEx7.Imputed.bed&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">CONFIG</span><span class="p">[</span><span class="s1">&#39;wd&#39;</span><span class="p">])</span>
<span class="kp">sos_run</span><span class="p">(</span><span class="s2">&quot;DW.variants_filter+DW.plink_to_hdf5_batch&quot;</span><span class="p">,</span> 
        <span class="kp">workdir</span> <span class="o">=</span> <span class="n">CONFIG</span><span class="p">[</span><span class="s1">&#39;wd&#39;</span><span class="p">],</span> 
        <span class="n">include</span> <span class="o">=</span> <span class="n">original_variants</span><span class="p">,</span>
        <span class="n">ann</span> <span class="o">=</span> <span class="n">gene_annotation</span><span class="p">)</span>

<span class="p">[</span><span class="n">covariate_preparation</span><span class="p">]</span>
<span class="c1"># Covariates are: sex, platform, 3 PC and PEER factors</span>
<span class="kn">parameter:</span> <span class="n">peer_factors</span> <span class="o">=</span> <span class="n">glob</span><span class="o">.</span><span class="n">glob</span><span class="p">(</span><span class="s2">&quot;{}/*_PEER_covariates.txt&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="s2">&quot;${CONFIG[&#39;wd&#39;]!a}&quot;</span><span class="p">))</span>
<span class="kn">parameter:</span> <span class="n">pc_file</span> <span class="o">=</span> <span class="s2">&quot;{}/GTEx7.Imputed.prune.pc.ped&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">CONFIG</span><span class="p">[</span><span class="s1">&#39;wd&#39;</span><span class="p">])</span>
<span class="kn">parameter:</span> <span class="n">attr_file</span> <span class="o">=</span> <span class="n">CONFIG</span><span class="p">[</span><span class="s1">&#39;sample_attr&#39;</span><span class="p">]</span>
<span class="kn">parameter:</span> <span class="n">covar_file</span> <span class="o">=</span> <span class="n">CONFIG</span><span class="p">[</span><span class="s1">&#39;phenotype&#39;</span><span class="p">]</span>
<span class="kn">parameter:</span> <span class="n">expression_file</span> <span class="o">=</span> <span class="n">CONFIG</span><span class="p">[</span><span class="s1">&#39;expression_db&#39;</span><span class="p">]</span>
<span class="kp">sos_run</span><span class="p">(</span><span class="s2">&quot;DW.recode_platform + DW.covariates_to_HDF5&quot;</span><span class="p">,</span>
        <span class="kp">workdir</span> <span class="o">=</span> <span class="n">CONFIG</span><span class="p">[</span><span class="s1">&#39;wd&#39;</span><span class="p">],</span>
        <span class="n">peer_factors</span> <span class="o">=</span> <span class="n">peer_factors</span><span class="p">,</span>
        <span class="n">pc_file</span> <span class="o">=</span> <span class="n">pc_file</span><span class="p">,</span>
        <span class="n">attr_file</span> <span class="o">=</span> <span class="n">attr_file</span><span class="p">,</span>
        <span class="n">covar_file</span> <span class="o">=</span> <span class="n">covar_file</span><span class="p">,</span>
        <span class="n">output_file</span> <span class="o">=</span> <span class="s2">&quot;{}/GTEx7.Imputed.covariates.h5&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">CONFIG</span><span class="p">[</span><span class="s1">&#39;wd&#39;</span><span class="p">]))</span>

<span class="p">[</span><span class="n">make_toy</span><span class="p">]</span>
<span class="c1"># Create a toy example</span>
<span class="kp">sos_run</span><span class="p">(</span><span class="s2">&quot;DW.subset_HDF5_data&quot;</span><span class="p">,</span>
        <span class="kp">workdir</span> <span class="o">=</span> <span class="n">CONFIG</span><span class="p">[</span><span class="s1">&#39;wd&#39;</span><span class="p">],</span>
        <span class="n">ann_file</span> <span class="o">=</span> <span class="s2">&quot;${CONFIG[&#39;rna_cnts&#39;]!n}.annotation&quot;</span><span class="p">,</span>
        <span class="n">geno_file</span> <span class="o">=</span> <span class="s2">&quot;{}/GTEx7.Imputed.genotyped.filtered.cis.h5&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">CONFIG</span><span class="p">[</span><span class="s1">&#39;wd&#39;</span><span class="p">]),</span>
        <span class="n">expr_file</span> <span class="o">=</span> <span class="s2">&quot;{}/${CONFIG[&#39;rna_rpkm&#39;]!bnn}.qnorm.std.h5&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">CONFIG</span><span class="p">[</span><span class="s1">&#39;wd&#39;</span><span class="p">]),</span>
        <span class="n">toy_file</span> <span class="o">=</span> <span class="n">CONFIG</span><span class="p">[</span><span class="s1">&#39;toy_prefix&#39;</span><span class="p">],</span>
        <span class="n">gene_list</span> <span class="o">=</span> <span class="n">CONFIG</span><span class="p">[</span><span class="s1">&#39;toy_gene_list&#39;</span><span class="p">])</span>
</pre></div>

</div>
</div>
</div>

<div class="output_wrapper">
<div class="output">


<div class="output_area">
<div class="prompt"></div>


<div class="output_html rendered_html output_subarea ">
Workflow saved to <a href="prep.sos" target="_blank">prep.sos</a>
</div>

</div>

</div>
</div>

</div>
<div class="cell border-box-sizing text_cell rendered">
<div class="prompt input_prompt">
</div>
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h3 id="Prepare-computational-resource">Prepare computational resource<a class="anchor-link" href="#Prepare-computational-resource">&#182;</a></h3><p>This will download / install (most) software necessary for the analysis pipeline</p>

</div>
</div>
</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[&nbsp;]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-sos"><pre><span></span><span class="err">!</span><span class="o">./</span><span class="n">prep</span><span class="o">.</span><span class="n">sos</span> <span class="kp">download</span> <span class="o">-</span><span class="n">c</span> <span class="n">conf</span><span class="o">/</span><span class="mf">20170507.</span><span class="n">conf</span> <span class="o">-</span><span class="n">b</span> <span class="o">~/</span><span class="n">Documents</span><span class="o">/</span><span class="n">GTEx</span><span class="o">/</span><span class="nb">bin</span> <span class="o">-</span><span class="n">J</span> <span class="mi">8</span> <span class="o">-</span><span class="n">j</span> <span class="mi">1</span>
</pre></div>

</div>
</div>
</div>

</div>
<div class="cell border-box-sizing text_cell rendered">
<div class="prompt input_prompt">
</div>
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h3 id="Data-summary">Data summary<a class="anchor-link" href="#Data-summary">&#182;</a></h3><p>Summary statistics on data, such as missingness statistics on genotypes. It also includes various diagnostic plots. More features will be added to this workflow as analysis develops.</p>

</div>
</div>
</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[&nbsp;]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-sos"><pre><span></span><span class="err">!</span><span class="o">./</span><span class="n">prep</span><span class="o">.</span><span class="n">sos</span> <span class="n">data_summary</span> <span class="o">-</span><span class="n">c</span> <span class="n">conf</span><span class="o">/</span><span class="mf">20170507.</span><span class="n">conf</span> <span class="o">-</span><span class="n">b</span> <span class="o">~/</span><span class="n">Documents</span><span class="o">/</span><span class="n">GTEx</span><span class="o">/</span><span class="nb">bin</span> <span class="o">-</span><span class="n">J</span> <span class="mi">8</span> <span class="o">-</span><span class="n">j</span> <span class="mi">1</span>
</pre></div>

</div>
</div>
</div>

</div>
<div class="cell border-box-sizing text_cell rendered">
<div class="prompt input_prompt">
</div>
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h3 id="Genotype-imputation">Genotype imputation<a class="anchor-link" href="#Genotype-imputation">&#182;</a></h3>
</div>
</div>
</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[&nbsp;]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-sos"><pre><span></span><span class="err">!</span><span class="o">./</span><span class="n">prep</span><span class="o">.</span><span class="n">sos</span> <span class="n">genotype_preprocessing</span> <span class="o">-</span><span class="n">c</span> <span class="n">conf</span><span class="o">/</span><span class="mf">20170507.</span><span class="n">conf</span> <span class="o">-</span><span class="n">b</span> <span class="o">~/</span><span class="n">Documents</span><span class="o">/</span><span class="n">GTEx</span><span class="o">/</span><span class="nb">bin</span> <span class="o">-</span><span class="n">J</span> <span class="mi">8</span> <span class="o">-</span><span class="n">j</span> <span class="mi">1</span>
</pre></div>

</div>
</div>
</div>

</div>
<div class="cell border-box-sizing text_cell rendered">
<div class="prompt input_prompt">
</div>
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p>The imputation step was done with <a href="https://imputationserver.sph.umich.edu">Michigan Imputation Server</a> because it uses Haplotype Reference Consortium (32,914 samples) reference panel which is not publicly available otherwise. <a href="https://imputationserver.sph.umich.edu/start.html#!pages/help">Here is how to prepare data</a> for this service. The prepared files are uploaded to <a href="https://imputationserver.sph.umich.edu">Michigan imputation server</a>. <a href="https://gaow.github.io/mvarbvs/img/UMichImputation.png">Here is configuration</a> of imputation job on UMich server, and <a href="https://gaow.github.io/mvarbvs/img/UMichImputationResult.pdf">here</a> is summary of the outcome.</p>

</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered">
<div class="prompt input_prompt">
</div>
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h3 id="RNA-seq-preprocessing">RNA-seq preprocessing<a class="anchor-link" href="#RNA-seq-preprocessing">&#182;</a></h3><p>This workflow includes data normalization and PEER factor analysis. It results in 4 <strong>analysis ready expression data files</strong> in HDF5 format of different versions / organizations of the same information: emperical quantile normalized and standard normal quantile normalized, saved as a flat file or grouped by tissues. This is so far the most computational intensive step</p>

</div>
</div>
</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[&nbsp;]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-sos"><pre><span></span><span class="err">!</span><span class="o">./</span><span class="n">prep</span><span class="o">.</span><span class="n">sos</span> <span class="n">rna_preprocessing</span> <span class="o">-</span><span class="n">c</span> <span class="n">conf</span><span class="o">/</span><span class="mf">20170507.</span><span class="n">conf</span> <span class="o">-</span><span class="n">b</span> <span class="o">~/</span><span class="n">Documents</span><span class="o">/</span><span class="n">GTEx</span><span class="o">/</span><span class="nb">bin</span> <span class="o">-</span><span class="n">J</span> <span class="mi">38</span> <span class="o">-</span><span class="n">j</span> <span class="mi">1</span>
</pre></div>

</div>
</div>
</div>

</div>
<div class="cell border-box-sizing text_cell rendered">
<div class="prompt input_prompt">
</div>
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h3 id="Genotype-filtering-and-formatting,-followed-by-global-ancestry-analysis">Genotype filtering and formatting, followed by global ancestry analysis<a class="anchor-link" href="#Genotype-filtering-and-formatting,-followed-by-global-ancestry-analysis">&#182;</a></h3><p><code>genotype_pca_umich_filtered</code> has to be executed after both <code>genotype_pca_umich</code> and <code>genotype_pca_broad</code> are executed.</p>
<p>Genotypes are converted to PLINK binary format. During the conversion, imputed sites are removed and variant ID for tri-allelic sites are fixed. Also both the original and imputed data are converted for PCA analysis. The <a href="https://github.com/gaow/mvarbvs/issues/15#issuecomment-303814249">results are compared</a>.</p>

</div>
</div>
</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[&nbsp;]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-sos"><pre><span></span><span class="err">!</span><span class="o">./</span><span class="n">prep</span><span class="o">.</span><span class="n">sos</span> <span class="n">genotype_pca_umich</span> <span class="o">-</span><span class="n">c</span> <span class="n">conf</span><span class="o">/</span><span class="mf">20170507.</span><span class="n">conf</span> <span class="o">-</span><span class="n">b</span> <span class="o">~/</span><span class="n">Documents</span><span class="o">/</span><span class="n">GTEx</span><span class="o">/</span><span class="nb">bin</span><span class="o">/</span> <span class="o">-</span><span class="n">J</span> <span class="mi">4</span>
<span class="err">!</span><span class="o">./</span><span class="n">prep</span><span class="o">.</span><span class="n">sos</span> <span class="n">genotype_pca_broad</span> <span class="o">-</span><span class="n">c</span> <span class="n">conf</span><span class="o">/</span><span class="mf">20170507.</span><span class="n">conf</span> <span class="o">-</span><span class="n">b</span> <span class="o">~/</span><span class="n">Documents</span><span class="o">/</span><span class="n">GTEx</span><span class="o">/</span><span class="nb">bin</span><span class="o">/</span> <span class="o">-</span><span class="n">J</span> <span class="mi">4</span>
<span class="err">!</span><span class="o">./</span><span class="n">prep</span><span class="o">.</span><span class="n">sos</span> <span class="n">pca_plot_broad</span>
<span class="err">!</span><span class="o">./</span><span class="n">prep</span><span class="o">.</span><span class="n">sos</span> <span class="n">gene_annotation</span> <span class="o">-</span><span class="n">c</span> <span class="n">conf</span><span class="o">/</span><span class="mf">20170507.</span><span class="n">conf</span> <span class="o">-</span><span class="n">b</span> <span class="o">~/</span><span class="n">Documents</span><span class="o">/</span><span class="n">GTEx</span><span class="o">/</span><span class="nb">bin</span> <span class="o">-</span><span class="n">J</span> <span class="mi">6</span> <span class="o">-</span><span class="n">j</span> <span class="mi">1</span>
<span class="err">!</span><span class="o">./</span><span class="n">prep</span><span class="o">.</span><span class="n">sos</span> <span class="n">genotype_formatting</span> <span class="o">-</span><span class="n">c</span> <span class="n">conf</span><span class="o">/</span><span class="mf">20170507.</span><span class="n">conf</span> <span class="o">-</span><span class="n">b</span> <span class="o">~/</span><span class="n">Documents</span><span class="o">/</span><span class="n">GTEx</span><span class="o">/</span><span class="nb">bin</span> <span class="o">-</span><span class="n">J</span> <span class="mi">6</span> <span class="o">-</span><span class="n">j</span> <span class="mi">1</span>
<span class="err">!</span><span class="o">./</span><span class="n">prep</span><span class="o">.</span><span class="n">sos</span> <span class="n">genotype_pca_umich_filtered</span> <span class="o">-</span><span class="n">c</span> <span class="n">conf</span><span class="o">/</span><span class="mf">20170507.</span><span class="n">conf</span> <span class="o">-</span><span class="n">b</span> <span class="o">~/</span><span class="n">Documents</span><span class="o">/</span><span class="n">GTEx</span><span class="o">/</span><span class="nb">bin</span><span class="o">/</span> <span class="o">-</span><span class="n">J</span> <span class="mi">6</span>
</pre></div>

</div>
</div>
</div>

</div>
<div class="cell border-box-sizing text_cell rendered">
<div class="prompt input_prompt">
</div>
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h3 id="Variants-annotation,-cis-SNP-selection-and-genotype-formatting">Variants annotation, cis-SNP selection and genotype formatting<a class="anchor-link" href="#Variants-annotation,-cis-SNP-selection-and-genotype-formatting">&#182;</a></h3><p>Genes are annotated to chromosomal positions, and annotate variants to genes. Then for each gene, variants are selected 2MB of a gene's TSS. It results a <strong>single analysis ready file</strong> in HDF5 format containing ~50K groups of genotype data (gene-names).</p>

</div>
</div>
</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[&nbsp;]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-sos"><pre><span></span><span class="err">!</span><span class="o">./</span><span class="n">prep</span><span class="o">.</span><span class="n">sos</span> <span class="n">gene_annotation</span> <span class="o">-</span><span class="n">c</span> <span class="n">conf</span><span class="o">/</span><span class="mf">20170507.</span><span class="n">conf</span> <span class="o">-</span><span class="n">b</span> <span class="o">~/</span><span class="n">Documents</span><span class="o">/</span><span class="n">GTEx</span><span class="o">/</span><span class="nb">bin</span> <span class="o">-</span><span class="n">J</span> <span class="mi">6</span> <span class="o">-</span><span class="n">j</span> <span class="mi">1</span>
<span class="err">!</span><span class="o">./</span><span class="n">prep</span><span class="o">.</span><span class="n">sos</span> <span class="n">genotype_formatting</span> <span class="o">-</span><span class="n">c</span> <span class="n">conf</span><span class="o">/</span><span class="mf">20170507.</span><span class="n">conf</span> <span class="o">-</span><span class="n">b</span> <span class="o">~/</span><span class="n">Documents</span><span class="o">/</span><span class="n">GTEx</span><span class="o">/</span><span class="nb">bin</span> <span class="o">-</span><span class="n">J</span> <span class="mi">6</span> <span class="o">-</span><span class="n">j</span> <span class="mi">1</span>
</pre></div>

</div>
</div>
</div>

</div>
<div class="cell border-box-sizing text_cell rendered">
<div class="prompt input_prompt">
</div>
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h3 id="Merge-covariates-info">Merge covariates info<a class="anchor-link" href="#Merge-covariates-info">&#182;</a></h3><p>Covariates for analysis we've got so far include sample phenotypes (sex), sample attributes (genotyping platform), first principle components for population structure, and PEER factors. All saved in various files.</p>
<p>This workflow consolidates these files and generates a <strong>single analysis ready covariate file</strong> in HDF5 format.</p>

</div>
</div>
</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[&nbsp;]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-sos"><pre><span></span><span class="err">!</span><span class="o">./</span><span class="n">prep</span><span class="o">.</span><span class="n">sos</span> <span class="n">covariate_preparation</span> <span class="o">-</span><span class="n">c</span> <span class="n">conf</span><span class="o">/</span><span class="mf">20170507.</span><span class="n">conf</span> <span class="o">-</span><span class="n">b</span> <span class="o">~/</span><span class="n">Documents</span><span class="o">/</span><span class="n">GTEx</span><span class="o">/</span><span class="nb">bin</span><span class="o">/</span>
</pre></div>

</div>
</div>
</div>

</div>
<div class="cell border-box-sizing text_cell rendered">
<div class="prompt input_prompt">
</div>
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h3 id="Generate-a-toy-data-set">Generate a toy data-set<a class="anchor-link" href="#Generate-a-toy-data-set">&#182;</a></h3><p>Finally, a toy data-set is created from the data bundle. This toy can be used for methods / pipeline development. Genes selected for the toy are the same as the <a href="https://stephenslab.github.io/gtexresults_mash/TwoSNP/2SNP.sos">LD show-case in the mash paper</a> (although the workflow itself takes an arbitary list of genes). See <a href="https://stephenslab.github.io/gtexresults_mash/TwoSNP/">this table</a> for motivation that these genes get selected.</p>

</div>
</div>
</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[&nbsp;]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-sos"><pre><span></span><span class="err">!</span><span class="o">./</span><span class="n">prep</span><span class="o">.</span><span class="n">sos</span> <span class="n">make_toy</span> <span class="o">-</span><span class="n">c</span> <span class="n">conf</span><span class="o">/</span><span class="mf">20170507.</span><span class="n">conf</span> <span class="o">-</span><span class="n">b</span> <span class="o">~/</span><span class="n">Documents</span><span class="o">/</span><span class="n">GTEx</span><span class="o">/</span><span class="nb">bin</span> <span class="o">-</span><span class="n">J</span> <span class="mi">6</span> <span class="o">-</span><span class="n">j</span> <span class="mi">1</span>
</pre></div>

</div>
</div>
</div>

</div>
<div class="cell border-box-sizing text_cell rendered">
<div class="prompt input_prompt">
</div>
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h2 id="Simulations">Simulations<a class="anchor-link" href="#Simulations">&#182;</a></h2><p>Please see <a href="../documentation/MR-ASH-Simulation.html">this notebook</a> for interactive codes simulating expression data for given genotypes.</p>

</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered">
<div class="prompt input_prompt">
</div>
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h2 id="Univariate-eQTL-analysis">Univariate eQTL analysis<a class="anchor-link" href="#Univariate-eQTL-analysis">&#182;</a></h2><p>Here I perform eQTL analysis using</p>

</div>
</div>
</div>
    </div>
  </div>
</body>

 


</html>
