
<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="ipynb_website:version" content="0.9.6" />
<meta name="viewport" content="width=device-width, initial-scale=1" />

<link rel="stylesheet" type="text/css" href="../css/jt.css">

<link rel="stylesheet" type="text/css" href="../css/toc2.css">

<link href="../site_libs/jqueryui-1.11.4/jquery-ui.css">
<link rel="stylesheet" href="../site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<link rel="stylesheet" href="../site_libs/font-awesome-4.5.0/css/font-awesome.min.css" rel="stylesheet" />
<script type="text/javascript" src="https://ajax.googleapis.com/ajax/libs/jquery/1.9.1/jquery.min.js"></script>
<script type="text/javascript" src="https://ajax.googleapis.com/ajax/libs/jqueryui/1.9.1/jquery-ui.min.js"></script>
<script src="../site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="../site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="../site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>

<link rel="stylesheet"
      href="../site_libs/highlightjs/null.min.css"
      type="text/css" />

<script src="../site_libs/highlightjs/highlight.pack.js"></script>
<script>hljs.initHighlightingOnLoad();</script>
<script type="text/javascript">
if (window.hljs && document.readyState && document.readyState === "complete") {
   window.setTimeout(function() {
      hljs.initHighlighting();
   }, 0);
}
</script>

<script src="../js/doc_toc.js"></script>
<script src="../js/docs.js"></script>

<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.2/MathJax.js?config=TeX-MML-AM_CHTML"></script>
<script>
    MathJax.Hub.Config({
        extensions: ["tex2jax.js"],
        jax: ["input/TeX", "output/HTML-CSS"],
        tex2jax: {
        inlineMath: [ ['$','$'], ["\\(","\\)"] ],
        displayMath: [ ['$$','$$'], ["\\[","\\]"] ],
        processEscapes: true
        },
        "HTML-CSS": {
            preferredFont: "TeX",
            availableFonts: ["TeX"],
            styles: {
                scale: 110,
                ".MathJax_Display": {
                    "font-size": "110%",
                }
            }
        }
    });
</script>
<script>
function filterDataFrame(id) {
    var input = document.getElementById("search_" + id);
    var filter = input.value.toUpperCase();
    var table = document.getElementById("dataframe_" + id);
    var tr = table.getElementsByTagName("tr");
    // Loop through all table rows, and hide those who don't match the search query
    for (var i = 1; i < tr.length; i++) {
        for (var j = 0; j < tr[i].cells.length; ++j) {
            var matched = false;
            if (tr[i].cells[j].innerHTML.toUpperCase().indexOf(filter) != -1) {
                tr[i].style.display = "";
                matched = true
                break;
            }
            if (!matched)
                tr[i].style.display = "none";
        }
    }
}
function sortDataFrame(id, n, dtype) {
    var table = document.getElementById("dataframe_" + id);
    var tb = table.tBodies[0]; // use `<tbody>` to ignore `<thead>` and `<tfoot>` rows
    var tr = Array.prototype.slice.call(tb.rows, 0); // put rows into array
    if (dtype === 'numeric') {
        var fn = function(a, b) { 
            return parseFloat(a.cells[n].textContent) <= parseFloat(b.cells[n].textContent) ? -1 : 1;
        }
    } else {
        var fn = function(a, b) {
            var c = a.cells[n].textContent.trim().localeCompare(b.cells[n].textContent.trim()); 
            return c > 0 ? 1 : (c < 0 ? -1 : 0) }
    }
    var isSorted = function(array, fn) {
        if (array.length < 2)
            return 1;
        var direction = fn(array[0], array[1]); 
        for (var i = 1; i < array.length - 1; ++i) {
            var d = fn(array[i], array[i+1]);
            if (d == 0)
                continue;
            else if (direction == 0)
                direction = d;
            else if (direction != d)
                return 0;
            }
        return direction;
    }
    var sorted = isSorted(tr, fn);
    if (sorted == 1 || sorted == -1) {
        // if sorted already, reverse it
        for(var i = tr.length - 1; i >= 0; --i)
            tb.appendChild(tr[i]); // append each row in order
    } else {
        tr = tr.sort(fn);
        for(var i = 0; i < tr.length; ++i)
            tb.appendChild(tr[i]); // append each row in order
    }
}
</script>

<script>
$( document ).ready(function(){
            var cfg={'threshold':3,     // depth of toc (number of levels)
             'number_sections': false,
             'toc_cell': false,          // useless here
             'toc_window_display': true, // display the toc window
             "toc_section_display": "block", // display toc contents in the window
             'sideBar':true,       // sidebar or floating window
             'navigate_menu':false       // navigation menu (only in liveNotebook -- do not change)
            }
            var st={};                  // some variables used in the script
            st.rendering_toc_cell = false;
            st.config_loaded = false;
            st.extension_initialized=false;
            st.nbcontainer_marginleft = $('#notebook-container').css('margin-left')
            st.nbcontainer_marginright = $('#notebook-container').css('margin-right')
            st.nbcontainer_width = $('#notebook-container').css('width')
            st.oldTocHeight = undefined
            st.cell_toc = undefined;
            st.toc_index=0;
            // fire the main function with these parameters
            table_of_contents(cfg, st);
            var file=analysisDict[$("h1:first").attr("id")];
            $("#toc-level0 a").css("color","#126dce");
            $('a[href="#'+$("h1:first").attr("id")+'"]').hide()
            var docs=analysisArray;
            var docs_map=analysisArrayMap;
            var pos=analysisArray.indexOf(file);
            for (var a=pos;a>=0;a--){
                  $('<li><a href="'+docs[a]+'.html"><font color="#073642"><b>'+docs_map[docs[a]].replace(/_/g," ")+'</b></font></a></li>').insertBefore("#toc-level0 li:eq(0)");
            }
            $('a[href="'+file+'.html'+'"]').css("color","#126dce");
            for (var a=pos+1;a<docs.length;a++){
                  $(".toc #toc-level0").append('<li><a href="'+docs[a]+'.html"><font color="#073642"><b>'+docs_map[docs[a]].replace(/_/g," ")+'</b></font></a></li>');
            }
            // $("#toc-header").hide(); // comment out because it prevents search bar from displaying
    });
</script>

<script>
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');
  // mark it active
  menuAnchor.parent().addClass('active');
  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');
});
</script>
<div class="container-fluid main-container">
<!-- tabsets -->
<script src="../site_libs/navigation-1.1/tabsets.js"></script>
<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});
</script>



<title>m&m ash</title>

<style type = "text/css">
body {
  font-family: "Droid Sans";
  padding-top: 66px;
  padding-bottom: 40px;
}
</style>
</head>

<body>
<div tabindex="-1" id="notebook" class="border-box-sizing">
<div class="container" id="notebook-container">

<!-- code folding -->

<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="../index.html">m&m ash</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        
<li>
  <a href="../index.html">Overview</a>
</li>
        
<li>
  <a href="../details.html">Details</a>
</li>
        
<li>
  <a href="../analysis.html">Analysis</a>
</li>
        
<li>
  <a href="../prototype.html">Prototype</a>
</li>
        
<li>
  <a href="../dsc.html">Dsc</a>
</li>
        
<li>
  <a href="../writeup.html">Writeup</a>
</li>
        
      </ul>
        
<ul class="nav navbar-nav navbar-right">
<li>
   <a href="http://github.com/gaow/mvarbvs"> <span class="fa fa-github"></span> </a>
</li>
</ul>
        
      </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->
<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div>
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h1 id="Extract-per-gene-dataset">Extract per gene dataset<a class="anchor-link" href="#Extract-per-gene-dataset">&#182;</a></h1>
</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div>
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p>This generates per gene data-set and surveys how many SNPs in each generated dataset.</p>

</div>
</div>
</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[1]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-sos"><pre><span></span><span class="o">%</span><span class="n">revisions</span> <span class="o">-</span><span class="n">s</span> <span class="o">-</span><span class="n">n</span> <span class="mi">10</span>
</pre></div>

</div>
</div>
</div>

<div class="output_wrapper">
<div class="output">


<div class="output_area">

<div class="prompt"></div>



<div class="output_html rendered_html output_subarea ">

        <table class="revision_table">
        <tr>
        <th>Revision</th>
        <th>Author</th>
        <th>Date</th>
        <th>Message</th>
        <tr>
        <tr><td><a target="_blank" href="http://github.com/gaow/mvarbvs/blob/d4e8a8cd3ad151e98bb6644d7fe5ddf721fd9e5b/analysis/20180515_Extract_Benchmark_Data.ipynb"><span class="revision_id">d4e8a8c<span></a></td>
<td>Gao Wang</td>
<td>2018-05-24</td>
<td>Update documentation</td></tr></table>
</div>

</div>

</div>
</div>

</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[&nbsp;]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-sos"><pre><span></span><span class="p">[</span><span class="k">global</span><span class="p">]</span>
<span class="kn">parameter:</span> <span class="n">cwd</span> <span class="o">=</span> <span class="n">path</span><span class="p">(</span><span class="s1">&#39;~/Documents/GTExV8&#39;</span><span class="p">)</span>
<span class="kn">parameter:</span> <span class="n">genotype_file</span> <span class="o">=</span> <span class="n">f</span><span class="s2">&quot;{cwd:a}/GTExV7.genotype.cis.h5&quot;</span>
<span class="kn">parameter:</span> <span class="n">expr_file</span> <span class="o">=</span> <span class="n">f</span><span class="s2">&quot;{cwd:a}/GTExV7.expression.h5&quot;</span>
<span class="kn">parameter:</span> <span class="n">cov_file</span> <span class="o">=</span> <span class="n">f</span><span class="s2">&quot;{cwd:a}/GTExV7.covariates.h5&quot;</span>
<span class="kn">parameter:</span> <span class="n">outdir</span> <span class="o">=</span> <span class="n">f</span><span class="s1">&#39;{cwd:a}/Toys&#39;</span>
<span class="c1"># we can extract single or multiple tissue data</span>
<span class="kn">parameter:</span> <span class="n">tissues</span> <span class="o">=</span> <span class="p">[]</span>
<span class="c1"># list of genes to extract data for</span>
<span class="kn">parameter:</span> <span class="n">genes</span> <span class="o">=</span> <span class="p">[]</span>
<span class="c1"># if list of genes is not provided we&#39;ll get from a fixed file `n` genes</span>
<span class="kn">parameter:</span> <span class="n">n</span> <span class="o">=</span> <span class="mi">200</span>

<span class="n">tissues_default</span> <span class="o">=</span> <span class="s1">&#39;&#39;&#39;</span>
<span class="s1">Adipose_Subcutaneous</span>
<span class="s1">Adipose_Visceral_Omentum</span>
<span class="s1">Adrenal_Gland</span>
<span class="s1">Artery_Aorta</span>
<span class="s1">Artery_Coronary</span>
<span class="s1">Artery_Tibial</span>
<span class="s1">Brain_Amygdala</span>
<span class="s1">Brain_Anterior_cingulate_cortex_BA24</span>
<span class="s1">Brain_Caudate_basal_ganglia</span>
<span class="s1">Brain_Cerebellar_Hemisphere</span>
<span class="s1">Brain_Cerebellum</span>
<span class="s1">Brain_Cortex</span>
<span class="s1">Brain_Frontal_Cortex_BA9</span>
<span class="s1">Brain_Hippocampus</span>
<span class="s1">Brain_Hypothalamus</span>
<span class="s1">Brain_Nucleus_accumbens_basal_ganglia</span>
<span class="s1">Brain_Putamen_basal_ganglia</span>
<span class="s1">Brain_Spinal_cord_cervical_c-1</span>
<span class="s1">Brain_Substantia_nigra</span>
<span class="s1">Breast_Mammary_Tissue</span>
<span class="s1">Cells_Cultured_fibroblasts</span>
<span class="s1">Cells_EBV-transformed_lymphocytes</span>
<span class="s1">Colon_Sigmoid</span>
<span class="s1">Colon_Transverse</span>
<span class="s1">Esophagus_Gastroesophageal_Junction</span>
<span class="s1">Esophagus_Mucosa</span>
<span class="s1">Esophagus_Muscularis</span>
<span class="s1">Heart_Atrial_Appendage</span>
<span class="s1">Heart_Left_Ventricle</span>
<span class="s1">Kidney_Cortex</span>
<span class="s1">Liver</span>
<span class="s1">Lung</span>
<span class="s1">Minor_Salivary_Gland</span>
<span class="s1">Muscle_Skeletal</span>
<span class="s1">Nerve_Tibial</span>
<span class="s1">Ovary</span>
<span class="s1">Pancreas</span>
<span class="s1">Pituitary</span>
<span class="s1">Prostate</span>
<span class="s1">Skin_Not_Sun_Exposed_Suprapubic</span>
<span class="s1">Skin_Sun_Exposed_Lower_leg</span>
<span class="s1">Small_Intestine_Terminal_Ileum</span>
<span class="s1">Spleen</span>
<span class="s1">Stomach</span>
<span class="s1">Testis</span>
<span class="s1">Thyroid</span>
<span class="s1">Uterus</span>
<span class="s1">Vagina</span>
<span class="s1">Whole_Blood</span>
<span class="s1">&#39;&#39;&#39;</span>

<span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">tissues</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
    <span class="n">tissues</span> <span class="o">=</span> <span class="n">tissues_default</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>
</pre></div>

</div>
</div>
</div>

</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[&nbsp;]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-sos"><pre><span></span><span class="p">[</span><span class="n">extract</span><span class="p">]</span>
<span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">genes</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
    <span class="n">genes</span> <span class="o">=</span> <span class="n">get_output</span><span class="p">(</span><span class="n">f</span><span class="s1">&#39;shuf {cwd:a}/all_genes.txt | sort -u | head -{n} &#39;</span><span class="p">)</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span>
<span class="n">tissue_tables</span> <span class="o">=</span> <span class="s1">&#39;c(&#39;</span> <span class="o">+</span> <span class="s1">&#39;,&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span><span class="n">f</span><span class="s1">&#39;&quot;/{item}&quot;&#39;</span> <span class="k">for</span> <span class="n">item</span> <span class="ow">in</span> <span class="n">tissues</span><span class="p">])</span> <span class="o">+</span> <span class="s1">&#39;)&#39;</span>
<span class="kn">depends:</span> <span class="n">R_library</span><span class="p">(</span><span class="s1">&#39;rhdf5&#39;</span><span class="p">),</span> <span class="n">R_library</span><span class="p">(</span><span class="s1">&#39;plyr&#39;</span><span class="p">)</span>
<span class="kn">input:</span> <span class="n">for_each</span> <span class="o">=</span> <span class="s1">&#39;genes&#39;</span><span class="p">,</span> <span class="kp">concurrent</span> <span class="o">=</span> <span class="bp">True</span>
<span class="kn">output:</span> <span class="n">f</span><span class="s2">&quot;{outdir}/{tissues[0] if len(tissues) == 1 else &#39;Multi_Tissues&#39;}.{path(_genes):b}.RDS&quot;</span>
<span class="kn">R:</span> <span class="n">expand</span> <span class="o">=</span> <span class="s1">&#39;${ }&#39;</span>
    <span class="n">library</span><span class="p">(</span><span class="n">rhdf5</span><span class="p">)</span>
    <span class="n">load_genotype</span> <span class="o">=</span> <span class="n">function</span><span class="p">(</span><span class="n">genotype_file</span><span class="p">,</span> <span class="n">geno_table</span><span class="p">)</span> <span class="p">{</span>
            <span class="n">geno</span> <span class="o">=</span> <span class="n">h5read</span><span class="p">(</span><span class="n">genotype_file</span><span class="p">,</span> <span class="n">geno_table</span><span class="p">)</span>
            <span class="n">gdata</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">frame</span><span class="p">(</span><span class="n">geno</span><span class="err">$</span><span class="n">block0_values</span><span class="p">)</span>
            <span class="n">colnames</span><span class="p">(</span><span class="n">gdata</span><span class="p">)</span> <span class="o">=</span> <span class="n">geno</span><span class="err">$</span><span class="n">axis1</span>
            <span class="n">rownames</span><span class="p">(</span><span class="n">gdata</span><span class="p">)</span> <span class="o">=</span> <span class="n">geno</span><span class="err">$</span><span class="n">axis0</span>
            <span class="k">return</span><span class="p">(</span><span class="n">gdata</span><span class="p">)</span>
    <span class="p">}</span>
    <span class="n">load_expression</span> <span class="o">=</span> <span class="n">function</span><span class="p">(</span><span class="n">expr_file</span><span class="p">,</span> <span class="n">expr_table</span><span class="p">,</span> <span class="n">gene</span><span class="p">)</span> <span class="p">{</span>
            <span class="n">expr</span> <span class="o">=</span> <span class="n">h5read</span><span class="p">(</span><span class="n">expr_file</span><span class="p">,</span> <span class="n">expr_table</span><span class="p">)</span>
            <span class="n">edata</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">frame</span><span class="p">(</span><span class="n">expr</span><span class="err">$</span><span class="n">block0_values</span><span class="p">)</span>
            <span class="n">colnames</span><span class="p">(</span><span class="n">edata</span><span class="p">)</span> <span class="o">=</span> <span class="n">tools</span><span class="p">::</span><span class="n">file_path_sans_ext</span><span class="p">(</span><span class="n">expr</span><span class="err">$</span><span class="n">axis1</span><span class="p">)</span>
            <span class="n">edata</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">frame</span><span class="p">(</span><span class="n">edata</span><span class="p">[,</span> <span class="n">gene</span><span class="p">])</span>
            <span class="n">rownames</span><span class="p">(</span><span class="n">edata</span><span class="p">)</span> <span class="o">=</span> <span class="n">expr</span><span class="err">$</span><span class="n">axis0</span>
            <span class="n">colnames</span><span class="p">(</span><span class="n">edata</span><span class="p">)</span> <span class="o">=</span> <span class="n">basename</span><span class="p">(</span><span class="n">expr_table</span><span class="p">)</span>
            <span class="k">return</span><span class="p">(</span><span class="n">edata</span><span class="p">)</span>
    <span class="p">}</span>
    <span class="n">load_covariates</span> <span class="o">=</span> <span class="n">function</span><span class="p">(</span><span class="n">cov_file</span><span class="p">,</span> <span class="n">cov_table</span><span class="p">)</span> <span class="p">{</span>
            <span class="n">covariate</span> <span class="o">=</span> <span class="n">h5read</span><span class="p">(</span><span class="n">cov_file</span><span class="p">,</span> <span class="n">cov_table</span><span class="p">)</span>
            <span class="n">cdata</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">frame</span><span class="p">(</span><span class="n">covariate</span><span class="err">$</span><span class="n">block0_values</span><span class="p">)</span>
            <span class="n">colnames</span><span class="p">(</span><span class="n">cdata</span><span class="p">)</span> <span class="o">=</span> <span class="n">covariate</span><span class="err">$</span><span class="n">axis1</span>
            <span class="n">rownames</span><span class="p">(</span><span class="n">cdata</span><span class="p">)</span> <span class="o">=</span> <span class="n">covariate</span><span class="err">$</span><span class="n">axis0</span>
            <span class="n">cdata</span> <span class="o">=</span> <span class="n">t</span><span class="p">(</span><span class="n">cdata</span><span class="p">)</span>
            <span class="k">return</span><span class="p">(</span><span class="n">cdata</span><span class="p">)</span>
    <span class="p">}</span>
    <span class="n">load_data_single</span> <span class="o">=</span> <span class="n">function</span><span class="p">(</span><span class="n">genotype_file</span><span class="p">,</span> <span class="n">expr_file</span><span class="p">,</span> <span class="n">cov_file</span><span class="p">,</span> <span class="n">geno_table</span><span class="p">,</span> <span class="n">expr_table</span><span class="p">,</span> <span class="n">cov_table</span><span class="p">)</span> <span class="p">{</span>
            <span class="n">gdata</span> <span class="o">=</span> <span class="n">load_genotype</span><span class="p">(</span><span class="n">genotype_file</span><span class="p">,</span> <span class="n">geno_table</span><span class="p">)</span>
            <span class="n">edata</span> <span class="o">=</span> <span class="n">load_expression</span><span class="p">(</span><span class="n">expr_file</span><span class="p">,</span> <span class="n">expr_table</span><span class="p">,</span> <span class="n">basename</span><span class="p">(</span><span class="n">geno_table</span><span class="p">))</span>
            <span class="n">cdata</span> <span class="o">=</span> <span class="n">load_covariates</span><span class="p">(</span><span class="n">cov_file</span><span class="p">,</span> <span class="n">cov_table</span><span class="p">)</span>
            <span class="c1"># reorder</span>
            <span class="n">edata</span> <span class="o">=</span> <span class="n">edata</span><span class="p">[</span><span class="n">match</span><span class="p">(</span><span class="n">rownames</span><span class="p">(</span><span class="n">cdata</span><span class="p">),</span> <span class="n">rownames</span><span class="p">(</span><span class="n">edata</span><span class="p">)),]</span>
            <span class="n">gdata</span> <span class="o">=</span> <span class="n">gdata</span><span class="p">[</span><span class="n">rownames</span><span class="p">(</span><span class="n">cdata</span><span class="p">),]</span>
            <span class="k">return</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="n">X</span> <span class="o">=</span> <span class="k">as</span><span class="o">.</span><span class="n">matrix</span><span class="p">(</span><span class="n">gdata</span><span class="p">),</span> <span class="n">y</span> <span class="o">=</span> <span class="k">as</span><span class="o">.</span><span class="n">vector</span><span class="p">(</span><span class="n">edata</span><span class="p">),</span> <span class="n">Z</span> <span class="o">=</span> <span class="k">as</span><span class="o">.</span><span class="n">matrix</span><span class="p">(</span><span class="n">cdata</span><span class="p">),</span> <span class="n">chrom</span> <span class="o">=</span> <span class="n">strsplit</span><span class="p">(</span><span class="n">colnames</span><span class="p">(</span><span class="n">gdata</span><span class="p">)[</span><span class="mi">1</span><span class="p">],</span> <span class="s1">&#39;_&#39;</span><span class="p">)[[</span><span class="mi">1</span><span class="p">]][</span><span class="mi">1</span><span class="p">],</span> 
                        <span class="n">pos</span> <span class="o">=</span> <span class="k">as</span><span class="o">.</span><span class="n">integer</span><span class="p">(</span><span class="n">do</span><span class="o">.</span><span class="n">call</span><span class="p">(</span><span class="n">rbind</span><span class="p">,</span> <span class="n">strsplit</span><span class="p">(</span><span class="n">colnames</span><span class="p">(</span><span class="n">gdata</span><span class="p">),</span> <span class="s1">&#39;_&#39;</span><span class="p">))[,</span><span class="mi">2</span><span class="p">])))</span>
    <span class="p">}</span>
    <span class="n">load_data_multi</span> <span class="o">=</span> <span class="n">function</span><span class="p">(</span><span class="n">genotype_file</span><span class="p">,</span> <span class="n">expr_file</span><span class="p">,</span> <span class="n">cov_file</span><span class="p">,</span> <span class="n">geno_table</span><span class="p">,</span> <span class="n">expr_tables</span><span class="p">,</span> <span class="n">cov_tables</span><span class="p">)</span> <span class="p">{</span>
            <span class="n">gdata</span> <span class="o">=</span> <span class="n">load_genotype</span><span class="p">(</span><span class="n">genotype_file</span><span class="p">,</span> <span class="n">geno_table</span><span class="p">)</span>
            <span class="n">edata</span> <span class="o">=</span> <span class="n">lapply</span><span class="p">(</span><span class="mi">1</span><span class="p">:</span><span class="n">length</span><span class="p">(</span><span class="n">expr_tables</span><span class="p">),</span> <span class="n">function</span><span class="p">(</span><span class="n">i</span><span class="p">)</span> <span class="n">load_expression</span><span class="p">(</span><span class="n">expr_file</span><span class="p">,</span> <span class="n">expr_tables</span><span class="p">[[</span><span class="n">i</span><span class="p">]],</span> <span class="n">basename</span><span class="p">(</span><span class="n">geno_table</span><span class="p">)))</span>
            <span class="n">cdata</span> <span class="o">=</span> <span class="n">lapply</span><span class="p">(</span><span class="mi">1</span><span class="p">:</span><span class="n">length</span><span class="p">(</span><span class="n">cov_tables</span><span class="p">),</span> <span class="n">function</span><span class="p">(</span><span class="n">i</span><span class="p">)</span> <span class="n">load_covariates</span><span class="p">(</span><span class="n">cov_file</span><span class="p">,</span> <span class="n">cov_tables</span><span class="p">[[</span><span class="n">i</span><span class="p">]]))</span>
            <span class="n">y_res</span> <span class="o">=</span> <span class="nb">list</span><span class="p">()</span>
            <span class="c1"># reorder and remove covariates</span>
            <span class="k">for</span> <span class="p">(</span><span class="n">i</span> <span class="ow">in</span> <span class="mi">1</span><span class="p">:</span><span class="n">length</span><span class="p">(</span><span class="n">edata</span><span class="p">))</span> <span class="p">{</span>
                <span class="n">idx</span> <span class="o">=</span> <span class="n">match</span><span class="p">(</span><span class="n">rownames</span><span class="p">(</span><span class="n">cdata</span><span class="p">[[</span><span class="n">i</span><span class="p">]]),</span> <span class="n">rownames</span><span class="p">(</span><span class="n">edata</span><span class="p">[[</span><span class="n">i</span><span class="p">]]))</span>
                <span class="n">row_str</span> <span class="o">=</span> <span class="n">rownames</span><span class="p">(</span><span class="n">edata</span><span class="p">[[</span><span class="n">i</span><span class="p">]])[</span><span class="n">idx</span><span class="p">]</span>
                <span class="n">col_str</span> <span class="o">=</span> <span class="n">colnames</span><span class="p">(</span><span class="n">edata</span><span class="p">[[</span><span class="n">i</span><span class="p">]])</span>
                <span class="n">edata</span><span class="p">[[</span><span class="n">i</span><span class="p">]]</span> <span class="o">=</span> <span class="n">edata</span><span class="p">[[</span><span class="n">i</span><span class="p">]][</span><span class="n">idx</span><span class="p">,]</span>
                <span class="n">Z</span> <span class="o">=</span> <span class="n">scale</span><span class="p">(</span><span class="n">cdata</span><span class="p">[[</span><span class="n">i</span><span class="p">]],</span> <span class="n">center</span><span class="o">=</span><span class="n">TRUE</span><span class="p">,</span> <span class="n">scale</span><span class="o">=</span><span class="n">FALSE</span><span class="p">)</span>
                <span class="n">y_res</span><span class="p">[[</span><span class="n">i</span><span class="p">]]</span> <span class="o">=</span> <span class="o">.</span><span class="n">lm</span><span class="o">.</span><span class="n">fit</span><span class="p">(</span><span class="n">Z</span><span class="p">,</span> <span class="n">edata</span><span class="p">[[</span><span class="n">i</span><span class="p">]])</span><span class="err">$</span><span class="n">residuals</span>
                <span class="n">y_res</span><span class="p">[[</span><span class="n">i</span><span class="p">]]</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">frame</span><span class="p">(</span><span class="n">matrix</span><span class="p">(</span><span class="n">y_res</span><span class="p">[[</span><span class="n">i</span><span class="p">]],</span> <span class="n">length</span><span class="p">(</span><span class="n">y_res</span><span class="p">[[</span><span class="n">i</span><span class="p">]]),</span> <span class="mi">1</span><span class="p">))</span>
                <span class="n">y_res</span><span class="p">[[</span><span class="n">i</span><span class="p">]]</span> <span class="o">=</span> <span class="n">cbind</span><span class="p">(</span><span class="n">row_str</span><span class="p">,</span> <span class="n">y_res</span><span class="p">[[</span><span class="n">i</span><span class="p">]])</span>
                <span class="n">colnames</span><span class="p">(</span><span class="n">y_res</span><span class="p">[[</span><span class="n">i</span><span class="p">]])</span> <span class="o">=</span> <span class="n">c</span><span class="p">(</span><span class="s1">&#39;sample.id&#39;</span><span class="p">,</span> <span class="n">col_str</span><span class="p">)</span>
            <span class="p">}</span>
            <span class="c1"># merge residuals into one matrix</span>
            <span class="c1"># need to ensure same rows </span>
            <span class="c1"># are matched and unmatched entries are filled with NA</span>
            <span class="n">y_res_mat</span> <span class="o">=</span> <span class="n">plyr</span><span class="p">::</span><span class="n">join_all</span><span class="p">(</span><span class="n">y_res</span><span class="p">,</span> <span class="n">by</span> <span class="o">=</span> <span class="s2">&quot;sample.id&quot;</span><span class="p">,</span> <span class="nb">type</span> <span class="o">=</span> <span class="s2">&quot;full&quot;</span><span class="p">,</span> <span class="n">match</span> <span class="o">=</span> <span class="s2">&quot;all&quot;</span><span class="p">)</span>
            <span class="n">sample</span><span class="o">.</span><span class="n">names</span> <span class="o">=</span> <span class="n">y_res_mat</span><span class="p">[,</span><span class="mi">1</span><span class="p">]</span>
            <span class="n">y_res_mat</span> <span class="o">=</span> <span class="k">as</span><span class="o">.</span><span class="n">matrix</span><span class="p">(</span><span class="n">y_res_mat</span><span class="p">[,</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span>
            <span class="n">rownames</span><span class="p">(</span><span class="n">y_res_mat</span><span class="p">)</span> <span class="o">=</span> <span class="n">sample</span><span class="o">.</span><span class="n">names</span>
            <span class="c1"># and match its order with genotype data</span>
            <span class="c1">#gdata = gdata[rownames(y_res_mat),]</span>
            <span class="k">return</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="n">X</span> <span class="o">=</span> <span class="k">as</span><span class="o">.</span><span class="n">matrix</span><span class="p">(</span><span class="n">gdata</span><span class="p">),</span> <span class="n">y</span> <span class="o">=</span> <span class="n">edata</span><span class="p">,</span> <span class="n">Z</span> <span class="o">=</span> <span class="n">cdata</span><span class="p">,</span> <span class="n">y_res</span> <span class="o">=</span> <span class="n">y_res_mat</span><span class="p">,</span>
                        <span class="n">chrom</span> <span class="o">=</span> <span class="n">strsplit</span><span class="p">(</span><span class="n">colnames</span><span class="p">(</span><span class="n">gdata</span><span class="p">)[</span><span class="mi">1</span><span class="p">],</span> <span class="s1">&#39;_&#39;</span><span class="p">)[[</span><span class="mi">1</span><span class="p">]][</span><span class="mi">1</span><span class="p">],</span> 
                        <span class="n">pos</span> <span class="o">=</span> <span class="k">as</span><span class="o">.</span><span class="n">integer</span><span class="p">(</span><span class="n">do</span><span class="o">.</span><span class="n">call</span><span class="p">(</span><span class="n">rbind</span><span class="p">,</span> <span class="n">strsplit</span><span class="p">(</span><span class="n">colnames</span><span class="p">(</span><span class="n">gdata</span><span class="p">),</span> <span class="s1">&#39;_&#39;</span><span class="p">))[,</span><span class="mi">2</span><span class="p">])))</span>
    <span class="p">}</span>
    <span class="c1">#</span>
    <span class="n">geno_table</span> <span class="o">=</span> <span class="s2">&quot;${_genes}&quot;</span>
    <span class="n">expr_table</span> <span class="o">=</span> <span class="n">cov_table</span> <span class="o">=</span> <span class="err">$</span><span class="p">{</span><span class="n">tissue_tables</span><span class="p">}</span>
    <span class="k">if</span> <span class="p">(</span><span class="n">length</span><span class="p">(</span><span class="n">expr_table</span><span class="p">)</span> <span class="o">==</span> <span class="mi">1</span><span class="p">)</span> <span class="p">{</span>
        <span class="n">dat</span> <span class="o">=</span> <span class="n">load_data_single</span><span class="p">(</span><span class="s2">&quot;${genotype_file}&quot;</span><span class="p">,</span> <span class="s2">&quot;${expr_file}&quot;</span><span class="p">,</span> <span class="s2">&quot;${cov_file}&quot;</span><span class="p">,</span> <span class="n">geno_table</span><span class="p">,</span> <span class="n">expr_table</span><span class="p">,</span> <span class="n">cov_table</span><span class="p">)</span>
    <span class="p">}</span> <span class="k">else</span> <span class="p">{</span>
        <span class="n">dat</span> <span class="o">=</span> <span class="n">load_data_multi</span><span class="p">(</span><span class="s2">&quot;${genotype_file}&quot;</span><span class="p">,</span> <span class="s2">&quot;${expr_file}&quot;</span><span class="p">,</span> <span class="s2">&quot;${cov_file}&quot;</span><span class="p">,</span> <span class="n">geno_table</span><span class="p">,</span> <span class="n">expr_table</span><span class="p">,</span> <span class="n">cov_table</span><span class="p">)</span>        
    <span class="p">}</span>
    <span class="n">saveRDS</span><span class="p">(</span><span class="n">dat</span><span class="p">,</span> <span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">:</span><span class="n">r</span><span class="p">})</span>
</pre></div>

</div>
</div>
</div>

</div>
<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div>
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p>For (one) example,</p>

</div>
</div>
</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[&nbsp;]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-sos"><pre><span></span><span class="n">sos</span> <span class="n">run</span> <span class="n">analysis</span><span class="o">/</span><span class="mi">20180515</span><span class="n">_Extract_Benchmark_Data</span><span class="o">.</span><span class="n">ipynb</span> <span class="n">extract</span> <span class="o">--</span><span class="n">genes</span> <span class="o">/</span><span class="n">chr4</span><span class="o">/</span><span class="n">ENSG00000145214</span>
</pre></div>

</div>
</div>
</div>

</div>
<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div>
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p>To get an idea how many variants are there:</p>

</div>
</div>
</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[&nbsp;]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-sos"><pre><span></span><span class="p">[</span><span class="n">count_var</span><span class="p">]</span>
<span class="kn">input:</span> <span class="n">glob</span><span class="o">.</span><span class="n">glob</span><span class="p">(</span><span class="s1">&#39;~/Documents/GTExV8/Toys/*.RDS&#39;</span><span class="p">),</span> <span class="n">group_by</span> <span class="o">=</span> <span class="mi">1</span>
<span class="kn">output:</span> <span class="s1">&#39;~/Documents/GTExV8/Toys/counts.txt&#39;</span>
<span class="kn">R:</span> <span class="n">expand</span> <span class="o">=</span> <span class="bp">True</span><span class="p">,</span> <span class="n">stdout</span> <span class="o">=</span> <span class="n">_output</span>
  <span class="n">cat</span><span class="p">(</span><span class="n">paste0</span><span class="p">({</span><span class="n">_input</span><span class="p">:</span><span class="n">br</span><span class="p">},</span> <span class="n">ncol</span><span class="p">(</span><span class="n">readRDS</span><span class="p">({</span><span class="n">_input</span><span class="p">:</span><span class="n">r</span><span class="p">})</span><span class="err">$</span><span class="n">X</span><span class="p">),</span><span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">))</span>
</pre></div>

</div>
</div>
</div>

</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[1]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-sos"><pre><span></span><span class="o">%</span><span class="n">cd</span> <span class="o">~/</span><span class="n">Documents</span><span class="o">/</span><span class="n">GTExV8</span><span class="o">/</span><span class="n">Toys</span>
</pre></div>

</div>
</div>
</div>

<div class="output_wrapper">
<div class="output">


<div class="output_area">

<div class="prompt"></div>


<div class="output_subarea output_stream output_stdout output_text">
<pre>/home/gaow/Documents/GTExV8/Toys</pre>
</div>
</div>

</div>
</div>

</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[2]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-sos"><pre><span></span><span class="n">dat</span> <span class="o">=</span> <span class="n">scan</span><span class="p">(</span><span class="s1">&#39;counts.txt&#39;</span><span class="p">)</span>
<span class="n">summary</span><span class="p">(</span><span class="n">dat</span><span class="p">)</span>
</pre></div>

</div>
</div>
</div>

<div class="output_wrapper">
<div class="output">


<div class="output_area">

<div class="prompt"></div>




<div class="output_text output_subarea ">
<pre>   Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
   3022    6450    7226    7217    8033   11999 </pre>
</div>

</div>

</div>
</div>

</div>
<hr>
Copyright &copy 2016-2018 Gao Wang et al at Stephens Lab, University of Chicago
<p><small>Exported from <a href="http://github.com/gaow/mvarbvs/blob/1bf5d8546ea3ad3f3f90e92eb7d06209f32edcf9/analysis/20180515_Extract_Benchmark_Data.ipynb"><code>analysis/20180515_Extract_Benchmark_Data.ipynb</code></a> committed by Gao Wang on Mon May 13 08:57:38 2019 <a href="http://github.com/gaow/mvarbvs/commit/1bf5d8546ea3ad3f3f90e92eb7d06209f32edcf9">revision 7, 1bf5d85</a> <a href="https://stephenslab.github.io/ipynb-website/notes.html#Note-about-commit-ids"><span class="fa fa-question-circle"></span></a></small></p>
</div>
</div>
</body>
</html>
