
<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="ipynb_website:version" content="0.9.4" />
<meta name="viewport" content="width=device-width, initial-scale=1" />

<link rel="stylesheet" type="text/css" href="../css/jt.css">

<link rel="stylesheet" type="text/css" href="../css/toc2.css">

<link href="../site_libs/jqueryui-1.11.4/jquery-ui.css">
<link rel="stylesheet" href="../site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<link rel="stylesheet" href="../site_libs/font-awesome-4.5.0/css/font-awesome.min.css" rel="stylesheet" />
<script type="text/javascript" src="https://ajax.googleapis.com/ajax/libs/jquery/1.9.1/jquery.min.js"></script>
<script type="text/javascript" src="https://ajax.googleapis.com/ajax/libs/jqueryui/1.9.1/jquery-ui.min.js"></script>
<script src="../site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="../site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="../site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>

<link rel="stylesheet"
      href="../site_libs/highlightjs/null.min.css"
      type="text/css" />

<script src="../site_libs/highlightjs/highlight.pack.js"></script>
<script>hljs.initHighlightingOnLoad();</script>
<script type="text/javascript">
if (window.hljs && document.readyState && document.readyState === "complete") {
   window.setTimeout(function() {
      hljs.initHighlighting();
   }, 0);
}
</script>

<script src="../js/doc_toc.js"></script>
<script src="../js/docs.js"></script>

<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.2/MathJax.js?config=TeX-MML-AM_CHTML"></script>
<script>
    MathJax.Hub.Config({
        extensions: ["tex2jax.js"],
        jax: ["input/TeX", "output/HTML-CSS"],
        tex2jax: {
        inlineMath: [ ['$','$'], ["\\(","\\)"] ],
        displayMath: [ ['$$','$$'], ["\\[","\\]"] ],
        processEscapes: true
        },
        "HTML-CSS": {
            preferredFont: "TeX",
            availableFonts: ["TeX"],
            styles: {
                scale: 110,
                ".MathJax_Display": {
                    "font-size": "110%",
                }
            }
        }
    });
</script>
<script>
function filterDataFrame(id) {
    var input = document.getElementById("search_" + id);
    var filter = input.value.toUpperCase();
    var table = document.getElementById("dataframe_" + id);
    var tr = table.getElementsByTagName("tr");
    // Loop through all table rows, and hide those who don't match the search query
    for (var i = 1; i < tr.length; i++) {
        for (var j = 0; j < tr[i].cells.length; ++j) {
            var matched = false;
            if (tr[i].cells[j].innerHTML.toUpperCase().indexOf(filter) != -1) {
                tr[i].style.display = "";
                matched = true
                break;
            }
            if (!matched)
                tr[i].style.display = "none";
        }
    }
}
function sortDataFrame(id, n, dtype) {
    var table = document.getElementById("dataframe_" + id);
    var tb = table.tBodies[0]; // use `<tbody>` to ignore `<thead>` and `<tfoot>` rows
    var tr = Array.prototype.slice.call(tb.rows, 0); // put rows into array
    if (dtype === 'numeric') {
        var fn = function(a, b) { 
            return parseFloat(a.cells[n].textContent) <= parseFloat(b.cells[n].textContent) ? -1 : 1;
        }
    } else {
        var fn = function(a, b) {
            var c = a.cells[n].textContent.trim().localeCompare(b.cells[n].textContent.trim()); 
            return c > 0 ? 1 : (c < 0 ? -1 : 0) }
    }
    var isSorted = function(array, fn) {
        if (array.length < 2)
            return 1;
        var direction = fn(array[0], array[1]); 
        for (var i = 1; i < array.length - 1; ++i) {
            var d = fn(array[i], array[i+1]);
            if (d == 0)
                continue;
            else if (direction == 0)
                direction = d;
            else if (direction != d)
                return 0;
            }
        return direction;
    }
    var sorted = isSorted(tr, fn);
    if (sorted == 1 || sorted == -1) {
        // if sorted already, reverse it
        for(var i = tr.length - 1; i >= 0; --i)
            tb.appendChild(tr[i]); // append each row in order
    } else {
        tr = tr.sort(fn);
        for(var i = 0; i < tr.length; ++i)
            tb.appendChild(tr[i]); // append each row in order
    }
}
</script>

<script>
$( document ).ready(function(){
            var cfg={'threshold':3,     // depth of toc (number of levels)
             'number_sections': false,
             'toc_cell': false,          // useless here
             'toc_window_display': true, // display the toc window
             "toc_section_display": "block", // display toc contents in the window
             'sideBar':true,       // sidebar or floating window
             'navigate_menu':false       // navigation menu (only in liveNotebook -- do not change)
            }
            var st={};                  // some variables used in the script
            st.rendering_toc_cell = false;
            st.config_loaded = false;
            st.extension_initialized=false;
            st.nbcontainer_marginleft = $('#notebook-container').css('margin-left')
            st.nbcontainer_marginright = $('#notebook-container').css('margin-right')
            st.nbcontainer_width = $('#notebook-container').css('width')
            st.oldTocHeight = undefined
            st.cell_toc = undefined;
            st.toc_index=0;
            // fire the main function with these parameters
            table_of_contents(cfg, st);
            var file=analysisDict[$("h1:first").attr("id")];
            $("#toc-level0 a").css("color","#126dce");
            $('a[href="#'+$("h1:first").attr("id")+'"]').hide()
            var docs=analysisArray;
            var docs_map=analysisArrayMap;
            var pos=analysisArray.indexOf(file);
            for (var a=pos;a>=0;a--){
                  $('<li><a href="'+docs[a]+'.html"><font color="#073642"><b>'+docs_map[docs[a]].replace(/_/g," ")+'</b></font></a></li>').insertBefore("#toc-level0 li:eq(0)");
            }
            $('a[href="'+file+'.html'+'"]').css("color","#126dce");
            for (var a=pos+1;a<docs.length;a++){
                  $(".toc #toc-level0").append('<li><a href="'+docs[a]+'.html"><font color="#073642"><b>'+docs_map[docs[a]].replace(/_/g," ")+'</b></font></a></li>');
            }
            // $("#toc-header").hide(); // comment out because it prevents search bar from displaying
    });
</script>

<script>
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');
  // mark it active
  menuAnchor.parent().addClass('active');
  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');
});
</script>
<div class="container-fluid main-container">
<!-- tabsets -->
<script src="../site_libs/navigation-1.1/tabsets.js"></script>
<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});
</script>



<title>m&m ash</title>

<style type = "text/css">
body {
  font-family: "Droid Sans";
  padding-top: 66px;
  padding-bottom: 40px;
}
</style>
</head>

<body>
<div tabindex="-1" id="notebook" class="border-box-sizing">
<div class="container" id="notebook-container">

<!-- code folding -->

<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="../index.html">m&m ash</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        
<li>
  <a href="../index.html">Overview</a>
</li>
        
<li>
  <a href="../details.html">Details</a>
</li>
        
<li>
  <a href="../analysis.html">Analysis</a>
</li>
        
<li>
  <a href="../prototype.html">Prototype</a>
</li>
        
<li>
  <a href="../dsc.html">Dsc</a>
</li>
        
<li>
  <a href="../writeup.html">Writeup</a>
</li>
        
      </ul>
        
<ul class="nav navbar-nav navbar-right">
<li>
   <a href="http://github.com/gaow/mvarbvs"> <span class="fa fa-github"></span> </a>
</li>
</ul>
        
      </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->
<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div>
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h1 id="Enrichment-analysis-workflow-for-molecular-QTL-results">Enrichment analysis workflow for molecular QTL results<a class="anchor-link" href="#Enrichment-analysis-workflow-for-molecular-QTL-results">&#182;</a></h1>
</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div>
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p>For molecular QTL analysis results we obtained we'd like to see if:</p>
<ol>
<li>PIP are higher on average in certain annotation groups than in the rest of genome</li>
<li>Is there an enrichment for variables both in CS and in some annotation groups<ul>
<li>Specifically, whether or not there is an enrichment in the secondary CS that we capture</li>
</ul>
</li>
</ol>
<p>We focus only on the results that has 1 or more CS identified.</p>
<h2 id="Annotations-used-for-splice-QTL">Annotations used for splice QTL<a class="anchor-link" href="#Annotations-used-for-splice-QTL">&#182;</a></h2><p>We use the following types of annotations:</p>
<ol>
<li>Histone marks for LCLs from Epigenome Roadmap (download link see pipeline below)</li>
<li><a href="http://www.ebi.ac.uk/birney-srv/CTCF-QTL/">CTCF binding sites</a></li>
<li>POL II ChIP-seq (download link see pipeline below)</li>
<li>Intron and exon regions from UCSC</li>
<li>Extended splice site, defined as 5bp up/down intron start site and 15bp up/down intron end site (based on UCSC intron data)</li>
</ol>
<p>I will first test for enrichment with the primary CS (defined as the set with highest purity), then test for the non-primary CS. The idea is to discover enrichment of annotations in the primary CS, then also show that additional CS has similar (or interestingly different!) pattern of enrichment.</p>

</div>
</div>
</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[10]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-sos"><pre><span></span><span class="o">%</span><span class="n">revisions</span> <span class="o">-</span><span class="n">s</span> <span class="o">-</span><span class="n">n</span> <span class="mi">10</span>
</pre></div>

</div>
</div>
</div>

<div class="output_wrapper">
<div class="output">


<div class="output_area">

<div class="prompt"></div>



<div class="output_html rendered_html output_subarea ">

        <table class="revision_table">
        <tr>
        <th>Revision</th>
        <th>Author</th>
        <th>Date</th>
        <th>Message</th>
        <tr>
        <tr><td><a target="_blank" href="https://github.com/gaow/mvarbvs/blob/7f148480375c188a1af4353a6fe227c985571ac7/analysis/20180712_Enrichment_Workflow.ipynb"><span class="revision_id">7f14848<span></a></td>
<td>Gao Wang</td>
<td>2018-07-24</td>
<td>Add GREGOR for matched enrichment analysis</td></tr><tr><td><a target="_blank" href="https://github.com/gaow/mvarbvs/blob/24b7bf304421316d1c70d0cab18fe420ea96e9ed/analysis/20180712_Enrichment_Workflow.ipynb"><span class="revision_id">24b7bf3<span></a></td>
<td>Gao Wang</td>
<td>2018-07-13</td>
<td>Update documentation</td></tr><tr><td><a target="_blank" href="https://github.com/gaow/mvarbvs/blob/9531187446835c456a7a6fca586c890d8203aad4/analysis/20180712_Enrichment_Workflow.ipynb"><span class="revision_id">9531187<span></a></td>
<td>Gao Wang</td>
<td>2018-07-12</td>
<td>Add CS based enrichment</td></tr><tr><td><a target="_blank" href="https://github.com/gaow/mvarbvs/blob/e7227739946a66d644df8999a32690f93106caba/analysis/20180712_Enrichment_Workflow.ipynb"><span class="revision_id">e722773<span></a></td>
<td>Gao Wang</td>
<td>2018-07-12</td>
<td>Update enrichment results</td></tr><tr><td><a target="_blank" href="https://github.com/gaow/mvarbvs/blob/305751a80925bf3b47f0af237b73856a6b0a25e2/analysis/20180712_Enrichment_Workflow.ipynb"><span class="revision_id">305751a<span></a></td>
<td>Gao Wang</td>
<td>2018-07-12</td>
<td>Add enrichment results</td></tr></table>
</div>

</div>

</div>
</div>

</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[2]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-sos"><pre><span></span><span class="err">!</span> <span class="n">sos</span> <span class="n">run</span> <span class="mi">20180712</span><span class="n">_Enrichment_Workflow</span><span class="o">.</span><span class="n">ipynb</span> <span class="o">-</span><span class="n">h</span>
</pre></div>

</div>
</div>
</div>

<div class="output_wrapper">
<div class="output">


<div class="output_area">

<div class="prompt"></div>


<div class="output_subarea output_stream output_stdout output_text">
<pre>usage: sos run 20180712_Enrichment_Workflow.ipynb
               [workflow_name | -t targets] [options] [workflow_options]
  workflow_name:        Single or combined workflows defined in this script
  targets:              One or more targets to generate
  options:              Single-hyphen sos parameters (see &#34;sos run -h&#34; for details)
  workflow_options:     Double-hyphen workflow-specific parameters

Workflows:
  extract_sumstats
  zscore2bed
  get_variants
  range2var_annotation
  pip_rank_test
  cs_fisher_test

Global Workflow Options:
  --y-data . (as path)
                        Y data, the phenotype file paths
  --trait VAL (required)
                        Trait name
  --cwd  path(f&#39;{y_data:d}/{trait}_output&#39;)

                        Specify work / output directory
  --annotation-dir /home/gaow/Documents/GIT/LargeFiles/he_lab_annotations_bed (as path)
                        Path to directory of annotation
                        files
  --single-annot . (as path)
                        Path to list of single annotations
                        to use
  --max-dist 100000 (as int)
                        Maximum distance to site of
                        interest, set to eg. 100Kb or 1Mb
                        up/downstream to start site of
                        analysis unit

Sections
  extract_sumstats_1:   Extract summary stats from RDS files
                        to plain text
  extract_sumstats_2:   Consolidate results to one file
  zscore2bed_1:         Auxiliary step to get variant in bed
                        format based on variant ID in
                        z-score file
    Workflow Options:
      --in-file . (as path)
      --chr-prefix &#39;&#39;
  zscore2bed_2:
  get_variants:
  range2var_annotation_1: Get variants in data that falls in
                        target region
  range2var_annotation_2: Make binary annotation file
  pip_rank_test_1:      Test if PIP is larger in annotations
  pip_rank_test_2, cs_fisher_test_2: Consolidate results to
                        one file
  cs_fisher_test_1:     Test if CS is enriched with
                        annotations
</pre>
</div>
</div>

</div>
</div>

</div>
<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div>
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h2 id="Annotation-input">Annotation input<a class="anchor-link" href="#Annotation-input">&#182;</a></h2><p>Annotation files are in bed format (for example: <code>Coding_UCSC.bed</code>).</p>

<pre><code>chr1    69090   70008
chr1    367658  368597
chr1    621095  622034
......
chr9    141121357       141121553
chr9    141124188       141124276
chr9    141134069       141134172</code></pre>
<p>There are many annotations one can use. For this workflow one should prepare a list of annotations for input, like this:</p>

<pre><code># my annotation
Coding_UCSC
CTCF_binding
E116-DNase.macs2
E116-H2A.Z
E116-H3K27ac
E116-H3K27me3
E116-H3K36me3
E116-H3K4me1
E116-H3K4me2
E116-H3K4me3
E116-H3K79me2
E116-H3K9ac
E116-H3K9me3
E116-H4K20me1
E117-DNase.macs2
Intron_UCSC
PolII_binding</code></pre>
<p>comment symbol <code>#</code> is allowed.</p>

</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div>
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h2 id="Data-input">Data input<a class="anchor-link" href="#Data-input">&#182;</a></h2>
</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div>
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p>Need to format the data to:</p>

</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div>
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">

<pre><code>Variant_ID PIP z_score CS_ID</code></pre>
<p>Where <code>CS_ID</code> is 0 if variant is not in SuSiE CS, 1 if in CS 1, 2 in CS 2, etc. <code>Variant_ID</code> carries information of chrom and pos. eg, <code>rs10131831_chr14_20905250_G_A</code></p>

</div>
</div>
</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[&nbsp;]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-sos"><pre><span></span><span class="p">[</span><span class="k">global</span><span class="p">]</span>
<span class="c1"># Y data, the phenotype file paths</span>
<span class="kn">parameter:</span> <span class="n">y_data</span> <span class="o">=</span> <span class="n">path</span><span class="p">()</span>
<span class="c1"># Trait name</span>
<span class="kn">parameter:</span> <span class="n">trait</span> <span class="o">=</span> <span class="bp">None</span>
<span class="c1"># Specify work / output directory</span>
<span class="kn">parameter:</span> <span class="n">cwd</span> <span class="o">=</span> <span class="n">path</span><span class="p">(</span><span class="n">f</span><span class="s1">&#39;{y_data:d}/{trait}_output&#39;</span><span class="p">)</span>
<span class="c1"># Path to directory of annotation files</span>
<span class="kn">parameter:</span> <span class="n">annotation_dir</span> <span class="o">=</span> <span class="n">path</span><span class="p">(</span><span class="s1">&#39;~/Documents/GIT/LargeFiles/lcl_annotation&#39;</span><span class="p">)</span>
<span class="c1"># Path to list of single annotations to use</span>
<span class="kn">parameter:</span> <span class="n">single_annot</span> <span class="o">=</span> <span class="n">path</span><span class="p">()</span> <span class="c1">#parameter: single_annot = path(&quot;data/all_annotations.txt&quot;)</span>
<span class="c1"># Maximum distance to site of interest, set to eg. 100Kb or 1Mb up/downstream to start site of analysis unit</span>
<span class="kn">parameter:</span> <span class="n">max_dist</span> <span class="o">=</span> <span class="mi">100000</span>
<span class="n">fail_if</span><span class="p">(</span><span class="ow">not</span> <span class="n">y_data</span><span class="o">.</span><span class="n">is_file</span><span class="p">(),</span> <span class="n">msg</span> <span class="o">=</span> <span class="s1">&#39;Please provide valid ``--y-data``!&#39;</span><span class="p">)</span>
<span class="kn">parameter:</span> <span class="n">z_score</span> <span class="o">=</span> <span class="n">path</span><span class="p">(</span><span class="n">f</span><span class="s1">&#39;{cwd}/{y_data:bnn}_{int(max_dist/1000)}Kb/enrichment/SuSiE_loci.sumstats.gz&#39;</span><span class="p">)</span>
<span class="n">out_dir</span> <span class="o">=</span> <span class="n">f</span><span class="s1">&#39;{cwd}/{y_data:bnn}_{int(max_dist/1000)}Kb/enrichment/{z_score:bn}&#39;</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s1">&#39;.&#39;</span><span class="p">,</span> <span class="s1">&#39;_&#39;</span><span class="p">)</span>
<span class="kn">try:</span>
    <span class="n">single_anno</span> <span class="o">=</span> <span class="p">[</span><span class="n">f</span><span class="s2">&quot;{annotation_dir}/{x.split()[0]}.bed&quot;</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="nb">open</span><span class="p">(</span><span class="n">single_annot</span><span class="p">)</span><span class="o">.</span><span class="n">readlines</span><span class="p">()</span> <span class="k">if</span> <span class="ow">not</span> <span class="n">x</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s1">&#39;#&#39;</span><span class="p">)]</span>
<span class="k">except</span> <span class="p">(</span><span class="n">FileNotFoundError</span><span class="p">,</span> <span class="n">IsADirectoryError</span><span class="p">):</span>
    <span class="n">single_anno</span> <span class="o">=</span> <span class="p">[]</span>
</pre></div>

</div>
</div>
</div>

</div>
<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div>
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h2 id="Prepare-annotations">Prepare annotations<a class="anchor-link" href="#Prepare-annotations">&#182;</a></h2><p>Histone marks</p>

<pre><code>sos run analysis/20180712_Enrichment_Workflow.ipynb download_histone_annotation --trait AS \
    --y-data ~/Documents/GIT/LargeFiles/JointLCL/fastqtl_qqnorm_ASintron_RNAseqGeuvadis_YangVCF.txt.gz</code></pre>

</div>
</div>
</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[&nbsp;]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-sos"><pre><span></span><span class="p">[</span><span class="n">download_histone_annotation</span><span class="p">]</span>
<span class="n">dataset</span> <span class="o">=</span> <span class="n">paths</span><span class="p">([</span><span class="s2">&quot;E116-DNase.macs2.narrowPeak.gz&quot;</span><span class="p">,</span>
<span class="s2">&quot;E116-H2A.Z.narrowPeak.gz&quot;</span><span class="p">,</span>
<span class="s2">&quot;E116-H3K4me1.narrowPeak.gz&quot;</span><span class="p">,</span>
<span class="s2">&quot;E116-H3K4me2.narrowPeak.gz&quot;</span><span class="p">,</span>
<span class="s2">&quot;E116-H3K4me3.narrowPeak.gz&quot;</span><span class="p">,</span>
<span class="s2">&quot;E116-H3K9ac.narrowPeak.gz&quot;</span><span class="p">,</span>
<span class="s2">&quot;E116-H3K9me3.narrowPeak.gz&quot;</span><span class="p">,</span>
<span class="s2">&quot;E116-H3K27ac.narrowPeak.gz&quot;</span><span class="p">,</span>
<span class="s2">&quot;E116-H3K27me3.narrowPeak.gz&quot;</span><span class="p">,</span>
<span class="s2">&quot;E116-H3K36me3.narrowPeak.gz&quot;</span><span class="p">,</span>
<span class="s2">&quot;E116-H3K79me2.narrowPeak.gz&quot;</span><span class="p">,</span>
<span class="s2">&quot;E116-H4K20me1.narrowPeak.gz&quot;</span><span class="p">])</span>
<span class="kn">input:</span> <span class="kp">for_each</span> <span class="o">=</span> <span class="s1">&#39;dataset&#39;</span><span class="p">,</span> <span class="kp">concurrent</span> <span class="o">=</span> <span class="bp">True</span>
<span class="kn">output:</span> <span class="n">f</span><span class="s1">&#39;{annotation_dir}/{_dataset:nn}.bed&#39;</span>
<span class="kn">download:</span> <span class="n">dest_dir</span> <span class="o">=</span> <span class="n">annotation_dir</span><span class="p">,</span> <span class="n">expand</span> <span class="o">=</span> <span class="bp">True</span>
    <span class="n">https</span><span class="p">:</span><span class="o">//</span><span class="n">egg2</span><span class="o">.</span><span class="n">wustl</span><span class="o">.</span><span class="n">edu</span><span class="o">/</span><span class="n">roadmap</span><span class="o">/</span><span class="n">data</span><span class="o">/</span><span class="n">byFileType</span><span class="o">/</span><span class="n">peaks</span><span class="o">/</span><span class="n">consolidated</span><span class="o">/</span><span class="n">narrowPeak</span><span class="o">/</span><span class="p">{</span><span class="n">_dataset</span><span class="p">}</span>
<span class="kn">bash:</span> <span class="n">expand</span> <span class="o">=</span> <span class="bp">True</span><span class="p">,</span> <span class="kp">workdir</span> <span class="o">=</span> <span class="n">annotation_dir</span><span class="p">,</span> <span class="n">container</span> <span class="o">=</span> <span class="s1">&#39;gaow/atac-gwas&#39;</span>
    <span class="n">zcat</span> <span class="p">{</span><span class="n">_dataset</span><span class="p">}</span> <span class="o">|</span> <span class="n">cut</span> <span class="o">-</span><span class="n">f1</span><span class="o">-</span><span class="mi">3</span> <span class="o">|</span> <span class="n">grep</span> <span class="o">-</span><span class="n">v</span> <span class="n">chrM</span> <span class="o">|</span> <span class="n">sort</span><span class="o">-</span><span class="n">bed</span> <span class="o">-</span> <span class="o">&gt;</span> <span class="p">{</span><span class="n">_output</span><span class="p">}</span>
</pre></div>

</div>
</div>
</div>

</div>
<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div>
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p>CTCF binding:</p>

<pre><code>sos run analysis/20180712_Enrichment_Workflow.ipynb download_ctcf_annotation --trait AS \
    --y-data ~/Documents/GIT/LargeFiles/JointLCL/fastqtl_qqnorm_ASintron_RNAseqGeuvadis_YangVCF.txt.gz</code></pre>

</div>
</div>
</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[&nbsp;]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-sos"><pre><span></span><span class="p">[</span><span class="n">download_ctcf_annotation</span><span class="p">]</span>
<span class="kn">output:</span> <span class="n">f</span><span class="s1">&#39;{annotation_dir}/CTCF_binding.bed&#39;</span>
<span class="kn">download:</span> <span class="n">dest_dir</span> <span class="o">=</span> <span class="n">annotation_dir</span>
    <span class="n">https</span><span class="p">:</span><span class="o">//</span><span class="n">www</span><span class="o">.</span><span class="n">ebi</span><span class="o">.</span><span class="n">ac</span><span class="o">.</span><span class="n">uk</span><span class="o">/</span><span class="n">birney</span><span class="o">-</span><span class="n">srv</span><span class="o">/</span><span class="n">CTCF</span><span class="o">-</span><span class="n">QTL</span><span class="o">/</span><span class="n">phenotypes</span><span class="o">/</span><span class="n">bindings</span><span class="o">.</span><span class="n">csv</span>
<span class="kn">bash:</span> <span class="n">expand</span> <span class="o">=</span> <span class="bp">True</span><span class="p">,</span> <span class="kp">workdir</span> <span class="o">=</span> <span class="n">annotation_dir</span><span class="p">,</span> <span class="n">container</span> <span class="o">=</span> <span class="s1">&#39;gaow/atac-gwas&#39;</span>
    <span class="n">cut</span> <span class="o">-</span><span class="n">f1</span><span class="o">-</span><span class="mi">3</span> <span class="o">-</span><span class="n">d</span> <span class="s2">&quot;,&quot;</span> <span class="n">bindings</span><span class="o">.</span><span class="n">csv</span> <span class="o">|</span> <span class="n">tail</span> <span class="o">-</span><span class="n">n</span><span class="o">+</span><span class="mi">2</span> <span class="o">|</span> <span class="n">sed</span> <span class="s1">&#39;s/,/</span><span class="se">\t</span><span class="s1">/g&#39;</span> <span class="o">|</span> <span class="n">sed</span> <span class="s1">&#39;s/^/chr/g&#39;</span> <span class="o">|</span> <span class="n">grep</span> <span class="o">-</span><span class="n">v</span> <span class="n">chrM</span> <span class="o">|</span> <span class="n">sort</span><span class="o">-</span><span class="n">bed</span> <span class="o">-</span> <span class="o">&gt;</span> <span class="p">{</span><span class="n">_output</span><span class="p">}</span>
</pre></div>

</div>
</div>
</div>

</div>
<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div>
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p>PolyII binding:</p>

<pre><code>sos run analysis/20180712_Enrichment_Workflow.ipynb download_polII_annotation --trait AS \
    --y-data ~/Documents/GIT/LargeFiles/JointLCL/fastqtl_qqnorm_ASintron_RNAseqGeuvadis_YangVCF.txt.gz</code></pre>
<p>Need to <code>liftOver</code>:</p>

<pre><code>conda config --add channels bioconda
conda install ucsc-liftover</code></pre>

</div>
</div>
</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[&nbsp;]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-sos"><pre><span></span><span class="p">[</span><span class="n">download_polII_annotation</span><span class="p">]</span>
<span class="n">dataset</span> <span class="o">=</span> <span class="s2">&quot;GSM487431_GM12878_Pol2_narrowPeak.bed.gz&quot;</span>
<span class="kn">output:</span> <span class="n">f</span><span class="s1">&#39;{annotation_dir}/PolII_binding.bed&#39;</span>
<span class="kn">download:</span> <span class="n">dest_dir</span> <span class="o">=</span> <span class="n">annotation_dir</span><span class="p">,</span> <span class="n">expand</span><span class="o">=</span><span class="bp">True</span>
    <span class="n">ftp</span><span class="p">:</span><span class="o">//</span><span class="n">ftp</span><span class="o">.</span><span class="n">ncbi</span><span class="o">.</span><span class="n">nlm</span><span class="o">.</span><span class="n">nih</span><span class="o">.</span><span class="n">gov</span><span class="o">/</span><span class="n">geo</span><span class="o">/</span><span class="n">samples</span><span class="o">/</span><span class="n">GSM487nnn</span><span class="o">/</span><span class="n">GSM487431</span><span class="o">/</span><span class="n">suppl</span><span class="o">/</span><span class="p">{</span><span class="n">dataset</span><span class="p">}</span>
<span class="kn">download:</span> <span class="n">dest_dir</span> <span class="o">=</span> <span class="n">annotation_dir</span>
    <span class="n">http</span><span class="p">:</span><span class="o">//</span><span class="n">hgdownload</span><span class="o">.</span><span class="n">cse</span><span class="o">.</span><span class="n">ucsc</span><span class="o">.</span><span class="n">edu</span><span class="o">/</span><span class="n">goldenpath</span><span class="o">/</span><span class="n">hg18</span><span class="o">/</span><span class="n">liftOver</span><span class="o">/</span><span class="n">hg18ToHg19</span><span class="o">.</span><span class="n">over</span><span class="o">.</span><span class="n">chain</span><span class="o">.</span><span class="n">gz</span>
<span class="kn">bash:</span> <span class="n">expand</span> <span class="o">=</span> <span class="bp">True</span><span class="p">,</span> <span class="kp">workdir</span> <span class="o">=</span> <span class="n">annotation_dir</span><span class="p">,</span> <span class="n">container</span> <span class="o">=</span> <span class="s1">&#39;gaow/atac-gwas&#39;</span>
    <span class="n">zcat</span> <span class="p">{</span><span class="n">dataset</span><span class="p">}</span> <span class="o">|</span> <span class="n">cut</span> <span class="o">-</span><span class="n">f1</span><span class="o">-</span><span class="mi">3</span>  <span class="o">|</span> <span class="n">grep</span> <span class="o">-</span><span class="n">v</span> <span class="n">chrM</span> <span class="o">|</span> <span class="n">sort</span><span class="o">-</span><span class="n">bed</span> <span class="o">-</span> <span class="o">&gt;</span> <span class="p">{</span><span class="n">_output</span><span class="p">:</span><span class="n">n</span><span class="p">}</span><span class="o">.</span><span class="n">hg18</span><span class="o">.</span><span class="n">bed</span>
<span class="kn">bash:</span> <span class="n">expand</span> <span class="o">=</span> <span class="bp">True</span><span class="p">,</span> <span class="kp">workdir</span> <span class="o">=</span> <span class="n">annotation_dir</span>
    <span class="n">liftOver</span> <span class="p">{</span><span class="n">_output</span><span class="p">:</span><span class="n">n</span><span class="p">}</span><span class="o">.</span><span class="n">hg18</span><span class="o">.</span><span class="n">bed</span> <span class="n">hg18ToHg19</span><span class="o">.</span><span class="n">over</span><span class="o">.</span><span class="n">chain</span><span class="o">.</span><span class="n">gz</span> <span class="p">{</span><span class="n">_output</span><span class="p">}</span> <span class="n">unmapped</span>
</pre></div>

</div>
</div>
</div>

</div>
<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div>
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p>Gene regions:</p>

<pre><code>sos run analysis/20180712_Enrichment_Workflow.ipynb gene_regions --trait AS \
    --y-data ~/Documents/GIT/LargeFiles/JointLCL/fastqtl_qqnorm_ASintron_RNAseqGeuvadis_YangVCF.txt.gz</code></pre>

</div>
</div>
</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[&nbsp;]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-sos"><pre><span></span><span class="p">[</span><span class="n">gene_regions_refgene</span><span class="p">]</span>
<span class="kn">output:</span> <span class="n">f</span><span class="s1">&#39;{annotation_dir}/genes.bed&#39;</span>
<span class="kn">download:</span> <span class="n">dest_dir</span> <span class="o">=</span> <span class="n">annotation_dir</span>
    <span class="n">http</span><span class="p">:</span><span class="o">//</span><span class="n">hgdownload</span><span class="o">.</span><span class="n">cse</span><span class="o">.</span><span class="n">ucsc</span><span class="o">.</span><span class="n">edu</span><span class="o">/</span><span class="n">goldenPath</span><span class="o">/</span><span class="n">hg19</span><span class="o">/</span><span class="n">database</span><span class="o">/</span><span class="n">refGene</span><span class="o">.</span><span class="n">txt</span><span class="o">.</span><span class="n">gz</span>
<span class="kn">python:</span> <span class="n">expand</span> <span class="o">=</span> <span class="s1">&#39;${ }&#39;</span><span class="p">,</span> <span class="kp">workdir</span> <span class="o">=</span> <span class="n">annotation_dir</span>
    <span class="kn">import</span> <span class="nn">pandas</span> <span class="kn">as</span> <span class="nn">pd</span>
    <span class="n">ref_gene</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_table</span><span class="p">(</span><span class="s1">&#39;refGene.txt.gz&#39;</span><span class="p">,</span> <span class="n">compression</span><span class="o">=</span><span class="s2">&quot;gzip&quot;</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span> 
                         <span class="n">header</span> <span class="o">=</span> <span class="bp">None</span><span class="p">,</span> <span class="n">usecols</span><span class="o">=</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span><span class="mi">2</span><span class="p">,</span><span class="mi">4</span><span class="p">,</span><span class="mi">5</span><span class="p">,</span><span class="mi">12</span><span class="p">),</span> 
                         <span class="n">names</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;tx_name&quot;</span><span class="p">,</span> <span class="s2">&quot;chrom&quot;</span><span class="p">,</span> <span class="s2">&quot;tx_start&quot;</span><span class="p">,</span> <span class="s2">&quot;tx_end&quot;</span><span class="p">,</span> <span class="s2">&quot;gene_name&quot;</span><span class="p">])</span>
    <span class="n">ref_gene</span> <span class="o">=</span> <span class="n">ref_gene</span><span class="o">.</span><span class="n">drop_duplicates</span><span class="p">(</span><span class="n">subset</span><span class="o">=</span><span class="p">(</span><span class="s2">&quot;chrom&quot;</span><span class="p">,</span> <span class="s2">&quot;tx_start&quot;</span><span class="p">,</span> <span class="s2">&quot;tx_end&quot;</span><span class="p">))</span>
    <span class="n">ref_gene</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">ref_gene</span><span class="o">.</span><span class="n">groupby</span><span class="p">([</span><span class="s1">&#39;chrom&#39;</span><span class="p">,</span><span class="s1">&#39;gene_name&#39;</span><span class="p">])</span><span class="o">.</span><span class="n">agg</span><span class="p">({</span><span class="s1">&#39;tx_start&#39;</span><span class="p">:</span> <span class="s1">&#39;min&#39;</span><span class="p">,</span> <span class="s1">&#39;tx_end&#39;</span><span class="p">:</span> <span class="s1">&#39;max&#39;</span><span class="p">})</span><span class="o">.</span><span class="n">to_records</span><span class="p">())</span>
    <span class="n">ref_gene</span><span class="o">.</span><span class="n">iloc</span><span class="p">[:,[</span><span class="mi">0</span><span class="p">,</span><span class="mi">2</span><span class="p">,</span><span class="mi">3</span><span class="p">]]</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">:</span><span class="n">r</span><span class="p">},</span> <span class="n">sep</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">header</span> <span class="o">=</span> <span class="bp">False</span><span class="p">,</span> <span class="n">index</span> <span class="o">=</span> <span class="bp">False</span><span class="p">)</span>
</pre></div>

</div>
</div>
</div>

</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[&nbsp;]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-sos"><pre><span></span><span class="p">[</span><span class="n">gene_regions</span><span class="p">]</span>
<span class="kn">input:</span> <span class="n">f</span><span class="s1">&#39;{annotation_dir}/gene_map.txt&#39;</span>
<span class="kn">output:</span> <span class="n">f</span><span class="s1">&#39;{annotation_dir}/gene_map.bed&#39;</span>
<span class="kn">bash:</span> <span class="n">expand</span> <span class="o">=</span> <span class="bp">True</span><span class="p">,</span> <span class="kp">workdir</span> <span class="o">=</span> <span class="n">annotation_dir</span><span class="p">,</span> <span class="n">container</span> <span class="o">=</span> <span class="s1">&#39;gaow/atac-gwas&#39;</span>
    <span class="n">cut</span> <span class="p">{</span><span class="n">_input</span><span class="p">}</span> <span class="o">-</span><span class="n">f2</span><span class="o">-</span><span class="mi">4</span> <span class="o">|</span> <span class="n">sed</span> <span class="s1">&#39;s/^/chr/g&#39;</span> <span class="o">|</span> <span class="n">grep</span> <span class="o">-</span><span class="n">v</span> <span class="n">chrM</span> <span class="o">|</span> <span class="n">sort</span><span class="o">-</span><span class="n">bed</span> <span class="o">-</span> <span class="o">&gt;</span> <span class="p">{</span><span class="n">_output</span><span class="p">}</span>
</pre></div>

</div>
</div>
</div>

</div>
<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div>
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p>Extended splice sites:</p>

<pre><code>sos run analysis/20180712_Enrichment_Workflow.ipynb extended_splice_site --trait AS \
    --y-data ~/Documents/GIT/LargeFiles/JointLCL/fastqtl_qqnorm_ASintron_RNAseqGeuvadis_YangVCF.txt.gz</code></pre>

</div>
</div>
</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[&nbsp;]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-sos"><pre><span></span><span class="p">[</span><span class="n">extended_splice_site</span><span class="p">]</span>
<span class="kn">input:</span> <span class="n">f</span><span class="s2">&quot;{annotation_dir}/Intron_UCSC.bed&quot;</span>
<span class="kn">output:</span> <span class="n">f</span><span class="s2">&quot;{annotation_dir}/Extended_splice_site.bed&quot;</span>
<span class="kn">python:</span> <span class="n">expand</span> <span class="o">=</span> <span class="s2">&quot;${ }&quot;</span><span class="p">,</span> <span class="kp">workdir</span> <span class="o">=</span> <span class="n">annotation_dir</span>
    <span class="n">lines</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span><span class="o">.</span><span class="n">split</span><span class="p">()</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="nb">open</span><span class="p">(</span><span class="err">$</span><span class="p">{</span><span class="n">_input</span><span class="p">:</span><span class="n">r</span><span class="p">})</span><span class="o">.</span><span class="n">readlines</span><span class="p">()]</span>
    <span class="n">out</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">lines</span><span class="p">:</span>
        <span class="n">out</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">f</span><span class="s1">&#39;{line[0]}</span><span class="se">\t</span><span class="s1">{int(line[1])-5}</span><span class="se">\t</span><span class="s1">{int(line[1])+5}&#39;</span><span class="p">)</span>
        <span class="n">out</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">f</span><span class="s1">&#39;{line[0]}</span><span class="se">\t</span><span class="s1">{int(line[2])-15}</span><span class="se">\t</span><span class="s1">{int(line[2])+15}&#39;</span><span class="p">)</span>
    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">:</span><span class="n">r</span><span class="p">},</span> <span class="s1">&#39;w&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
        <span class="n">f</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">out</span><span class="p">)</span> <span class="o">+</span> <span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span>
<span class="kn">bash:</span> <span class="n">expand</span> <span class="o">=</span> <span class="bp">True</span><span class="p">,</span> <span class="kp">workdir</span> <span class="o">=</span> <span class="n">annotation_dir</span><span class="p">,</span> <span class="n">container</span> <span class="o">=</span> <span class="s1">&#39;gaow/atac-gwas&#39;</span>
    <span class="n">sort</span><span class="o">-</span><span class="n">bed</span> <span class="p">{</span><span class="n">_output</span><span class="p">}</span> <span class="o">&gt;</span> <span class="p">{</span><span class="n">_output</span><span class="p">:</span><span class="n">n</span><span class="p">}</span><span class="o">.</span><span class="n">sorted</span><span class="o">.</span><span class="n">txt</span>
    <span class="n">mv</span> <span class="p">{</span><span class="n">_output</span><span class="p">:</span><span class="n">n</span><span class="p">}</span><span class="o">.</span><span class="n">sorted</span><span class="o">.</span><span class="n">txt</span> <span class="p">{</span><span class="n">_output</span><span class="p">}</span>
</pre></div>

</div>
</div>
</div>

</div>
<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div>
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h2 id="Prepare-summary-statistics-file">Prepare summary statistics file<a class="anchor-link" href="#Prepare-summary-statistics-file">&#182;</a></h2>
<pre><code>sos run analysis/20180712_Enrichment_Workflow.ipynb extract_sumstats \
    --y-data ~/Documents/GIT/LargeFiles/JointLCL/fastqtl_qqnorm_ASintron_RNAseqGeuvadis_YangVCF.txt.gz \
    --trait AS -j 16</code></pre>

</div>
</div>
</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[&nbsp;]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-sos"><pre><span></span><span class="c1"># Extract summary stats from RDS files to plain text</span>
<span class="p">[</span><span class="n">extract_sumstats_1</span><span class="p">]</span>
<span class="kn">input:</span> <span class="n">glob</span><span class="o">.</span><span class="n">glob</span><span class="p">(</span><span class="n">f</span><span class="s1">&#39;{cwd}/{y_data:bnn}_{int(max_dist/1000)}Kb/SuSiE_CS_[1-9]/*.rds&#39;</span><span class="p">),</span> <span class="kp">group_by</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="kp">concurrent</span> <span class="o">=</span> <span class="bp">True</span>
<span class="kn">output:</span> <span class="n">f</span><span class="s1">&#39;{_input:dd}/enrichment/{_input:bn}.sumstats.gz&#39;</span>
<span class="kn">R:</span> <span class="n">expand</span> <span class="o">=</span> <span class="s1">&#39;${ }&#39;</span>
    <span class="n">dat</span> <span class="o">=</span> <span class="n">readRDS</span><span class="p">(</span><span class="err">$</span><span class="p">{</span><span class="n">_input</span><span class="p">:</span><span class="n">r</span><span class="p">})</span>
    <span class="n">pip</span> <span class="o">=</span> <span class="n">dat</span><span class="err">$</span><span class="n">pip</span>
    <span class="n">names</span> <span class="o">=</span> <span class="n">names</span><span class="p">(</span><span class="n">readRDS</span><span class="p">(</span><span class="n">dat</span><span class="err">$</span><span class="nb">input</span><span class="p">)[[</span><span class="n">dat</span><span class="err">$</span><span class="n">idx</span><span class="p">]]</span><span class="err">$</span><span class="n">z_score</span><span class="p">)</span>
    <span class="n">zscore</span> <span class="o">=</span> <span class="n">readRDS</span><span class="p">(</span><span class="n">dat</span><span class="err">$</span><span class="nb">input</span><span class="p">)[[</span><span class="n">dat</span><span class="err">$</span><span class="n">idx</span><span class="p">]]</span><span class="err">$</span><span class="n">z_score</span>
    <span class="c1"># For earlier version of susieR</span>
    <span class="n">ordering</span> <span class="o">=</span> <span class="n">order</span><span class="p">(</span><span class="n">dat</span><span class="err">$</span><span class="n">sets</span><span class="err">$</span><span class="n">purity</span><span class="err">$</span><span class="nb">min</span><span class="o">.</span><span class="n">abs</span><span class="o">.</span><span class="n">corr</span><span class="p">,</span> <span class="n">decreasing</span><span class="o">=</span><span class="n">T</span><span class="p">)</span>
    <span class="n">dat</span><span class="err">$</span><span class="n">sets</span><span class="err">$</span><span class="n">cs</span> <span class="o">=</span> <span class="n">dat</span><span class="err">$</span><span class="n">sets</span><span class="err">$</span><span class="n">cs</span><span class="p">[</span><span class="n">ordering</span><span class="p">]</span>
    <span class="n">dat</span><span class="err">$</span><span class="n">sets</span><span class="err">$</span><span class="n">purity</span> <span class="o">=</span> <span class="n">dat</span><span class="err">$</span><span class="n">sets</span><span class="err">$</span><span class="n">purity</span><span class="p">[</span><span class="n">ordering</span><span class="p">,]</span>
    <span class="n">cs_id</span> <span class="o">=</span> <span class="n">rep</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">length</span><span class="p">(</span><span class="n">pip</span><span class="p">))</span>
    <span class="k">for</span> <span class="p">(</span><span class="n">i</span> <span class="ow">in</span> <span class="mi">1</span><span class="p">:</span><span class="n">length</span><span class="p">(</span><span class="n">dat</span><span class="err">$</span><span class="n">sets</span><span class="err">$</span><span class="n">cs</span><span class="p">))</span> <span class="p">{</span>
        <span class="n">cs_id</span><span class="p">[</span><span class="n">dat</span><span class="err">$</span><span class="n">sets</span><span class="err">$</span><span class="n">cs</span><span class="p">[</span><span class="n">i</span><span class="p">]]</span> <span class="o">=</span> <span class="n">i</span>
    <span class="p">}</span>
    <span class="n">write</span><span class="o">.</span><span class="n">table</span><span class="p">(</span><span class="n">cbind</span><span class="p">(</span><span class="n">names</span><span class="p">,</span><span class="n">pip</span><span class="p">,</span><span class="n">zscore</span><span class="p">,</span><span class="n">cs_id</span><span class="p">),</span> <span class="n">gzfile</span><span class="p">(</span><span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">:</span><span class="n">r</span><span class="p">}),</span> <span class="n">quote</span><span class="o">=</span><span class="n">FALSE</span><span class="p">,</span> <span class="n">col</span><span class="o">.</span><span class="n">names</span><span class="o">=</span><span class="n">FALSE</span><span class="p">,</span> <span class="n">row</span><span class="o">.</span><span class="n">names</span><span class="o">=</span><span class="n">FALSE</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span>

<span class="c1"># Consolidate results to one file</span>
<span class="p">[</span><span class="n">extract_sumstats_2</span><span class="p">]</span>
<span class="kn">output:</span> <span class="n">f</span><span class="s1">&#39;{cwd}/{y_data:bnn}_{int(max_dist/1000)}Kb/enrichment/SuSiE_loci.sumstats.gz&#39;</span>
<span class="kn">bash:</span> <span class="n">expand</span> <span class="o">=</span> <span class="bp">True</span>
    <span class="n">zcat</span> <span class="p">{</span><span class="n">_input</span><span class="p">}</span> <span class="o">|</span> <span class="n">gzip</span> <span class="o">--</span><span class="n">best</span> <span class="o">&gt;</span> <span class="p">{</span><span class="n">_output</span><span class="p">}</span>
<span class="n">_input</span><span class="o">.</span><span class="n">zap</span><span class="p">()</span>
</pre></div>

</div>
</div>
</div>

</div>
<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div>
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">

<pre><code>[GW] zcat /home/gaow/GIT/LargeFiles/JointLCL/AS_output/fastqtl_qqnorm_ASintron_RNAseqGeuvadis_YangVCF_100Kb/enrichment/SuSiE_loci.sumstats.gz | wc -l
1289796
[GW] zcat /home/gaow/GIT/LargeFiles/JointLCL/AS_output/fastqtl_qqnorm_ASintron_RNAseqGeuvadis_YangVCF_100Kb/enrichment/SuSiE_loci.sumstats.gz | cut -f 4 | grep 1 | wc -l
37911
[GW] zcat /home/gaow/GIT/LargeFiles/JointLCL/AS_output/fastqtl_qqnorm_ASintron_RNAseqGeuvadis_YangVCF_100Kb/enrichment/SuSiE_loci.sumstats.gz | cut -f 4 | grep 2 | wc -l
1881
[GW] zcat /home/gaow/GIT/LargeFiles/JointLCL/AS_output/fastqtl_qqnorm_ASintron_RNAseqGeuvadis_YangVCF_100Kb/enrichment/SuSiE_loci.sumstats.gz | cut -f 4 | grep 3 | wc -l
91</code></pre>
<p>So we have total of 1289796 variants, 37911 in CS 1, 1881 in CS 2, etc.</p>

</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div>
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h2 id="Convert-variants-from-summary-statistics-to-bed-format">Convert variants from summary statistics to bed format<a class="anchor-link" href="#Convert-variants-from-summary-statistics-to-bed-format">&#182;</a></h2>
</div>
</div>
</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[&nbsp;]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-sos"><pre><span></span><span class="c1"># Auxiliary step to get variant in bed format based on variant ID in z-score file</span>
<span class="p">[</span><span class="n">zscore2bed_1</span><span class="p">]</span>
<span class="kn">parameter:</span> <span class="n">in_file</span> <span class="o">=</span> <span class="n">path</span><span class="p">()</span>
<span class="kn">parameter:</span> <span class="n">chr_prefix</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
<span class="kn">input:</span> <span class="n">in_file</span>
<span class="kn">output:</span> <span class="n">f</span><span class="s1">&#39;{_input:n}.bed.unsorted&#39;</span>
<span class="kn">R:</span> <span class="n">expand</span> <span class="o">=</span> <span class="s2">&quot;${ }&quot;</span><span class="p">,</span> <span class="n">container</span> <span class="o">=</span> <span class="s1">&#39;gaow/atac-gwas&#39;</span><span class="p">,</span> <span class="kp">workdir</span> <span class="o">=</span> <span class="n">cwd</span><span class="p">,</span> <span class="n">stdout</span> <span class="o">=</span> <span class="n">f</span><span class="s1">&#39;{_output:n}.stdout&#39;</span>
    <span class="n">library</span><span class="p">(</span><span class="n">readr</span><span class="p">)</span>
    <span class="n">library</span><span class="p">(</span><span class="n">stringr</span><span class="p">)</span>
    <span class="n">library</span><span class="p">(</span><span class="n">dplyr</span><span class="p">)</span>
    <span class="n">var_file</span> <span class="o">&lt;-</span> <span class="err">$</span><span class="p">{</span><span class="n">_input</span><span class="p">:</span><span class="n">r</span><span class="p">}</span>
    <span class="n">out_file</span> <span class="o">&lt;-</span> <span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">:</span><span class="n">r</span><span class="p">}</span>

    <span class="n">variants</span> <span class="o">&lt;-</span> <span class="n">read_tsv</span><span class="p">(</span><span class="n">var_file</span><span class="p">)</span>
    <span class="n">colnames</span><span class="p">(</span><span class="n">variants</span><span class="p">)</span> <span class="o">=</span> <span class="n">c</span><span class="p">(</span><span class="s1">&#39;variant&#39;</span><span class="p">,</span> <span class="s1">&#39;pip&#39;</span><span class="p">,</span> <span class="s1">&#39;zscore&#39;</span><span class="p">,</span> <span class="s1">&#39;cs&#39;</span><span class="p">)</span>
    <span class="n">var_info</span> <span class="o">&lt;-</span> <span class="n">str_split</span><span class="p">(</span><span class="n">variants</span><span class="err">$</span><span class="n">variant</span><span class="p">,</span> <span class="s2">&quot;_&quot;</span><span class="p">)</span>
    <span class="n">variants</span> <span class="o">&lt;-</span> <span class="n">mutate</span><span class="p">(</span><span class="n">variants</span><span class="p">,</span> <span class="nb">chr</span> <span class="o">=</span> <span class="n">paste0</span><span class="p">(</span><span class="s2">&quot;${chr_prefix}&quot;</span><span class="p">,</span> <span class="n">sapply</span><span class="p">(</span><span class="n">var_info</span><span class="p">,</span> <span class="n">function</span><span class="p">(</span><span class="n">x</span><span class="p">){</span><span class="n">x</span><span class="p">[</span><span class="mi">2</span><span class="p">]})),</span> 
                                 <span class="n">pos</span> <span class="o">=</span> <span class="n">sapply</span><span class="p">(</span><span class="n">var_info</span><span class="p">,</span> <span class="n">function</span><span class="p">(</span><span class="n">x</span><span class="p">){</span><span class="n">x</span><span class="p">[</span><span class="mi">3</span><span class="p">]}))</span> <span class="o">%&gt;%</span>
                <span class="n">mutate</span><span class="p">(</span><span class="n">start</span> <span class="o">=</span> <span class="k">as</span><span class="o">.</span><span class="n">numeric</span><span class="p">(</span><span class="n">pos</span><span class="p">),</span> <span class="n">stop</span><span class="o">=</span><span class="k">as</span><span class="o">.</span><span class="n">numeric</span><span class="p">(</span><span class="n">pos</span><span class="p">)</span>  <span class="o">+</span> <span class="mi">1</span><span class="p">)</span> <span class="o">%&gt;%</span>
                <span class="n">select</span><span class="p">(</span><span class="nb">chr</span><span class="p">,</span> <span class="n">start</span><span class="p">,</span> <span class="n">stop</span><span class="p">,</span> <span class="n">variant</span><span class="p">)</span>
    <span class="n">options</span><span class="p">(</span><span class="n">scipen</span><span class="o">=</span><span class="mi">1000</span><span class="p">)</span> <span class="c1"># So that positions are always fully written out)</span>
    <span class="n">write</span><span class="o">.</span><span class="n">table</span><span class="p">(</span><span class="n">variants</span><span class="p">,</span> <span class="nb">file</span><span class="o">=</span><span class="n">out_file</span><span class="p">,</span> <span class="n">quote</span><span class="o">=</span><span class="n">FALSE</span><span class="p">,</span> <span class="n">col</span><span class="o">.</span><span class="n">names</span><span class="o">=</span><span class="n">FALSE</span><span class="p">,</span> <span class="n">row</span><span class="o">.</span><span class="n">names</span><span class="o">=</span><span class="n">FALSE</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span>

<span class="p">[</span><span class="n">zscore2bed_2</span><span class="p">]</span>
<span class="kn">output:</span> <span class="n">f</span><span class="s1">&#39;{_input:n}&#39;</span>
<span class="kn">bash:</span> <span class="n">expand</span> <span class="o">=</span> <span class="bp">True</span><span class="p">,</span> <span class="n">container</span> <span class="o">=</span> <span class="s1">&#39;gaow/atac-gwas&#39;</span><span class="p">,</span> <span class="kp">workdir</span> <span class="o">=</span> <span class="n">cwd</span>
     <span class="n">sort</span><span class="o">-</span><span class="n">bed</span> <span class="p">{</span><span class="n">_input</span><span class="p">}</span> <span class="o">&gt;</span> <span class="p">{</span><span class="n">_output</span><span class="p">}</span>
<span class="n">_input</span><span class="o">.</span><span class="n">zap</span><span class="p">()</span>

<span class="p">[</span><span class="n">get_variants</span><span class="p">:</span> <span class="kp">provides</span> <span class="o">=</span> <span class="s1">&#39;{data}.bed&#39;</span><span class="p">]</span>
<span class="kn">output:</span> <span class="n">f</span><span class="s1">&#39;{data}.bed&#39;</span>
<span class="n">sos_run</span><span class="p">(</span><span class="s1">&#39;zscore2bed&#39;</span><span class="p">,</span> <span class="n">in_file</span> <span class="o">=</span> <span class="n">f</span><span class="s1">&#39;{_output:n}.gz&#39;</span><span class="p">)</span>
</pre></div>

</div>
</div>
</div>

</div>
<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div>
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h2 id="Apply-ranged-based-annotations">Apply ranged based annotations<a class="anchor-link" href="#Apply-ranged-based-annotations">&#182;</a></h2>
<pre><code>sos run analysis/20180712_Enrichment_Workflow.ipynb range2var_annotation \
    --y-data ~/Documents/GIT/LargeFiles/JointLCL/fastqtl_qqnorm_ASintron_RNAseqGeuvadis_YangVCF.txt.gz \
    --trait AS -j 16 \
    --single-annot data/annotation.list</code></pre>

</div>
</div>
</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[&nbsp;]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-sos"><pre><span></span><span class="c1"># Get variants in data that falls in target region</span>
<span class="p">[</span><span class="n">range2var_annotation_1</span><span class="p">]</span>
<span class="kn">depends:</span> <span class="n">f</span><span class="s1">&#39;{z_score:n}.bed&#39;</span>
<span class="kn">input:</span> <span class="nb">set</span><span class="p">(</span><span class="n">paths</span><span class="p">(</span><span class="n">single_anno</span><span class="p">)),</span> <span class="kp">group_by</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="kp">concurrent</span> <span class="o">=</span> <span class="bp">True</span>
<span class="kn">output:</span> <span class="n">f</span><span class="s1">&#39;{out_dir}/{_input:bn}.{z_score:bn}.bed&#39;</span>
<span class="kn">bash:</span> <span class="n">expand</span> <span class="o">=</span> <span class="bp">True</span><span class="p">,</span> <span class="n">container</span> <span class="o">=</span> <span class="s1">&#39;gaow/atac-gwas&#39;</span><span class="p">,</span> <span class="kp">workdir</span> <span class="o">=</span> <span class="n">cwd</span><span class="p">,</span> <span class="n">volumes</span> <span class="o">=</span> <span class="n">f</span><span class="s1">&#39;{annotation_dir}:{annotation_dir}&#39;</span>
    <span class="n">bedops</span> <span class="o">-</span><span class="n">e</span> <span class="p">{</span><span class="n">z_score</span><span class="p">:</span><span class="n">n</span><span class="p">}</span><span class="o">.</span><span class="n">bed</span> <span class="p">{</span><span class="n">_input</span><span class="p">}</span> <span class="o">&gt;</span> <span class="p">{</span><span class="n">_output</span><span class="p">}</span>
</pre></div>

</div>
</div>
</div>

</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[&nbsp;]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-sos"><pre><span></span><span class="c1"># Make binary annotation file</span>
<span class="p">[</span><span class="n">range2var_annotation_2</span><span class="p">]</span>
<span class="kn">depends:</span> <span class="n">z_score</span>
<span class="kn">input:</span> <span class="kp">group_by</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="kp">concurrent</span> <span class="o">=</span> <span class="bp">True</span>
<span class="kn">output:</span> <span class="n">f</span><span class="s1">&#39;{_input:n}.gz&#39;</span>
<span class="kn">R:</span> <span class="n">expand</span> <span class="o">=</span> <span class="s2">&quot;${ }&quot;</span><span class="p">,</span> <span class="n">container</span> <span class="o">=</span> <span class="s1">&#39;gaow/atac-gwas&#39;</span><span class="p">,</span> <span class="kp">workdir</span> <span class="o">=</span> <span class="n">cwd</span><span class="p">,</span> <span class="n">stdout</span> <span class="o">=</span> <span class="n">f</span><span class="s1">&#39;{_output:n}.stdout&#39;</span>
    <span class="n">library</span><span class="p">(</span><span class="n">readr</span><span class="p">)</span>
    <span class="n">library</span><span class="p">(</span><span class="n">dplyr</span><span class="p">)</span>
    <span class="n">library</span><span class="p">(</span><span class="n">stringr</span><span class="p">)</span>

    <span class="n">variant_tsv</span> <span class="o">&lt;-</span> <span class="err">$</span><span class="p">{</span><span class="n">z_score</span><span class="p">:</span><span class="n">r</span><span class="p">}</span>
    <span class="n">annotation_var_bed</span> <span class="o">&lt;-</span> <span class="err">$</span><span class="p">{</span><span class="n">_input</span><span class="p">:</span><span class="n">r</span><span class="p">}</span>
    <span class="n">annot_name</span> <span class="o">&lt;-</span> <span class="err">$</span><span class="p">{</span><span class="n">_input</span><span class="p">:</span><span class="n">bnr</span><span class="p">}</span> <span class="o">%&gt;%</span> <span class="n">str_replace</span><span class="p">(</span><span class="n">paste0</span><span class="p">(</span><span class="s2">&quot;.&quot;</span><span class="p">,</span><span class="err">$</span><span class="p">{</span><span class="n">z_score</span><span class="p">:</span><span class="n">bnr</span><span class="p">}),</span> <span class="s2">&quot;&quot;</span><span class="p">)</span>
    <span class="n">out_name</span> <span class="o">&lt;-</span> <span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">:</span><span class="n">r</span><span class="p">}</span>

    <span class="nb">vars</span> <span class="o">&lt;-</span> <span class="n">read_tsv</span><span class="p">(</span><span class="n">variant_tsv</span><span class="p">,</span> <span class="n">col_names</span><span class="o">=</span><span class="n">FALSE</span><span class="p">)[,</span><span class="mi">1</span><span class="p">]</span>
    <span class="n">annot_vars</span> <span class="o">=</span> <span class="n">read_tsv</span><span class="p">(</span><span class="n">annotation_var_bed</span><span class="p">,</span> <span class="n">col_names</span><span class="o">=</span><span class="n">FALSE</span><span class="p">)</span>
    <span class="n">names</span><span class="p">(</span><span class="nb">vars</span><span class="p">)</span> <span class="o">&lt;-</span> <span class="s2">&quot;SNP&quot;</span>
    <span class="nb">vars</span> <span class="o">&lt;-</span> <span class="nb">vars</span> <span class="o">%&gt;%</span>
            <span class="n">mutate</span><span class="p">(</span><span class="n">annot_d</span> <span class="o">=</span> <span class="n">case_when</span><span class="p">(</span><span class="n">SNP</span> <span class="o">%</span><span class="ow">in</span><span class="o">%</span> <span class="n">annot_vars</span><span class="err">$</span><span class="n">X4</span> <span class="o">~</span> <span class="mi">1</span><span class="p">,</span>
                                                        <span class="n">TRUE</span> <span class="o">~</span> <span class="mi">0</span><span class="p">))</span>
    <span class="c1"># remove possible duplicate rows</span>
    <span class="nb">vars</span> <span class="o">&lt;-</span> <span class="n">unique</span><span class="p">(</span><span class="nb">vars</span><span class="p">)</span>    
    <span class="n">names</span><span class="p">(</span><span class="nb">vars</span><span class="p">)[</span><span class="mi">2</span><span class="p">]</span> <span class="o">&lt;-</span> <span class="n">annot_name</span>
    <span class="n">write</span><span class="o">.</span><span class="n">table</span><span class="p">(</span><span class="nb">vars</span><span class="p">,</span> <span class="nb">file</span><span class="o">=</span><span class="n">gzfile</span><span class="p">(</span><span class="n">out_name</span><span class="p">),</span>
                <span class="n">col</span><span class="o">.</span><span class="n">names</span><span class="o">=</span><span class="n">TRUE</span><span class="p">,</span> <span class="n">row</span><span class="o">.</span><span class="n">names</span><span class="o">=</span><span class="n">FALSE</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">quote</span><span class="o">=</span><span class="n">FALSE</span><span class="p">)</span>
</pre></div>

</div>
</div>
</div>

</div>
<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div>
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h2 id="Matched-enrichment-analysis-via-GREGOR">Matched enrichment analysis via GREGOR<a class="anchor-link" href="#Matched-enrichment-analysis-via-GREGOR">&#182;</a></h2><p>To properly perform enrichment analysis we want to match the control SNPs with the SNPs of interest -- that is, SNPs inside CS -- in terms of LD, distance to nearest gene and MAF. The <a href="http://csg.sph.umich.edu/GREGOR/index.php/site/download">GREGOR</a> software can generate list of matched SNPs. I will use SNPs inside CS as input and expect a list of output SNPs matching these inputs.</p>
<p>GREGOR is release under University of Michigan license so I'll not make it into a docker image. So path to GREGOR directory is required. Also we need reference files, prepared by:</p>

<pre><code>cat \
    GREGOR.AFR.ref.r2.greater.than.0.7.tar.gz.part.00 \
    GREGOR.AFR.ref.r2.greater.than.0.7.tar.gz.part.01 \
    &gt; GREGOR.AFR.ref.r2.greater.than.0.7.tar.gz
tar zxvf GREGOR.AFR.ref.r2.greater.than.0.7.tar.gz</code></pre>
<p>MD5SUM check:</p>

<pre><code>AFR.part.0  ( MD5: 9926904128dd58d6bf1ad4f1e90638af )
AFR.part.1  ( MD5: c1d30aff89a584bfa8c1fa1bdc197f21 )</code></pre>

</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div>
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p>The input file can be a list of <code>rs</code> ID's. In <code>SuSiE_loci.sumstats.gz</code> we do have this information.</p>

</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div>
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">

<pre><code>sos run analysis/20180712_Enrichment_Workflow.ipynb gregor --cs 1 \
    --y-data ~/Documents/GIT/LargeFiles/JointLCL/fastqtl_qqnorm_ASintron_RNAseqGeuvadis_YangVCF.txt.gz \
    --trait AS -j 16 \
    --single-annot data/annotation.list 

sos run analysis/20180712_Enrichment_Workflow.ipynb gregor --cs 2 \
    --y-data ~/Documents/GIT/LargeFiles/JointLCL/fastqtl_qqnorm_ASintron_RNAseqGeuvadis_YangVCF.txt.gz \
    --trait AS -j 16 \
    --single-annot data/annotation.list</code></pre>

</div>
</div>
</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[&nbsp;]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-sos"><pre><span></span><span class="p">[</span><span class="n">gregor_1</span> <span class="p">(</span><span class="n">make</span> <span class="n">SNP</span> <span class="n">index</span><span class="p">)]</span>
<span class="c1"># which set to test. 1 to test primary test, otherwise non-primary set</span>
<span class="kn">parameter:</span> <span class="n">cs</span> <span class="o">=</span> <span class="mi">1</span>
<span class="kn">input:</span> <span class="n">z_score</span>
<span class="kn">output:</span> <span class="n">f</span><span class="s1">&#39;{_input:nn}.set{cs}.txt&#39;</span><span class="p">,</span> <span class="n">f</span><span class="s1">&#39;{_input:nn}.annotations.list&#39;</span>
<span class="kn">bash:</span> <span class="n">expand</span> <span class="o">=</span> <span class="s1">&#39;${ }&#39;</span>
    <span class="n">zcat</span> <span class="err">$</span><span class="p">{</span><span class="n">_input</span><span class="p">}</span> <span class="o">|</span> <span class="n">awk</span> <span class="s1">&#39;{if (${&quot;$4==1 &amp;&amp; $2 &gt; 0.05&quot; if cs == 1 else &quot;$4&gt;1 &amp;&amp; $2 &gt; 0.05&quot;}) print $0}&#39;</span> <span class="o">|</span> <span class="n">cut</span> <span class="o">-</span><span class="n">f</span> <span class="mi">2</span><span class="p">,</span><span class="mi">3</span> <span class="o">-</span><span class="n">d</span> <span class="s2">&quot;_&quot;</span> <span class="o">|</span> <span class="n">sed</span> <span class="s1">&#39;s/_/:/g&#39;</span> <span class="o">|</span> <span class="n">sort</span> <span class="o">-</span><span class="n">u</span> <span class="o">&gt;</span> <span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]}</span>

<span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">_output</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span> <span class="s1">&#39;w&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
    <span class="n">f</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">single_anno</span><span class="p">))</span>

<span class="p">[</span><span class="n">gregor_2</span> <span class="p">(</span><span class="n">make</span> <span class="n">configuration</span> <span class="nb">file</span><span class="p">)]</span>
<span class="kn">parameter:</span> <span class="n">gregor_db</span> <span class="o">=</span> <span class="n">path</span><span class="p">(</span><span class="s1">&#39;/data/GREGOR_DB&#39;</span><span class="p">)</span>
<span class="kn">parameter:</span> <span class="n">pop</span> <span class="o">=</span> <span class="s1">&#39;AFR&#39;</span>
<span class="kn">parameter:</span> <span class="n">ncpu</span> <span class="o">=</span> <span class="mi">10</span>
<span class="kn">output:</span> <span class="n">f</span><span class="s1">&#39;{_input[0]:n}.gregor.conf&#39;</span>
<span class="kn">report:</span> <span class="n">output</span> <span class="o">=</span> <span class="n">f</span><span class="s1">&#39;{_output}&#39;</span><span class="p">,</span> <span class="n">expand</span> <span class="o">=</span> <span class="bp">True</span>
    <span class="c1">##############################################################################</span>
    <span class="c1"># CHIPSEQ ENRICHMENT CONFIGURATION FILE</span>
    <span class="c1"># This configuration file contains run-time configuration of</span>
    <span class="c1"># CHIP_SEQ ENRICHMENT</span>
    <span class="c1">###############################################################################</span>
    <span class="c1">## KEY ELEMENTS TO CONFIGURE : NEED TO MODIFY</span>
    <span class="c1">###############################################################################</span>
    <span class="n">INDEX_SNP_FILE</span> <span class="o">=</span> <span class="p">{</span><span class="n">_input</span><span class="p">[</span><span class="mi">0</span><span class="p">]}</span>
    <span class="n">BED_FILE_INDEX</span> <span class="o">=</span> <span class="p">{</span><span class="n">_input</span><span class="p">[</span><span class="mi">1</span><span class="p">]}</span> 
    <span class="n">REF_DIR</span> <span class="o">=</span> <span class="p">{</span><span class="n">gregor_db</span><span class="p">}</span>
    <span class="n">R2THRESHOLD</span> <span class="o">=</span> <span class="mf">0.7</span> <span class="c1">## must be greater than 0.7</span>
    <span class="n">LDWINDOWSIZE</span> <span class="o">=</span> <span class="mi">10000</span> <span class="c1">## must be less than 1MB; these two values define LD buddies</span>
    <span class="n">OUT_DIR</span> <span class="o">=</span> <span class="p">{</span><span class="n">_output</span><span class="p">:</span><span class="n">nn</span><span class="p">}</span><span class="n">_gregor_output</span>
    <span class="n">MIN_NEIGHBOR_NUM</span> <span class="o">=</span> <span class="mi">10</span> <span class="c1">## define the size of neighborhood</span>
    <span class="n">BEDFILE_IS_SORTED</span> <span class="o">=</span> <span class="n">true</span>  <span class="c1">## false, if the bed files are not sorted</span>
    <span class="n">POPULATION</span> <span class="o">=</span> <span class="p">{</span><span class="n">pop</span><span class="p">}</span>  <span class="c1">## define the population, you can specify EUR, AFR, AMR or ASN</span>
    <span class="n">TOPNBEDFILES</span> <span class="o">=</span> <span class="mi">2</span> 
    <span class="n">JOBNUMBER</span> <span class="o">=</span> <span class="p">{</span><span class="n">ncpu</span><span class="p">}</span>
    <span class="c1">###############################################################################</span>
    <span class="c1">#BATCHTYPE = mosix ##  submit jobs on MOSIX</span>
    <span class="c1">#BATCHOPTS = -E/tmp -i -m2000 -j10,11,12,13,14,15,16,17,18,19,120,122,123,124,125 sh -c</span>
    <span class="c1">###############################################################################</span>
    <span class="c1">#BATCHTYPE = slurm   ##  submit jobs on SLURM</span>
    <span class="c1">#BATCHOPTS = --partition=broadwl --account=pi-mstephens --time=0:30:0</span>
    <span class="c1">###############################################################################</span>
    <span class="n">BATCHTYPE</span> <span class="o">=</span> <span class="n">local</span> <span class="c1">##  run jobs on local machine</span>

<span class="kn">bash:</span> <span class="n">expand</span> <span class="o">=</span> <span class="bp">True</span>
    <span class="n">sed</span> <span class="o">-</span><span class="n">i</span> <span class="s1">&#39;/^$/d&#39;</span> <span class="p">{</span><span class="n">_output</span><span class="p">}</span>
</pre></div>

</div>
</div>
</div>

</div>
<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div>
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p>GREGOR is written in <code>perl</code>. Some libraries are required to run GREGOR:</p>

<pre><code>sudo apt-get install libdbi-perl libswitch-perl libdbd-sqlite3-perl</code></pre>

</div>
</div>
</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[&nbsp;]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-sos"><pre><span></span><span class="p">[</span><span class="n">gregor_3</span> <span class="p">(</span><span class="n">run</span> <span class="n">gregor</span><span class="p">)]</span>
<span class="kn">parameter:</span> <span class="n">gregor_path</span> <span class="o">=</span> <span class="n">path</span><span class="p">(</span><span class="s1">&#39;/opt/GREGOR&#39;</span><span class="p">)</span>
<span class="kn">output:</span> <span class="n">f</span><span class="s1">&#39;{_input:nn}_gregor_output/StatisticSummaryFile.txt&#39;</span>
<span class="kn">bash:</span> <span class="n">expand</span> <span class="o">=</span> <span class="bp">True</span>
    <span class="n">rm</span> <span class="o">-</span><span class="n">rf</span> <span class="p">{</span><span class="n">_input</span><span class="p">:</span><span class="n">nn</span><span class="p">}</span><span class="n">_gregor_output</span> <span class="o">&amp;&amp;</span> <span class="n">perl</span> <span class="p">{</span><span class="n">gregor_path</span><span class="p">}</span><span class="o">/</span><span class="n">script</span><span class="o">/</span><span class="n">GREGOR</span><span class="o">.</span><span class="n">pl</span> <span class="o">--</span><span class="n">conf</span> <span class="p">{</span><span class="n">_input</span><span class="p">}</span> <span class="o">&amp;&amp;</span> <span class="n">touch</span> <span class="p">{</span><span class="n">_output</span><span class="p">}</span>

<span class="p">[</span><span class="n">gregor_4</span> <span class="p">(</span><span class="n">format</span> <span class="n">output</span><span class="p">)]</span>
<span class="kn">output:</span> <span class="n">f</span><span class="s1">&#39;{_input:n}.csv&#39;</span>
<span class="kn">bash:</span> <span class="n">expand</span> <span class="o">=</span> <span class="bp">True</span>
    <span class="n">sed</span> <span class="s1">&#39;s/</span><span class="se">\t</span><span class="s1">/,/g&#39;</span> <span class="p">{</span><span class="n">_input</span><span class="p">}</span> <span class="o">&gt;</span> <span class="p">{</span><span class="n">_output</span><span class="p">}</span>
</pre></div>

</div>
</div>
</div>

</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[11]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-sos"><pre><span></span><span class="n">dat</span> <span class="o">=</span> <span class="n">read</span><span class="o">.</span><span class="n">table</span><span class="p">(</span><span class="s1">&#39;/home/gaow/Documents/GIT/LargeFiles/JointLCL/AS_output/fastqtl_qqnorm_ASintron_RNAseqGeuvadis_YangVCF_100Kb/enrichment/SuSiE_loci_gregor_output/StatisticSummaryFile.csv&#39;</span><span class="p">,</span> <span class="n">head</span><span class="o">=</span><span class="n">T</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s1">&#39;,&#39;</span><span class="p">)</span>
<span class="n">dat</span><span class="p">[</span><span class="n">order</span><span class="p">(</span><span class="n">dat</span><span class="err">$</span><span class="n">PValue</span><span class="p">),]</span>
</pre></div>

</div>
</div>
</div>

<div class="output_wrapper">
<div class="output">


<div class="output_area">

<div class="prompt"></div>



<div class="output_html rendered_html output_subarea ">
<table>
<thead><tr><th></th><th scope=col>Bed_File</th><th scope=col>InBed_Index_SNP</th><th scope=col>ExpectNum_of_InBed_SNP</th><th scope=col>PValue</th></tr></thead>
<tbody>
	<tr><th scope=row>1</th><td>E116-H3K36me3.bed       </td><td>2883                    </td><td> 770.67361              </td><td> 0.000000e+00           </td></tr>
	<tr><th scope=row>2</th><td>E116-H3K4me3.bed        </td><td>2162                    </td><td> 650.85295              </td><td> 0.000000e+00           </td></tr>
	<tr><th scope=row>7</th><td>PolII_binding.bed       </td><td>1353                    </td><td> 252.57187              </td><td> 0.000000e+00           </td></tr>
	<tr><th scope=row>9</th><td>E116-H3K9ac.bed         </td><td>2112                    </td><td> 591.04726              </td><td> 0.000000e+00           </td></tr>
	<tr><th scope=row>10</th><td>E116-H3K79me2.bed       </td><td>2668                    </td><td> 752.76032              </td><td> 0.000000e+00           </td></tr>
	<tr><th scope=row>11</th><td>E116-H3K4me2.bed        </td><td>2809                    </td><td>1024.89199              </td><td> 0.000000e+00           </td></tr>
	<tr><th scope=row>12</th><td>Coding_UCSC.bed         </td><td>1932                    </td><td> 791.45368              </td><td> 0.000000e+00           </td></tr>
	<tr><th scope=row>13</th><td>E116-H3K27ac.bed        </td><td>2424                    </td><td> 767.45946              </td><td> 0.000000e+00           </td></tr>
	<tr><th scope=row>6</th><td>E116-H3K4me1.bed        </td><td>2049                    </td><td> 900.11623              </td><td>2.817817e-303           </td></tr>
	<tr><th scope=row>8</th><td>E116-H2A.Z.bed          </td><td>1512                    </td><td> 591.64615              </td><td>5.953945e-269           </td></tr>
	<tr><th scope=row>3</th><td>Extended_splice_site.bed</td><td> 522                    </td><td> 131.50573              </td><td>1.491773e-164           </td></tr>
	<tr><th scope=row>5</th><td>E116-H4K20me1.bed       </td><td> 294                    </td><td>  73.90964              </td><td> 1.072375e-89           </td></tr>
	<tr><th scope=row>4</th><td>CTCF_binding.bed        </td><td> 779                    </td><td> 432.57289              </td><td> 5.859478e-59           </td></tr>
	<tr><th scope=row>14</th><td>Intron_UCSC.bed         </td><td>5879                    </td><td>5724.81885              </td><td> 5.798512e-13           </td></tr>
</tbody>
</table>

</div>

</div>

</div>
</div>

</div>
<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div>
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h2 id="Unmatched-enrichment-analysis">Unmatched enrichment analysis<a class="anchor-link" href="#Unmatched-enrichment-analysis">&#182;</a></h2>
</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div>
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p>In Li 2016 manuscript the analysis was not matched by LD, MAF and TSS for sQTL. The only constraint for control SNPs was that they should be within intron clusters. To focus analysis on these SNPs:</p>

<pre><code>sos run analysis/20180712_Enrichment_Workflow.ipynb overlap_cluster \
    --y-data ~/Documents/GIT/LargeFiles/JointLCL/fastqtl_qqnorm_ASintron_RNAseqGeuvadis_YangVCF.txt.gz \
    --trait AS</code></pre>

</div>
</div>
</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[&nbsp;]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-sos"><pre><span></span><span class="c1"># Overlap z-scores with intron clusters</span>
<span class="p">[</span><span class="n">overlap_cluster</span><span class="p">]</span>
<span class="kn">input:</span> <span class="n">f</span><span class="s2">&quot;{z_score:n}.bed&quot;</span>
<span class="kn">output:</span> <span class="n">f</span><span class="s1">&#39;{out_dir}/{y_data:bnn}.{z_score:bn}.overlap.bed&#39;</span>
<span class="kn">bash:</span> <span class="n">expand</span> <span class="o">=</span> <span class="bp">True</span><span class="p">,</span> <span class="n">container</span> <span class="o">=</span> <span class="s1">&#39;gaow/atac-gwas&#39;</span><span class="p">,</span> <span class="kp">workdir</span> <span class="o">=</span> <span class="n">cwd</span><span class="p">,</span> <span class="n">volumes</span> <span class="o">=</span> <span class="p">[</span><span class="n">f</span><span class="s1">&#39;{annotation_dir}:{annotation_dir}&#39;</span><span class="p">,</span> <span class="n">f</span><span class="s2">&quot;{y_data:d}:{y_data:d}&quot;</span><span class="p">]</span>
    <span class="c1"># bedops -e {_input} {annotation_dir}/gene_map.bed | sort -u &gt; {_output}</span>
    <span class="n">bedops</span> <span class="o">-</span><span class="n">e</span> <span class="p">{</span><span class="n">_input</span><span class="p">}</span> <span class="o">&lt;</span><span class="p">(</span><span class="n">cat</span> <span class="p">{</span><span class="n">annotation_dir</span><span class="p">}</span><span class="o">/</span><span class="n">Coding_UCSC</span><span class="o">.</span><span class="n">bed</span> <span class="p">{</span><span class="n">annotation_dir</span><span class="p">}</span><span class="o">/</span><span class="n">Intron_UCSC</span><span class="o">.</span><span class="n">bed</span> <span class="o">|</span> <span class="n">sort</span><span class="o">-</span><span class="n">bed</span> <span class="o">-</span><span class="p">)</span> <span class="o">|</span> <span class="n">sort</span> <span class="o">-</span><span class="n">u</span> <span class="o">&gt;</span> <span class="p">{</span><span class="n">_output</span><span class="p">}</span>
</pre></div>

</div>
</div>
</div>

</div>
<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div>
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p>We ended up having 348K variants:</p>

<pre><code>[GW] wc -l /home/gaow/Documents/GIT/LargeFiles/JointLCL/AS_output/fastqtl_qqnorm_ASintron_RNAseqGeuvadis_YangVCF_100Kb/enrichment/SuSiE_loci_sumstats/fastqtl_qqnorm_ASintron_RNAseqGeuvadis_YangVCF.SuSiE_loci.sumstats.overlap.bed
348694</code></pre>

</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div>
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p>Test for larger PIP in annotation vs outside it.</p>

<pre><code>sos run analysis/20180712_Enrichment_Workflow.ipynb pip_rank_test \
    --y-data ~/Documents/GIT/LargeFiles/JointLCL/fastqtl_qqnorm_ASintron_RNAseqGeuvadis_YangVCF.txt.gz \
    --trait AS -j 16 \
    --single-annot data/annotation.list</code></pre>

</div>
</div>
</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[&nbsp;]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-sos"><pre><span></span><span class="c1"># Subset data</span>
<span class="p">[</span><span class="n">pip_rank_test_1</span><span class="p">,</span> <span class="n">cs_fisher_test_1</span><span class="p">]</span>
<span class="kn">depends:</span> <span class="n">z_score</span><span class="p">,</span> <span class="n">f</span><span class="s1">&#39;{out_dir}/{y_data:bnn}.{z_score:bn}.overlap.bed&#39;</span>
<span class="n">input_files</span> <span class="o">=</span> <span class="p">[</span><span class="n">f</span><span class="s1">&#39;{out_dir}/{value:bn}.{z_score:bn}.gz&#39;</span> <span class="k">for</span> <span class="n">value</span> <span class="ow">in</span> <span class="n">paths</span><span class="p">(</span><span class="n">single_anno</span><span class="p">)]</span>
<span class="kn">input:</span> <span class="n">input_files</span><span class="p">,</span> <span class="kp">group_by</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="kp">concurrent</span> <span class="o">=</span> <span class="bp">True</span>
<span class="kn">output:</span> <span class="n">f</span><span class="s1">&#39;{_input:n}.data.rds&#39;</span>
<span class="kn">R:</span> <span class="n">expand</span> <span class="o">=</span> <span class="s1">&#39;${ }&#39;</span><span class="p">,</span> <span class="n">container</span> <span class="o">=</span> <span class="s1">&#39;gaow/atac-gwas&#39;</span><span class="p">,</span> <span class="kp">workdir</span> <span class="o">=</span> <span class="n">cwd</span><span class="p">,</span> <span class="n">stdout</span> <span class="o">=</span> <span class="n">f</span><span class="s1">&#39;{_output:n}.stdout&#39;</span>
    <span class="nb">set</span><span class="o">.</span><span class="n">seed</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span>
    <span class="n">library</span><span class="p">(</span><span class="n">readr</span><span class="p">)</span>
    <span class="n">library</span><span class="p">(</span><span class="n">dplyr</span><span class="p">)</span>
    <span class="n">variants</span> <span class="o">&lt;-</span> <span class="n">read_tsv</span><span class="p">(</span><span class="err">$</span><span class="p">{</span><span class="n">z_score</span><span class="p">:</span><span class="n">r</span><span class="p">},</span> <span class="n">col_names</span><span class="o">=</span><span class="n">FALSE</span><span class="p">)</span>
    <span class="n">colnames</span><span class="p">(</span><span class="n">variants</span><span class="p">)</span> <span class="o">=</span> <span class="n">c</span><span class="p">(</span><span class="s1">&#39;SNP&#39;</span><span class="p">,</span> <span class="s1">&#39;PIP&#39;</span><span class="p">,</span> <span class="s1">&#39;Z&#39;</span><span class="p">,</span> <span class="s1">&#39;CS&#39;</span><span class="p">)</span>
    <span class="n">annotation</span> <span class="o">&lt;-</span> <span class="n">read_tsv</span><span class="p">(</span><span class="err">$</span><span class="p">{</span><span class="n">_input</span><span class="p">:</span><span class="n">r</span><span class="p">},</span> <span class="n">col_names</span><span class="o">=</span><span class="n">TRUE</span><span class="p">)</span>
    <span class="kp">name</span> <span class="o">=</span> <span class="n">colnames</span><span class="p">(</span><span class="n">annotation</span><span class="p">)[</span><span class="mi">2</span><span class="p">]</span>
    <span class="n">colnames</span><span class="p">(</span><span class="n">annotation</span><span class="p">)</span> <span class="o">=</span> <span class="n">c</span><span class="p">(</span><span class="s1">&#39;SNP&#39;</span><span class="p">,</span> <span class="s1">&#39;GROUP&#39;</span><span class="p">)</span>
    <span class="n">cluster_subset</span> <span class="o">&lt;-</span> <span class="n">read_tsv</span><span class="p">(</span><span class="s1">&#39;${out_dir}/${y_data:bnn}.${z_score:bn}.overlap.bed&#39;</span><span class="p">,</span> <span class="n">col_names</span><span class="o">=</span><span class="n">FALSE</span><span class="p">)</span>
    <span class="n">colnames</span><span class="p">(</span><span class="n">cluster_subset</span><span class="p">)</span> <span class="o">=</span> <span class="n">c</span><span class="p">(</span><span class="s2">&quot;CHR&quot;</span><span class="p">,</span> <span class="s2">&quot;START&quot;</span><span class="p">,</span> <span class="s2">&quot;END&quot;</span><span class="p">,</span> <span class="s2">&quot;SNP&quot;</span><span class="p">)</span>
    <span class="c1"># restrict analysis to within genes</span>
    <span class="n">annotation</span> <span class="o">=</span> <span class="n">inner_join</span><span class="p">(</span><span class="n">cluster_subset</span><span class="p">,</span> <span class="n">annotation</span><span class="p">,</span> <span class="n">by</span> <span class="o">=</span> <span class="s2">&quot;SNP&quot;</span><span class="p">)[,</span><span class="n">c</span><span class="p">(</span><span class="s2">&quot;SNP&quot;</span><span class="p">,</span> <span class="s2">&quot;GROUP&quot;</span><span class="p">)]</span>
    <span class="c1"># remove duplicates in variants but keep the largest CS label when possible</span>
    <span class="n">variants</span> <span class="o">=</span> <span class="k">as</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">frame</span><span class="p">(</span><span class="n">variants</span><span class="p">)</span>
    <span class="n">variants</span> <span class="o">=</span> <span class="n">variants</span><span class="p">[</span><span class="n">order</span><span class="p">(</span><span class="n">variants</span><span class="p">[,</span><span class="s1">&#39;SNP&#39;</span><span class="p">],</span><span class="o">-</span><span class="n">variants</span><span class="p">[,</span><span class="s1">&#39;CS&#39;</span><span class="p">]),]</span>
    <span class="n">variants</span> <span class="o">=</span> <span class="n">variants</span><span class="p">[</span><span class="err">!</span><span class="n">duplicated</span><span class="p">(</span><span class="n">variants</span><span class="p">[,</span><span class="s1">&#39;SNP&#39;</span><span class="p">]),]</span>
    <span class="n">variants</span> <span class="o">=</span> <span class="n">inner_join</span><span class="p">(</span><span class="n">annotation</span><span class="p">,</span> <span class="n">variants</span><span class="p">,</span> <span class="n">by</span> <span class="o">=</span> <span class="s2">&quot;SNP&quot;</span><span class="p">)</span>
    <span class="n">saveRDS</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="kp">name</span><span class="o">=</span><span class="kp">name</span><span class="p">,</span> <span class="n">variants</span><span class="o">=</span><span class="n">variants</span><span class="p">),</span> <span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">:</span><span class="n">r</span><span class="p">})</span>
</pre></div>

</div>
</div>
</div>

</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[&nbsp;]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-sos"><pre><span></span><span class="c1"># Test if PIP is larger in annotations</span>
<span class="p">[</span><span class="n">pip_rank_test_2</span><span class="p">]</span>
<span class="kn">input:</span> <span class="kp">group_by</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="kp">concurrent</span> <span class="o">=</span> <span class="bp">True</span>
<span class="kn">output:</span> <span class="n">f</span><span class="s1">&#39;{_input:n}.pip_rank_test.csv&#39;</span>
<span class="kn">R:</span> <span class="n">expand</span> <span class="o">=</span> <span class="s1">&#39;${ }&#39;</span><span class="p">,</span> <span class="n">container</span> <span class="o">=</span> <span class="s1">&#39;gaow/atac-gwas&#39;</span><span class="p">,</span> <span class="kp">workdir</span> <span class="o">=</span> <span class="n">cwd</span><span class="p">,</span> <span class="n">stdout</span> <span class="o">=</span> <span class="n">f</span><span class="s1">&#39;{_output:n}.stdout&#39;</span>
    <span class="n">dat</span> <span class="o">=</span> <span class="n">readRDS</span><span class="p">(</span><span class="err">$</span><span class="p">{</span><span class="n">_input</span><span class="p">:</span><span class="n">r</span><span class="p">})</span>
    <span class="n">variants</span> <span class="o">=</span> <span class="n">dat</span><span class="err">$</span><span class="n">variants</span>
    <span class="kp">name</span> <span class="o">=</span> <span class="n">dat</span><span class="err">$</span><span class="kp">name</span>
    <span class="k">if</span> <span class="p">(</span><span class="n">length</span><span class="p">(</span><span class="n">unique</span><span class="p">(</span><span class="n">variants</span><span class="err">$</span><span class="n">GROUP</span><span class="p">))</span> <span class="o">==</span> <span class="mi">1</span><span class="p">)</span> <span class="p">{</span>
      <span class="n">write</span><span class="p">(</span><span class="n">paste</span><span class="p">(</span><span class="kp">name</span><span class="p">,</span> <span class="n">NA</span><span class="p">,</span> <span class="n">NA</span><span class="p">,</span> <span class="n">NA</span><span class="p">,</span> <span class="n">NA</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s1">&#39;,&#39;</span><span class="p">),</span> <span class="nb">file</span> <span class="o">=</span> <span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">:</span><span class="n">r</span><span class="p">})</span>
    <span class="p">}</span> <span class="k">else</span> <span class="p">{</span>
    <span class="n">test</span> <span class="o">=</span> <span class="n">wilcox</span><span class="o">.</span><span class="n">test</span><span class="p">(</span><span class="n">PIP</span> <span class="o">~</span> <span class="n">GROUP</span><span class="p">,</span> <span class="n">data</span><span class="o">=</span><span class="n">variants</span><span class="p">,</span> <span class="n">conf</span><span class="o">.</span><span class="n">int</span><span class="o">=</span><span class="n">TRUE</span><span class="p">)</span>
    <span class="n">write</span><span class="p">(</span><span class="n">paste</span><span class="p">(</span><span class="kp">name</span><span class="p">,</span> <span class="n">test</span><span class="err">$</span><span class="n">p</span><span class="o">.</span><span class="n">value</span><span class="p">,</span> <span class="n">test</span><span class="err">$</span><span class="n">estimate</span><span class="p">,</span> <span class="n">test</span><span class="err">$</span><span class="n">conf</span><span class="o">.</span><span class="n">int</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span> <span class="n">test</span><span class="err">$</span><span class="n">conf</span><span class="o">.</span><span class="n">int</span><span class="p">[</span><span class="mi">2</span><span class="p">],</span> <span class="n">sep</span><span class="o">=</span><span class="s1">&#39;,&#39;</span><span class="p">),</span> <span class="nb">file</span> <span class="o">=</span> <span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">:</span><span class="n">r</span><span class="p">})</span>
    <span class="p">}</span>

<span class="c1"># Consolidate results to one file</span>
<span class="p">[</span><span class="n">pip_rank_test_3</span><span class="p">,</span><span class="n">cs_fisher_test_3</span><span class="p">]</span>
<span class="kn">output:</span> <span class="n">f</span><span class="s2">&quot;{cwd}/{y_data:bnn}_{int(max_dist/1000)}Kb/enrichment/SuSiE_loci.sumstats.{step_name.rsplit(&#39;_&#39;, 1)[0]}.csv&quot;</span>
<span class="kn">bash:</span> <span class="n">expand</span> <span class="o">=</span> <span class="bp">True</span>
    <span class="n">cat</span> <span class="p">{</span><span class="n">_input</span><span class="p">}</span> <span class="o">&gt;</span> <span class="p">{</span><span class="n">_output</span><span class="p">}</span>
<span class="n">_input</span><span class="o">.</span><span class="n">zap</span><span class="p">()</span>
</pre></div>

</div>
</div>
</div>

</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[9]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-sos"><pre><span></span><span class="n">dat</span> <span class="o">=</span> <span class="n">read</span><span class="o">.</span><span class="n">table</span><span class="p">(</span><span class="s1">&#39;/home/gaow/Documents/GIT/LargeFiles/JointLCL/AS_output/fastqtl_qqnorm_ASintron_RNAseqGeuvadis_YangVCF_100Kb/enrichment/SuSiE_loci.sumstats.pip_rank_test.csv&#39;</span><span class="p">,</span> <span class="n">head</span><span class="o">=</span><span class="n">F</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s1">&#39;,&#39;</span><span class="p">)</span>
<span class="n">colnames</span><span class="p">(</span><span class="n">dat</span><span class="p">)</span> <span class="o">=</span> <span class="n">c</span><span class="p">(</span><span class="s1">&#39;annotation&#39;</span><span class="p">,</span> <span class="s1">&#39;p.value&#39;</span><span class="p">,</span> <span class="s1">&#39;estimate&#39;</span><span class="p">,</span> <span class="s1">&#39;conf.int.1&#39;</span><span class="p">,</span> <span class="s2">&quot;conf.int.2&quot;</span><span class="p">)</span>
<span class="n">dat</span><span class="p">[</span><span class="n">order</span><span class="p">(</span><span class="n">dat</span><span class="p">[,</span><span class="mi">2</span><span class="p">]),]</span>
</pre></div>

</div>
</div>
</div>

<div class="output_wrapper">
<div class="output">


<div class="output_area">

<div class="prompt"></div>



<div class="output_html rendered_html output_subarea ">
<table>
<thead><tr><th></th><th scope=col>annotation</th><th scope=col>p.value</th><th scope=col>estimate</th><th scope=col>conf.int.1</th><th scope=col>conf.int.2</th></tr></thead>
<tbody>
	<tr><th scope=row>5</th><td>E116-H3K36me3       </td><td>3.502158e-30        </td><td>-5.129676e-06       </td><td> 1.775306e-05       </td><td>-3.221700e-05       </td></tr>
	<tr><th scope=row>9</th><td>E116-H3K79me2       </td><td>4.270451e-13        </td><td>-6.049179e-07       </td><td> 7.960476e-06       </td><td>-2.022996e-05       </td></tr>
	<tr><th scope=row>6</th><td>E116-H3K4me1        </td><td>1.697103e-09        </td><td>-9.792211e-08       </td><td>-1.556027e-06       </td><td> 1.362075e-05       </td></tr>
	<tr><th scope=row>1</th><td>Coding_UCSC         </td><td>2.509293e-05        </td><td>-4.857236e-06       </td><td>-1.807982e-06       </td><td> 1.770077e-05       </td></tr>
	<tr><th scope=row>12</th><td>Intron_UCSC         </td><td>8.206769e-05        </td><td> 9.223260e-07       </td><td> 1.164579e-05       </td><td>-3.812980e-05       </td></tr>
	<tr><th scope=row>7</th><td>E116-H3K4me2        </td><td>1.346789e-02        </td><td>-9.790462e-07       </td><td> 1.863036e-05       </td><td>-2.333847e-05       </td></tr>
	<tr><th scope=row>11</th><td>E116-H4K20me1       </td><td>2.504146e-02        </td><td> 9.040980e-06       </td><td> 3.240370e-05       </td><td>-1.051296e-05       </td></tr>
	<tr><th scope=row>3</th><td>E116-H2A.Z          </td><td>3.381491e-02        </td><td>-2.721414e-06       </td><td>-3.348304e-05       </td><td>-3.768027e-05       </td></tr>
	<tr><th scope=row>14</th><td>Extended_splice_site</td><td>4.238859e-02        </td><td>-4.047743e-06       </td><td>-2.379580e-05       </td><td>-1.682468e-05       </td></tr>
	<tr><th scope=row>10</th><td>E116-H3K9ac         </td><td>5.446758e-02        </td><td>-5.920611e-07       </td><td>-1.564686e-05       </td><td>-4.512561e-05       </td></tr>
	<tr><th scope=row>8</th><td>E116-H3K4me3        </td><td>1.027078e-01        </td><td>-2.603563e-06       </td><td>-2.694487e-05       </td><td> 9.840231e-06       </td></tr>
	<tr><th scope=row>13</th><td>PolII_binding       </td><td>1.918467e-01        </td><td>-3.929391e-06       </td><td> 7.624660e-06       </td><td>-2.928285e-05       </td></tr>
	<tr><th scope=row>2</th><td>CTCF_binding        </td><td>2.040274e-01        </td><td>-3.818905e-06       </td><td> 2.232505e-06       </td><td> 2.335322e-05       </td></tr>
	<tr><th scope=row>4</th><td>E116-H3K27ac        </td><td>4.087071e-01        </td><td>-1.722333e-07       </td><td> 8.590494e-06       </td><td>-2.235855e-05       </td></tr>
</tbody>
</table>

</div>

</div>

</div>
</div>

</div>
<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div>
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p>Test for enrichment of annotation in CS.</p>

<pre><code>sos run analysis/20180712_Enrichment_Workflow.ipynb cs_fisher_test \
    --y-data ~/Documents/GIT/LargeFiles/JointLCL/fastqtl_qqnorm_ASintron_RNAseqGeuvadis_YangVCF.txt.gz \
    --trait AS -j 16 \
    --single-annot data/annotation.list</code></pre>

</div>
</div>
</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[&nbsp;]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-sos"><pre><span></span><span class="c1"># Test if CS is enriched with annotations</span>
<span class="p">[</span><span class="n">cs_fisher_test_2</span><span class="p">]</span>
<span class="kn">input:</span> <span class="kp">group_by</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="kp">concurrent</span> <span class="o">=</span> <span class="bp">True</span>
<span class="kn">output:</span> <span class="n">f</span><span class="s1">&#39;{_input:n}.cs_fisher_test.csv&#39;</span>
<span class="kn">R:</span> <span class="n">expand</span> <span class="o">=</span> <span class="s1">&#39;${ }&#39;</span><span class="p">,</span> <span class="n">container</span> <span class="o">=</span> <span class="s1">&#39;gaow/atac-gwas&#39;</span><span class="p">,</span> <span class="kp">workdir</span> <span class="o">=</span> <span class="n">cwd</span><span class="p">,</span> <span class="n">stdout</span> <span class="o">=</span> <span class="n">f</span><span class="s1">&#39;{_output:n}.stdout&#39;</span>
    <span class="n">run_test</span> <span class="o">=</span> <span class="n">function</span><span class="p">(</span><span class="n">dat</span><span class="p">)</span> <span class="p">{</span>
        <span class="n">d1</span> <span class="o">=</span> <span class="n">dat</span><span class="p">[</span><span class="n">which</span><span class="p">(</span><span class="n">dat</span><span class="err">$</span><span class="n">CS</span> <span class="o">&lt;</span> <span class="mi">2</span><span class="p">),]</span>
        <span class="n">d1</span> <span class="o">=</span> <span class="n">d1</span><span class="p">[</span><span class="o">-</span><span class="n">which</span><span class="p">(</span><span class="n">d1</span><span class="err">$</span><span class="n">CS</span> <span class="o">==</span> <span class="mi">1</span> <span class="o">&amp;</span> <span class="n">d1</span><span class="err">$</span><span class="n">PIP</span> <span class="o">&lt;</span> <span class="mf">0.05</span><span class="p">),</span> <span class="p">]</span>
        <span class="n">d1</span><span class="err">$</span><span class="n">PIP</span> <span class="o">=</span> <span class="n">NULL</span>
        <span class="n">d1</span> <span class="o">=</span> <span class="n">table</span><span class="p">(</span><span class="n">d1</span><span class="p">)</span>
        <span class="n">test</span><span class="o">.</span><span class="n">d1</span> <span class="o">=</span> <span class="n">fisher</span><span class="o">.</span><span class="n">test</span><span class="p">(</span><span class="n">d1</span><span class="p">)</span>
        <span class="c1"># test for non-first CS only</span>
        <span class="n">d2</span> <span class="o">=</span> <span class="n">dat</span><span class="p">[</span><span class="n">which</span><span class="p">(</span><span class="n">dat</span><span class="err">$</span><span class="n">CS</span> <span class="o">!=</span> <span class="mi">1</span><span class="p">),]</span>
        <span class="n">d2</span><span class="err">$</span><span class="n">CS</span><span class="p">[</span><span class="n">which</span><span class="p">(</span><span class="n">d2</span><span class="err">$</span><span class="n">CS</span><span class="o">&gt;</span><span class="mi">0</span><span class="p">)]</span> <span class="o">=</span> <span class="mi">1</span>
        <span class="n">d2</span> <span class="o">=</span> <span class="n">d2</span><span class="p">[</span><span class="o">-</span><span class="n">which</span><span class="p">(</span><span class="n">d2</span><span class="err">$</span><span class="n">CS</span> <span class="o">==</span> <span class="mi">1</span> <span class="o">&amp;</span> <span class="n">d2</span><span class="err">$</span><span class="n">PIP</span> <span class="o">&lt;</span> <span class="mf">0.05</span><span class="p">),</span> <span class="p">]</span>
        <span class="n">d2</span><span class="err">$</span><span class="n">PIP</span> <span class="o">=</span> <span class="n">NULL</span>
        <span class="n">d2</span> <span class="o">=</span> <span class="n">table</span><span class="p">(</span><span class="n">d2</span><span class="p">)</span>
        <span class="n">test</span><span class="o">.</span><span class="n">d2</span> <span class="o">=</span> <span class="n">fisher</span><span class="o">.</span><span class="n">test</span><span class="p">(</span><span class="n">d2</span><span class="p">)</span>
        <span class="k">return</span><span class="p">(</span><span class="n">c</span><span class="p">(</span><span class="n">test</span><span class="o">.</span><span class="n">d1</span><span class="err">$</span><span class="n">p</span><span class="o">.</span><span class="n">value</span><span class="p">,</span> <span class="n">test</span><span class="o">.</span><span class="n">d1</span><span class="err">$</span><span class="n">estimate</span><span class="p">,</span> <span class="n">test</span><span class="o">.</span><span class="n">d1</span><span class="err">$</span><span class="n">conf</span><span class="o">.</span><span class="n">int</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span> <span class="n">test</span><span class="o">.</span><span class="n">d1</span><span class="err">$</span><span class="n">conf</span><span class="o">.</span><span class="n">int</span><span class="p">[</span><span class="mi">2</span><span class="p">],</span> <span class="n">test</span><span class="o">.</span><span class="n">d2</span><span class="err">$</span><span class="n">p</span><span class="o">.</span><span class="n">value</span><span class="p">,</span> <span class="n">test</span><span class="o">.</span><span class="n">d2</span><span class="err">$</span><span class="n">estimate</span><span class="p">,</span> <span class="n">test</span><span class="o">.</span><span class="n">d2</span><span class="err">$</span><span class="n">conf</span><span class="o">.</span><span class="n">int</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span> <span class="n">test</span><span class="o">.</span><span class="n">d2</span><span class="err">$</span><span class="n">conf</span><span class="o">.</span><span class="n">int</span><span class="p">[</span><span class="mi">2</span><span class="p">]))</span>
    <span class="p">}</span>
    <span class="n">dat</span> <span class="o">=</span> <span class="n">readRDS</span><span class="p">(</span><span class="err">$</span><span class="p">{</span><span class="n">_input</span><span class="p">:</span><span class="n">r</span><span class="p">})</span>
    <span class="n">variants</span> <span class="o">=</span> <span class="n">dat</span><span class="err">$</span><span class="n">variants</span>
    <span class="kp">name</span> <span class="o">=</span> <span class="n">dat</span><span class="err">$</span><span class="kp">name</span>
    <span class="k">if</span> <span class="p">(</span><span class="n">length</span><span class="p">(</span><span class="n">unique</span><span class="p">(</span><span class="n">variants</span><span class="err">$</span><span class="n">GROUP</span><span class="p">))</span> <span class="o">==</span> <span class="mi">1</span><span class="p">)</span> <span class="p">{</span>
      <span class="n">write</span><span class="p">(</span><span class="n">paste</span><span class="p">(</span><span class="kp">name</span><span class="p">,</span> <span class="n">NA</span><span class="p">,</span> <span class="n">NA</span><span class="p">,</span> <span class="n">NA</span><span class="p">,</span> <span class="n">NA</span><span class="p">,</span> <span class="n">NA</span><span class="p">,</span> <span class="n">NA</span><span class="p">,</span> <span class="n">NA</span><span class="p">,</span> <span class="n">NA</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s1">&#39;,&#39;</span><span class="p">),</span> <span class="nb">file</span> <span class="o">=</span> <span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">:</span><span class="n">r</span><span class="p">})</span>
    <span class="p">}</span> <span class="k">else</span> <span class="p">{</span>
    <span class="c1"># test against all CS</span>
    <span class="n">test</span> <span class="o">=</span> <span class="n">run_test</span><span class="p">(</span><span class="n">variants</span><span class="p">[,</span><span class="n">c</span><span class="p">(</span><span class="s1">&#39;CS&#39;</span><span class="p">,</span> <span class="s1">&#39;GROUP&#39;</span><span class="p">,</span> <span class="s1">&#39;PIP&#39;</span><span class="p">)])</span>
    <span class="n">write</span><span class="p">(</span><span class="n">paste</span><span class="p">(</span><span class="n">c</span><span class="p">(</span><span class="kp">name</span><span class="p">,</span> <span class="n">test</span><span class="p">),</span> <span class="n">collapse</span><span class="o">=</span><span class="s1">&#39;,&#39;</span><span class="p">),</span> <span class="nb">file</span> <span class="o">=</span> <span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">:</span><span class="n">r</span><span class="p">})</span>
    <span class="p">}</span>
</pre></div>

</div>
</div>
</div>

</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[10]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-sos"><pre><span></span><span class="n">dat</span> <span class="o">=</span> <span class="n">read</span><span class="o">.</span><span class="n">table</span><span class="p">(</span><span class="s1">&#39;/home/gaow/Documents/GIT/LargeFiles/JointLCL/AS_output/fastqtl_qqnorm_ASintron_RNAseqGeuvadis_YangVCF_100Kb/enrichment/SuSiE_loci.sumstats.cs_fisher_test.csv&#39;</span><span class="p">,</span> <span class="n">head</span><span class="o">=</span><span class="n">F</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s1">&#39;,&#39;</span><span class="p">)</span>
<span class="n">colnames</span><span class="p">(</span><span class="n">dat</span><span class="p">)</span> <span class="o">=</span> <span class="n">c</span><span class="p">(</span><span class="s1">&#39;annotation&#39;</span><span class="p">,</span> <span class="s1">&#39;p.value.1&#39;</span><span class="p">,</span> <span class="s1">&#39;est.1&#39;</span><span class="p">,</span> <span class="s1">&#39;conf.1a&#39;</span><span class="p">,</span> <span class="s2">&quot;conf.1b&quot;</span><span class="p">,</span>  <span class="s1">&#39;p.value.2&#39;</span><span class="p">,</span> <span class="s1">&#39;est.2&#39;</span><span class="p">,</span> <span class="s1">&#39;conf.2a&#39;</span><span class="p">,</span> <span class="s2">&quot;conf.2b&quot;</span><span class="p">)</span>
<span class="n">dat</span><span class="p">[</span><span class="n">order</span><span class="p">(</span><span class="n">dat</span><span class="p">[,</span><span class="mi">2</span><span class="p">]),]</span>
</pre></div>

</div>
</div>
</div>

<div class="output_wrapper">
<div class="output">


<div class="output_area">

<div class="prompt"></div>



<div class="output_html rendered_html output_subarea ">
<table>
<thead><tr><th></th><th scope=col>annotation</th><th scope=col>p.value.1</th><th scope=col>est.1</th><th scope=col>conf.1a</th><th scope=col>conf.1b</th><th scope=col>p.value.2</th><th scope=col>est.2</th><th scope=col>conf.2a</th><th scope=col>conf.2b</th></tr></thead>
<tbody>
	<tr><th scope=row>9</th><td>E116-H3K79me2       </td><td>6.213563e-53        </td><td>1.7733509           </td><td>1.6528902           </td><td>1.9012944           </td><td>7.045409e-10        </td><td>2.5640731           </td><td>1.9118226           </td><td>3.406334            </td></tr>
	<tr><th scope=row>8</th><td>E116-H3K4me3        </td><td>6.082080e-41        </td><td>1.9288253           </td><td>1.7620249           </td><td>2.1082689           </td><td>2.513862e-07        </td><td>2.7630391           </td><td>1.9089417           </td><td>3.906909            </td></tr>
	<tr><th scope=row>5</th><td>E116-H3K36me3       </td><td>7.558364e-41        </td><td>1.7262809           </td><td>1.5989467           </td><td>1.8620315           </td><td>4.346591e-05        </td><td>2.0319146           </td><td>1.4497790           </td><td>2.798520            </td></tr>
	<tr><th scope=row>13</th><td>PolII_binding       </td><td>1.124630e-40        </td><td>2.5990105           </td><td>2.2890045           </td><td>2.9413682           </td><td>5.745633e-04        </td><td>2.8025486           </td><td>1.5425794           </td><td>4.723309            </td></tr>
	<tr><th scope=row>10</th><td>E116-H3K9ac         </td><td>2.042655e-39        </td><td>1.9337026           </td><td>1.7625323           </td><td>2.1180509           </td><td>1.216273e-04        </td><td>2.2532773           </td><td>1.4933752           </td><td>3.295533            </td></tr>
	<tr><th scope=row>7</th><td>E116-H3K4me2        </td><td>4.015520e-37        </td><td>1.6912262           </td><td>1.5648070           </td><td>1.8261439           </td><td>5.155783e-06        </td><td>2.1827886           </td><td>1.5664095           </td><td>2.992412            </td></tr>
	<tr><th scope=row>4</th><td>E116-H3K27ac        </td><td>4.990333e-36        </td><td>1.7684080           </td><td>1.6243937           </td><td>1.9227320           </td><td>1.353474e-04        </td><td>2.0899986           </td><td>1.4378677           </td><td>2.965523            </td></tr>
	<tr><th scope=row>1</th><td>Coding_UCSC         </td><td>2.689713e-16        </td><td>1.5378655           </td><td>1.3919312           </td><td>1.6959596           </td><td>6.343610e-02        </td><td>1.5414558           </td><td>0.9569762           </td><td>2.373234            </td></tr>
	<tr><th scope=row>14</th><td>Extended_splice_site</td><td>2.953173e-15        </td><td>2.8019658           </td><td>2.2146418           </td><td>3.5018781           </td><td>6.427788e-02        </td><td>2.7080663           </td><td>0.7312262           </td><td>7.034874            </td></tr>
	<tr><th scope=row>3</th><td>E116-H2A.Z          </td><td>1.499169e-11        </td><td>1.6450145           </td><td>1.4307537           </td><td>1.8834071           </td><td>4.404253e-01        </td><td>1.2854915           </td><td>0.5804513           </td><td>2.483132            </td></tr>
	<tr><th scope=row>12</th><td>Intron_UCSC         </td><td>6.510039e-09        </td><td>0.7177286           </td><td>0.6445246           </td><td>0.8011805           </td><td>7.800842e-01        </td><td>1.1424304           </td><td>0.6415253           </td><td>2.243416            </td></tr>
	<tr><th scope=row>6</th><td>E116-H3K4me1        </td><td>3.041316e-04        </td><td>1.2066611           </td><td>1.0900523           </td><td>1.3330231           </td><td>5.125559e-02        </td><td>1.5192093           </td><td>0.9785290           </td><td>2.272590            </td></tr>
	<tr><th scope=row>11</th><td>E116-H4K20me1       </td><td>1.216984e-03        </td><td>1.5920187           </td><td>1.1971216           </td><td>2.0781821           </td><td>3.465422e-02        </td><td>2.8434502           </td><td>0.9141362           </td><td>6.738671            </td></tr>
	<tr><th scope=row>2</th><td>CTCF_binding        </td><td>6.975631e-01        </td><td>1.0396775           </td><td>0.8297934           </td><td>1.2874168           </td><td>1.000000e+00        </td><td>0.9460932           </td><td>0.2556027           </td><td>2.455165            </td></tr>
</tbody>
</table>

</div>

</div>

</div>
</div>

</div>
<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div>
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h2 id="Check-for-overlap-between-signals-identifyed-by-Li-2016">Check for overlap between signals identifyed by Li 2016<a class="anchor-link" href="#Check-for-overlap-between-signals-identifyed-by-Li-2016">&#182;</a></h2><p>Data download from the <a href="http://eqtl.uchicago.edu/jointLCL/output_ASintron_PC3.txt">paper's companion data website</a>.</p>

</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div>
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">

<pre><code>sos run analysis/20180712_Enrichment_Workflow.ipynb signal_overlap \
    --y-data ~/Documents/GIT/LargeFiles/JointLCL/fastqtl_qqnorm_ASintron_RNAseqGeuvadis_YangVCF.txt.gz \
    --trait AS -j 16</code></pre>

</div>
</div>
</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[&nbsp;]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-sos"><pre><span></span><span class="p">[</span><span class="n">signal_overlap_1</span> <span class="p">(</span><span class="n">download</span> <span class="n">Li</span> <span class="mi">2016</span> <span class="n">results</span><span class="p">)]</span>
<span class="kn">output:</span> <span class="n">f</span><span class="s1">&#39;{cwd}/output_ASintron_PC3.txt&#39;</span>
<span class="kn">download:</span> <span class="n">dest_dir</span> <span class="o">=</span> <span class="n">f</span><span class="s2">&quot;{_output:d}&quot;</span>
    <span class="n">http</span><span class="p">:</span><span class="o">//</span><span class="n">eqtl</span><span class="o">.</span><span class="n">uchicago</span><span class="o">.</span><span class="n">edu</span><span class="o">/</span><span class="n">jointLCL</span><span class="o">/</span><span class="n">output_ASintron_PC3</span><span class="o">.</span><span class="n">txt</span>
</pre></div>

</div>
</div>
</div>

</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[&nbsp;]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-sos"><pre><span></span><span class="p">[</span><span class="n">signal_overlap_2</span><span class="p">]</span>
<span class="kn">depends:</span> <span class="n">z_score</span>
<span class="n">fdr</span> <span class="o">=</span> <span class="mf">0.01</span>
<span class="kn">python:</span> <span class="n">expand</span> <span class="o">=</span> <span class="s2">&quot;${ }&quot;</span>
    <span class="kn">from</span> <span class="nn">sos.utils</span> <span class="kn">import</span> <span class="n">get_output</span>
    <span class="n">susie</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="n">get_output</span><span class="p">(</span><span class="s2">&quot;zcat ${z_score} | cut -f2,3 -d&#39;_&#39; | sed &#39;s/_/./g&#39;&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">))</span>
    <span class="n">yang</span> <span class="o">=</span> <span class="n">get_output</span><span class="p">(</span><span class="s2">&quot;awk &#39;{if ($5 &lt; ${fdr}) print $1,$2}&#39; ${_input}&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)[</span><span class="mi">1</span><span class="p">:]</span>
    <span class="n">yang_dict</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">()</span>
    <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">yang</span><span class="p">:</span>
        <span class="n">line</span> <span class="o">=</span> <span class="n">line</span><span class="o">.</span><span class="n">split</span><span class="p">()</span>
        <span class="k">if</span> <span class="n">line</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">yang_dict</span><span class="p">:</span>
            <span class="n">yang_dict</span><span class="p">[</span><span class="n">line</span><span class="p">[</span><span class="mi">1</span><span class="p">]]</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">yang_dict</span><span class="p">[</span><span class="n">line</span><span class="p">[</span><span class="mi">1</span><span class="p">]]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">line</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
    <span class="n">intersects</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">key</span> <span class="ow">in</span> <span class="n">yang_dict</span><span class="p">:</span>
        <span class="n">common</span> <span class="o">=</span> <span class="n">susie</span><span class="o">.</span><span class="n">intersection</span><span class="p">(</span><span class="n">yang_dict</span><span class="p">[</span><span class="n">key</span><span class="p">])</span>
        <span class="n">intersects</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">common</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">)</span>
    <span class="k">print</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">intersects</span><span class="p">))</span>
    <span class="k">print</span><span class="p">(</span><span class="nb">sum</span><span class="p">(</span><span class="n">intersects</span><span class="p">))</span>
</pre></div>

</div>
</div>
</div>

</div>
<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div>
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p>1906 out of 2437 clusters match.</p>

</div>
</div>
</div>
<hr>
Copyright &copy 2016-2018 Gao Wang et al at Stephens Lab, University of Chicago
<p><small>Exported from <a href="http://github.com/gaow/mvarbvs/blob/dd86c4bd1cd1add3bc596cb6de916949f64b1053/analysis/20180712_Enrichment_Workflow.ipynb"><code>analysis/20180712_Enrichment_Workflow.ipynb</code></a> committed by Gao Wang on Tue Aug 14 19:43:08 2018 <a href="http://github.com/gaow/mvarbvs/commit/dd86c4bd1cd1add3bc596cb6de916949f64b1053">revision 9, dd86c4b</a> <a href="https://stephenslab.github.io/ipynb-website/notes.html#Note-about-commit-ids"><span class="fa fa-question-circle"></span></a></small></p>
</div>
</div>
</body>
</html>
