<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="ipynb_website:version" content="0.9.2" />
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link rel="stylesheet" type="text/css" href="../css/jt.css">
<link rel="stylesheet" type="text/css" href="../css/toc2.css">
<link href="../site_libs/jqueryui-1.11.4/jquery-ui.css">
<link rel="stylesheet" href="../site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<link rel="stylesheet" href="../site_libs/font-awesome-4.5.0/css/font-awesome.min.css" rel="stylesheet" />
<script type="text/javascript" src="https://ajax.googleapis.com/ajax/libs/jquery/1.9.1/jquery.min.js"></script>
<script type="text/javascript" src="https://ajax.googleapis.com/ajax/libs/jqueryui/1.9.1/jquery-ui.min.js"></script>
<script src="../site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="../site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="../site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<link rel="stylesheet"
      href="../site_libs/highlight/textmate.css"
      type="text/css" />
<script src="../site_libs/highlight/highlight.js"></script>
<script type="text/javascript">
if (window.hljs && document.readyState && document.readyState === "complete") {
   window.setTimeout(function() {
      hljs.initHighlighting();
   }, 0);
}
</script>
<script src="../js/toc2.js"></script>
<script src="../js/docs.js"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.1/MathJax.js?config=TeX-AMS_HTML"></script>
<script>
    MathJax.Hub.Config({
        extensions: ["tex2jax.js"],
        jax: ["input/TeX", "output/HTML-CSS"],
        tex2jax: {
        inlineMath: [ ['$','$'], ["\\(","\\)"] ],
        displayMath: [ ['$$','$$'], ["\\[","\\]"] ],
        processEscapes: true
        },
        "HTML-CSS": {
            preferredFont: "TeX",
            availableFonts: ["TeX"],
            styles: {
                scale: 110,
                ".MathJax_Display": {
                    "font-size": "110%",
                }
            }
        }
    });
</script>
<script>
$( document ).ready(function(){
            var cfg={'threshold':3,     // depth of toc (number of levels)
             'number_sections': false,
             'toc_cell': false,          // useless here
             'toc_window_display': true, // display the toc window
             "toc_section_display": "block", // display toc contents in the window
             'sideBar':true,       // sidebar or floating window
             'navigate_menu':false       // navigation menu (only in liveNotebook -- do not change)
            }
            var st={};                  // some variables used in the script
            st.rendering_toc_cell = false;
            st.config_loaded = false;
            st.extension_initialized=false;
            st.nbcontainer_marginleft = $('#notebook-container').css('margin-left')
            st.nbcontainer_marginright = $('#notebook-container').css('margin-right')
            st.nbcontainer_width = $('#notebook-container').css('width')
            st.oldTocHeight = undefined
            st.cell_toc = undefined;
            st.toc_index=0;
            // fire the main function with these parameters
            table_of_contents(cfg, st);
            var file=analysisDict[$("h1:first").attr("id")];
            $("#toc-level0 a").css("color","#126dce");
            $('a[href="#'+$("h1:first").attr("id")+'"]').hide()
            var docs=analysisArray;
            var pos=analysisArray.indexOf(file);
            for (var a=pos;a>=0;a--){
                  var name=docs[a]
                  $('<li><a href="'+name+'.html"><font color="#073642"><b>'+name.replace(/_/g," ")+'</b></font></a></li>').insertBefore("#toc-level0 li:eq(0)");
            }
            $('a[href="'+file+'.html'+'"]').css("color","#126dce");
            for (var a=pos+1;a<docs.length;a++){
                  var name=docs[a]
                  $(".toc #toc-level0").append('<li><a href="'+name+'.html"><font color="#073642"><b>'+name.replace(/_/g," ")+'</b></font></a></li>');
            }
            $("#toc-header").hide();
    });
</script>
<script>
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');
  // mark it active
  menuAnchor.parent().addClass('active');
  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');
});
</script>
<div class="container-fluid main-container">
<!-- tabsets -->
<script src="../site_libs/navigation-1.1/tabsets.js"></script>
<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});
</script><title>m&m ash</title>
<style type = "text/css">
body {
  font-family: "Droid Sans";
  padding-top: 66px;
  padding-bottom: 40px;
}
</style>
</head>
<body>
<div tabindex="-1" id="notebook" class="border-box-sizing">
<div class="container" id="notebook-container">
<!-- code folding -->
<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="../index.html">m&m ash</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">Overview</a>
</li>
<li>
  <a href="../analysis.html">Analysis</a>
</li>
<li>
  <a href="../pipeline.html">Pipeline</a>
</li>
<li>
  <a href="../writeup.html">Writeup</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
   <a href="http://github.com/gaow/mvarbvs"> <span class="fa fa-github"></span> </a>
</li>
</ul>
      </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->
<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div>
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h1 id="Simulation-of-quantitative-phenotype-given-genotypes">Simulation of quantitative phenotype given genotypes<a class="anchor-link" href="#Simulation-of-quantitative-phenotype-given-genotypes">&#182;</a></h1>
</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div>
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p>Here we simulate effect size from mixture gaussian distribution and match strong effects with "heavily LD convoluted" SNPs. See below for details.</p>
<h2 id="Core-functions">Core functions<a class="anchor-link" href="#Core-functions">&#182;</a></h2>
</div>
</div>
</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[3]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="o">%%</span><span class="k">writefile</span> src/SimUtils.py
import pandas as pd
import numpy as np
import os, random
def shuffle(df, axis=1):
        return df.apply(np.random.permutation, axis=axis)
class PhenotypeSimulator:
    def __init__(self, genotype_file):
        self.gfile = genotype_file
        self.phenotype = {}
        self.beta = {}
        self.pid = None
    def get_genes(self, limit = 5):
        res = pd.HDFStore(self.gfile).keys()
        if len(res) &gt; limit:
            res = res[:limit]
        return res
    def get_X(self, table):
        return pd.read_hdf(self.gfile, table)
    def permute_X(self, tables, save_to):
        if os.path.isfile(save_to):
            os.remove(save_to)
        for table in tables:
            X = pd.read_hdf(self.gfile, table)
            X = shuffle(X)
            X.to_hdf(save_to, table, mode = &#39;a&#39;, complevel = 9, complib = &#39;zlib&#39;)
        self.gfile = save_to
    def get_ld(self, tables, save_to = None):
        &#39;&#39;&#39;r^2 based LD calculation&#39;&#39;&#39;
        ld = {table: pd.read_hdf(self.gfile, table).transpose().corr(method = &#39;pearson&#39;) for table in tables}
        ld = {key: (np.power(value, 2) * np.sign(value)).astype(np.float16) for key, value in ld.items()}
        if save_to is not None:
            if os.path.isfile(save_to):
                os.remove(save_to)
            for key in ld:
                ld[key].to_hdf(save_to, key, mode = &#39;a&#39;, complevel = 9, complib = &#39;zlib&#39;)
        return ld
    def load_ld(self, tables, fn):
        ld = {}
        for table in tables:
            ld[table] = pd.read_hdf(fn, table)
        return ld
    def ld_heatmap(self, corrmat, out):
        import seaborn as sns
        import matplotlib.pyplot as plt
        fig, ax = plt.subplots()
        sns.heatmap(corrmat, ax = ax, vmin=-1, vmax=1, square=True, xticklabels = False, yticklabels = False)
        plt.savefig(out, dpi = 500)
    def generate_betamix(self, nbeta, mus, sigmas, pis, pi0 = 0):
        &#39;&#39;&#39;beta ~ \pi_0\delta_0 + \sum \pi_i N(0, sigma_i)
        sigma here is a nbeta list or nbeta * nbeta matrix
        &#39;&#39;&#39;
        if isinstance(sigmas, list):
            sigmas = np.diag(sigmas)
        assert (len(pis), len(pis)) == sigmas.shape
        masks = np.random.multinomial(1, pis, size = nbeta)
        mix = np.random.multivariate_normal(mus, sigmas, nbeta)
        return np.sum(mix * masks, axis = 1) * np.random.binomial(1, 1 - pi0, nbeta)
    def generate_y(self, X, beta, sigma, force = False):
        if self.pid in self.phenotype and force is not True:
            print(&#39;Name &quot;{}&quot; already exists. Use &quot;force = True&quot; to overwrite it&#39;.format(self.pid))
            return self.phenotype[self.pid]
        assert X.shape[0] == len(beta)
        self.beta[self.pid] = beta.tolist()
        beta.reshape((len(beta),1))
        y = np.dot(X.T, beta) + np.random.normal(0, sigma, X.shape[1])
        y.reshape(len(y), 1)
        y = pd.DataFrame(data = y, columns = [self.pid], index = X.columns).transpose()
        self.phenotype[self.pid] = y
        return y
    def select_convoluted_snps(self, ld, cutoff1 = 0.8, cutoff2 = 10, cutoff3 = 0.01):
        &#39;&#39;&#39;based on LD matrix select SNPs in strong LD with other SNPs 
        yet are independent between themselves&#39;&#39;&#39;
        print(&#39;Count strong LD&#39;)
        strong_ld_count = ((np.absolute(ld) &gt; cutoff1) * ld).sum(axis = 0).sort_values(ascending = False)
        strong_ld_count = strong_ld_count[strong_ld_count &gt; cutoff2]
        print(&#39;Filter by LD&#39;)
        exclude = []
        for x in strong_ld_count.index:
            if x in exclude:
                continue
            for y in strong_ld_count.index:
                if y in exclude or y == x:
                    continue
                if np.absolute(ld[x][y]) &gt; cutoff3:
                    exclude.append(y)
        print(&#39;Done&#39;)
        return [i for i, x in enumerate(strong_ld_count.index) if not x in exclude]
    def swap_beta(self, beta, strength_index):
        &#39;&#39;&#39;Set tops of beta to tops in strength_index&#39;&#39;&#39;
        nb = [0] * len(beta)
        beta = sorted(beta, key=abs, reverse=True)
        for item in strength_index:
            nb[item] = beta.pop(0)
        random.shuffle(beta)
        for idx in range(len(nb)):
            if not idx in strength_index:
                nb[idx] = beta.pop(0)
        assert len(beta) == 0
        return np.array(nb)
    def set_id(self, name):
        self.pid = name
class BetaDist:
    &#39;&#39;&#39;Reproducing simulated distributions of Stephens 2017 (ASH paper)&#39;&#39;&#39;
    def __init__(self):
        self.pi0 = 0
        self.pis = [None]
        self.mus = [None]
        self.sigmas = [None]
    def set_pi0(self, pi0):
        self.pi0 = pi0
    def set_spiky(self):
        self.pis = [0.4,0.2,0.2,0.2]
        self.mus = [0,0,0,0]
        self.sigmas = [0.25,0.5,1,2]
    def set_near_normal(self):
        self.pis = [2/3,1/3]
        self.mus = [0,0]
        self.sigmas = [1,2]
    def set_flat_top(self):
        self.pis = [1/7] * 7
        self.mus = [-1.5, -1, -.5 , 0, .5, 1, 1.5]
        self.sigmas = [0.5] * 7
    def set_skew(self):
        self.pis = [1/4,1/4,1/3,1/6]
        self.mus = [-2,-1,0,1]
        self.sigmas = [2,1.5,1,1]
    def set_big_normal(self):
        self.pis = [1]
        self.mus = [0]
        self.sigmas = [4]
    def set_bimodal(self):
        self.pis = [0.5, 0.5]
        self.mus = [-2, 2]
        self.sigmas = [1, 1]
    def __str__(self):
        params = &#39; + &#39;.join([&quot;{} N({}, {}^2)&quot;.format(x,y,z) for x, y, z in zip(self.pis, self.mus, self.sigmas)])
        return &#39;{:.3f} \delta_0 + {:.3f} [{}]&#39;.format(self.pi0, 1 - self.pi0, params)
</pre></div>
</div>
</div>
</div>
<div class="output_wrapper">
<div class="output">
<div class="hidden_output">
<div class="output_area">
<div class="prompt"></div>
<div class="output_subarea output_stream output_stdout output_text">
<pre>Overwriting src/SimUtils.py
</pre>
</div>
</div>
        </div>
</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div>
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h2 id="Load-data">Load data<a class="anchor-link" href="#Load-data">&#182;</a></h2>
</div>
</div>
</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[2]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="kn">from</span> <span class="nn">src.SimUtils</span> <span class="k">import</span> <span class="n">PhenotypeSimulator</span><span class="p">,</span> <span class="n">BetaDist</span>
<span class="n">ms</span> <span class="o">=</span> <span class="n">PhenotypeSimulator</span><span class="p">(</span><span class="s2">&quot;/home/gaow/Documents/GTEx/ToyExample/TY.genotype.h5&quot;</span><span class="p">)</span>
<span class="n">tables</span> <span class="o">=</span> <span class="n">ms</span><span class="o">.</span><span class="n">get_genes</span><span class="p">()</span>
</pre></div>
</div>
</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div>
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h2 id="Compute-and-save-LD">Compute and save LD<a class="anchor-link" href="#Compute-and-save-LD">&#182;</a></h2>
</div>
</div>
</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[3]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">ld</span> <span class="o">=</span> <span class="n">ms</span><span class="o">.</span><span class="n">get_ld</span><span class="p">(</span><span class="n">tables</span><span class="p">,</span> <span class="n">save_to</span> <span class="o">=</span> <span class="s2">&quot;/home/gaow/Documents/GTEx/ToyExample/TY.ld.h5&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div>
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h2 id="Putting-all-together">Putting all together<a class="anchor-link" href="#Putting-all-together">&#182;</a></h2><p>If you just want to get simulated data without knowing the details you can run the following code and find the output <code>/home/gaow/Documents/GTEx/ToyExample/TY.expr_simulated.h5</code>. Otherwise you should read on for more details.</p>
</div>
</div>
</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[5]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="kn">import</span> <span class="nn">os</span><span class="o">,</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">import</span> <span class="nn">json</span>
<span class="n">expr_file</span> <span class="o">=</span> <span class="s1">&#39;/home/gaow/Documents/GTEx/ToyExample/TY.expr_simulated.h5&#39;</span>
<span class="k">if</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">isfile</span><span class="p">(</span><span class="n">expr_file</span><span class="p">):</span>
    <span class="n">os</span><span class="o">.</span><span class="n">remove</span><span class="p">(</span><span class="n">expr_file</span><span class="p">)</span>
<span class="n">param</span> <span class="o">=</span> <span class="n">BetaDist</span><span class="p">()</span>
<span class="n">param</span><span class="o">.</span><span class="n">set_pi0</span><span class="p">(</span><span class="mf">0.998</span><span class="p">)</span>
<span class="c1"># Can use other settings</span>
<span class="n">param</span><span class="o">.</span><span class="n">set_spiky</span><span class="p">()</span>
<span class="nb">print</span><span class="p">(</span><span class="n">param</span><span class="p">)</span>
<span class="k">for</span> <span class="n">table</span> <span class="ow">in</span> <span class="n">tables</span><span class="p">:</span>
    <span class="n">ms</span><span class="o">.</span><span class="n">set_id</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">basename</span><span class="p">(</span><span class="n">table</span><span class="p">))</span>
    <span class="n">nbeta</span> <span class="o">=</span> <span class="n">ld</span><span class="p">[</span><span class="n">table</span><span class="p">]</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
    <span class="n">beta</span> <span class="o">=</span> <span class="n">ms</span><span class="o">.</span><span class="n">generate_betamix</span><span class="p">(</span><span class="n">nbeta</span><span class="o">=</span><span class="n">nbeta</span><span class="p">,</span><span class="n">pi0</span><span class="o">=</span><span class="n">param</span><span class="o">.</span><span class="n">pi0</span><span class="p">,</span> <span class="n">pis</span><span class="o">=</span><span class="n">param</span><span class="o">.</span><span class="n">pis</span><span class="p">,</span> <span class="n">mus</span> <span class="o">=</span> <span class="n">param</span><span class="o">.</span><span class="n">mus</span><span class="p">,</span> <span class="n">sigmas</span><span class="o">=</span><span class="n">param</span><span class="o">.</span><span class="n">sigmas</span><span class="p">)</span>
    <span class="n">strong_snps_idx</span> <span class="o">=</span> <span class="n">ms</span><span class="o">.</span><span class="n">select_convoluted_snps</span><span class="p">(</span><span class="n">ld</span><span class="p">[</span><span class="n">table</span><span class="p">])</span>
    <span class="n">beta</span> <span class="o">=</span> <span class="n">ms</span><span class="o">.</span><span class="n">swap_beta</span><span class="p">(</span><span class="n">beta</span><span class="p">,</span> <span class="n">strong_snps_idx</span><span class="p">)</span>
    <span class="n">X</span> <span class="o">=</span> <span class="n">ms</span><span class="o">.</span><span class="n">get_X</span><span class="p">(</span><span class="n">table</span><span class="o">=</span><span class="n">table</span><span class="p">)</span>
    <span class="n">y</span> <span class="o">=</span> <span class="n">ms</span><span class="o">.</span><span class="n">generate_y</span><span class="p">(</span><span class="n">beta</span><span class="o">=</span><span class="n">beta</span><span class="p">,</span><span class="n">sigma</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">X</span><span class="o">=</span><span class="n">X</span><span class="p">,</span> <span class="n">force</span> <span class="o">=</span> <span class="kc">True</span><span class="p">)</span>
<span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">(</span><span class="n">ms</span><span class="o">.</span><span class="n">phenotype</span><span class="o">.</span><span class="n">values</span><span class="p">())</span><span class="o">.</span><span class="n">to_hdf</span><span class="p">(</span><span class="n">expr_file</span><span class="p">,</span> <span class="s1">&#39;/simulated&#39;</span><span class="p">,</span> <span class="n">mode</span> <span class="o">=</span> <span class="s1">&#39;a&#39;</span><span class="p">,</span> <span class="n">complevel</span> <span class="o">=</span> <span class="mi">9</span><span class="p">,</span> <span class="n">complib</span> <span class="o">=</span> <span class="s1">&#39;zlib&#39;</span><span class="p">)</span>
<span class="n">meta</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;pi&#39;</span><span class="p">:</span> <span class="n">param</span><span class="o">.</span><span class="n">pis</span><span class="p">,</span> <span class="s1">&#39;pi0&#39;</span><span class="p">:</span> <span class="n">param</span><span class="o">.</span><span class="n">pi0</span><span class="p">,</span> <span class="s1">&#39;sigma&#39;</span><span class="p">:</span> <span class="n">param</span><span class="o">.</span><span class="n">sigmas</span><span class="p">,</span> <span class="s1">&#39;beta&#39;</span><span class="p">:</span> <span class="n">ms</span><span class="o">.</span><span class="n">beta</span><span class="p">}</span>
<span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="s2">&quot;/home/gaow/Documents/GTEx/ToyExample/TY.meta_simulation.json&quot;</span><span class="p">,</span> <span class="s1">&#39;w&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">fp</span><span class="p">:</span>
    <span class="n">json</span><span class="o">.</span><span class="n">dump</span><span class="p">(</span><span class="n">meta</span><span class="p">,</span> <span class="n">fp</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="output_wrapper">
<div class="output">
<div class="hidden_output">
<div class="output_area">
<div class="prompt"></div>
<div class="output_subarea output_stream output_stdout output_text">
<pre>0.998 \delta_0 + 0.0020000000000000018 [0.4 N(0, 0.25^2) + 0.2 N(0, 0.5^2) + 0.2 N(0, 1^2) + 0.2 N(0, 2^2)]
Count strong LD
Filter by LD
Done
Count strong LD
Filter by LD
Done
Count strong LD
Filter by LD
Done
</pre>
</div>
</div>
        </div>
</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div>
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h2 id="View-and-select-LD-structure">View and select LD structure<a class="anchor-link" href="#View-and-select-LD-structure">&#182;</a></h2><p>Take gene <code>ENSG00000264247</code> for example:</p>
</div>
</div>
</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[&nbsp;]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">ld</span><span class="p">[</span><span class="s1">&#39;/chr18/ENSG00000264247&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</pre></div>
</div>
</div>
</div>
</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[9]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">ms</span><span class="o">.</span><span class="n">ld_heatmap</span><span class="p">(</span><span class="n">ld</span><span class="p">[</span><span class="s1">&#39;/chr18/ENSG00000264247&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">iloc</span><span class="p">[:</span><span class="mi">1000</span><span class="p">,:</span><span class="mi">1000</span><span class="p">],</span> <span class="s1">&#39;img/ENSG00000264247.ld.png&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="output_wrapper">
<div class="output">
<div class="hidden_output">
<div class="output_area">
<div class="prompt"></div>
<div class="output_subarea output_stream output_stderr output_text">
<pre>/opt/miniconda3/lib/python3.6/site-packages/IPython/html.py:14: ShimWarning: The `IPython.html` package has been deprecated since IPython 4.0. You should import from `notebook` instead. `IPython.html.widgets` has moved to `ipywidgets`.
  &#34;`IPython.html.widgets` has moved to `ipywidgets`.&#34;, ShimWarning)
</pre>
</div>
</div>
        </div>
</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div>
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p><img src="img/ENSG00000264247.ld.png" alt=""></p>
</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div>
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h2 id="Simulating-effect-size">Simulating effect size<a class="anchor-link" href="#Simulating-effect-size">&#182;</a></h2><p>Now let's degress to effect size simulation. Effect size refers to $\beta$ in the linear model $ Y = X \beta + E$ where for simplicity we assume $E_{ij} \sim N(0,1)$. We sample $\beta$ from a mixture of gaussian distribution and a point mass at zero.</p>
<p>Here I start with a simple 3 components mixture under the alternative. 98% data will be the null (take that point mass at zero).</p>
</div>
</div>
</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[&nbsp;]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">nbeta</span> <span class="o">=</span> <span class="n">ld</span><span class="p">[</span><span class="s1">&#39;/chr18/ENSG00000264247&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
<span class="n">pis</span> <span class="o">=</span> <span class="p">[</span><span class="mf">0.25</span><span class="p">,</span> <span class="mf">0.3</span><span class="p">,</span> <span class="mf">0.45</span><span class="p">]</span>
<span class="n">pi0</span> <span class="o">=</span> <span class="mf">0.98</span>
<span class="n">sigmas</span> <span class="o">=</span> <span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mf">0.4</span><span class="p">,</span> <span class="mi">3</span><span class="p">]</span>
<span class="n">beta</span> <span class="o">=</span> <span class="n">ms</span><span class="o">.</span><span class="n">generate_betamix</span><span class="p">(</span><span class="n">nbeta</span><span class="o">=</span><span class="n">nbeta</span><span class="p">,</span><span class="n">pi0</span><span class="o">=</span><span class="n">pi0</span><span class="p">,</span><span class="n">pis</span><span class="o">=</span><span class="n">pis</span><span class="p">,</span><span class="n">sigmas</span><span class="o">=</span><span class="n">sigmas</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div>
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h2 id="Swap-big-effect-size-to-most-LD-convoluted-SNPs">Swap big effect size to most LD-convoluted SNPs<a class="anchor-link" href="#Swap-big-effect-size-to-most-LD-convoluted-SNPs">&#182;</a></h2><p>To better illustrate the potential of mr-ash I identify from genotype matrix potentially the more LD-convoluted SNPs and assign them the largest effect size. Specifically, I rank SNPs by their number of having LD with other SNPs greater than 0.9 (unsigned), and filter the top ranked SNPs until there is no strong LD between them ($r^2&lt;0.1$). Then I swap $\beta$s so that these SNPs have large effect size.</p>
</div>
</div>
</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[&nbsp;]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">strong_snps_idx</span> <span class="o">=</span> <span class="n">ms</span><span class="o">.</span><span class="n">select_convoluted_snps</span><span class="p">(</span><span class="n">ld</span><span class="p">[</span><span class="s1">&#39;/chr18/ENSG00000264247&#39;</span><span class="p">])</span>
</pre></div>
</div>
</div>
</div>
</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[&nbsp;]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">beta</span> <span class="o">=</span> <span class="n">ms</span><span class="o">.</span><span class="n">swap_beta</span><span class="p">(</span><span class="n">beta</span><span class="p">,</span> <span class="n">strong_snps_idx</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div>
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h2 id="Simulate-phenotypes">Simulate phenotypes<a class="anchor-link" href="#Simulate-phenotypes">&#182;</a></h2>
</div>
</div>
</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[&nbsp;]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">X</span> <span class="o">=</span> <span class="n">ms</span><span class="o">.</span><span class="n">get_X</span><span class="p">(</span><span class="n">table</span><span class="o">=</span><span class="s1">&#39;/chr18/ENSG00000264247&#39;</span><span class="p">)</span>
<span class="n">y</span> <span class="o">=</span> <span class="n">ms</span><span class="o">.</span><span class="n">generate_y</span><span class="p">(</span><span class="n">beta</span><span class="o">=</span><span class="n">beta</span><span class="p">,</span><span class="n">sigma</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">X</span><span class="o">=</span><span class="n">X</span><span class="p">,</span> <span class="n">name</span> <span class="o">=</span> <span class="s1">&#39;ENSG00000264247&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[&nbsp;]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">y</span>
</pre></div>
</div>
</div>
</div>
</div>
<hr>
&copy 2016-2017 Gao Wang et al at Stephens Lab, University of Chicago
<p><small>Exported from <a href="http://github.com/gaow/mvarbvs/blob/d5a5350a22e6fdd2e700bdf8d77d3b779638a03f/analysis/MR_ASH_Simulation.ipynb"><code>analysis/MR_ASH_Simulation.ipynb</code></a> committed by Gao Wang on Thu Aug 24 16:53:58 2017 <a href="http://github.com/gaow/mvarbvs/commit/d5a5350a22e6fdd2e700bdf8d77d3b779638a03f">revision 232, d5a5350</a> <a href="https://stephenslab.github.io/ipynb-website/notes.html#Note-about-commit-ids"><span class="fa fa-question-circle"></span></a></small></p>
</div>
</div>
</body>
</html>
