<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="ipynb_website:version" content="0.9.2" />
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link rel="stylesheet" type="text/css" href="../css/jt.css">
<link rel="stylesheet" type="text/css" href="../css/toc2.css">
<link href="../site_libs/jqueryui-1.11.4/jquery-ui.css">
<link rel="stylesheet" href="../site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<link rel="stylesheet" href="../site_libs/font-awesome-4.5.0/css/font-awesome.min.css" rel="stylesheet" />
<script type="text/javascript" src="https://ajax.googleapis.com/ajax/libs/jquery/1.9.1/jquery.min.js"></script>
<script type="text/javascript" src="https://ajax.googleapis.com/ajax/libs/jqueryui/1.9.1/jquery-ui.min.js"></script>
<script src="../site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="../site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="../site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<link rel="stylesheet"
      href="../site_libs/highlight/textmate.css"
      type="text/css" />
<script src="../site_libs/highlight/highlight.js"></script>
<script type="text/javascript">
if (window.hljs && document.readyState && document.readyState === "complete") {
   window.setTimeout(function() {
      hljs.initHighlighting();
   }, 0);
}
</script>
<script src="../js/toc2.js"></script>
<script src="../js/docs.js"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.1/MathJax.js?config=TeX-AMS_HTML"></script>
<script>
    MathJax.Hub.Config({
        extensions: ["tex2jax.js"],
        jax: ["input/TeX", "output/HTML-CSS"],
        tex2jax: {
        inlineMath: [ ['$','$'], ["\\(","\\)"] ],
        displayMath: [ ['$$','$$'], ["\\[","\\]"] ],
        processEscapes: true
        },
        "HTML-CSS": {
            preferredFont: "TeX",
            availableFonts: ["TeX"],
            styles: {
                scale: 110,
                ".MathJax_Display": {
                    "font-size": "110%",
                }
            }
        }
    });
</script>
<script>
$( document ).ready(function(){
            var cfg={'threshold':3,     // depth of toc (number of levels)
             'number_sections': false,
             'toc_cell': false,          // useless here
             'toc_window_display': true, // display the toc window
             "toc_section_display": "block", // display toc contents in the window
             'sideBar':true,       // sidebar or floating window
             'navigate_menu':false       // navigation menu (only in liveNotebook -- do not change)
            }
            var st={};                  // some variables used in the script
            st.rendering_toc_cell = false;
            st.config_loaded = false;
            st.extension_initialized=false;
            st.nbcontainer_marginleft = $('#notebook-container').css('margin-left')
            st.nbcontainer_marginright = $('#notebook-container').css('margin-right')
            st.nbcontainer_width = $('#notebook-container').css('width')
            st.oldTocHeight = undefined
            st.cell_toc = undefined;
            st.toc_index=0;
            // fire the main function with these parameters
            table_of_contents(cfg, st);
            var file=analysisDict[$("h1:first").attr("id")];
            $("#toc-level0 a").css("color","#126dce");
            $('a[href="#'+$("h1:first").attr("id")+'"]').hide()
            var docs=analysisArray;
            var pos=analysisArray.indexOf(file);
            for (var a=pos;a>=0;a--){
                  var name=docs[a]
                  $('<li><a href="'+name+'.html"><font color="#073642"><b>'+name.replace(/_/g," ")+'</b></font></a></li>').insertBefore("#toc-level0 li:eq(0)");
            }
            $('a[href="'+file+'.html'+'"]').css("color","#126dce");
            for (var a=pos+1;a<docs.length;a++){
                  var name=docs[a]
                  $(".toc #toc-level0").append('<li><a href="'+name+'.html"><font color="#073642"><b>'+name.replace(/_/g," ")+'</b></font></a></li>');
            }
            $("#toc-header").hide();
    });
</script>
<script>
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');
  // mark it active
  menuAnchor.parent().addClass('active');
  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');
});
</script>
<div class="container-fluid main-container">
<!-- tabsets -->
<script src="../site_libs/navigation-1.1/tabsets.js"></script>
<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});
</script><style type="text/css">
table {
   padding: 0;
   border-collapse: collapse; }
thead {
    border-bottom-width: 1px;
    border-bottom-color: rgb(0,0,0);
    border-bottom-style: solid;
}
table tr {
   border: none;
   background-color: white;
   margin: 0;
   padding: 0; }
table tr:nth-child(2n) {
   background-color: #f8f8f8; }
table tr th {
   font-weight: bold;
   border: none;
   margin: 0;
   padding: 6px 13px; }
table tr td {
   border: none;
   margin: 0;
   padding: 6px 13px; }
table tr th :first-child, table tr td :first-child {
   margin-top: 0; }
table tr th :last-child, table tr td :last-child {
   margin-bottom: 0; }
.dataframe_container { max-height: 400px }
.dataframe_input {
    border: 1px solid #ddd;
    margin-bottom: 5px;
}
.rendered_html table {
  border: none;
}
.sos_hint {
  color: rgba(0,0,0,.4);
  font-family: monospace;
  display: none;
}
.output_stderr {
    display: none;
}
div.input {
    display: none;
}
.hidden_output {
    display: none;
}
.input_prompt {
    display: none;
}
.output_area .prompt {
    display: none;
}
.output_prompt {
    display: none;
}
#nextsteps {
   color: blue;
}
.scatterplot_by_rowname div.xAxis div.tickLabel {
    transform: translateY(15px) translateX(15px) rotate(45deg);
    -ms-transform: translateY(15px) translateX(15px) rotate(45deg);
    -moz-transform: translateY(15px) translateX(15px) rotate(45deg);
    -webkit-transform: translateY(15px) translateX(15px) rotate(45deg);
    -o-transform: translateY(15px) translateX(15px) rotate(45deg);
    /*rotation-point:50% 50%;*/
    /*rotation:270deg;*/
}
div.cell {
    padding: 0pt;
    border-width: 0pt;
}
.sos_dataframe td, .sos_dataframe th, .sos_dataframe tr {
    white-space: nowrap;
    border: none;
}
.sos_dataframe tr:hover {
    background-color: #e6f2ff;
}
.display_control_panel  {
    padding: 10pt;
    left: 5px;
    top: 5px;
    position: fixed;
    z-index: 1000;
}
.display_control_panel:hover {
    background: rgb(224, 234, 241);
}
.display_checkboxes {
    margin-top: 5pt;
}
.display_control_panel:hover .display_control {
    display: block;
    opacity: 100;
}
.display_control_panel .display_control {
    opacity: 0;
}.lan_SoS {}.lan_Python3 .input_prompt { background-color: #FFE771 !important }.lan_R .input_prompt { background-color: #DCDCDA !important }</style>
<script>
function filterDataFrame(id) {
    var input = document.getElementById("search_" + id);
    var filter = input.value.toUpperCase();
    var table = document.getElementById("dataframe_" + id);
    var tr = table.getElementsByTagName("tr");
    // Loop through all table rows, and hide those who don't match the search query
    for (var i = 1; i < tr.length; i++) {
        for (var j = 0; j < tr[i].cells.length; ++j) {
            var matched = false;
            if (tr[i].cells[j].innerHTML.toUpperCase().indexOf(filter) != -1) {
                tr[i].style.display = "";
                matched = true
                break;
            }
            if (!matched)
                tr[i].style.display = "none";
        } 
    }
}
function sortDataFrame(id, n, dtype) {
    var table = document.getElementById("dataframe_" + id);
    var tb = table.tBodies[0]; // use `<tbody>` to ignore `<thead>` and `<tfoot>` rows
    var tr = Array.prototype.slice.call(tb.rows, 0); // put rows into array
    if (dtype === 'numeric') {
        var fn = function(a, b) { 
            return parseFloat(a.cells[n].textContent) <= parseFloat(b.cells[n].textContent) ? -1 : 1;
        }
    } else {
        var fn = function(a, b) {
            var c = a.cells[n].textContent.trim().localeCompare(b.cells[n].textContent.trim()); 
            return c > 0 ? 1 : (c < 0 ? -1 : 0) }
    }
    var isSorted = function(array, fn) {
        if (array.length < 2)
            return 1;
        var direction = fn(array[0], array[1]); 
        for (var i = 1; i < array.length - 1; ++i) {
            var d = fn(array[i], array[i+1]);
            if (d == 0)
                continue;
            else if (direction == 0)
                direction = d;
            else if (direction != d)
                return 0;
            }
        return direction;
    }
    var sorted = isSorted(tr, fn);
    if (sorted == 1 || sorted == -1) {
        // if sorted already, reverse it
        for(var i = tr.length - 1; i >= 0; --i)
            tb.appendChild(tr[i]); // append each row in order
    } else {
        tr = tr.sort(fn);
        for(var i = 0; i < tr.length; ++i)
            tb.appendChild(tr[i]); // append each row in order
    }
}
function toggle_source() {
    var btn = document.getElementById("show_cells");
    if (btn.checked) {
        $('div.input').css('display', 'flex');
        $('.hidden_output').show();
        // this somehow does not work.
        $('div.cell').css('padding', '5pt').css('border-width', '1pt');
    } else {
        $('div.input').hide();
        $('.hidden_output').hide();
        $('div.cell').css('padding', '0pt').css('border-width', '0pt');
    }
}
function toggle_prompt() {
    var btn = document.getElementById("show_prompt");
    if (btn.checked) {
        $('.output_prompt').show();
        $('.input_prompt').show();
        $('.output_area .prompt').show();
    } else {
        $('.output_prompt').hide();
        $('.input_prompt').hide();
        $('.output_area .prompt').hide();
    }
}
function toggle_messages() {
    var btn = document.getElementById("show_messages");
    if (btn.checked) {
        $('.sos_hint').show();
        $('.output_stderr').show();
    } else {
        $('.output_stderr').hide();
        $('.sos_hint').hide();
    }
}
</script>
<div class='display_control_panel'>
    <div class="display_control">
Display content:<br>
<div class="display_checkboxes">
<input type="checkbox" id="show_cells" name="show_cells" onclick="toggle_source()">
<label for="show_cells">All cells</label>
<br>
<input type="checkbox" id="show_prompt" name="show_prompt" onclick="toggle_prompt()">
<label for="show_prompt">Prompt</label>
<br>
<input type="checkbox" id="show_messages" name="show_messages" onclick="toggle_messages()">
<label for="show_messages">Messages</label>
</div>
   </div>
</div>
</script><title>m&m ash</title>
<style type = "text/css">
body {
  font-family: "Droid Sans";
  padding-top: 66px;
  padding-bottom: 40px;
}
</style>
</head>
<body>
<div tabindex="-1" id="notebook" class="border-box-sizing">
<div class="container" id="notebook-container">
<!-- code folding -->
<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="../index.html">m&m ash</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">Overview</a>
</li>
<li>
  <a href="../analysis.html">Analysis</a>
</li>
<li>
  <a href="../pipeline.html">Pipeline</a>
</li>
<li>
  <a href="../writeup.html">Writeup</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
   <a href="http://github.com/gaow/mvarbvs"> <span class="fa fa-github"></span> </a>
</li>
</ul>
      </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->
<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div>
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h1 id="m&amp;m-ash-pipeline-execution-interface">m&amp;m ash pipeline execution interface<a class="anchor-link" href="#m&amp;m-ash-pipeline-execution-interface">&#182;</a></h1>
</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div>
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h2 id="Preparation-&amp;-preprocessing-GTEx-data">Preparation &amp; preprocessing GTEx data<a class="anchor-link" href="#Preparation-&amp;-preprocessing-GTEx-data">&#182;</a></h2><p>See <a href="../writeup/GTEx7_Analysis_Plan.html#Preprocessing">this page</a> and <a href="../writeup/Meetings.html#Project-meeting-20170518">this meeting note</a> for details.</p>
<p><strong>Update</strong>: on Aug 14 2017 updated eQTL pipeline for <a href="../img/eQTL_pipeline_AWG_call_081417.pdf">V7 &amp; 8 are released</a>. Major changes involve normalization procedure and PEER factor analysis.</p>
</div>
</div>
</div><div class="rendered lan_SoS">
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[2]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-sos"><pre><span></span><span class="o">%</span><span class="n">save</span> <span class="n">prep</span><span class="o">.</span><span class="n">sos</span> <span class="o">-</span><span class="n">f</span> <span class="o">-</span><span class="n">x</span>
<span class="c1">#!/usr/bin/env sos-runner</span>
<span class="c1">#fileformat=SOS1.0</span>
<span class="c1"># Usage:</span>
<span class="c1"># ./prep.sos download</span>
<span class="c1"># ./prep.sos </span>
<span class="c1">#%include ResourceManagement as RM</span>
<span class="c1">#%include Misc as MC</span>
<span class="c1">#%include DataWrestling as DW</span>
<span class="p">[</span><span class="k">global</span><span class="p">]</span>
<span class="c1">#</span>
<span class="c1"># Auxiliary steps</span>
<span class="c1">#</span>
<span class="p">[</span><span class="kp">download</span><span class="p">]</span>
<span class="c1"># Resource preparation</span>
<span class="kp">sos_run</span><span class="p">(</span><span class="s1">&#39;RM.plink&#39;</span><span class="p">,</span> <span class="kp">workdir</span> <span class="o">=</span> <span class="n">CONFIG</span><span class="p">[</span><span class="s1">&#39;wd&#39;</span><span class="p">])</span>
<span class="kp">sos_run</span><span class="p">(</span><span class="s1">&#39;RM.minimac3&#39;</span><span class="p">,</span> <span class="kp">workdir</span> <span class="o">=</span> <span class="n">CONFIG</span><span class="p">[</span><span class="s1">&#39;wd&#39;</span><span class="p">])</span>
<span class="kp">sos_run</span><span class="p">(</span><span class="s1">&#39;RM.vcftools&#39;</span><span class="p">,</span> <span class="kp">workdir</span> <span class="o">=</span> <span class="n">CONFIG</span><span class="p">[</span><span class="s1">&#39;wd&#39;</span><span class="p">])</span>
<span class="kp">sos_run</span><span class="p">(</span><span class="s1">&#39;RM.peer&#39;</span><span class="p">,</span> <span class="kp">workdir</span> <span class="o">=</span> <span class="n">CONFIG</span><span class="p">[</span><span class="s1">&#39;wd&#39;</span><span class="p">])</span>
<span class="kp">sos_run</span><span class="p">(</span><span class="s1">&#39;RM.king&#39;</span><span class="p">,</span> <span class="kp">workdir</span> <span class="o">=</span> <span class="n">CONFIG</span><span class="p">[</span><span class="s1">&#39;wd&#39;</span><span class="p">])</span>
<span class="c1">#</span>
<span class="c1"># Workhorse</span>
<span class="c1">#</span>
<span class="p">[</span><span class="n">data_summary</span><span class="p">]</span>
<span class="kn">input:</span> <span class="n">CONFIG</span><span class="p">[</span><span class="s1">&#39;genotype&#39;</span><span class="p">]</span>
<span class="kp">sos_run</span><span class="p">(</span><span class="s2">&quot;MC.genotype_stats&quot;</span><span class="p">,</span> <span class="kp">workdir</span> <span class="o">=</span> <span class="n">CONFIG</span><span class="p">[</span><span class="s1">&#39;wd&#39;</span><span class="p">])</span>
<span class="p">[</span><span class="n">genotype_preprocessing</span><span class="p">]</span>
<span class="kn">input:</span> <span class="n">CONFIG</span><span class="p">[</span><span class="s1">&#39;genotype&#39;</span><span class="p">]</span>
<span class="kp">sos_run</span><span class="p">(</span><span class="s2">&quot;DW.vcf_by_chrom&quot;</span><span class="p">,</span> <span class="kp">workdir</span> <span class="o">=</span> <span class="n">CONFIG</span><span class="p">[</span><span class="s1">&#39;wd&#39;</span><span class="p">])</span>
<span class="p">[</span><span class="n">rna_preprocessing</span><span class="p">]</span>
<span class="kn">input:</span> <span class="n">CONFIG</span><span class="p">[</span><span class="s1">&#39;rna_rpkm&#39;</span><span class="p">],</span> <span class="n">CONFIG</span><span class="p">[</span><span class="s1">&#39;rna_cnts&#39;</span><span class="p">],</span> <span class="n">CONFIG</span><span class="p">[</span><span class="s1">&#39;genotype&#39;</span><span class="p">],</span> <span class="n">CONFIG</span><span class="p">[</span><span class="s1">&#39;sample_attr&#39;</span><span class="p">]</span> 
<span class="kp">sos_run</span><span class="p">(</span><span class="s2">&quot;MC.rnaseq&quot;</span><span class="p">,</span> <span class="kp">workdir</span> <span class="o">=</span>  <span class="n">CONFIG</span><span class="p">[</span><span class="s1">&#39;wd&#39;</span><span class="p">])</span>
<span class="p">[</span><span class="n">pca_plot_broad</span><span class="p">]</span>
<span class="kn">input:</span> <span class="s1">&#39;/tmp/GTExPCA.ped&#39;</span>
<span class="kp">sos_run</span><span class="p">(</span><span class="s2">&quot;MC.global_ancestry:2&quot;</span><span class="p">,</span> <span class="kp">workdir</span> <span class="o">=</span> <span class="s1">&#39;/tmp&#39;</span><span class="p">)</span>
<span class="p">[</span><span class="n">genotype_pca_broad</span><span class="p">]</span>
<span class="kn">parameter:</span> <span class="n">project_name</span> <span class="o">=</span> <span class="s2">&quot;GTEx7.dbGaP&quot;</span>
<span class="kn">input:</span> <span class="n">CONFIG</span><span class="p">[</span><span class="s1">&#39;genotype&#39;</span><span class="p">]</span>
<span class="kp">sos_run</span><span class="p">(</span><span class="s2">&quot;DW.vcf_by_chrom+DW.broad_to_plink+MC.LD_pruning+MC.global_ancestry&quot;</span><span class="p">,</span> 
        <span class="kp">workdir</span> <span class="o">=</span> <span class="n">CONFIG</span><span class="p">[</span><span class="s1">&#39;wd&#39;</span><span class="p">],</span>
        <span class="n">project_name</span> <span class="o">=</span> <span class="n">project_name</span><span class="p">)</span>
<span class="p">[</span><span class="n">genotype_pca_umich</span><span class="p">]</span>
<span class="kn">parameter:</span> <span class="n">project_name</span> <span class="o">=</span> <span class="s2">&quot;GTEx7.Imputed&quot;</span>
<span class="kn">input:</span> <span class="n">CONFIG</span><span class="p">[</span><span class="s1">&#39;imputed_genotype&#39;</span><span class="p">]</span>
<span class="kp">sos_run</span><span class="p">(</span><span class="s2">&quot;DW.umich_to_plink+MC.LD_pruning+MC.global_ancestry&quot;</span><span class="p">,</span> 
        <span class="kp">workdir</span> <span class="o">=</span> <span class="n">CONFIG</span><span class="p">[</span><span class="s1">&#39;wd&#39;</span><span class="p">],</span>
        <span class="n">project_name</span> <span class="o">=</span> <span class="n">project_name</span><span class="p">)</span>
<span class="p">[</span><span class="n">genotype_pca_umich_filtered</span><span class="p">]</span>
<span class="c1"># Filtered imputation data removing imputed sites</span>
<span class="kn">input:</span> <span class="s2">&quot;{}/GTEx7.Imputed.genotyped.filtered.bed&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">CONFIG</span><span class="p">[</span><span class="s1">&#39;wd&#39;</span><span class="p">])</span>
<span class="kp">sos_run</span><span class="p">(</span><span class="s2">&quot;MC.LD_pruning+MC.global_ancestry&quot;</span><span class="p">,</span> 
        <span class="kp">workdir</span> <span class="o">=</span> <span class="n">CONFIG</span><span class="p">[</span><span class="s1">&#39;wd&#39;</span><span class="p">])</span>
<span class="p">[</span><span class="n">gene_annotation</span><span class="p">:</span> <span class="kp">provides</span> <span class="o">=</span> <span class="s2">&quot;${CONFIG[&#39;rna_cnts&#39;]!n}.annotation&quot;</span><span class="p">]</span>
<span class="kn">input:</span> <span class="s2">&quot;${CONFIG[&#39;rna_cnts&#39;]}&quot;</span>
<span class="kn">output:</span> <span class="s2">&quot;${CONFIG[&#39;rna_cnts&#39;]!n}.annotation&quot;</span>
<span class="kp">sos_run</span><span class="p">(</span><span class="s2">&quot;MC.ensembl_annotation&quot;</span><span class="p">,</span> <span class="kp">workdir</span> <span class="o">=</span> <span class="n">CONFIG</span><span class="p">[</span><span class="s1">&#39;wd&#39;</span><span class="p">])</span>
<span class="p">[</span><span class="n">genotype_formatting</span><span class="p">]</span>
<span class="kn">parameter:</span> <span class="n">original_variants</span> <span class="o">=</span> <span class="s2">&quot;{}/GTEx7.dbGaP.bed&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">CONFIG</span><span class="p">[</span><span class="s1">&#39;wd&#39;</span><span class="p">])</span>
<span class="kn">parameter:</span> <span class="n">gene_annotation</span> <span class="o">=</span> <span class="s2">&quot;${CONFIG[&#39;rna_cnts&#39;]!n}.annotation&quot;</span>
<span class="kn">depends:</span> <span class="n">original_variants</span>
<span class="kn">input:</span> <span class="s2">&quot;{}/GTEx7.Imputed.bed&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">CONFIG</span><span class="p">[</span><span class="s1">&#39;wd&#39;</span><span class="p">])</span>
<span class="kp">sos_run</span><span class="p">(</span><span class="s2">&quot;DW.variants_filter+DW.plink_to_hdf5_batch&quot;</span><span class="p">,</span> 
        <span class="kp">workdir</span> <span class="o">=</span> <span class="n">CONFIG</span><span class="p">[</span><span class="s1">&#39;wd&#39;</span><span class="p">],</span> 
        <span class="n">include</span> <span class="o">=</span> <span class="n">original_variants</span><span class="p">,</span>
        <span class="n">ann</span> <span class="o">=</span> <span class="n">gene_annotation</span><span class="p">)</span>
<span class="p">[</span><span class="n">covariate_preparation</span><span class="p">]</span>
<span class="c1"># Covariates are: sex, platform, 3 PC and PEER factors</span>
<span class="kn">parameter:</span> <span class="n">peer_factors</span> <span class="o">=</span> <span class="n">glob</span><span class="o">.</span><span class="n">glob</span><span class="p">(</span><span class="s2">&quot;{}/*_PEER_covariates.txt&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="s2">&quot;${CONFIG[&#39;wd&#39;]!a}&quot;</span><span class="p">))</span>
<span class="kn">parameter:</span> <span class="n">pc_file</span> <span class="o">=</span> <span class="s2">&quot;{}/GTEx7.Imputed.prune.pc.ped&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">CONFIG</span><span class="p">[</span><span class="s1">&#39;wd&#39;</span><span class="p">])</span>
<span class="kn">parameter:</span> <span class="n">attr_file</span> <span class="o">=</span> <span class="n">CONFIG</span><span class="p">[</span><span class="s1">&#39;sample_attr&#39;</span><span class="p">]</span>
<span class="kn">parameter:</span> <span class="n">covar_file</span> <span class="o">=</span> <span class="n">CONFIG</span><span class="p">[</span><span class="s1">&#39;phenotype&#39;</span><span class="p">]</span>
<span class="kn">parameter:</span> <span class="n">expression_file</span> <span class="o">=</span> <span class="n">CONFIG</span><span class="p">[</span><span class="s1">&#39;expression_db&#39;</span><span class="p">]</span>
<span class="kp">sos_run</span><span class="p">(</span><span class="s2">&quot;DW.recode_platform + DW.covariates_to_HDF5&quot;</span><span class="p">,</span>
        <span class="kp">workdir</span> <span class="o">=</span> <span class="n">CONFIG</span><span class="p">[</span><span class="s1">&#39;wd&#39;</span><span class="p">],</span>
        <span class="n">peer_factors</span> <span class="o">=</span> <span class="n">peer_factors</span><span class="p">,</span>
        <span class="n">pc_file</span> <span class="o">=</span> <span class="n">pc_file</span><span class="p">,</span>
        <span class="n">attr_file</span> <span class="o">=</span> <span class="n">attr_file</span><span class="p">,</span>
        <span class="n">covar_file</span> <span class="o">=</span> <span class="n">covar_file</span><span class="p">,</span>
        <span class="n">output_file</span> <span class="o">=</span> <span class="s2">&quot;{}/GTEx7.Imputed.covariates.h5&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">CONFIG</span><span class="p">[</span><span class="s1">&#39;wd&#39;</span><span class="p">]))</span>
<span class="p">[</span><span class="n">make_toy</span><span class="p">]</span>
<span class="c1"># Create a toy example</span>
<span class="kp">sos_run</span><span class="p">(</span><span class="s2">&quot;DW.subset_HDF5_data&quot;</span><span class="p">,</span>
        <span class="kp">workdir</span> <span class="o">=</span> <span class="n">CONFIG</span><span class="p">[</span><span class="s1">&#39;wd&#39;</span><span class="p">],</span>
        <span class="n">ann_file</span> <span class="o">=</span> <span class="s2">&quot;${CONFIG[&#39;rna_cnts&#39;]!n}.annotation&quot;</span><span class="p">,</span>
        <span class="n">geno_file</span> <span class="o">=</span> <span class="s2">&quot;{}/GTEx7.Imputed.genotyped.filtered.cis.h5&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">CONFIG</span><span class="p">[</span><span class="s1">&#39;wd&#39;</span><span class="p">]),</span>
        <span class="n">expr_file</span> <span class="o">=</span> <span class="s2">&quot;{}/${CONFIG[&#39;rna_rpkm&#39;]!bnn}.qnorm.std.h5&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">CONFIG</span><span class="p">[</span><span class="s1">&#39;wd&#39;</span><span class="p">]),</span>
        <span class="n">toy_file</span> <span class="o">=</span> <span class="n">CONFIG</span><span class="p">[</span><span class="s1">&#39;toy_prefix&#39;</span><span class="p">],</span>
        <span class="n">gene_list</span> <span class="o">=</span> <span class="n">CONFIG</span><span class="p">[</span><span class="s1">&#39;toy_gene_list&#39;</span><span class="p">])</span>
</pre></div>
</div>
</div>
</div>
<div class="output_wrapper">
<div class="output">
<div class="hidden_output">
<div class="output_area">
<div class="prompt"></div>
<div class="output_html rendered_html output_subarea ">
<div class="sos_hint">Cell content saved to <a href="prep.sos" target="_blank">prep.sos</a></div>
</div>
</div>
        </div>
</div>
</div>
</div>
    </div>
<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div>
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h3 id="Prepare-computational-resource">Prepare computational resource<a class="anchor-link" href="#Prepare-computational-resource">&#182;</a></h3><p>This will download / install (most) software necessary for the analysis pipeline</p>
</div>
</div>
</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[&nbsp;]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-sos"><pre><span></span><span class="err">!</span><span class="o">./</span><span class="n">prep</span><span class="o">.</span><span class="n">sos</span> <span class="kp">download</span> <span class="o">-</span><span class="n">c</span> <span class="n">conf</span><span class="o">/</span><span class="mf">20170507.</span><span class="n">conf</span> <span class="o">-</span><span class="n">b</span> <span class="o">~/</span><span class="n">Documents</span><span class="o">/</span><span class="n">GTEx</span><span class="o">/</span><span class="nb">bin</span> <span class="o">-</span><span class="n">J</span> <span class="mi">8</span> <span class="o">-</span><span class="n">j</span> <span class="mi">1</span>
</pre></div>
</div>
</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div>
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h3 id="Data-summary">Data summary<a class="anchor-link" href="#Data-summary">&#182;</a></h3><p>Summary statistics on data, such as missingness statistics on genotypes. It also includes various diagnostic plots. More features will be added to this workflow as analysis develops.</p>
</div>
</div>
</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[&nbsp;]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-sos"><pre><span></span><span class="err">!</span><span class="o">./</span><span class="n">prep</span><span class="o">.</span><span class="n">sos</span> <span class="n">data_summary</span> <span class="o">-</span><span class="n">c</span> <span class="n">conf</span><span class="o">/</span><span class="mf">20170507.</span><span class="n">conf</span> <span class="o">-</span><span class="n">b</span> <span class="o">~/</span><span class="n">Documents</span><span class="o">/</span><span class="n">GTEx</span><span class="o">/</span><span class="nb">bin</span> <span class="o">-</span><span class="n">J</span> <span class="mi">8</span> <span class="o">-</span><span class="n">j</span> <span class="mi">1</span>
</pre></div>
</div>
</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div>
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h3 id="Genotype-imputation">Genotype imputation<a class="anchor-link" href="#Genotype-imputation">&#182;</a></h3>
</div>
</div>
</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[&nbsp;]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-sos"><pre><span></span><span class="err">!</span><span class="o">./</span><span class="n">prep</span><span class="o">.</span><span class="n">sos</span> <span class="n">genotype_preprocessing</span> <span class="o">-</span><span class="n">c</span> <span class="n">conf</span><span class="o">/</span><span class="mf">20170507.</span><span class="n">conf</span> <span class="o">-</span><span class="n">b</span> <span class="o">~/</span><span class="n">Documents</span><span class="o">/</span><span class="n">GTEx</span><span class="o">/</span><span class="nb">bin</span> <span class="o">-</span><span class="n">J</span> <span class="mi">8</span> <span class="o">-</span><span class="n">j</span> <span class="mi">1</span>
</pre></div>
</div>
</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div>
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p>The imputation step was done with <a href="https://imputationserver.sph.umich.edu">Michigan Imputation Server</a> because it uses Haplotype Reference Consortium (32,914 samples) reference panel which is not publicly available otherwise. <a href="https://imputationserver.sph.umich.edu/start.html#!pages/help">Here is how to prepare data</a> for this service. The prepared files are uploaded to <a href="https://imputationserver.sph.umich.edu">Michigan imputation server</a>. <a href="img/UMichImputation.png">Here is configuration</a> of imputation job on UMich server, and <a href="img/UMichImputationResult.pdf">here</a> is summary of the outcome.</p>
</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div>
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h3 id="RNA-seq-preprocessing">RNA-seq preprocessing<a class="anchor-link" href="#RNA-seq-preprocessing">&#182;</a></h3><p>This workflow includes data normalization and PEER factor analysis. It results in 4 <strong>analysis ready expression data files</strong> in HDF5 format of different versions / organizations of the same information: emperical quantile normalized and standard normal quantile normalized, saved as a flat file or grouped by tissues. This is so far the most computational intensive step</p>
</div>
</div>
</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[&nbsp;]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-sos"><pre><span></span><span class="err">!</span><span class="o">./</span><span class="n">prep</span><span class="o">.</span><span class="n">sos</span> <span class="n">rna_preprocessing</span> <span class="o">-</span><span class="n">c</span> <span class="n">conf</span><span class="o">/</span><span class="mf">20170507.</span><span class="n">conf</span> <span class="o">-</span><span class="n">b</span> <span class="o">~/</span><span class="n">Documents</span><span class="o">/</span><span class="n">GTEx</span><span class="o">/</span><span class="nb">bin</span> <span class="o">-</span><span class="n">J</span> <span class="mi">38</span> <span class="o">-</span><span class="n">j</span> <span class="mi">1</span>
</pre></div>
</div>
</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div>
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h3 id="Genotype-filtering-and-formatting,-followed-by-global-ancestry-analysis">Genotype filtering and formatting, followed by global ancestry analysis<a class="anchor-link" href="#Genotype-filtering-and-formatting,-followed-by-global-ancestry-analysis">&#182;</a></h3><p><code>genotype_pca_umich_filtered</code> has to be executed after both <code>genotype_pca_umich</code> and <code>genotype_pca_broad</code> are executed.</p>
<p>Genotypes are converted to PLINK binary format. During the conversion, imputed sites are removed and variant ID for tri-allelic sites are fixed. Also both the original and imputed data are converted for PCA analysis. The <a href="https://github.com/gaow/mvarbvs/issues/15#issuecomment-303814249">results are compared</a>.</p>
</div>
</div>
</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[&nbsp;]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-sos"><pre><span></span><span class="err">!</span><span class="o">./</span><span class="n">prep</span><span class="o">.</span><span class="n">sos</span> <span class="n">genotype_pca_umich</span> <span class="o">-</span><span class="n">c</span> <span class="n">conf</span><span class="o">/</span><span class="mf">20170507.</span><span class="n">conf</span> <span class="o">-</span><span class="n">b</span> <span class="o">~/</span><span class="n">Documents</span><span class="o">/</span><span class="n">GTEx</span><span class="o">/</span><span class="nb">bin</span><span class="o">/</span> <span class="o">-</span><span class="n">J</span> <span class="mi">4</span>
<span class="err">!</span><span class="o">./</span><span class="n">prep</span><span class="o">.</span><span class="n">sos</span> <span class="n">genotype_pca_broad</span> <span class="o">-</span><span class="n">c</span> <span class="n">conf</span><span class="o">/</span><span class="mf">20170507.</span><span class="n">conf</span> <span class="o">-</span><span class="n">b</span> <span class="o">~/</span><span class="n">Documents</span><span class="o">/</span><span class="n">GTEx</span><span class="o">/</span><span class="nb">bin</span><span class="o">/</span> <span class="o">-</span><span class="n">J</span> <span class="mi">4</span>
<span class="err">!</span><span class="o">./</span><span class="n">prep</span><span class="o">.</span><span class="n">sos</span> <span class="n">pca_plot_broad</span>
<span class="err">!</span><span class="o">./</span><span class="n">prep</span><span class="o">.</span><span class="n">sos</span> <span class="n">gene_annotation</span> <span class="o">-</span><span class="n">c</span> <span class="n">conf</span><span class="o">/</span><span class="mf">20170507.</span><span class="n">conf</span> <span class="o">-</span><span class="n">b</span> <span class="o">~/</span><span class="n">Documents</span><span class="o">/</span><span class="n">GTEx</span><span class="o">/</span><span class="nb">bin</span> <span class="o">-</span><span class="n">J</span> <span class="mi">6</span> <span class="o">-</span><span class="n">j</span> <span class="mi">1</span>
<span class="err">!</span><span class="o">./</span><span class="n">prep</span><span class="o">.</span><span class="n">sos</span> <span class="n">genotype_formatting</span> <span class="o">-</span><span class="n">c</span> <span class="n">conf</span><span class="o">/</span><span class="mf">20170507.</span><span class="n">conf</span> <span class="o">-</span><span class="n">b</span> <span class="o">~/</span><span class="n">Documents</span><span class="o">/</span><span class="n">GTEx</span><span class="o">/</span><span class="nb">bin</span> <span class="o">-</span><span class="n">J</span> <span class="mi">6</span> <span class="o">-</span><span class="n">j</span> <span class="mi">1</span>
<span class="err">!</span><span class="o">./</span><span class="n">prep</span><span class="o">.</span><span class="n">sos</span> <span class="n">genotype_pca_umich_filtered</span> <span class="o">-</span><span class="n">c</span> <span class="n">conf</span><span class="o">/</span><span class="mf">20170507.</span><span class="n">conf</span> <span class="o">-</span><span class="n">b</span> <span class="o">~/</span><span class="n">Documents</span><span class="o">/</span><span class="n">GTEx</span><span class="o">/</span><span class="nb">bin</span><span class="o">/</span> <span class="o">-</span><span class="n">J</span> <span class="mi">6</span>
</pre></div>
</div>
</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div>
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h3 id="Variants-annotation,-cis-SNP-selection-and-genotype-formatting">Variants annotation, cis-SNP selection and genotype formatting<a class="anchor-link" href="#Variants-annotation,-cis-SNP-selection-and-genotype-formatting">&#182;</a></h3><p>Genes are annotated to chromosomal positions, and annotate variants to genes. Then for each gene, variants are selected 2MB of a gene's TSS. It results a <strong>single analysis ready file</strong> in HDF5 format containing ~50K groups of genotype data (gene-names).</p>
</div>
</div>
</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[&nbsp;]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-sos"><pre><span></span><span class="err">!</span><span class="o">./</span><span class="n">prep</span><span class="o">.</span><span class="n">sos</span> <span class="n">gene_annotation</span> <span class="o">-</span><span class="n">c</span> <span class="n">conf</span><span class="o">/</span><span class="mf">20170507.</span><span class="n">conf</span> <span class="o">-</span><span class="n">b</span> <span class="o">~/</span><span class="n">Documents</span><span class="o">/</span><span class="n">GTEx</span><span class="o">/</span><span class="nb">bin</span> <span class="o">-</span><span class="n">J</span> <span class="mi">6</span> <span class="o">-</span><span class="n">j</span> <span class="mi">1</span>
<span class="err">!</span><span class="o">./</span><span class="n">prep</span><span class="o">.</span><span class="n">sos</span> <span class="n">genotype_formatting</span> <span class="o">-</span><span class="n">c</span> <span class="n">conf</span><span class="o">/</span><span class="mf">20170507.</span><span class="n">conf</span> <span class="o">-</span><span class="n">b</span> <span class="o">~/</span><span class="n">Documents</span><span class="o">/</span><span class="n">GTEx</span><span class="o">/</span><span class="nb">bin</span> <span class="o">-</span><span class="n">J</span> <span class="mi">6</span> <span class="o">-</span><span class="n">j</span> <span class="mi">1</span>
</pre></div>
</div>
</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div>
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h3 id="Merge-covariates-info">Merge covariates info<a class="anchor-link" href="#Merge-covariates-info">&#182;</a></h3><p>Covariates for analysis we've got so far include sample phenotypes (sex), sample attributes (genotyping platform), first principle components for population structure, and PEER factors. All saved in various files.</p>
<p>This workflow consolidates these files and generates a <strong>single analysis ready covariate file</strong> in HDF5 format.</p>
</div>
</div>
</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[&nbsp;]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-sos"><pre><span></span><span class="err">!</span><span class="o">./</span><span class="n">prep</span><span class="o">.</span><span class="n">sos</span> <span class="n">covariate_preparation</span> <span class="o">-</span><span class="n">c</span> <span class="n">conf</span><span class="o">/</span><span class="mf">20170507.</span><span class="n">conf</span> <span class="o">-</span><span class="n">b</span> <span class="o">~/</span><span class="n">Documents</span><span class="o">/</span><span class="n">GTEx</span><span class="o">/</span><span class="nb">bin</span><span class="o">/</span>
</pre></div>
</div>
</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div>
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h3 id="Generate-a-toy-data-set">Generate a toy data-set<a class="anchor-link" href="#Generate-a-toy-data-set">&#182;</a></h3><p>Finally, a toy data-set is created from the data bundle. This toy can be used for methods / pipeline development. Genes selected for the toy are the same as the <a href="https://stephenslab.github.io/gtexresults_mash/TwoSNP/2SNP.sos">LD show-case in the mash paper</a> (although the workflow itself takes an arbitary list of genes). See <a href="https://stephenslab.github.io/gtexresults_mash/TwoSNP/">this table</a> for motivation that these genes get selected.</p>
</div>
</div>
</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[&nbsp;]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-sos"><pre><span></span><span class="err">!</span><span class="o">./</span><span class="n">prep</span><span class="o">.</span><span class="n">sos</span> <span class="n">make_toy</span> <span class="o">-</span><span class="n">c</span> <span class="n">conf</span><span class="o">/</span><span class="mf">20170507.</span><span class="n">conf</span> <span class="o">-</span><span class="n">b</span> <span class="o">~/</span><span class="n">Documents</span><span class="o">/</span><span class="n">GTEx</span><span class="o">/</span><span class="nb">bin</span> <span class="o">-</span><span class="n">J</span> <span class="mi">6</span> <span class="o">-</span><span class="n">j</span> <span class="mi">1</span>
</pre></div>
</div>
</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div>
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h2 id="Simulations">Simulations<a class="anchor-link" href="#Simulations">&#182;</a></h2><p>Please see <a href="../analysis/MR-ASH-Simulation.html">this notebook</a> for interactive codes simulating expression data for given genotypes, and this <a href="../analysis/MR-ASH-Example.html">notebook</a> for a toy analysis. This pipeline is a more formal version of those exploratory analysis.</p>
</div>
</div>
</div><div class="rendered lan_SoS">
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[20]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-sos"><pre><span></span><span class="o">%</span><span class="n">save</span> <span class="n">simulation</span><span class="o">.</span><span class="n">sos</span> <span class="o">-</span><span class="n">f</span> <span class="o">-</span><span class="n">x</span>
<span class="c1">#!/usr/bin/env sos-runner</span>
<span class="c1">#fileformat=SOS1.0</span>
<span class="c1">#%include Misc as MC</span>
<span class="p">[</span><span class="n">phenotype_original_genotype</span><span class="p">]</span>
<span class="kp">sos_run</span><span class="p">(</span><span class="s1">&#39;MC.genotype_LD&#39;</span><span class="p">,</span> 
        <span class="n">genotype_data</span> <span class="o">=</span> <span class="n">CONFIG</span><span class="p">[</span><span class="s1">&#39;genotype&#39;</span><span class="p">],</span>
        <span class="kp">workdir</span> <span class="o">=</span> <span class="s2">&quot;${CONFIG[&#39;wd&#39;]!a}&quot;</span><span class="p">,</span>
        <span class="n">src</span> <span class="o">=</span> <span class="s2">&quot;${CONFIG[&#39;src&#39;]!a}&quot;</span><span class="p">)</span>
<span class="kp">sos_run</span><span class="p">(</span><span class="s1">&#39;MC.simulation&#39;</span><span class="p">,</span>
        <span class="n">genotype_data</span> <span class="o">=</span> <span class="n">CONFIG</span><span class="p">[</span><span class="s1">&#39;genotype&#39;</span><span class="p">],</span>
        <span class="kp">workdir</span> <span class="o">=</span> <span class="s2">&quot;${CONFIG[&#39;wd&#39;]!a}&quot;</span><span class="p">,</span>
        <span class="n">src</span> <span class="o">=</span> <span class="s2">&quot;${CONFIG[&#39;src&#39;]!a}&quot;</span><span class="p">,</span>
        <span class="n">pi0</span> <span class="o">=</span> <span class="n">CONFIG</span><span class="p">[</span><span class="s1">&#39;pi0&#39;</span><span class="p">],</span>
        <span class="n">shape</span> <span class="o">=</span> <span class="n">CONFIG</span><span class="p">[</span><span class="s1">&#39;shape&#39;</span><span class="p">],</span>
        <span class="n">n_rep</span> <span class="o">=</span> <span class="n">CONFIG</span><span class="p">[</span><span class="s1">&#39;n_rep&#39;</span><span class="p">])</span>
<span class="p">[</span><span class="n">phenotype_permuted_genotype</span><span class="p">]</span>
<span class="kp">sos_run</span><span class="p">(</span><span class="s1">&#39;MC.genotype_LD&#39;</span><span class="p">,</span>
        <span class="n">genotype_data</span> <span class="o">=</span> <span class="n">CONFIG</span><span class="p">[</span><span class="s1">&#39;genotype&#39;</span><span class="p">],</span>
        <span class="kp">workdir</span> <span class="o">=</span> <span class="s2">&quot;${CONFIG[&#39;wd&#39;]!a}&quot;</span><span class="p">,</span>
        <span class="n">src</span> <span class="o">=</span> <span class="s2">&quot;${CONFIG[&#39;src&#39;]!a}&quot;</span><span class="p">,</span>
        <span class="n">permute_genotype</span> <span class="o">=</span> <span class="s2">&quot;True&quot;</span><span class="p">)</span>
<span class="kp">sos_run</span><span class="p">(</span><span class="s1">&#39;MC.simulation&#39;</span><span class="p">,</span>
        <span class="n">genotype_data</span> <span class="o">=</span> <span class="n">CONFIG</span><span class="p">[</span><span class="s1">&#39;genotype&#39;</span><span class="p">],</span>
        <span class="kp">workdir</span> <span class="o">=</span> <span class="s2">&quot;${CONFIG[&#39;wd&#39;]!a}&quot;</span><span class="p">,</span>
        <span class="n">src</span> <span class="o">=</span> <span class="s2">&quot;${CONFIG[&#39;src&#39;]!a}&quot;</span><span class="p">,</span>
        <span class="n">permuted_genotype</span> <span class="o">=</span> <span class="s2">&quot;True&quot;</span><span class="p">,</span>
        <span class="n">pi0</span> <span class="o">=</span> <span class="n">CONFIG</span><span class="p">[</span><span class="s1">&#39;pi0&#39;</span><span class="p">],</span>
        <span class="n">shape</span> <span class="o">=</span> <span class="n">CONFIG</span><span class="p">[</span><span class="s1">&#39;shape&#39;</span><span class="p">],</span>
        <span class="n">n_rep</span> <span class="o">=</span> <span class="n">CONFIG</span><span class="p">[</span><span class="s1">&#39;n_rep&#39;</span><span class="p">])</span>
</pre></div>
</div>
</div>
</div>
<div class="output_wrapper">
<div class="output">
<div class="hidden_output">
<div class="output_area">
<div class="prompt"></div>
<div class="output_html rendered_html output_subarea ">
<div class="sos_hint">Cell content saved to <a href="simulation.sos" target="_blank">simulation.sos</a></div>
</div>
</div>
        </div>
</div>
</div>
</div>
    </div>
<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div>
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h3 id="Generate-phenotype-from-original-genotype">Generate phenotype from original genotype<a class="anchor-link" href="#Generate-phenotype-from-original-genotype">&#182;</a></h3><p>One replicate multiple scenarios</p>
<div class="highlight"><pre><span></span>./simulation.sos phenotype_original_genotype -J <span class="m">38</span> -c conf/simulate-20170630.conf
./simulation.sos phenotype_permuted_genotype -J <span class="m">38</span> -c conf/simulate-20170630.conf
</pre></div>
<p>One scenario multiple replicates</p>
<div class="highlight"><pre><span></span>./simulation.sos phenotype_original_genotype -J <span class="m">38</span> -c conf/simulate-20170630-reps.conf
./simulation.sos phenotype_permuted_genotype -J <span class="m">38</span> -c conf/simulate-20170630-reps.conf
</pre></div>
</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div>
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h2 id="Univariate-eQTL-analysis">Univariate eQTL analysis<a class="anchor-link" href="#Univariate-eQTL-analysis">&#182;</a></h2><h3 id="Single-SNP">Single SNP<a class="anchor-link" href="#Single-SNP">&#182;</a></h3><p>A <a href="http://www.bios.unc.edu/research/genomic_software/Matrix_eQTL/">fastEQTL</a> based pipeline.</p>
</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div>
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h3 id="Multi-SNP">Multi-SNP<a class="anchor-link" href="#Multi-SNP">&#182;</a></h3>
</div>
</div>
</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[&nbsp;]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-sos"><pre><span></span><span class="p">[</span><span class="n">mr_ash</span><span class="p">]</span>
<span class="kn">parameter:</span> <span class="n">cwd</span> <span class="o">=</span> <span class="s2">&quot;./&quot;</span>
<span class="kn">parameter:</span> <span class="n">genotype_file</span> <span class="o">=</span> <span class="s1">&#39;&#39;</span>
<span class="kn">parameter:</span> <span class="n">expr_file</span> <span class="o">=</span> <span class="s1">&#39;&#39;</span>
<span class="kn">parameter:</span> <span class="n">covar_file</span> <span class="o">=</span> <span class="s1">&#39;&#39;</span>
<span class="kn">parameter:</span> <span class="n">genes</span> <span class="o">=</span> <span class="p">[]</span>
<span class="kn">parameter:</span> <span class="n">tissues</span> <span class="o">=</span> <span class="p">[]</span>
<span class="kn">depends:</span> <span class="n">R_library</span><span class="p">(</span><span class="s1">&#39;rhdf5&#39;</span><span class="p">),</span> <span class="n">R_library</span><span class="p">(</span><span class="s1">&#39;tools&#39;</span><span class="p">)</span>
<span class="kn">input:</span> <span class="kp">for_each</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;genes&#39;</span><span class="p">,</span> <span class="s1">&#39;tissues&#39;</span><span class="p">]</span>
<span class="kn">output:</span> <span class="s2">&quot;${cwd!a}/${_tissues}_${_genes!b}.rds&quot;</span>
<span class="kn">task:</span>
<span class="kn">R:</span>
<span class="n">load_data</span> <span class="o">=</span> <span class="n">function</span><span class="p">(</span><span class="n">genotype_file</span><span class="p">,</span> <span class="n">expr_file</span><span class="p">,</span> <span class="n">cov_file</span><span class="p">,</span> <span class="n">geno_table</span><span class="p">,</span> <span class="n">expr_table</span><span class="p">,</span><span class="n">cov_table</span><span class="p">)</span> <span class="p">{</span>
  <span class="n">geno</span> <span class="o">=</span> <span class="n">h5read</span><span class="p">(</span><span class="n">genotype_file</span><span class="p">,</span> <span class="n">geno_table</span><span class="p">)</span>
  <span class="n">gdata</span> <span class="o">=</span> <span class="n">geno</span><span class="err">$</span><span class="n">block0_values</span>
  <span class="n">colnames</span><span class="p">(</span><span class="n">gdata</span><span class="p">)</span> <span class="o">=</span> <span class="n">geno</span><span class="err">$</span><span class="n">axis1</span>
  <span class="n">rownames</span><span class="p">(</span><span class="n">gdata</span><span class="p">)</span> <span class="o">=</span> <span class="n">geno</span><span class="err">$</span><span class="n">axis0</span>
  <span class="n">expr</span> <span class="o">=</span> <span class="n">h5read</span><span class="p">(</span><span class="n">expr_file</span><span class="p">,</span> <span class="n">expr_table</span><span class="p">)</span>
  <span class="n">edata</span> <span class="o">=</span> <span class="n">expr</span><span class="err">$</span><span class="n">block0_values</span>
  <span class="c1"># colnames(edata) = expr$axis1</span>
  <span class="n">colnames</span><span class="p">(</span><span class="n">edata</span><span class="p">)</span> <span class="o">=</span> <span class="n">tools</span><span class="p">::</span><span class="n">file_path_sans_ext</span><span class="p">(</span><span class="n">expr</span><span class="err">$</span><span class="n">axis1</span><span class="p">)</span>
  <span class="c1"># rownames(edata) = expr$axis0</span>
  <span class="n">rownames</span><span class="p">(</span><span class="n">edata</span><span class="p">)</span> <span class="o">=</span> <span class="nb">apply</span><span class="p">(</span><span class="n">sapply</span><span class="p">(</span><span class="n">strsplit</span><span class="p">(</span><span class="n">expr</span><span class="err">$</span><span class="n">axis0</span><span class="p">,</span><span class="s2">&quot;-&quot;</span><span class="p">),</span> <span class="sb">`[`</span><span class="p">,</span> <span class="n">c</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span><span class="mi">2</span><span class="p">)),</span> <span class="mi">2</span><span class="p">,</span> <span class="n">function</span><span class="p">(</span><span class="n">x</span><span class="p">)</span> <span class="n">paste</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">collapse</span> <span class="o">=</span> <span class="s1">&#39;-&#39;</span><span class="p">))</span>
  <span class="n">index</span> <span class="o">=</span> <span class="n">which</span><span class="p">(</span><span class="n">duplicated</span><span class="p">(</span><span class="n">row</span><span class="o">.</span><span class="n">names</span><span class="p">(</span><span class="n">edata</span><span class="p">)))</span>
  <span class="n">edata</span> <span class="o">=</span> <span class="n">edata</span><span class="p">[</span><span class="o">-</span><span class="n">index</span><span class="p">,]</span>
  <span class="c1"># edata = data.frame(edata)</span>
  <span class="n">covariate</span> <span class="o">&lt;-</span> <span class="n">h5read</span><span class="p">(</span><span class="n">cov_file</span><span class="p">,</span> <span class="n">cov_table</span><span class="p">)</span>
  <span class="n">cdata</span> <span class="o">=</span> <span class="n">covariate</span><span class="err">$</span><span class="n">block0_values</span>
  <span class="n">colnames</span><span class="p">(</span><span class="n">cdata</span><span class="p">)</span> <span class="o">=</span> <span class="nb">apply</span><span class="p">(</span><span class="n">sapply</span><span class="p">(</span><span class="n">strsplit</span><span class="p">(</span><span class="n">covariate</span><span class="err">$</span><span class="n">axis1</span><span class="p">,</span><span class="s2">&quot;-&quot;</span><span class="p">),</span> <span class="sb">`[`</span><span class="p">,</span> <span class="n">c</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span><span class="mi">2</span><span class="p">)),</span> <span class="mi">2</span><span class="p">,</span> <span class="n">function</span><span class="p">(</span><span class="n">x</span><span class="p">)</span> <span class="n">paste</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">collapse</span> <span class="o">=</span> <span class="s1">&#39;-&#39;</span><span class="p">))</span>
  <span class="c1"># rownames(edata) = expr$axis0</span>
  <span class="n">rownames</span><span class="p">(</span><span class="n">cdata</span><span class="p">)</span> <span class="o">=</span> <span class="n">covariate</span><span class="err">$</span><span class="n">axis0</span> 
  <span class="n">cdata</span> <span class="o">=</span> <span class="n">t</span><span class="p">(</span><span class="n">cdata</span><span class="p">)[</span><span class="o">-</span><span class="n">index</span><span class="p">,]</span>
  <span class="n">gdata</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">frame</span><span class="p">(</span><span class="n">gdata</span><span class="p">)</span>
  <span class="c1"># I want to use merge but later</span>
  <span class="c1"># index_overlap = which(row.names(gdata) %in% row.names(edata))</span>
  <span class="n">edata</span> <span class="o">=</span> <span class="n">edata</span><span class="p">[,</span> <span class="n">basename</span><span class="p">(</span><span class="n">geno_table</span><span class="p">)]</span>
  <span class="n">edata</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">frame</span><span class="p">(</span><span class="n">edata</span><span class="p">)</span>
  <span class="n">edata</span><span class="err">$</span><span class="n">ID</span> <span class="o">=</span> <span class="n">rownames</span><span class="p">(</span><span class="n">edata</span><span class="p">)</span>
  <span class="n">gdata</span><span class="err">$</span><span class="n">ID</span> <span class="o">=</span> <span class="n">rownames</span><span class="p">(</span><span class="n">gdata</span><span class="p">)</span>
  <span class="n">output</span> <span class="o">=</span> <span class="n">merge</span><span class="p">(</span><span class="n">x</span> <span class="o">=</span> <span class="n">edata</span><span class="p">,</span> <span class="n">y</span> <span class="o">=</span> <span class="n">gdata</span><span class="p">,</span> <span class="n">by</span> <span class="o">=</span> <span class="s2">&quot;ID&quot;</span><span class="p">,</span> <span class="nb">all</span><span class="o">.</span><span class="n">x</span> <span class="o">=</span> <span class="n">TRUE</span><span class="p">)</span>
  <span class="c1"># gdata = gdata[index_overlap,]</span>
  <span class="k">return</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="n">X</span><span class="o">=</span><span class="k">as</span><span class="o">.</span><span class="n">matrix</span><span class="p">(</span><span class="n">output</span><span class="p">[,</span><span class="o">-</span><span class="n">c</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span><span class="mi">2</span><span class="p">)]),</span> <span class="n">y</span> <span class="o">=</span> <span class="k">as</span><span class="o">.</span><span class="n">vector</span><span class="p">(</span><span class="n">output</span><span class="err">$</span><span class="n">edata</span><span class="p">),</span> <span class="n">Z</span> <span class="o">=</span> <span class="k">as</span><span class="o">.</span><span class="n">matrix</span><span class="p">(</span><span class="n">cdata</span><span class="p">)))</span>
<span class="p">}</span>
<span class="n">autoselect</span><span class="o">.</span><span class="n">mixsd</span> <span class="o">=</span> <span class="n">function</span><span class="p">(</span><span class="n">betahat</span><span class="p">,</span><span class="n">sebetahat</span><span class="p">,</span><span class="n">mult</span> <span class="o">=</span> <span class="n">sqrt</span><span class="p">(</span><span class="mi">2</span><span class="p">)){</span>
  <span class="n">sebetahat</span><span class="o">=</span><span class="n">sebetahat</span><span class="p">[</span><span class="n">sebetahat</span><span class="o">!=</span><span class="mi">0</span><span class="p">]</span> <span class="c1">#To avoid exact measure causing (usually by mistake)</span>
  <span class="n">sigmaamin</span> <span class="o">=</span> <span class="nb">min</span><span class="p">(</span><span class="n">sebetahat</span><span class="p">)</span><span class="o">/</span><span class="mi">10</span> <span class="c1">#so that the minimum is small compared with measurement precision</span>
  <span class="k">if</span><span class="p">(</span><span class="nb">all</span><span class="p">(</span><span class="n">betahat</span><span class="o">^</span><span class="mi">2</span><span class="o">&lt;=</span><span class="n">sebetahat</span><span class="o">^</span><span class="mi">2</span><span class="p">)){</span>
    <span class="n">sigmaamax</span> <span class="o">=</span> <span class="mi">8</span><span class="o">*</span><span class="n">sigmaamin</span> <span class="c1">#to deal with the occassional odd case where this could happen; 8 is arbitrary</span>
  <span class="p">}</span><span class="k">else</span><span class="p">{</span>
    <span class="n">sigmaamax</span> <span class="o">=</span> <span class="mi">2</span><span class="o">*</span><span class="n">sqrt</span><span class="p">(</span><span class="nb">max</span><span class="p">(</span><span class="n">betahat</span><span class="o">^</span><span class="mi">2</span><span class="o">-</span><span class="n">sebetahat</span><span class="o">^</span><span class="mi">2</span><span class="p">))</span> <span class="c1">#this computes a rough largest value you&#39;d want to use, based on idea that sigmaamax^2 + sebetahat^2 should be at least betahat^2   </span>
  <span class="p">}</span>
  <span class="k">if</span><span class="p">(</span><span class="n">mult</span><span class="o">==</span><span class="mi">0</span><span class="p">){</span>
    <span class="k">return</span><span class="p">(</span><span class="n">c</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="n">sigmaamax</span><span class="o">/</span><span class="mi">2</span><span class="p">))</span>
  <span class="p">}</span><span class="k">else</span><span class="p">{</span>
    <span class="n">npoint</span> <span class="o">=</span> <span class="n">ceiling</span><span class="p">(</span><span class="n">log2</span><span class="p">(</span><span class="n">sigmaamax</span><span class="o">/</span><span class="n">sigmaamin</span><span class="p">)</span><span class="o">/</span><span class="n">log2</span><span class="p">(</span><span class="n">mult</span><span class="p">))</span>
    <span class="k">return</span><span class="p">(</span><span class="n">mult</span><span class="o">^</span><span class="p">((</span><span class="o">-</span><span class="n">npoint</span><span class="p">):</span><span class="mi">0</span><span class="p">)</span> <span class="o">*</span> <span class="n">sigmaamax</span><span class="p">)</span>
  <span class="p">}</span>
<span class="p">}</span>
<span class="n">initial_step</span> <span class="o">=</span> <span class="n">function</span><span class="p">(</span><span class="n">X</span><span class="p">,</span><span class="n">y</span><span class="p">,</span><span class="n">Z</span> <span class="o">=</span> <span class="n">NULL</span><span class="p">){</span>
  <span class="n">P</span> <span class="o">=</span> <span class="n">dim</span><span class="p">(</span><span class="n">X</span><span class="p">)[</span><span class="mi">2</span><span class="p">]</span>
  <span class="n">output</span> <span class="o">=</span> <span class="n">matrix</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="n">nrow</span> <span class="o">=</span> <span class="n">P</span><span class="p">,</span><span class="n">ncol</span> <span class="o">=</span> <span class="mi">2</span><span class="p">)</span>
  <span class="k">for</span><span class="p">(</span><span class="n">i</span> <span class="ow">in</span> <span class="mi">1</span><span class="p">:</span><span class="n">P</span><span class="p">){</span>
    <span class="k">if</span><span class="p">(</span><span class="ow">is</span><span class="o">.</span><span class="n">null</span><span class="p">(</span><span class="n">Z</span><span class="p">)){</span>
      <span class="n">g</span> <span class="o">=</span> <span class="n">summary</span><span class="p">(</span><span class="n">lm</span><span class="p">(</span><span class="n">y</span><span class="o">~</span><span class="n">X</span><span class="p">[,</span><span class="n">i</span><span class="p">]))</span>
    <span class="p">}</span> <span class="k">else</span><span class="p">{</span>
      <span class="n">g</span> <span class="o">=</span> <span class="n">summary</span><span class="p">(</span><span class="n">lm</span><span class="p">(</span><span class="n">y</span><span class="o">~</span><span class="n">X</span><span class="p">[,</span><span class="n">i</span><span class="p">]</span><span class="o">+</span><span class="n">Z</span><span class="p">))</span>
    <span class="p">}</span>
    <span class="n">output</span><span class="p">[</span><span class="n">i</span><span class="p">,]</span> <span class="o">=</span> <span class="n">g</span><span class="err">$</span><span class="n">coefficients</span><span class="p">[</span><span class="mi">2</span><span class="p">,</span><span class="mi">1</span><span class="p">:</span><span class="mi">2</span><span class="p">]</span>
  <span class="p">}</span>
  <span class="k">return</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="n">betahat</span> <span class="o">=</span> <span class="n">output</span><span class="p">[,</span><span class="mi">1</span><span class="p">],</span><span class="n">sebetahat</span> <span class="o">=</span> <span class="n">output</span><span class="p">[,</span><span class="mi">2</span><span class="p">]))</span>
<span class="p">}</span>
<span class="n">analyze</span> <span class="o">=</span> <span class="n">function</span><span class="p">(</span><span class="n">genename</span> <span class="o">=</span> <span class="s1">&#39;/chr4/ENSG00000145214&#39;</span><span class="p">,</span> <span class="n">tissue</span> <span class="o">=</span> <span class="s1">&#39;/Lung&#39;</span><span class="p">,</span> <span class="n">out</span> <span class="o">=</span> <span class="s1">&#39;test.rds&#39;</span><span class="p">){</span>
  <span class="n">library</span><span class="p">(</span><span class="n">rhdf5</span><span class="p">)</span>
  <span class="n">genotype_file</span> <span class="o">=</span> <span class="err">$</span><span class="p">{</span><span class="n">genotype_file</span><span class="err">!</span><span class="n">ar</span><span class="p">}</span>
  <span class="n">expr_file</span> <span class="o">=</span> <span class="err">$</span><span class="p">{</span><span class="n">expr_file</span><span class="err">!</span><span class="n">ar</span><span class="p">}</span>
  <span class="n">geno_table</span> <span class="o">=</span> <span class="n">genename</span>
  <span class="n">expr_table</span> <span class="o">=</span> <span class="n">tissue</span> 
  <span class="n">gene</span> <span class="o">=</span> <span class="n">basename</span><span class="p">(</span><span class="n">geno_table</span><span class="p">)</span>
  <span class="n">cov_file</span> <span class="o">=</span> <span class="err">$</span><span class="p">{</span><span class="n">covar_file</span><span class="err">!</span><span class="n">ar</span><span class="p">}</span>
  <span class="n">cov_table</span> <span class="o">=</span> <span class="n">expr_table</span>
  <span class="n">dat</span> <span class="o">=</span> <span class="n">load_data</span><span class="p">(</span><span class="n">genotype_file</span> <span class="o">=</span> <span class="n">genotype_file</span><span class="p">,</span>
                  <span class="n">expr_file</span> <span class="o">=</span> <span class="n">expr_file</span><span class="p">,</span>
                  <span class="n">cov_file</span> <span class="o">=</span> <span class="n">cov_file</span><span class="p">,</span>
                  <span class="n">geno_table</span> <span class="o">=</span> <span class="n">geno_table</span><span class="p">,</span>
                  <span class="n">expr_table</span> <span class="o">=</span> <span class="n">expr_table</span><span class="p">,</span>
                  <span class="n">cov_table</span> <span class="o">=</span> <span class="n">cov_table</span><span class="p">)</span>
  <span class="n">X</span> <span class="o">=</span> <span class="k">as</span><span class="o">.</span><span class="n">matrix</span><span class="p">(</span><span class="n">dat</span><span class="err">$</span><span class="n">X</span><span class="p">)</span>
  <span class="n">X</span> <span class="o">=</span> <span class="n">X</span><span class="p">[,</span><span class="n">which</span><span class="p">(</span><span class="n">colSums</span><span class="p">(</span><span class="n">X</span><span class="p">)</span><span class="o">!=</span><span class="mi">0</span><span class="p">)]</span>
  <span class="k">if</span> <span class="p">((</span><span class="n">nrow</span><span class="p">(</span><span class="n">X</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">)</span> <span class="o">||</span> <span class="p">(</span><span class="n">ncol</span><span class="p">(</span><span class="n">X</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">))</span> <span class="p">{</span>
  	<span class="n">saveRDS</span><span class="p">(</span><span class="nb">list</span><span class="p">(),</span> <span class="n">out</span><span class="p">)</span>
  <span class="p">}</span> <span class="k">else</span> <span class="p">{</span>
  <span class="n">storage</span><span class="o">.</span><span class="n">mode</span><span class="p">(</span><span class="n">X</span><span class="p">)</span> <span class="o">&lt;-</span> <span class="s2">&quot;double&quot;</span>
  <span class="n">y</span> <span class="o">=</span> <span class="k">as</span><span class="o">.</span><span class="n">vector</span><span class="p">(</span><span class="n">dat</span><span class="err">$</span><span class="n">y</span><span class="p">)</span>
  <span class="n">Z</span> <span class="o">=</span> <span class="k">as</span><span class="o">.</span><span class="n">matrix</span><span class="p">(</span><span class="n">dat</span><span class="err">$</span><span class="n">Z</span><span class="p">)</span>
  <span class="n">initial</span> <span class="o">=</span> <span class="n">initial_step</span><span class="p">(</span><span class="n">X</span><span class="p">,</span><span class="n">y</span><span class="p">,</span><span class="n">Z</span><span class="p">)</span>
  <span class="n">mixsd</span> <span class="o">=</span> <span class="n">autoselect</span><span class="o">.</span><span class="n">mixsd</span><span class="p">(</span><span class="n">initial</span><span class="err">$</span><span class="n">betahat</span><span class="p">,</span><span class="n">initial</span><span class="err">$</span><span class="n">sebetahat</span><span class="p">)</span>
  <span class="n">logdata</span> <span class="o">=</span> <span class="n">capture</span><span class="o">.</span><span class="n">output</span><span class="p">({</span><span class="n">res</span> <span class="o">=</span> <span class="n">varbvs</span><span class="p">::</span><span class="n">varbvsmix</span><span class="p">(</span><span class="n">X</span><span class="p">,</span> <span class="n">Z</span><span class="p">,</span> <span class="n">y</span><span class="p">,</span> <span class="n">sa</span> <span class="o">=</span> <span class="n">c</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="n">mixsd</span><span class="o">^</span><span class="mi">2</span><span class="p">))</span> <span class="p">})</span>
  <span class="n">betahat</span> <span class="o">=</span> <span class="n">rowSums</span><span class="p">(</span><span class="n">res</span><span class="err">$</span><span class="n">alpha</span> <span class="o">*</span> <span class="n">res</span><span class="err">$</span><span class="n">mu</span><span class="p">)</span>
  <span class="n">names</span><span class="p">(</span><span class="n">betahat</span><span class="p">)</span> <span class="o">=</span> <span class="n">colnames</span><span class="p">(</span><span class="n">X</span><span class="p">)</span>
  <span class="n">mrash_out</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">betahat</span> <span class="o">=</span> <span class="n">betahat</span><span class="p">,</span> <span class="n">lfsr</span> <span class="o">=</span> <span class="n">res</span><span class="err">$</span><span class="n">lfsr</span><span class="p">)</span>
  <span class="n">ash_out</span> <span class="o">=</span> <span class="n">ashr</span><span class="p">::</span><span class="n">ash</span><span class="p">(</span><span class="n">initial</span><span class="err">$</span><span class="n">betahat</span><span class="p">,</span><span class="n">initial</span><span class="err">$</span><span class="n">sebetahat</span><span class="p">,</span><span class="n">mixcompdist</span> <span class="o">=</span> <span class="s2">&quot;normal&quot;</span><span class="p">)</span>
  <span class="n">saveRDS</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="n">ash</span> <span class="o">=</span> <span class="n">ash_out</span><span class="p">,</span> <span class="n">uni</span> <span class="o">=</span> <span class="n">initial</span><span class="p">,</span> <span class="n">mr_ash</span> <span class="o">=</span> <span class="n">mrash_out</span><span class="p">,</span> <span class="n">logdata</span> <span class="o">=</span> <span class="n">logdata</span><span class="p">),</span> <span class="n">out</span><span class="p">)</span>
<span class="p">}</span>
<span class="p">}</span>
<span class="n">analyze</span><span class="p">(</span><span class="n">genename</span> <span class="o">=</span> <span class="s2">&quot;/${_genes}&quot;</span><span class="p">,</span> <span class="n">tissue</span> <span class="o">=</span> <span class="s2">&quot;/${_tissues}&quot;</span><span class="p">,</span> <span class="n">out</span> <span class="o">=</span> <span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="err">!</span><span class="n">r</span><span class="p">})</span>
</pre></div>
</div>
</div>
</div>
</div>
<hr>
&copy 2016-2017 Gao Wang et al at Stephens Lab, University of Chicago
<p><small>Exported from <a href="http://github.com/gaow/mvarbvs/blob/d5a5350a22e6fdd2e700bdf8d77d3b779638a03f/analysis/utils.ipynb"><code>analysis/utils.ipynb</code></a> committed by Gao Wang on Thu Aug 24 16:53:58 2017 <a href="http://github.com/gaow/mvarbvs/commit/d5a5350a22e6fdd2e700bdf8d77d3b779638a03f">revision 232, d5a5350</a> <a href="https://stephenslab.github.io/ipynb-website/notes.html#Note-about-commit-ids"><span class="fa fa-question-circle"></span></a></small></p>
</div>
</div>
</body>
</html>
