<!DOCTYPE html>
<html>
<head><meta charset="utf-8" />
<title>GTEx7_Analysis_Plan</title><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.1.10/require.min.js"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/2.0.3/jquery.min.js"></script>








<!-- Custom stylesheet, it must be in the same directory as the html file -->
<link rel="stylesheet" href="custom.css">

<!-- Loading mathjax macro -->
<!-- Load mathjax -->
    <script src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS_HTML"></script>
    <!-- MathJax configuration -->
    <script type="text/x-mathjax-config">
    MathJax.Hub.Config({
        tex2jax: {
            inlineMath: [ ['$','$'], ["\\(","\\)"] ],
            displayMath: [ ['$$','$$'], ["\\[","\\]"] ],
            processEscapes: true,
            processEnvironments: true
        },
        // Center justify equations in code and markdown cells. Elsewhere
        // we use CSS to left justify single line equations in code cells.
        displayAlign: 'center',
        "HTML-CSS": {
            styles: {'.MathJax_Display': {"margin": 0}},
            linebreaks: { automatic: true }
        }
    });
    </script>
    <!-- End of mathjax configuration --></head>

 <link rel="stylesheet" href="https://code.jquery.com/ui/1.11.4/themes/smoothness/jquery-ui.css">

<script type="text/javascript" src="https://ajax.googleapis.com/ajax/libs/jquery/1.9.1/jquery.min.js"></script>
<script type="text/javascript" src="https://ajax.googleapis.com/ajax/libs/jqueryui/1.9.1/jquery-ui.min.js"></script>

<style>  /* defined here in case the main.css below cannot be loaded */
.lev1 {margin-left: 80px}
.lev2 {margin-left: 100px}
.lev3 {margin-left: 120px}
.lev4 {margin-left: 140px}
.lev5 {margin-left: 160px}
.lev6 {margin-left: 180px}
</style>


<link rel="stylesheet" type="text/css" href="../../css/jt.css">
<link rel="stylesheet" type="text/css" href="../../css/toc2.css">

<script src="../../js/toc2.js"></script>
<script src="../../js/docs.js"></script>

<script>
    MathJax.Hub.Config({
        extensions: ["tex2jax.js"],
        jax: ["input/TeX", "output/HTML-CSS"],
        tex2jax: {
        inlineMath: [ ['$','$'], ["\\(","\\)"] ],
        displayMath: [ ['$$','$$'], ["\\[","\\]"] ],
        processEscapes: true
        },
        "HTML-CSS": {
            preferredFont: "TeX",
            availableFonts: ["STIX","TeX"],
            styles: {
                scale: 110,
                ".MathJax_Display": {
                    "font-size": "110%",
                }
            }
        }
    });
</script>

<script>
$( document ).ready(function(){

            var cfg={'threshold':3,     // depth of toc (number of levels)
             'number_sections': false,  // sections numbering
             'toc_cell': false,          // useless here
             'toc_window_display': true, // display the toc window
             "toc_section_display": "block", // display toc contents in the window
             'sideBar':true,       // sidebar or floating window
             'navigate_menu':false       // navigation menu (only in liveNotebook -- do not change)
            }

            var st={};                  // some variables used in the script
            st.rendering_toc_cell = false;
            st.config_loaded = false;
            st.extension_initialized=false;
            st.nbcontainer_marginleft = $('#notebook-container').css('margin-left')
            st.nbcontainer_marginright = $('#notebook-container').css('margin-right')
            st.nbcontainer_width = $('#notebook-container').css('width')
            st.oldTocHeight = undefined
            st.cell_toc = undefined;
            st.toc_index=0;

            // fire the main function with these parameters


            table_of_contents(cfg,st);


            var file=writeupDict[$("h1:first").attr("id")];
            $("#toc-level0 a").css("color","#126dce");
            $('a[href="#'+$("h1:first").attr("id")+'"]').hide()
            
            var tuts=writeup;
            var pos=writeup.indexOf(file);
        
            for (var a=pos;a>=0;a--){
                  var name=tuts[a]
                  $('<li><a href="'+name+'.html">'+name.replace(/_/g," ")+'</a></li>').insertBefore("#toc-level0 li:eq(0)");
            }
            $('a[href="'+file+'.html'+'"]').css("color","#126dce");

            $('<li id="indexHome"><a href="../../writeup.html"><b>Writeup Home<b></a></li>').insertBefore("#toc-level0 li:eq(0)");
            for (var a=pos+1;a<tuts.length;a++){
                  var name=tuts[a]
                  $(".toc #toc-level0").append('<li><a href="'+name+'.html">'+name.replace(/_/g," ")+'</a></li>');
            }
            $("#toc-header").hide();


    });
</script>
<body>
  <div tabindex="-1" id="notebook" class="border-box-sizing">
    <div class="container" id="notebook-container">

<div class="cell border-box-sizing text_cell rendered">
<div class="prompt input_prompt">
</div>
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h1 id="GTEx-V7-eQTL-data-analysis-plan">GTEx V7 eQTL data analysis plan<a class="anchor-link" href="#GTEx-V7-eQTL-data-analysis-plan">&#182;</a></h1>
</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered">
<div class="prompt input_prompt">
</div>
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h2 id="Preprocessing">Preprocessing<a class="anchor-link" href="#Preprocessing">&#182;</a></h2>
</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered">
<div class="prompt input_prompt">
</div>
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h3 id="What-we-have">What we have<a class="anchor-link" href="#What-we-have">&#182;</a></h3><ul>
<li>Genotype data, filted and phased, yet not imputed</li>
<li>RNA-Seq data of gene expression for all sample tissues, available both the counts and RPKM</li>
<li>Sample phenotype / covariates</li>
</ul>
<h3 id="What-we-need-to-do">What we need to do<a class="anchor-link" href="#What-we-need-to-do">&#182;</a></h3><h4 id="Genotype-data">Genotype data<a class="anchor-link" href="#Genotype-data">&#182;</a></h4><ul>
<li>We need to impute missing genotypes. We will use <code>impute2</code> with reference data provided by the software (mostly 1000 genomes data).</li>
<li>Do some PLINK QC, including <ul>
<li>Keep variants with at least 10 samples having the minor allele</li>
</ul>
</li>
<li>Genotype samples have to match RNA-Seq samples by ID<ul>
<li>Renaming, via PLINK</li>
</ul>
</li>
</ul>
<h4 id="RNA-Seq-data">RNA-Seq data<a class="anchor-link" href="#RNA-Seq-data">&#182;</a></h4><ul>
<li>Genes should have at least 10 samples with RPKM &gt; 0.1 and read counts &gt; 6</li>
<li>Normalization VOOM + log(CPM) / gene_length, or simply log(RPKM)?<ul>
<li>Using quantile normalization would be fine for now (@mstephens. But why @mstephens perfers it? Need another discussion)</li>
</ul>
</li>
<li>Make them into HDF5</li>
</ul>
<h4 id="Covariates">Covariates<a class="anchor-link" href="#Covariates">&#182;</a></h4><ul>
<li>3 PCs, gender, genotyping platform, and PEER factors</li>
<li>PEER factors are generated using the top 1000 expressed genes per tissue, before normalization (or after??). <ul>
<li>Do it after normalization as V6 paper did</li>
</ul>
</li>
<li>Number of PEER depends on sample size N. <ul>
<li>$N < 150$, use 15 PEERs, $150 \le N < 250$, use 30 PEERs, $N \ge 250$ use 35 PEERs</li>
</ul>
</li>
<li>Regress out these factors separately for each tissue, save the residual to HDF5 as the Y for later analysis</li>
</ul>

</div>
</div>
</div>
    </div>
  </div>
</body>

 


</html>
