<!DOCTYPE html>
<html>
<head><meta charset="utf-8" />
<title>Meetings</title><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.1.10/require.min.js"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/2.0.3/jquery.min.js"></script>








<!-- Custom stylesheet, it must be in the same directory as the html file -->
<link rel="stylesheet" href="custom.css">

<!-- Loading mathjax macro -->
<!-- Load mathjax -->
    <script src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS_HTML"></script>
    <!-- MathJax configuration -->
    <script type="text/x-mathjax-config">
    MathJax.Hub.Config({
        tex2jax: {
            inlineMath: [ ['$','$'], ["\\(","\\)"] ],
            displayMath: [ ['$$','$$'], ["\\[","\\]"] ],
            processEscapes: true,
            processEnvironments: true
        },
        // Center justify equations in code and markdown cells. Elsewhere
        // we use CSS to left justify single line equations in code cells.
        displayAlign: 'center',
        "HTML-CSS": {
            styles: {'.MathJax_Display': {"margin": 0}},
            linebreaks: { automatic: true }
        }
    });
    </script>
    <!-- End of mathjax configuration --></head>

 <link rel="stylesheet" href="https://code.jquery.com/ui/1.11.4/themes/smoothness/jquery-ui.css">

<script type="text/javascript" src="https://ajax.googleapis.com/ajax/libs/jquery/1.9.1/jquery.min.js"></script>
<script type="text/javascript" src="https://ajax.googleapis.com/ajax/libs/jqueryui/1.9.1/jquery-ui.min.js"></script>

<style>  /* defined here in case the main.css below cannot be loaded */
.lev1 {margin-left: 80px}
.lev2 {margin-left: 100px}
.lev3 {margin-left: 120px}
.lev4 {margin-left: 140px}
.lev5 {margin-left: 160px}
.lev6 {margin-left: 180px}
</style>


<link rel="stylesheet" type="text/css" href="../../css/jt.css">
<link rel="stylesheet" type="text/css" href="../../css/toc2.css">

<script src="../../js/toc2.js"></script>
<script src="../../js/docs.js"></script>

<script>
    MathJax.Hub.Config({
        extensions: ["tex2jax.js"],
        jax: ["input/TeX", "output/HTML-CSS"],
        tex2jax: {
        inlineMath: [ ['$','$'], ["\\(","\\)"] ],
        displayMath: [ ['$$','$$'], ["\\[","\\]"] ],
        processEscapes: true
        },
        "HTML-CSS": {
            preferredFont: "TeX",
            availableFonts: ["STIX","TeX"],
            styles: {
                scale: 110,
                ".MathJax_Display": {
                    "font-size": "110%",
                }
            }
        }
    });
</script>

<script>
$( document ).ready(function(){

            var cfg={'threshold':3,     // depth of toc (number of levels)
             'number_sections': false,  // sections numbering
             'toc_cell': false,          // useless here
             'toc_window_display': true, // display the toc window
             "toc_section_display": "block", // display toc contents in the window
             'sideBar':true,       // sidebar or floating window
             'navigate_menu':false       // navigation menu (only in liveNotebook -- do not change)
            }

            var st={};                  // some variables used in the script
            st.rendering_toc_cell = false;
            st.config_loaded = false;
            st.extension_initialized=false;
            st.nbcontainer_marginleft = $('#notebook-container').css('margin-left')
            st.nbcontainer_marginright = $('#notebook-container').css('margin-right')
            st.nbcontainer_width = $('#notebook-container').css('width')
            st.oldTocHeight = undefined
            st.cell_toc = undefined;
            st.toc_index=0;

            // fire the main function with these parameters


            table_of_contents(cfg,st);


            var file=writeupDict[$("h1:first").attr("id")];
            $("#toc-level0 a").css("color","#126dce");
            $('a[href="#'+$("h1:first").attr("id")+'"]').hide()
            
            var tuts=writeup;
            var pos=writeup.indexOf(file);
        
            for (var a=pos;a>=0;a--){
                  var name=tuts[a]
                  $('<li><a href="'+name+'.html">'+name.replace(/_/g," ")+'</a></li>').insertBefore("#toc-level0 li:eq(0)");
            }
            $('a[href="'+file+'.html'+'"]').css("color","#126dce");

            $('<li id="indexHome"><a href="../../writeup.html"><b>Writeup Home<b></a></li>').insertBefore("#toc-level0 li:eq(0)");
            for (var a=pos+1;a<tuts.length;a++){
                  var name=tuts[a]
                  $(".toc #toc-level0").append('<li><a href="'+name+'.html">'+name.replace(/_/g," ")+'</a></li>');
            }
            $("#toc-header").hide();


    });
</script>
<body>
  <div tabindex="-1" id="notebook" class="border-box-sizing">
    <div class="container" id="notebook-container">

<div class="cell border-box-sizing text_cell rendered">
<div class="prompt input_prompt">
</div>
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h1 id="Meetings">Meetings<a class="anchor-link" href="#Meetings">&#182;</a></h1>
</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered">
<div class="prompt input_prompt">
</div>
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h2 id="Project-meeting-20170518">Project meeting 20170518<a class="anchor-link" href="#Project-meeting-20170518">&#182;</a></h2><p>With Matthew on issues with GTEx V7 preprocessing.</p>
<p>Questions:</p>
<ul>
<li>Normalization: why qnorm? should I standardize it? should I do per-tissue qnorm then standardize them?</li>
<li>PEER: should I do per-tissue or altogether? If altogether should I add tissue as a covariate? <ul>
<li>Variable number of PEER with GTEx V6 -- what's the deal with V7 now?</li>
</ul>
</li>
<li>Analysis: should I remove PEER + covariates first for each gene and then run mr-ash?</li>
<li>Any additional concerns with imputed genotypes?</li>
</ul>
<p>Feedback:</p>
<ul>
<li>We do quantile normalization because it is the most robust. And our model assumes normal.<ul>
<li>potential issue with RPKM: if some genes are extremely highly expressed they will impact other genes's RPKM</li>
<li>Potential issue with qnorm: may over-correct but the gain in power is well worth the possible loss</li>
</ul>
</li>
<li>No good answer to whether or not we adopt alternative stretagy. We should stick to one defensible strategy which easily is the <a href="https://gtexportal.org/home/documentationPage#staticTextAnalysisMethods">official GTEx guideline</a>.</li>
</ul>
<h2 id="Project-meeting-20170502">Project meeting 20170502<a class="anchor-link" href="#Project-meeting-20170502">&#182;</a></h2><p>With Matthew and Wei. We discussed mostly the fine-mapping step of m&amp;m, and some mr-ash related issues. First and foremost, we make it clear that focus of m&amp;m should be constrained to eQTL fine mapping for now because it is an important problem that has not been answered in the multivariate framework we propose.</p>
<p>Although we are interested in fine mapping eventually, it is perhaps too premature to make very concrete plan until we see the data analysis outcome from Step 2 the mash step (this comment was made in response to my initial request to finalize the fine mapping MCMC algorithm, and instead we brainstormed on what can be possibly done to perform fine mapping).</p>
<h3 id="How-to-summarize-fine-mapping-results?">How to summarize fine mapping results?<a class="anchor-link" href="#How-to-summarize-fine-mapping-results?">&#182;</a></h3><p>Sparse vs non-sparse result:</p>
<ul>
<li>If we adopt a none sparse model it is not good idea to evaluate whether $\beta=0$ -- we may want to look at correctness of the sign instead. </li>
<li>Result under a sparse model is potentially computationally easy. Example see Wen 2016 DAP paper. Or we can even look at one eQTL per gene model, or two eQTL per gene (thus a combination of ~2000 cis SNPs choose 2 pairs), at the fine-mapping step.</li>
</ul>
<p>What should be the output of fine mapping?</p>
<ul>
<li>One version is to output each SNP's LD with the eQTL</li>
<li>We can also output the 95% CI for the LD estimate with eQTL and see if it covers complete LD<ul>
<li>But which one is the eQTL?</li>
</ul>
</li>
<li>Or we can output the set of SNPs that we have 95% confidence that the set includes the eQTL (see Eskin)</li>
<li>We want to show that for given SNP, what is the nearest eQTL: what is the distance of this SNP to the nearest eQTL? what is the highest LD between this SNP and an eQTL?</li>
<li>Given MCMC result, how do we summarize it? <ul>
<li>What should we do if we want to use the results for data intergration? </li>
<li>We may want a mode, not a flat PIP from MCMC</li>
<li>We can take samples from posterior, then for each sample perform intergation analysis, and access the variance / sensitivity of results (how robust the conclusion is). And somehow combine the results. This is similar to the idea of multiple imputation.</li>
</ul>
</li>
</ul>
<p>What can we learn directly from Step 2 the mash step?</p>
<ul>
<li>If we go a bit further computing the lfsr and posterior mean of effects, we can get the "mode" candidate for eQTL, ie, we can then perform fine mapping around the mode. </li>
</ul>
<p>It is also perhaps too early to make any meaningful envision on what to do with finemapping, before looking into the data:</p>
<ul>
<li>How uncertain are the results we'll end up getting? Maybe LD is not a big issue in many of the new eQTLs we identify, which by itself would be exciting results.</li>
</ul>
<h3 id="Where-do-we-stop-for-mr-ash-application">Where do we stop for mr-ash application<a class="anchor-link" href="#Where-do-we-stop-for-mr-ash-application">&#182;</a></h3><p>We want to start the mr-ash paper by saying that there is great interest in introducing sparsity in regression, and recently there is a method called ash that introduces it in a "smart" way, and how it is relatively straightforward to use the ash idea in the context of regression.</p>
<p>The strength of mr-ash is the computational efficiency (VEM) and the flexibility (ASH compared to spike-slab), but there are disadvantages (PIP is too concentrated, see Carbonatto 2012).</p>
<p>In data application we can show how different the distribution of effect size of eQTL is, across genes. We can focus on a single tissue, fit separate mr-ash for each gene, and comment on interesting patterns that emerges; or we can use meta-analysis for multiple tissues on genes of interest, if there is not enough power from single tissue analysis.</p>
<h2 id="Project-meeting-20170417">Project meeting 20170417<a class="anchor-link" href="#Project-meeting-20170417">&#182;</a></h2><p>Meeting with Matthew. We went through the m&amp;m procedure on overleaf, revisited issue 8 on github and talked about next steps on data analysis + simulations. The discussion has led to minor changes in the overleaf write up.</p>
<p>Additionally we decide that the next step should be getting <em>Step 1</em> done, ie, mr-ash on GTEx data. We will start with analyzing GTEx V6 and verify with mash result, then move on to V7 data. <em>Step 1</em> would be an interesting application by itself as it is some form of univariate fine-mapping.</p>
<h2 id="Project-meeting-20170330">Project meeting 20170330<a class="anchor-link" href="#Project-meeting-20170330">&#182;</a></h2><p>Meeting with Matthew and Wei, to revive the project, by looking at what we have and what to be done.</p>
<h3 id="Tentative-agenda">Tentative agenda<a class="anchor-link" href="#Tentative-agenda">&#182;</a></h3><h4 id="Connected-work">Connected work<a class="anchor-link" href="#Connected-work">&#182;</a></h4><ul>
<li>varbvs</li>
<li>rss</li>
<li>ash</li>
<li>mash</li>
<li>mrash</li>
<li>BMASS</li>
</ul>
<h3 id="Questions">Questions<a class="anchor-link" href="#Questions">&#182;</a></h3><ul>
<li>How can m&amp;m ash generalize all theses work<ul>
<li>We have to think carefully what to incorperate in the generalized framework, and how to incorporate them</li>
<li>In particular how can we combine mrash / rss + mash?</li>
</ul>
</li>
<li>Do we start from summary statistics or full data?</li>
<li>Do we need MCMC in addition to VEM?</li>
</ul>
<p>Implementation-wise, shall we not write any code until we finalize on how the generalized framework is formulated? We should think "modularly" and we make contributions directly to other modules whenever possible, then build m&amp;m ash with these modules.</p>
<h3 id="What-to-do-with-mrash-as-a-standalone-work?">What to do with <code>mrash</code> as a standalone work?<a class="anchor-link" href="#What-to-do-with-mrash-as-a-standalone-work?">&#182;</a></h3><p>If we start from full data then finalizing <code>mrash</code> is a natural first step. It is then just a discussion of whether to create a separate package or to make it part of varbvs.</p>
<h3 id="Minutes">Minutes<a class="anchor-link" href="#Minutes">&#182;</a></h3><p>The meeting has outlined the approach we take towards a modularized m&amp;m. Most items on the agenda has been covered. See <a href="Modular_MNMASH.html">this document</a> for details.</p>
<h2 id="Project-meeting-20161103">Project meeting 20161103<a class="anchor-link" href="#Project-meeting-20161103">&#182;</a></h2><p>Meeting with Matthew. We started from recap on the motivation of project, then discussed the M&amp;M ASH model with practical considerations.</p>
<h3 id="Motivation">Motivation<a class="anchor-link" href="#Motivation">&#182;</a></h3><p>M&amp;M ASH model is motivated by what we have noticed in the MASH project. We have observed effect of a SNP (eQTL) positive in one tissue yet negative in another tissue. This bothers us. We suspect this type of observation is most likely due to negative LD between two causal SNPs both having positive effect in two separate tissues yet if we make the one eQTL per gene assumption as made in MASH we will observe opposite effects. So if we assume SNPs are independent in association analysis we obtain $\hat{\beta}$ convoluted by LD of all SNPs.</p>
<p>Let's consider univariate association analysis for a moment. Because of LD, $g(.)$, the distribution of $\beta$ we estimate via univariate methods, would have long tails. In other words $g(.)$ is inflated by LD with other SNPs. Estimates of $g(.)$ from multiple regression with ASH prior via variantional EM (currently called MVASH) will not have this problem. However when we want to make inference on $\beta$ the effect size, there will be identifiability issue with MVASH because VEM can reach local optima and the effect size it reports for the SNP identified may not be the SNP that in fact has an effect. The solution to this problem is to use MCMC for fine mapping on selected regions via VEM. A hybrid approach is to estimate hyper-parameters via VEM and use MCMC to sample the posterior.</p>
<p>Now to solve the same issue in the context of multivariate regression, we propose the M&amp;M ASH model, which applies multiple regression using ASH prior on multiple responses. David Gerard has derived a VEM procedure for the M&amp;M model. Assumptions in David's derivations are:</p>
<ul>
<li>The residual variance of genes (after regressing out eQTL effect) is structured low rank + diagonal</li>
<li>There will be missing data in the response matrix</li>
<li>The mixture proportion can be estimated per test, or be estimated jointly for all tests</li>
</ul>
<h3 id="M&amp;M-ASH-with-diagonal-residual-covariance-structure">M&amp;M ASH with diagonal residual covariance structure<a class="anchor-link" href="#M&amp;M-ASH-with-diagonal-residual-covariance-structure">&#182;</a></h3><p>Matthew suggests we make this model simpler and make sure it works. For starters we should ignore correlation among tissues. That is, we assume residual variance a diagonal matrix. Here are a few points why we should start with diagonal and why at least as a first pass we should not make non-diagonal assumption in M&amp;M ASH:</p>
<ul>
<li>We are not sure yet if correlated residual will cause a problem to our inference -- unless we can show it empirically: we should find real data examples when correlation between tissues are due to correlation between genes, not due to similarity of tissues. This would raise a red flag that we should model such correlations.</li>
<li>Even if the problem is confirmed we should use MASH model to show we can solve it, before incorporating the solution to M&amp;M ASH. As MASH model is simpler, it will get us assessment from real data quickly and we'll decide if it worth to pursue the fix in M&amp;M ASH.</li>
<li>To do it in MASH we should assume this residual correlation is the same as the tissues' correlations (eQTL effect is relatively small) and we estimate the 44 by 44 matrix of covariance directly from expression data. This is not a trivial problem; many methods get estimates that shrink the structure to diagonal. But sparse factor analysis methods can be a good technique to do this, as shown by Wei's work. We then plug this estimate to MASH model<ul>
<li>The advantage of this approach (over making inference jointly as what David has done for M&amp;M ASH) is that this approach is modular and we can choose a good method (such as SFA, FLASH) to make this step of inference. The method may be biased (ignoring impact of eQTL) but has better variance</li>
</ul>
</li>
<li>The problem with this approach is that if eQTL induces correlation we'll wrongfully believe there is residual covariance when in fact there is not. That is, after removing effect from eQTL the residual covariance is diagonal. This observation would favor the joint approach over the modular approach. To assess if this is a problem, we can choose genes with large covariance matrix, and remove the effect of top eQTL then see if the residual covariance matrix still retains correlations or is mostly diagonal.</li>
</ul>
<h3 id="Next-steps">Next steps<a class="anchor-link" href="#Next-steps">&#182;</a></h3><p>We should start with the simplest version (that residual covariance is diagonal) and make it work. The hard part is computation. Using summary data whenever possible may help with computation. Additionally in updating mixture components we can use noiser estimates, that is, estimates from randomly sampled \beta{hat} instead of 20K genes <em> 1000 SNPs </em> 50 conditions data points. We will have our next meeting (David and Gao with Matthew) after we get this simple version to work in practice.</p>
<h2 id="Project-meeting-20160921">Project meeting 20160921<a class="anchor-link" href="#Project-meeting-20160921">&#182;</a></h2><h3 id="Tentative-schedule">Tentative schedule<a class="anchor-link" href="#Tentative-schedule">&#182;</a></h3><p>Status</p>
<ul>
<li>The <code>m&amp;m ash</code> <a href="http://www.bioinformatics.org/labnotes/mnmash/mnmash-model.html">model</a> and <a href="https://github.com/gaow/mnmashr">implementation</a>.</li>
<li>Implementation is not working on data due to <a href="https://github.com/gaow/mnmashr/blob/master/src/mnmash.hpp#L41">data structure design</a><ul>
<li>When J is 40, P 2000, K 50 and L 20, the S and SI matrices will be of size 3.2G * 2 = 6.4G. Looping over such data is very slow.</li>
</ul>
</li>
<li>Correctness of implementation not tested</li>
</ul>
<p>Next steps</p>
<ul>
<li>Get implementation working</li>
<li>Run simulations</li>
<li>Put into OmicsBMA framework and do real data anlaysis</li>
</ul>
<h3 id="Minutes">Minutes<a class="anchor-link" href="#Minutes">&#182;</a></h3><ul>
<li>We should conceptually distinguish a model using original data from the one using summary data, though in the VB algorithm they are very similar.</li>
<li>Currently the model assumes $\Sigma_{J \times J}$ known. This is not easy to estimate because it would involve a non-trivial multiple regression. We should try to model $\Sigma_{J \times J}$ as unknown diagonal matrix and estimate it in the VB framework. For starters, write up the J = 1 case.</li>
</ul>

</div>
</div>
</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[&nbsp;]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-sos"><pre><span></span>
</pre></div>

</div>
</div>
</div>

</div>
    </div>
  </div>
</body>

 


</html>
