
<!DOCTYPE html PUBLIC "-//W3C//DTD HTML 4.01//EN" "http://www.w3.org/TR/html4/strict.dtd">
<html>
<head>
    <meta charset='utf-8'>
 <title>Misc.sos</title>
<meta http-equiv="content-type" content="text/html; charset=None">

<style type="text/css">
html {
  font-family: sans-serif;
  -ms-text-size-adjust: 100%;
  -webkit-text-size-adjust: 100%;
}

body {
  margin: 0;
}

a {
  background-color: transparent;
  -webkit-text-decoration-skip: objects;
}

a:active,
a:hover {
  outline-width: 0;
}

img {
  border-style: none;
}

pre {
  font-family: monospace, monospace;
  font-size: 1em;
}

button::-moz-focus-inner,
[type="button"]::-moz-focus-inner,
[type="reset"]::-moz-focus-inner,
[type="submit"]::-moz-focus-inner {
  border-style: none;
  padding: 0;
}

button:-moz-focusring,
[type="button"]:-moz-focusring,
[type="reset"]:-moz-focusring,
[type="submit"]:-moz-focusring {
  outline: 1px dotted ButtonText;
}

table {
  border-spacing: 0;
  border-collapse: collapse;
}

td {
  padding: 0;
}

* {
  box-sizing: border-box;
}

body {
  font: 13px / 1.4 Helvetica, arial, nimbussansl, liberationsans, freesans, clean, sans-serif, "Apple Color Emoji", "Segoe UI Emoji", "Segoe UI Symbol";
  color: #333;
  background-color: #fff;
}

a {
  color: #4078c0;
  text-decoration: none;
}

a:hover,
a:active {
  text-decoration: underline;
}

pre {
  margin-top: 0;
  margin-bottom: 0;
  font: 12px Consolas, "Liberation Mono", Menlo, Courier, monospace;
}

.container {
  width: 980px;
  margin-right: auto;
  margin-left: auto;
}

.container::before {
  display: table;
  content: "";
}

.container::after {
  display: table;
  clear: both;
  content: "";
}

.right {
  float: right !important;
}

::-moz-placeholder {
  color: #999;
}

:-ms-input-placeholder {
  color: #999;
}

::placeholder {
  color: #999;
}

.form-select::-ms-expand {
  opacity: 0;
}

.avatar {
  display: inline-block;
  overflow: hidden;
  line-height: 1;
  vertical-align: middle;
  border-radius: 3px;
}

.commit-tease {
  position: relative;
  padding: 10px;
  margin-bottom: -1px;
  line-height: 20px;
  color: #68777d;
  background-color: #f2f9fc;
  border: 1px solid #c9e6f2;
  border-radius: 3px;
  border-bottom-right-radius: 0;
  border-bottom-left-radius: 0;
}

.commit-tease .message {
  color: inherit;
}

.commit-tease .avatar {
  margin-top: -1px;
}

.commit-tease-sha {
  display: inline-block;
  font-family: Consolas, "Liberation Mono", Menlo, Courier, monospace;
  color: #445055;
}

.full-commit .btn-outline:not(:disabled):hover {
  color: #4078c0;
  border: 1px solid #4078c0;
}

.blob-wrapper {
  overflow-x: auto;
  overflow-y: hidden;
  border-bottom-right-radius: 3px;
  border-bottom-left-radius: 3px;
}

.blob-num {
  width: 1%;
  min-width: 50px;
  padding-right: 10px;
  padding-left: 10px;
  font-family: Consolas, "Liberation Mono", Menlo, Courier, monospace;
  font-size: 12px;
  line-height: 18px;
  color: rgba(0,0,0,0.3);
  text-align: right;
  white-space: nowrap;
  vertical-align: bottom;
  cursor: pointer;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  border: solid #eee;
  border-width: 0 1px 0 0;
}

.blob-num:hover {
  color: rgba(0,0,0,0.6);
}

.blob-num::before {
  content: attr(data-line-number);
}

.blob-code {
  position: relative;
  padding-right: 10px;
  padding-left: 10px;
  vertical-align: top;
  white-space: nowrap;
}

.sos-comment {
  /* background: #f7f7f7; */
}
.sos-report {
  background: #FAFAD2;
  padding-top: 5px;
}
.sos-header {
  margin-top: 5px;
  margin-bottom: 0px;
  padding-top: 5px;
  padding-bottom: 5px;
  background-color: #ddd;
  /* border-bottom-color: #dde4e6; */
  color: darkblue;
  border-top-style: solid;

}
.sos-directive {
  padding-top: 5px;
  /* background-color: #fff9ea; */
  border-color: #f1c0c0;
}
.sos-statement {
  padding-top: 5px;
  /* background-color: #f2f9fc; */
}
.sos-error {
  /* background: #ffdddd; */
  color: red;
}
.sos-script {
  position: relative;
  padding-right: 10px;
  padding-left: 30px;
  vertical-align: top;
  /* background-color: #eff0f1; */
}


.blob-code-inner {
  overflow: visible;
  font-family: Consolas, "Liberation Mono", Menlo, Courier, monospace;
  font-size: 12px;
  color: #333;
  word-wrap: normal;
  white-space: pre;
}

.blob-code-inner::before {
  content: "";
}

body {
  min-width: 1020px;
  word-wrap: break-word;
}

.user-mention {
  font-weight: bold;
  color: #333;
  white-space: nowrap;
}

.prose-diff>.markdown-body :not(li.moved).removed {
  color: #a33;
  text-decoration: line-through;
  background: #ffeaea;
}

.prose-diff>.markdown-body :not(.github-user-ins):not(li.moved).removed {
  text-decoration: line-through;
}

.prose-diff>.markdown-body :not(li.moved).added {
  background: #eaffea;
}

.prose-diff>.markdown-body :not(.github-user-del):not(li.moved).added li:not(.moved):not(.github-user-del).added {
  text-decoration: none;
}

:checked+.radio-label {
  position: relative;
  z-index: 1;
  border-color: #4078c0;
}

.select-menu-text-filter input::-moz-placeholder {
  color: #aaa;
}

.select-menu-text-filter input:-ms-input-placeholder {
  color: #aaa;
}

.select-menu-text-filter input::placeholder {
  color: #aaa;
}

.tab-size[data-tab-size="8"] {
  -moz-tab-size: 8;
  -o-tab-size: 8;
  tab-size: 8;
}

# override source
.source {
  background: auto;
}

.file {
  position: relative;
  margin-top: 20px;
  margin-bottom: 15px;
  border: 1px solid #ddd;
  border-radius: 3px;
}

.file-header {
  padding: 5px 10px;
  background-color: #f7f7f7;
  border-bottom: 1px solid #d8d8d8;
  border-top-left-radius: 2px;
  border-top-right-radius: 2px;
}

.file-header::before {
  display: table;
  content: "";
}

.file-header::after {
  display: table;
  clear: both;
  content: "";
}

.file-actions {
  float: right;
  padding-top: 3px;
}

.file-info {
  float: left;
  font-family: Consolas, "Liberation Mono", Menlo, Courier, monospace;
  font-size: 12px;
  line-height: 32px;
}

.file-info-divider {
  display: inline-block;
  width: 1px;
  height: 18px;
  margin-right: 3px;
  margin-left: 3px;
  vertical-align: middle;
  background-color: #ddd;
}

.header-search-input::-ms-clear {
  display: none;
}

.org-validate-group input::-ms-clear {
  display: none;
}


.source .c { color: #8f5902; font-style: italic } /* Comment */
.source .err { color: #a40000; border: 1px solid #ef2929 } /* Error */
.source .g { color: #000000 } /* Generic */
.source .k { color: #204a87; font-weight: bold } /* Keyword */
.source .l { color: #000000 } /* Literal */
.source .n { color: #000000 } /* Name */
.source .o { color: #ce5c00; font-weight: bold } /* Operator */
.source .x { color: #000000 } /* Other */
.source .p { color: #000000; font-weight: bold } /* Punctuation */
.source .ch { color: #8f5902; font-style: italic } /* Comment.Hashbang */
.source .cm { color: #8f5902; font-style: italic } /* Comment.Multiline */
.source .cp { color: #8f5902; font-style: italic } /* Comment.Preproc */
.source .cpf { color: #8f5902; font-style: italic } /* Comment.PreprocFile */
.source .c1 { color: #8f5902; font-style: italic } /* Comment.Single */
.source .cs { color: #8f5902; font-style: italic } /* Comment.Special */
.source .gd { color: #a40000 } /* Generic.Deleted */
.source .ge { color: #000000; font-style: italic } /* Generic.Emph */
.source .gr { color: #ef2929 } /* Generic.Error */
.source .gh { color: #000080; font-weight: bold } /* Generic.Heading */
.source .gi { color: #00A000 } /* Generic.Inserted */
.source .go { color: #000000; font-style: italic } /* Generic.Output */
.source .gp { color: #8f5902 } /* Generic.Prompt */
.source .gs { color: #000000; font-weight: bold } /* Generic.Strong */
.source .gu { color: #800080; font-weight: bold } /* Generic.Subheading */
.source .gt { color: #a40000; font-weight: bold } /* Generic.Traceback */
.source .kc { color: #204a87; font-weight: bold } /* Keyword.Constant */
.source .kd { color: #204a87; font-weight: bold } /* Keyword.Declaration */
.source .kn { color: #204a87; font-weight: bold } /* Keyword.Namespace */
.source .kp { color: #204a87; font-weight: bold } /* Keyword.Pseudo */
.source .kr { color: #204a87; font-weight: bold } /* Keyword.Reserved */
.source .kt { color: #204a87; font-weight: bold } /* Keyword.Type */
.source .ld { color: #000000 } /* Literal.Date */
.source .m { color: #0000cf; font-weight: bold } /* Literal.Number */
.source .s { color: #4e9a06 } /* Literal.String */
.source .na { color: #c4a000 } /* Name.Attribute */
.source .nb { color: #204a87 } /* Name.Builtin */
.source .nc { color: #000000 } /* Name.Class */
.source .no { color: #000000 } /* Name.Constant */
.source .nd { color: #5c35cc; font-weight: bold } /* Name.Decorator */
.source .ni { color: #ce5c00 } /* Name.Entity */
.source .ne { color: #cc0000; font-weight: bold } /* Name.Exception */
.source .nf { color: #000000 } /* Name.Function */
.source .nl { color: #f57900 } /* Name.Label */
.source .nn { color: #000000 } /* Name.Namespace */
.source .nx { color: #000000 } /* Name.Other */
.source .py { color: #000000 } /* Name.Property */
.source .nt { color: #204a87; font-weight: bold } /* Name.Tag */
.source .nv { color: #000000 } /* Name.Variable */
.source .ow { color: #204a87; font-weight: bold } /* Operator.Word */
.source .w { color: #f8f8f8; text-decoration: underline } /* Text.Whitespace */
.source .mb { color: #0000cf; font-weight: bold } /* Literal.Number.Bin */
.source .mf { color: #0000cf; font-weight: bold } /* Literal.Number.Float */
.source .mh { color: #0000cf; font-weight: bold } /* Literal.Number.Hex */
.source .mi { color: #0000cf; font-weight: bold } /* Literal.Number.Integer */
.source .mo { color: #0000cf; font-weight: bold } /* Literal.Number.Oct */
.source .sa { color: #4e9a06 } /* Literal.String.Affix */
.source .sb { color: #4e9a06 } /* Literal.String.Backtick */
.source .sc { color: #4e9a06 } /* Literal.String.Char */
.source .dl { color: #4e9a06 } /* Literal.String.Delimiter */
.source .sd { color: #8f5902; font-style: italic } /* Literal.String.Doc */
.source .s2 { color: #4e9a06 } /* Literal.String.Double */
.source .se { color: #4e9a06 } /* Literal.String.Escape */
.source .sh { color: #4e9a06 } /* Literal.String.Heredoc */
.source .si { color: #4e9a06 } /* Literal.String.Interpol */
.source .sx { color: #4e9a06 } /* Literal.String.Other */
.source .sr { color: #4e9a06 } /* Literal.String.Regex */
.source .s1 { color: #4e9a06 } /* Literal.String.Single */
.source .ss { color: #4e9a06 } /* Literal.String.Symbol */
.source .bp { color: #3465a4 } /* Name.Builtin.Pseudo */
.source .fm { color: #000000 } /* Name.Function.Magic */
.source .vc { color: #000000 } /* Name.Variable.Class */
.source .vg { color: #000000 } /* Name.Variable.Global */
.source .vi { color: #000000 } /* Name.Variable.Instance */
.source .vm { color: #000000 } /* Name.Variable.Magic */
.source .il { color: #0000cf; font-weight: bold } /* Literal.Number.Integer.Long */
 </style>
</head>
<body>



<div class="container new-discussion-timeline experiment-repo-nav">

 <div class="commit-tease">
      <span class="right">
      </span>
 <div>
        <a class="header-logo-invertocat" href="https://vatlab.github.io/SOS" data-hotkey="g d" aria-label="Homepage">
        <img src="http://vatlab.github.io/SOS/img/sos_icon.svg" alt="sos_icon">
        </a>

        <span class="user-mention">Misc.sos</span>
      </div>
      </div>

  <div class="repository-content">
  <div class="file">
  <div class="file-header">

  <div class="file-info">
      pipeline/Misc.sos
      <span class="file-info-divider"></span>
    13.9 KiB
  </div>
  <div class="file-actions">
  </div>
  </div>
 <div itemprop="text" class="blob-wrapper data type-python">

<table class="highlight tab-size js-file-line-container"><tr><td class="blob-num js-line-number"><div class="linenodiv"><pre>1</pre></div></td><td class="code"><div class="source blob-code sos-comment"><pre><span></span><span class="ch">#!/usr/bin/env sos-runner</span>
</pre></div>
</td></tr>
<tr><td class="blob-num js-line-number"><div class="linenodiv"><pre>2</pre></div></td><td class="code"><div class="source blob-code sos-comment"><pre><span></span><span class="c1">#fileformat=SOS1.0</span>
</pre></div>
</td></tr>
<tr><td class="blob-num js-line-number"><div class="linenodiv"><pre>3</pre></div></td><td class="code"><div class="source blob-code sos-comment"><pre><span></span> 
</pre></div>
</td></tr>
<tr><td class="blob-num js-line-number"><div class="linenodiv"><pre>4</pre></div></td><td class="code"><div class="source blob-code sos-comment"><pre><span></span><span class="c1"># Various calls to commands to accomplish some workflow</span>
</pre></div>
</td></tr>
<tr><td class="blob-num js-line-number"><div class="linenodiv"><pre>5</pre></div></td><td class="code"><div class="source blob-code sos-comment"><pre><span></span><span class="c1"># When some workflow gets complicated they will be separated into dedicated files</span>
</pre></div>
</td></tr>
<tr><td class="blob-num js-line-number"><div class="linenodiv"><pre>6</pre></div></td><td class="code"><div class="source blob-code sos-comment"><pre><span></span> 
</pre></div>
</td></tr><tr><td class="blob-num js-line-number"><div class="linenodiv"><pre>7
8</pre></div></td><td class="code"><div class="source blob-code sos-directive"><pre><span></span><span class="kn">parameter:</span> <span class="n">project_name</span> <span class="o">=</span> <span class="s1">&#39;GTEx&#39;</span>
 
</pre></div>
</td></tr>
<tr><td class="blob-num js-line-number"><div class="linenodiv"><pre>9</pre></div></td><td class="code"><div class="source blob-code sos-header"><pre><span></span><span class="p">[</span><span class="n">genotype_stats_1</span><span class="p">]</span>
</pre></div>
</td></tr>
<tr><td class="blob-num js-line-number"><div class="linenodiv"><pre>10</pre></div></td><td class="code"><div class="source blob-code sos-comment"><pre><span></span><span class="c1"># making genotype summary statistics via vcftools</span>
</pre></div>
</td></tr><tr><td class="blob-num js-line-number"><div class="linenodiv"><pre>11
12
13
14
15</pre></div></td><td class="code"><div class="source blob-code sos-directive"><pre><span></span><span class="kn">parameter:</span> <span class="kp">workdir</span> <span class="o">=</span> <span class="bp">None</span>
<span class="kn">depends:</span> <span class="n">executable</span><span class="p">(</span><span class="s2">&quot;vcftools&quot;</span><span class="p">)</span>
<span class="kn">input:</span> <span class="kp">pattern</span> <span class="o">=</span> <span class="s2">&quot;{name}.vcf.gz&quot;</span>
<span class="kn">output:</span> <span class="n">expand_pattern</span><span class="p">(</span><span class="s2">&quot;${CONFIG[&#39;wd&#39;]}/${name!b}.imiss&quot;</span><span class="p">),</span> <span class="n">expand_pattern</span><span class="p">(</span><span class="s2">&quot;${CONFIG[&#39;wd&#39;]}/${name!b}.lmiss&quot;</span><span class="p">)</span>
<span class="kn">task:</span> <span class="kp">workdir</span> <span class="o">=</span> <span class="kp">workdir</span>
</pre></div>
</td></tr>
<tr><td class="blob-num js-line-number"><div class="linenodiv"><pre>16</pre></div></td><td class="code"><div class="source blob-code sos-directive"><pre><span></span><span class="kn">run:</span> 
</pre></div>
</td></tr>
<tr><td class="blob-num js-line-number"><div class="linenodiv"><pre>17
18
19</pre></div></td><td class="code"><div class="source blob-code sos-script "><pre><span></span>vcftools --gzvcf <span class="si">${</span><span class="nv">input</span><span class="si">}</span> --out <span class="si">${</span><span class="nv">output</span><span class="p">[0]!n</span><span class="si">}</span> --missing-indv
vcftools --gzvcf <span class="si">${</span><span class="nv">input</span><span class="si">}</span> --out <span class="si">${</span><span class="nv">output</span><span class="p">[1]!n</span><span class="si">}</span> --missing-site
 
</pre></div>
</td></tr>
<tr><td class="blob-num js-line-number"><div class="linenodiv"><pre>20</pre></div></td><td class="code"><div class="source blob-code sos-header"><pre><span></span><span class="p">[</span><span class="n">genotype_stats_2</span><span class="p">]</span>
</pre></div>
</td></tr>
<tr><td class="blob-num js-line-number"><div class="linenodiv"><pre>21</pre></div></td><td class="code"><div class="source blob-code sos-comment"><pre><span></span><span class="c1"># making genotype summary statistics plot</span>
</pre></div>
</td></tr><tr><td class="blob-num js-line-number"><div class="linenodiv"><pre>22
23
24
25
26</pre></div></td><td class="code"><div class="source blob-code sos-directive"><pre><span></span><span class="kn">parameter:</span> <span class="kp">workdir</span> <span class="o">=</span> <span class="bp">None</span>
<span class="kn">depends:</span> <span class="n">Py_Module</span><span class="p">(</span><span class="s2">&quot;seaborn&quot;</span><span class="p">)</span>
<span class="kn">input:</span> <span class="kp">group_by</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="kp">pattern</span> <span class="o">=</span> <span class="s2">&quot;{name}.{ext}&quot;</span>
<span class="kn">output:</span> <span class="n">expand_pattern</span><span class="p">(</span><span class="s2">&quot;{_name}.{_ext}.pdf&quot;</span><span class="p">)</span>
<span class="kn">task:</span> <span class="kp">workdir</span> <span class="o">=</span> <span class="kp">workdir</span>
</pre></div>
</td></tr>
<tr><td class="blob-num js-line-number"><div class="linenodiv"><pre>27</pre></div></td><td class="code"><div class="source blob-code sos-directive"><pre><span></span><span class="kn">python:</span>
</pre></div>
</td></tr>
<tr><td class="blob-num js-line-number"><div class="linenodiv"><pre>28
29
30
31
32
33
34
35</pre></div></td><td class="code"><div class="source blob-code sos-script "><pre><span></span><span class="kn">import</span> <span class="nn">matplotlib.pyplot</span> <span class="kn">as</span> <span class="nn">plt</span><span class="o">,</span> <span class="nn">seaborn</span> <span class="kn">as</span> <span class="nn">sns</span><span class="o">,</span> <span class="nn">pandas</span> <span class="kn">as</span> <span class="nn">pd</span>
<span class="n">fig</span><span class="p">,</span> <span class="n">axs</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="n">ncols</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span>
<span class="n">data</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="err">$</span><span class="p">{</span><span class="n">_input</span><span class="err">!</span><span class="n">r</span><span class="p">},</span> <span class="n">sep</span> <span class="o">=</span> <span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">)</span>
<span class="n">sns</span><span class="o">.</span><span class="n">distplot</span><span class="p">(</span><span class="n">data</span><span class="p">[</span><span class="s2">&quot;F_MISS&quot;</span><span class="p">],</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">axs</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">kde</span> <span class="o">=</span> <span class="bp">False</span><span class="p">)</span>
<span class="n">sns</span><span class="o">.</span><span class="n">violinplot</span><span class="p">(</span><span class="n">data</span><span class="p">[</span><span class="s2">&quot;F_MISS&quot;</span><span class="p">],</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">axs</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="err">$</span><span class="p">{</span><span class="n">_input</span><span class="err">!</span><span class="n">br</span><span class="p">}</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;.&#39;</span><span class="p">)[</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span>
<span class="n">fig</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="err">!</span><span class="n">r</span><span class="p">})</span>
 
</pre></div>
</td></tr>
<tr><td class="blob-num js-line-number"><div class="linenodiv"><pre>36</pre></div></td><td class="code"><div class="source blob-code sos-header"><pre><span></span><span class="p">[</span><span class="n">rnaseq_1</span><span class="p">]</span>
</pre></div>
</td></tr>
<tr><td class="blob-num js-line-number"><div class="linenodiv"><pre>37</pre></div></td><td class="code"><div class="source blob-code sos-comment"><pre><span></span><span class="c1"># Quantile normalization of RNA-seq data</span>
</pre></div>
</td></tr>
<tr><td class="blob-num js-line-number"><div class="linenodiv"><pre>38</pre></div></td><td class="code"><div class="source blob-code sos-comment"><pre><span></span><span class="c1"># 1. expression values are quantile normalized to the average empirical distribution observed across samples</span>
</pre></div>
</td></tr>
<tr><td class="blob-num js-line-number"><div class="linenodiv"><pre>39</pre></div></td><td class="code"><div class="source blob-code sos-comment"><pre><span></span><span class="c1"># 2. for each gene, expression values are inverse quantile normalized to a standard normal distribution across samples</span>
</pre></div>
</td></tr>
<tr><td class="blob-num js-line-number"><div class="linenodiv"><pre>40</pre></div></td><td class="code"><div class="source blob-code sos-comment"><pre><span></span><span class="c1"># genes are selected based on expression thresholds of &gt;0.1 RPKM in &gt;10 samples and &gt;5 reads in &gt;10 samples</span>
</pre></div>
</td></tr>
<tr><td class="blob-num js-line-number"><div class="linenodiv"><pre>41</pre></div></td><td class="code"><div class="source blob-code sos-comment"><pre><span></span><span class="c1"># input are rpkm file (for normalization), count file (for QC) and vcf file (for removing samples that do not have genotypes)</span>
</pre></div>
</td></tr><tr><td class="blob-num js-line-number"><div class="linenodiv"><pre>42
43
44
45
46
47
48</pre></div></td><td class="code"><div class="source blob-code sos-directive"><pre><span></span><span class="kn">parameter:</span> <span class="n">rpkm_cutoff</span> <span class="o">=</span> <span class="mf">0.1</span>
<span class="kn">parameter:</span> <span class="n">read_cutoff</span> <span class="o">=</span> <span class="mi">5</span>
<span class="kn">parameter:</span> <span class="n">sample_cutoff</span> <span class="o">=</span> <span class="mi">10</span>
<span class="kn">parameter:</span> <span class="kp">script</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">abspath</span><span class="p">(</span><span class="s2">&quot;../src/normalize_expression.py&quot;</span><span class="p">)</span>
<span class="kn">parameter:</span> <span class="kp">workdir</span> <span class="o">=</span> <span class="bp">None</span>
<span class="kn">output:</span> <span class="s2">&quot;${workdir}/${input[0]!nnb}.qnorm.std.flat.h5&quot;</span><span class="p">,</span> <span class="s2">&quot;${workdir}/${input[0]!nnb}.qnorm.flat.h5&quot;</span><span class="p">,</span> <span class="s2">&quot;${workdir}/${input[0]!nnb}.qnorm.std.h5&quot;</span><span class="p">,</span> <span class="s2">&quot;${workdir}/${input[0]!nnb}.qnorm.h5&quot;</span>
<span class="kn">task:</span> <span class="kp">workdir</span> <span class="o">=</span> <span class="kp">workdir</span>
</pre></div>
</td></tr>
<tr><td class="blob-num js-line-number"><div class="linenodiv"><pre>49</pre></div></td><td class="code"><div class="source blob-code sos-directive"><pre><span></span><span class="kn">run:</span>
</pre></div>
</td></tr>
<tr><td class="blob-num js-line-number"><div class="linenodiv"><pre>50
51</pre></div></td><td class="code"><div class="source blob-code sos-script "><pre><span></span>python <span class="si">${</span><span class="nv">script</span><span class="si">}</span> <span class="si">${</span><span class="nv">input</span><span class="p">[0]</span><span class="si">}</span> <span class="si">${</span><span class="nv">input</span><span class="p">[1]</span><span class="si">}</span> <span class="si">${</span><span class="nv">input</span><span class="p">[2]</span><span class="si">}</span> <span class="si">${</span><span class="nv">input</span><span class="p">[3]</span><span class="si">}</span> <span class="si">${</span><span class="nv">input</span><span class="p">[0]!nnb</span><span class="si">}</span> --expression_threshold <span class="si">${</span><span class="nv">rpkm_cutoff</span><span class="si">}</span> --count_threshold <span class="si">${</span><span class="nv">read_cutoff</span><span class="si">}</span> --min_samples <span class="si">${</span><span class="nv">sample_cutoff</span><span class="si">}</span>
 
</pre></div>
</td></tr>
<tr><td class="blob-num js-line-number"><div class="linenodiv"><pre>52</pre></div></td><td class="code"><div class="source blob-code sos-header"><pre><span></span><span class="p">[</span><span class="n">rnaseq_2</span><span class="p">]</span>
</pre></div>
</td></tr>
<tr><td class="blob-num js-line-number"><div class="linenodiv"><pre>53</pre></div></td><td class="code"><div class="source blob-code sos-comment"><pre><span></span><span class="c1"># PEER analysis</span>
</pre></div>
</td></tr>
<tr><td class="blob-num js-line-number"><div class="linenodiv"><pre>54</pre></div></td><td class="code"><div class="source blob-code sos-comment"><pre><span></span><span class="c1"># There are a number of configuable parameters in this script</span>
</pre></div>
</td></tr>
<tr><td class="blob-num js-line-number"><div class="linenodiv"><pre>55</pre></div></td><td class="code"><div class="source blob-code sos-comment"><pre><span></span><span class="c1"># Using default values for now. Can be changed here or not depending on the analyst</span>
</pre></div>
</td></tr><tr><td class="blob-num js-line-number"><div class="linenodiv"><pre>56
57
58
59
60
61</pre></div></td><td class="code"><div class="source blob-code sos-directive"><pre><span></span><span class="kn">depends:</span> <span class="n">R_library</span><span class="p">(</span><span class="s1">&#39;rhdf5&#39;</span><span class="p">),</span> <span class="n">R_library</span><span class="p">(</span><span class="s1">&#39;peer&#39;</span><span class="p">)</span>
<span class="kn">parameter:</span> <span class="n">tissues</span> <span class="o">=</span> <span class="n">get_output</span><span class="p">(</span><span class="s2">&quot;h5ls ${input[2]} | awk &#39;{print $1}&#39;&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span>
<span class="kn">parameter:</span> <span class="kp">workdir</span> <span class="o">=</span> <span class="bp">None</span>
<span class="kn">input:</span> <span class="kp">for_each</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;tissues&#39;</span><span class="p">]</span>
<span class="kn">output:</span> <span class="s2">&quot;${workdir}/${_tissues}_PEER_covariates.txt&quot;</span><span class="p">,</span> <span class="s2">&quot;${workdir}/${_tissues}_PEER_alpha.txt&quot;</span><span class="p">,</span> <span class="s2">&quot;${workdir}/${_tissues}_PEER_residuals.txt&quot;</span>
<span class="kn">task:</span> <span class="kp">workdir</span> <span class="o">=</span> <span class="kp">workdir</span><span class="p">,</span> <span class="kp">walltime</span> <span class="o">=</span> <span class="s2">&quot;20:00:00&quot;</span><span class="p">,</span> <span class="kp">cores</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="kp">mem</span> <span class="o">=</span> <span class="s2">&quot;4G&quot;</span>
</pre></div>
</td></tr>
<tr><td class="blob-num js-line-number"><div class="linenodiv"><pre>62</pre></div></td><td class="code"><div class="source blob-code sos-directive"><pre><span></span><span class="kn">R:</span>
</pre></div>
</td></tr>
<tr><td class="blob-num js-line-number"><div class="linenodiv"><pre> 63
 64
 65
 66
 67
 68
 69
 70
 71
 72
 73
 74
 75
 76
 77
 78
 79
 80
 81
 82
 83
 84
 85
 86
 87
 88
 89
 90
 91
 92
 93
 94
 95
 96
 97
 98
 99
100
101
102
103
104
105
106
107
108
109
110
111
112
113
114
115
116
117
118
119
120
121
122
123
124
125
126
127
128
129
130
131
132
133
134
135
136
137
138</pre></div></td><td class="code"><div class="source blob-code sos-script "><pre><span></span> 
expr.h5 <span class="o">=</span> <span class="o">$</span><span class="p">{</span>input<span class="p">[</span><span class="m">2</span><span class="p">]</span><span class="o">!</span>r<span class="p">}</span>
prefix <span class="o">=</span> <span class="o">$</span><span class="p">{</span>_tissues<span class="o">!</span>r<span class="p">}</span>
alphaprior_a<span class="o">=</span><span class="m">0.001</span>
alphaprior_b<span class="o">=</span><span class="m">0.01</span>
epsprior_a<span class="o">=</span><span class="m">0.1</span>
epsprior_b<span class="o">=</span><span class="m">10</span>
max_iter<span class="o">=</span><span class="m">1000</span>
 
<span class="kn">library</span><span class="p">(</span>peer<span class="p">,</span> quietly<span class="o">=</span><span class="kc">TRUE</span><span class="p">)</span>  <span class="c1"># https://github.com/PMBio/peer</span>
<span class="kn">library</span><span class="p">(</span>rhdf5<span class="p">,</span> quietly<span class="o">=</span><span class="kc">TRUE</span><span class="p">)</span>
 
WriteTable <span class="o">&lt;-</span> <span class="kr">function</span><span class="p">(</span>data<span class="p">,</span> filename<span class="p">,</span> index.name<span class="p">)</span> <span class="p">{</span>
	datafile <span class="o">&lt;-</span> <span class="kp">file</span><span class="p">(</span>filename<span class="p">,</span> open <span class="o">=</span> <span class="s">&quot;wt&quot;</span><span class="p">)</span>
	<span class="kp">on.exit</span><span class="p">(</span><span class="kp">close</span><span class="p">(</span>datafile<span class="p">))</span>
	header <span class="o">&lt;-</span> <span class="kt">c</span><span class="p">(</span>index.name<span class="p">,</span> <span class="kp">colnames</span><span class="p">(</span>data<span class="p">))</span>
	<span class="kp">writeLines</span><span class="p">(</span><span class="kp">paste0</span><span class="p">(</span>header<span class="p">,</span> collapse<span class="o">=</span><span class="s">&quot;\t&quot;</span><span class="p">),</span> con<span class="o">=</span>datafile<span class="p">,</span> sep<span class="o">=</span><span class="s">&quot;\n&quot;</span><span class="p">)</span>
	write.table<span class="p">(</span>data<span class="p">,</span> datafile<span class="p">,</span> sep<span class="o">=</span><span class="s">&quot;\t&quot;</span><span class="p">,</span> col.names<span class="o">=</span><span class="kc">FALSE</span><span class="p">,</span> quote<span class="o">=</span><span class="kc">FALSE</span><span class="p">)</span>
<span class="p">}</span>
 
loadTable <span class="o">&lt;-</span> <span class="kr">function</span><span class="p">(</span>filename<span class="p">,</span> group<span class="p">,</span> auto_transpose <span class="o">=</span> <span class="kc">FALSE</span><span class="p">)</span> <span class="p">{</span>
  obj <span class="o">&lt;-</span> h5read<span class="p">(</span>filename<span class="p">,</span> group<span class="p">)</span>
  dat <span class="o">&lt;-</span> obj<span class="o">$</span>block0_values
  <span class="kp">rownames</span><span class="p">(</span>dat<span class="p">)</span> <span class="o">&lt;-</span> obj<span class="o">$</span>axis0
  <span class="kp">colnames</span><span class="p">(</span>dat<span class="p">)</span> <span class="o">&lt;-</span> obj<span class="o">$</span>axis1
  <span class="kr">if</span> <span class="p">(</span><span class="kp">ncol</span><span class="p">(</span>dat<span class="p">)</span> <span class="o">&gt;</span> <span class="kp">nrow</span><span class="p">(</span>dat<span class="p">)</span> <span class="o">&amp;&amp;</span> auto_transpose<span class="p">)</span> dat <span class="o">&lt;-</span> <span class="kp">t</span><span class="p">(</span>dat<span class="p">)</span>
  <span class="kr">return</span><span class="p">(</span>dat<span class="p">)</span>
<span class="p">}</span>
 
getNumPeer <span class="o">&lt;-</span> <span class="kr">function</span><span class="p">(</span>ss<span class="p">)</span> <span class="p">{</span>
  <span class="kr">if</span> <span class="p">(</span>ss<span class="o">&lt;</span><span class="m">150</span><span class="p">)</span> <span class="kr">return</span> <span class="p">(</span><span class="kp">min</span><span class="p">(</span><span class="m">15</span><span class="p">,</span> <span class="kp">ceiling</span><span class="p">(</span>ss <span class="o">/</span> <span class="m">5</span><span class="p">)))</span>
  <span class="kr">else</span> <span class="kr">if</span> <span class="p">(</span>ss <span class="o">&gt;=</span><span class="m">150</span> <span class="o">&amp;&amp;</span> ss <span class="o">&lt;</span> <span class="m">250</span><span class="p">)</span> <span class="kr">return</span><span class="p">(</span><span class="m">30</span><span class="p">)</span>
  <span class="kr">else</span> <span class="kr">return</span><span class="p">(</span><span class="m">35</span><span class="p">)</span>
<span class="p">}</span>
 
<span class="kp">cat</span><span class="p">(</span><span class="s">&quot;PEER: loading expression data ... &quot;</span><span class="p">)</span>
<span class="c1"># rows are number of samples. columns are number of genes</span>
M <span class="o">&lt;-</span> <span class="kp">as.matrix</span><span class="p">(</span>loadTable<span class="p">(</span>expr.h5<span class="p">,</span> <span class="s">&quot;/${_tissues}&quot;</span><span class="p">))</span>
n <span class="o">=</span> getNumPeer<span class="p">(</span><span class="kp">nrow</span><span class="p">(</span>M<span class="p">))</span>
<span class="kp">cat</span><span class="p">(</span><span class="s">&quot;done.\n&quot;</span><span class="p">)</span>
 
<span class="c1"># run PEER</span>
<span class="kp">cat</span><span class="p">(</span><span class="kp">paste0</span><span class="p">(</span><span class="s">&quot;PEER: estimating hidden confounders (&quot;</span><span class="p">,</span> n<span class="p">,</span> <span class="s">&quot; for tissue &quot;</span><span class="p">,</span> prefix <span class="p">,</span> <span class="s">&quot;)\n&quot;</span><span class="p">))</span>
model <span class="o">&lt;-</span> PEER<span class="p">()</span>
<span class="kp">invisible</span><span class="p">(</span>PEER_setNk<span class="p">(</span>model<span class="p">,</span> n<span class="p">))</span>
<span class="kp">invisible</span><span class="p">(</span>PEER_setPhenoMean<span class="p">(</span>model<span class="p">,</span> M<span class="p">))</span>
<span class="kp">invisible</span><span class="p">(</span>PEER_setPriorAlpha<span class="p">(</span>model<span class="p">,</span> alphaprior_a<span class="p">,</span> alphaprior_b<span class="p">))</span>
<span class="kp">invisible</span><span class="p">(</span>PEER_setPriorEps<span class="p">(</span>model<span class="p">,</span>epsprior_a<span class="p">,</span> epsprior_b<span class="p">))</span>
<span class="kp">invisible</span><span class="p">(</span>PEER_setNmax_iterations<span class="p">(</span>model<span class="p">,</span> max_iter<span class="p">))</span>
<span class="c1"># if(!is.null(covs)) {</span>
<span class="c1">#   invisible(PEER_setCovariates(model, covs))</span>
<span class="c1"># }</span>
time <span class="o">&lt;-</span> <span class="kp">system.time</span><span class="p">(</span>PEER_update<span class="p">(</span>model<span class="p">))</span>
 
X <span class="o">&lt;-</span> PEER_getX<span class="p">(</span>model<span class="p">)</span>  <span class="c1"># samples x PEER factors</span>
A <span class="o">&lt;-</span> PEER_getAlpha<span class="p">(</span>model<span class="p">)</span>  <span class="c1"># PEER factors x 1</span>
R <span class="o">&lt;-</span> <span class="kp">t</span><span class="p">(</span>PEER_getResiduals<span class="p">(</span>model<span class="p">))</span>  <span class="c1"># genes x samples</span>
 
<span class="c1"># add relevant row/column names</span>
<span class="kt">c</span> <span class="o">&lt;-</span> <span class="kp">paste0</span><span class="p">(</span><span class="s">&quot;InferredCov&quot;</span><span class="p">,</span><span class="m">1</span><span class="o">:</span><span class="kp">ncol</span><span class="p">(</span>X<span class="p">))</span>
<span class="kp">rownames</span><span class="p">(</span>X<span class="p">)</span> <span class="o">&lt;-</span> <span class="kp">rownames</span><span class="p">(</span>M<span class="p">)</span>
<span class="kp">colnames</span><span class="p">(</span>X<span class="p">)</span> <span class="o">&lt;-</span> <span class="kt">c</span>
<span class="kp">rownames</span><span class="p">(</span>A<span class="p">)</span> <span class="o">&lt;-</span> <span class="kt">c</span>
<span class="kp">colnames</span><span class="p">(</span>A<span class="p">)</span> <span class="o">&lt;-</span> <span class="s">&quot;Alpha&quot;</span>
A <span class="o">&lt;-</span> <span class="kp">as.data.frame</span><span class="p">(</span>A<span class="p">)</span>
A<span class="o">$</span>Relevance <span class="o">&lt;-</span> <span class="m">1.0</span> <span class="o">/</span> A<span class="o">$</span>Alpha
<span class="kp">rownames</span><span class="p">(</span>R<span class="p">)</span> <span class="o">&lt;-</span> <span class="kp">colnames</span><span class="p">(</span>M<span class="p">)</span>
<span class="kp">colnames</span><span class="p">(</span>R<span class="p">)</span> <span class="o">&lt;-</span> <span class="kp">rownames</span><span class="p">(</span>M<span class="p">)</span>
 
<span class="c1"># write results</span>
<span class="kp">cat</span><span class="p">(</span><span class="s">&quot;PEER: writing results ... &quot;</span><span class="p">)</span>
WriteTable<span class="p">(</span><span class="kp">t</span><span class="p">(</span>X<span class="p">),</span> <span class="kp">paste0</span><span class="p">(</span>prefix<span class="p">,</span> <span class="s">&quot;_PEER_covariates.txt&quot;</span><span class="p">),</span> <span class="s">&quot;ID&quot;</span><span class="p">)</span>  <span class="c1"># format(X, digits=6)</span>
WriteTable<span class="p">(</span>A<span class="p">,</span> <span class="kp">paste0</span><span class="p">(</span>prefix<span class="p">,</span> <span class="s">&quot;_PEER_alpha.txt&quot;</span><span class="p">),</span> <span class="s">&quot;ID&quot;</span><span class="p">)</span>
WriteTable<span class="p">(</span>R<span class="p">,</span> <span class="kp">paste0</span><span class="p">(</span>prefix<span class="p">,</span> <span class="s">&quot;_PEER_residuals.txt&quot;</span><span class="p">),</span> <span class="s">&quot;ID&quot;</span><span class="p">)</span>
<span class="kp">cat</span><span class="p">(</span><span class="s">&quot;done.\n&quot;</span><span class="p">)</span>
 
</pre></div>
</td></tr>
<tr><td class="blob-num js-line-number"><div class="linenodiv"><pre>139</pre></div></td><td class="code"><div class="source blob-code sos-header"><pre><span></span><span class="p">[</span><span class="n">LD_pruning_1</span><span class="p">]</span>
</pre></div>
</td></tr>
<tr><td class="blob-num js-line-number"><div class="linenodiv"><pre>140</pre></div></td><td class="code"><div class="source blob-code sos-comment"><pre><span></span><span class="c1"># Mark independent variants</span>
</pre></div>
</td></tr><tr><td class="blob-num js-line-number"><div class="linenodiv"><pre>141
142
143
144
145
146
147</pre></div></td><td class="code"><div class="source blob-code sos-directive"><pre><span></span><span class="kn">parameter:</span> <span class="kp">workdir</span> <span class="o">=</span> <span class="bp">None</span>
<span class="kn">parameter:</span> <span class="n">pairwise_ld_param</span> <span class="o">=</span> <span class="s1">&#39;50 5 0.2&#39;</span>
<span class="kn">parameter:</span> <span class="n">maf</span> <span class="o">=</span> <span class="mf">0.1</span>
<span class="kn">depends:</span> <span class="n">executable</span><span class="p">(</span><span class="s1">&#39;plink&#39;</span><span class="p">)</span>
<span class="kn">input:</span> <span class="kp">pattern</span> <span class="o">=</span> <span class="s1">&#39;{name}.{ext}&#39;</span>
<span class="kn">output:</span> <span class="n">expand_pattern</span><span class="p">(</span><span class="s1">&#39;{name}.prune.in&#39;</span><span class="p">)</span>
<span class="kn">task:</span> <span class="kp">workdir</span> <span class="o">=</span> <span class="kp">workdir</span> 
</pre></div>
</td></tr>
<tr><td class="blob-num js-line-number"><div class="linenodiv"><pre>148</pre></div></td><td class="code"><div class="source blob-code sos-directive"><pre><span></span><span class="kn">run:</span>
</pre></div>
</td></tr>
<tr><td class="blob-num js-line-number"><div class="linenodiv"><pre>149
150
151
152
153
154</pre></div></td><td class="code"><div class="source blob-code sos-script "><pre><span></span>plink --bfile <span class="si">${</span><span class="nv">input</span><span class="p">!n</span><span class="si">}</span> <span class="se">\</span>
      --allow-no-sex <span class="se">\</span>
      --maf <span class="si">${</span><span class="nv">maf</span><span class="si">}</span> <span class="se">\</span>
      --indep-pairwise <span class="si">${</span><span class="nv">pairwise_ld_param</span><span class="si">}</span> <span class="se">\</span>
      --out <span class="si">${</span><span class="nv">output</span><span class="p">!nn</span><span class="si">}</span>
 
</pre></div>
</td></tr>
<tr><td class="blob-num js-line-number"><div class="linenodiv"><pre>155</pre></div></td><td class="code"><div class="source blob-code sos-header"><pre><span></span><span class="p">[</span><span class="n">LD_pruning_2</span><span class="p">]</span>
</pre></div>
</td></tr>
<tr><td class="blob-num js-line-number"><div class="linenodiv"><pre>156</pre></div></td><td class="code"><div class="source blob-code sos-comment"><pre><span></span><span class="c1"># Select independent common variants</span>
</pre></div>
</td></tr><tr><td class="blob-num js-line-number"><div class="linenodiv"><pre>157
158
159
160
161</pre></div></td><td class="code"><div class="source blob-code sos-directive"><pre><span></span><span class="kn">parameter:</span> <span class="kp">workdir</span> <span class="o">=</span> <span class="bp">None</span>
<span class="kn">depends:</span> <span class="n">executable</span><span class="p">(</span><span class="s1">&#39;plink&#39;</span><span class="p">)</span>
<span class="kn">input:</span> <span class="kp">pattern</span> <span class="o">=</span> <span class="s1">&#39;{name}.prune.in&#39;</span>
<span class="kn">output:</span> <span class="n">expand_pattern</span><span class="p">(</span><span class="s1">&#39;{name}.prune.bed&#39;</span><span class="p">)</span>
<span class="kn">task:</span> <span class="kp">workdir</span> <span class="o">=</span> <span class="kp">workdir</span> 
</pre></div>
</td></tr>
<tr><td class="blob-num js-line-number"><div class="linenodiv"><pre>162</pre></div></td><td class="code"><div class="source blob-code sos-directive"><pre><span></span><span class="kn">run:</span>
</pre></div>
</td></tr>
<tr><td class="blob-num js-line-number"><div class="linenodiv"><pre>163
164
165
166
167
168</pre></div></td><td class="code"><div class="source blob-code sos-script "><pre><span></span>plink --bfile <span class="si">${</span><span class="nv">input</span><span class="p">!nn</span><span class="si">}</span> <span class="se">\</span>
      --extract <span class="si">${</span><span class="nv">input</span><span class="si">}</span> <span class="se">\</span>
      --no-sex --no-pheno --no-parents <span class="se">\</span>
      --make-bed <span class="se">\</span>
      --out <span class="si">${</span><span class="nv">output</span><span class="p">!n</span><span class="si">}</span>
 
</pre></div>
</td></tr>
<tr><td class="blob-num js-line-number"><div class="linenodiv"><pre>169</pre></div></td><td class="code"><div class="source blob-code sos-header"><pre><span></span><span class="p">[</span><span class="n">global_ancestry_1</span><span class="p">]</span>
</pre></div>
</td></tr>
<tr><td class="blob-num js-line-number"><div class="linenodiv"><pre>170</pre></div></td><td class="code"><div class="source blob-code sos-comment"><pre><span></span><span class="c1"># global ancestry analysis via KING</span>
</pre></div>
</td></tr>
<tr><td class="blob-num js-line-number"><div class="linenodiv"><pre>171</pre></div></td><td class="code"><div class="source blob-code sos-comment"><pre><span></span><span class="c1"># And fix family ID (replace by ancestry coding)</span>
</pre></div>
</td></tr><tr><td class="blob-num js-line-number"><div class="linenodiv"><pre>172
173
174
175
176
177</pre></div></td><td class="code"><div class="source blob-code sos-directive"><pre><span></span><span class="kn">parameter:</span> <span class="n">phenotype</span> <span class="o">=</span> <span class="s2">&quot;${CONFIG[&#39;phenotype&#39;]!a}&quot;</span>
<span class="kn">parameter:</span> <span class="kp">workdir</span> <span class="o">=</span> <span class="bp">None</span>
<span class="kn">depends:</span> <span class="n">executable</span><span class="p">(</span><span class="s1">&#39;king&#39;</span><span class="p">),</span> <span class="n">phenotype</span>
<span class="kn">input:</span> <span class="kp">pattern</span> <span class="o">=</span> <span class="s2">&quot;{name}.{ext}&quot;</span>
<span class="kn">output:</span> <span class="n">expand_pattern</span><span class="p">(</span><span class="s1">&#39;{name}.pc.ped&#39;</span><span class="p">)</span>
<span class="kn">task:</span> <span class="kp">workdir</span> <span class="o">=</span> <span class="kp">workdir</span>
</pre></div>
</td></tr>
<tr><td class="blob-num js-line-number"><div class="linenodiv"><pre>178</pre></div></td><td class="code"><div class="source blob-code sos-directive"><pre><span></span><span class="kn">run:</span>
</pre></div>
</td></tr>
<tr><td class="blob-num js-line-number"><div class="linenodiv"><pre>179
180
181
182
183</pre></div></td><td class="code"><div class="source blob-code sos-script "><pre><span></span><span class="c1"># king -b ${input!n}.bed --pca 20 --prefix ${input!n}.</span>
<span class="c1"># Would be much faster if I use MDS</span>
<span class="c1"># One should use PCA to be more comparable to GTEx official results, though</span>
king -b <span class="si">${</span><span class="nv">input</span><span class="p">!n</span><span class="si">}</span>.bed --mds --prefix <span class="si">${</span><span class="nv">input</span><span class="p">!n</span><span class="si">}</span>.
 
</pre></div>
</td></tr>
<tr><td class="blob-num js-line-number"><div class="linenodiv"><pre>184</pre></div></td><td class="code"><div class="source blob-code sos-directive"><pre><span></span><span class="kn">python:</span>
</pre></div>
</td></tr>
<tr><td class="blob-num js-line-number"><div class="linenodiv"><pre>185
186
187
188
189
190
191
192
193</pre></div></td><td class="code"><div class="source blob-code sos-script "><pre><span></span><span class="kn">import</span> <span class="nn">pandas</span> <span class="kn">as</span> <span class="nn">pd</span>
<span class="n">reference</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="err">$</span><span class="p">{</span><span class="n">phenotype</span><span class="err">!</span><span class="n">r</span><span class="p">},</span> <span class="n">dtype</span> <span class="o">=</span> <span class="nb">str</span><span class="p">,</span> <span class="n">usecols</span> <span class="o">=</span> <span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="mi">4</span><span class="p">),</span> <span class="n">sep</span> <span class="o">=</span> <span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">)</span>
<span class="n">reference</span><span class="p">[</span><span class="s1">&#39;RACE&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39;POP&#39;</span> <span class="o">+</span> <span class="n">reference</span><span class="p">[</span><span class="s1">&#39;RACE&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">str</span><span class="p">)</span>
<span class="n">fam</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="err">$</span><span class="p">{</span><span class="n">output</span><span class="err">!</span><span class="n">r</span><span class="p">},</span> <span class="n">header</span> <span class="o">=</span> <span class="bp">None</span><span class="p">,</span> <span class="n">sep</span> <span class="o">=</span> <span class="s1">&#39; &#39;</span><span class="p">,</span>
                  <span class="n">names</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;fid&#39;</span><span class="p">,</span> <span class="s1">&#39;sid&#39;</span><span class="p">,</span> <span class="s1">&#39;pid&#39;</span><span class="p">,</span> <span class="s1">&#39;mid&#39;</span><span class="p">,</span> <span class="s1">&#39;sex&#39;</span><span class="p">,</span> <span class="s1">&#39;phen&#39;</span><span class="p">]</span> <span class="o">+</span> <span class="p">[</span><span class="s1">&#39;PC{}&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">x</span><span class="o">+</span><span class="mi">1</span><span class="p">)</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">20</span><span class="p">)])</span>
<span class="n">dat</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">merge</span><span class="p">(</span><span class="n">reference</span><span class="p">,</span> <span class="n">fam</span><span class="p">,</span> <span class="n">left_on</span> <span class="o">=</span> <span class="s1">&#39;SUBJID&#39;</span><span class="p">,</span> <span class="n">right_on</span> <span class="o">=</span> <span class="s2">&quot;sid&quot;</span><span class="p">)</span>
<span class="n">dat</span><span class="o">.</span><span class="n">drop</span><span class="p">([</span><span class="s1">&#39;SUBJID&#39;</span><span class="p">,</span> <span class="s1">&#39;fid&#39;</span><span class="p">],</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">inplace</span> <span class="o">=</span> <span class="bp">True</span><span class="p">)</span>
<span class="n">dat</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="err">$</span><span class="p">{</span><span class="n">output</span><span class="err">!</span><span class="n">r</span><span class="p">},</span> <span class="n">sep</span> <span class="o">=</span> <span class="s1">&#39; &#39;</span><span class="p">,</span> <span class="n">header</span> <span class="o">=</span> <span class="bp">False</span><span class="p">,</span> <span class="n">index</span> <span class="o">=</span> <span class="bp">False</span><span class="p">)</span>
 
</pre></div>
</td></tr>
<tr><td class="blob-num js-line-number"><div class="linenodiv"><pre>194</pre></div></td><td class="code"><div class="source blob-code sos-header"><pre><span></span><span class="p">[</span><span class="n">global_ancestry_2</span><span class="p">]</span>
</pre></div>
</td></tr>
<tr><td class="blob-num js-line-number"><div class="linenodiv"><pre>195</pre></div></td><td class="code"><div class="source blob-code sos-comment"><pre><span></span><span class="c1"># scatter plot for global ancestry analysis result</span>
</pre></div>
</td></tr>
<tr><td class="blob-num js-line-number"><div class="linenodiv"><pre>196</pre></div></td><td class="code"><div class="source blob-code sos-comment"><pre><span></span><span class="c1"># R plotly implementation</span>
</pre></div>
</td></tr><tr><td class="blob-num js-line-number"><div class="linenodiv"><pre>197
198
199
200
201</pre></div></td><td class="code"><div class="source blob-code sos-directive"><pre><span></span><span class="kn">parameter:</span> <span class="kp">workdir</span> <span class="o">=</span> <span class="bp">None</span>
<span class="kn">depends:</span> <span class="n">R_library</span><span class="p">(</span><span class="s2">&quot;plotly&quot;</span><span class="p">)</span>
<span class="kn">input:</span> <span class="kp">pattern</span> <span class="o">=</span> <span class="s2">&quot;{name}.ped&quot;</span> 
<span class="kn">output:</span> <span class="n">expand_pattern</span><span class="p">(</span><span class="s1">&#39;{name}.html&#39;</span><span class="p">)</span>
<span class="kn">task:</span> <span class="kp">workdir</span> <span class="o">=</span> <span class="kp">workdir</span> 
</pre></div>
</td></tr>
<tr><td class="blob-num js-line-number"><div class="linenodiv"><pre>202</pre></div></td><td class="code"><div class="source blob-code sos-directive"><pre><span></span><span class="kn">R:</span> 
</pre></div>
</td></tr>
<tr><td class="blob-num js-line-number"><div class="linenodiv"><pre>203
204
205
206
207
208
209
210</pre></div></td><td class="code"><div class="source blob-code sos-script "><pre><span></span><span class="kn">library</span><span class="p">(</span>plotly<span class="p">)</span>
dat <span class="o">&lt;-</span> read.table<span class="p">(</span><span class="o">$</span><span class="p">{</span>input<span class="o">!</span>r<span class="p">},</span> sep <span class="o">=</span> <span class="s">&#39; &#39;</span><span class="p">)</span>
<span class="kr">for</span> <span class="p">(</span>i <span class="kr">in</span> <span class="m">1</span><span class="o">:</span><span class="p">(</span><span class="kp">ncol</span><span class="p">(</span>dat<span class="p">)</span> <span class="o">-</span> <span class="m">6</span> <span class="o">-</span> <span class="m">1</span><span class="p">))</span> <span class="p">{</span>
    title <span class="o">&lt;-</span> <span class="kp">paste</span><span class="p">(</span><span class="o">$</span><span class="p">{</span>input<span class="o">!</span>bnr<span class="p">},</span> <span class="s">&quot;PC&quot;</span><span class="p">,</span> i<span class="p">,</span> <span class="s">&quot;vs&quot;</span><span class="p">,</span> <span class="s">&quot;PC&quot;</span><span class="p">,</span> i <span class="o">+</span> <span class="m">1</span><span class="p">,</span> sep <span class="o">=</span> <span class="s">&quot;_&quot;</span><span class="p">)</span>
    p <span class="o">&lt;-</span> plot_ly<span class="p">(</span>dat<span class="p">,</span> x <span class="o">=</span> dat<span class="p">[,</span> <span class="m">6</span> <span class="o">+</span> i<span class="p">],</span> y <span class="o">=</span> dat<span class="p">[,</span> <span class="m">7</span> <span class="o">+</span> i<span class="p">],</span> text <span class="o">=</span> dat<span class="p">[,</span><span class="m">2</span><span class="p">],</span> color <span class="o">=</span> dat<span class="p">[,</span><span class="m">1</span><span class="p">],</span>
                 mode <span class="o">=</span> <span class="s">&quot;markers&quot;</span><span class="p">,</span> type <span class="o">=</span> <span class="s">&#39;scatter&#39;</span><span class="p">,</span> visible <span class="o">=</span> <span class="s">&#39;legendonly&#39;</span><span class="p">)</span> <span class="o">%&gt;%</span> layout<span class="p">(</span>title <span class="o">=</span> title<span class="p">)</span>
    htmlwidgets<span class="o">::</span>saveWidget<span class="p">(</span>as.widget<span class="p">(</span>p<span class="p">),</span> <span class="kp">paste0</span><span class="p">(</span>title<span class="p">,</span> <span class="s">&quot;.html&quot;</span><span class="p">))</span>
<span class="p">}</span>
</pre></div>
</td></tr>
<tr><td class="blob-num js-line-number"><div class="linenodiv"><pre>211</pre></div></td><td class="code"><div class="source blob-code sos-directive"><pre><span></span><span class="kn">run:</span>
</pre></div>
</td></tr>
<tr><td class="blob-num js-line-number"><div class="linenodiv"><pre>212
213
214
215
216
217
218
219</pre></div></td><td class="code"><div class="source blob-code sos-script "><pre><span></span>foo <span class="o">()</span> <span class="o">{</span>
  <span class="nb">echo</span> <span class="s1">&#39;&lt;html&gt;&lt;body&gt;&#39;</span>
  sed <span class="s1">&#39;s/^.*/&lt;a href=&quot;&amp;&quot;&gt;&amp;&lt;\/a&gt;&lt;br\/&gt;/&#39;</span>
  <span class="nb">echo</span> <span class="s1">&#39;&lt;/body&gt;&lt;/html&gt;&#39;</span>
<span class="o">}</span>
mkdir -p <span class="si">${</span><span class="nv">input</span><span class="p">!n</span><span class="si">}</span><span class="p">;</span> mv <span class="si">${</span><span class="nv">input</span><span class="p">!bn</span><span class="si">}</span>*.html <span class="si">${</span><span class="nv">input</span><span class="p">!n</span><span class="si">}</span>
ls <span class="si">${</span><span class="nv">input</span><span class="p">!n</span><span class="si">}</span>/*.html <span class="p">|</span> foo &gt; <span class="si">${</span><span class="nv">output</span><span class="si">}</span>
 
</pre></div>
</td></tr>
<tr><td class="blob-num js-line-number"><div class="linenodiv"><pre>220</pre></div></td><td class="code"><div class="source blob-code sos-header"><pre><span></span><span class="p">[</span><span class="n">ensembl_annotation</span><span class="p">]</span>
</pre></div>
</td></tr>
<tr><td class="blob-num js-line-number"><div class="linenodiv"><pre>221</pre></div></td><td class="code"><div class="source blob-code sos-comment"><pre><span></span><span class="c1"># Annotate gene to chromosomal positions</span>
</pre></div>
</td></tr>
<tr><td class="blob-num js-line-number"><div class="linenodiv"><pre>222</pre></div></td><td class="code"><div class="source blob-code sos-comment"><pre><span></span><span class="c1"># Input is GTEx expression file with first column the ENSG IDs</span>
</pre></div>
</td></tr><tr><td class="blob-num js-line-number"><div class="linenodiv"><pre>223
224
225</pre></div></td><td class="code"><div class="source blob-code sos-directive"><pre><span></span><span class="kn">depends:</span> <span class="n">R_library</span><span class="p">(</span><span class="s2">&quot;biomaRt&quot;</span><span class="p">)</span>
<span class="kn">output:</span> <span class="s2">&quot;${input!n}.annotation&quot;</span>
<span class="kn">task:</span> <span class="kp">workdir</span> <span class="o">=</span> <span class="kp">workdir</span>
</pre></div>
</td></tr>
<tr><td class="blob-num js-line-number"><div class="linenodiv"><pre>226</pre></div></td><td class="code"><div class="source blob-code sos-directive"><pre><span></span><span class="kn">R:</span>
</pre></div>
</td></tr>
<tr><td class="blob-num js-line-number"><div class="linenodiv"><pre>227
228
229
230
231
232</pre></div></td><td class="code"><div class="source blob-code sos-script "><pre><span></span><span class="kp">system</span><span class="p">(</span><span class="s">&quot;zcat ${input} | cut -f 1 | tail -n+4 | cut -f 1 -d &#39;.&#39; &gt; ${input!n}.gene_id&quot;</span><span class="p">)</span>
values <span class="o">=</span> <span class="kp">unlist</span><span class="p">(</span>read.table<span class="p">(</span><span class="s">&quot;${input!n}.gene_id&quot;</span><span class="p">,</span> stringsAsFactors<span class="o">=</span><span class="kc">FALSE</span><span class="p">))</span> 
ensembl <span class="o">=</span> biomaRt<span class="o">::</span>useMart<span class="p">(</span><span class="s">&quot;ensembl&quot;</span><span class="p">,</span> dataset<span class="o">=</span><span class="s">&quot;hsapiens_gene_ensembl&quot;</span><span class="p">,</span> host<span class="o">=</span><span class="s">&quot;www.ensembl.org&quot;</span><span class="p">)</span>
res <span class="o">=</span> biomaRt<span class="o">::</span>getBM<span class="p">(</span>attributes <span class="o">=</span> <span class="kt">c</span><span class="p">(</span><span class="s">&quot;chromosome_name&quot;</span><span class="p">,</span> <span class="s">&quot;start_position&quot;</span><span class="p">,</span> <span class="s">&quot;end_position&quot;</span><span class="p">,</span> <span class="s">&quot;ensembl_gene_id&quot;</span><span class="p">,</span> <span class="s">&quot;hgnc_symbol&quot;</span><span class="p">),</span> filters <span class="o">=</span> <span class="kt">c</span><span class="p">(</span><span class="s">&quot;ensembl_gene_id&quot;</span><span class="p">),</span> values <span class="o">=</span> values<span class="p">,</span>  mart <span class="o">=</span> ensembl<span class="p">)</span>
write.table<span class="p">(</span>res<span class="p">,</span> file <span class="o">=</span> <span class="o">$</span><span class="p">{</span>output<span class="o">!</span>r<span class="p">},</span> row.names<span class="o">=</span><span class="kc">FALSE</span><span class="p">,</span> na<span class="o">=</span> <span class="s">&quot;&lt;NA&gt;&quot;</span><span class="p">,</span> col.names <span class="o">=</span> <span class="kc">FALSE</span><span class="p">)</span>
 
</pre></div>
</td></tr>
<tr><td class="blob-num js-line-number"><div class="linenodiv"><pre>233</pre></div></td><td class="code"><div class="source blob-code sos-directive"><pre><span></span><span class="kn">run:</span>
</pre></div>
</td></tr>
<tr><td class="blob-num js-line-number"><div class="linenodiv"><pre>234
235
236</pre></div></td><td class="code"><div class="source blob-code sos-script "><pre><span></span>cat <span class="si">${</span><span class="nv">output</span><span class="si">}</span> <span class="p">|</span> sed -e <span class="s1">&#39;s/&quot;//g&#39;</span> <span class="p">|</span> sort -g -o <span class="si">${</span><span class="nv">output</span><span class="si">}</span>
rm -f <span class="si">${</span><span class="nv">input</span><span class="p">!n</span><span class="si">}</span>.gene_id
 
</pre></div>
</td></tr>
<tr><td class="blob-num js-line-number"><div class="linenodiv"><pre>237</pre></div></td><td class="code"><div class="source blob-code sos-header"><pre><span></span><span class="p">[</span><span class="n">genotype_LD_1</span><span class="p">]</span>
</pre></div>
</td></tr>
<tr><td class="blob-num js-line-number"><div class="linenodiv"><pre>238</pre></div></td><td class="code"><div class="source blob-code sos-comment"><pre><span></span><span class="c1"># Permute X &amp; Get LD structure</span>
</pre></div>
</td></tr><tr><td class="blob-num js-line-number"><div class="linenodiv"><pre>239
240
241
242
243
244
245</pre></div></td><td class="code"><div class="source blob-code sos-directive"><pre><span></span><span class="kn">parameter:</span> <span class="kp">workdir</span> <span class="o">=</span> <span class="bp">None</span>
<span class="kn">parameter:</span> <span class="n">src</span> <span class="o">=</span> <span class="bp">None</span>
<span class="kn">parameter:</span> <span class="n">genotype_data</span> <span class="o">=</span> <span class="bp">None</span>
<span class="kn">parameter:</span> <span class="n">permute_genotype</span> <span class="o">=</span> <span class="s2">&quot;False&quot;</span>
<span class="kn">input:</span> <span class="n">genotype_data</span><span class="p">,</span> <span class="kp">group_by</span> <span class="o">=</span> <span class="mi">1</span>
<span class="kn">output:</span> <span class="s2">&quot;${_input!an}.ld.h5&quot;</span> <span class="k">if</span> <span class="n">permute_genotype</span> <span class="o">==</span> <span class="s1">&#39;False&#39;</span> <span class="k">else</span> <span class="s2">&quot;${_input!an}.permuted.ld.h5&quot;</span>
<span class="kn">task:</span> <span class="kp">workdir</span> <span class="o">=</span> <span class="kp">workdir</span>
</pre></div>
</td></tr>
<tr><td class="blob-num js-line-number"><div class="linenodiv"><pre>246</pre></div></td><td class="code"><div class="source blob-code sos-directive"><pre><span></span><span class="kn">python:</span>
</pre></div>
</td></tr>
<tr><td class="blob-num js-line-number"><div class="linenodiv"><pre>247
248
249
250
251
252
253
254
255
256
257</pre></div></td><td class="code"><div class="source blob-code sos-script "><pre><span></span><span class="kn">import</span> <span class="nn">os</span><span class="o">,</span> <span class="nn">sys</span>
<span class="n">sys</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="err">$</span><span class="p">{</span><span class="n">src</span><span class="err">!</span><span class="n">ar</span><span class="p">})</span>
<span class="kn">from</span> <span class="nn">SimUtils</span> <span class="kn">import</span> <span class="n">PhenotypeSimulator</span><span class="p">,</span> <span class="n">BetaDist</span>
<span class="n">ms</span> <span class="o">=</span> <span class="n">PhenotypeSimulator</span><span class="p">(</span><span class="err">$</span><span class="p">{</span><span class="n">_input</span><span class="err">!</span><span class="n">ar</span><span class="p">})</span>
<span class="n">tables</span> <span class="o">=</span> <span class="n">ms</span><span class="o">.</span><span class="n">get_genes</span><span class="p">()</span>
<span class="k">if</span> <span class="err">$</span><span class="p">{</span><span class="n">permute_genotype</span><span class="p">}:</span>
   <span class="n">ms</span><span class="o">.</span><span class="n">permute_X</span><span class="p">(</span><span class="n">tables</span><span class="p">,</span> <span class="n">save_to</span> <span class="o">=</span> <span class="s1">&#39;${_output!nn}.h5&#39;</span><span class="p">)</span>
<span class="n">ld</span> <span class="o">=</span> <span class="n">ms</span><span class="o">.</span><span class="n">get_ld</span><span class="p">(</span><span class="n">tables</span><span class="p">,</span> <span class="n">save_to</span> <span class="o">=</span> <span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="err">!</span><span class="n">r</span><span class="p">})</span>
<span class="k">for</span> <span class="n">table</span> <span class="ow">in</span> <span class="n">tables</span><span class="p">:</span>
    <span class="n">ms</span><span class="o">.</span><span class="n">ld_heatmap</span><span class="p">(</span><span class="n">ld</span><span class="p">[</span><span class="n">table</span><span class="p">]</span><span class="o">.</span><span class="n">iloc</span><span class="p">[:</span><span class="mi">1000</span><span class="p">,:</span><span class="mi">1000</span><span class="p">],</span> <span class="s1">&#39;{}.ld.png&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">basename</span><span class="p">(</span><span class="n">table</span><span class="p">)</span> <span class="o">+</span> <span class="p">(</span><span class="s1">&#39;.permuted&#39;</span> <span class="k">if</span> <span class="err">$</span><span class="p">{</span><span class="n">permute_genotype</span><span class="p">}</span> <span class="k">else</span> <span class="s1">&#39;&#39;</span><span class="p">)))</span>
 
</pre></div>
</td></tr>
<tr><td class="blob-num js-line-number"><div class="linenodiv"><pre>258</pre></div></td><td class="code"><div class="source blob-code sos-header"><pre><span></span><span class="p">[</span><span class="n">simulation_1</span><span class="p">]</span>
</pre></div>
</td></tr>
<tr><td class="blob-num js-line-number"><div class="linenodiv"><pre>259</pre></div></td><td class="code"><div class="source blob-code sos-comment"><pre><span></span><span class="c1"># Generate simulated data</span>
</pre></div>
</td></tr><tr><td class="blob-num js-line-number"><div class="linenodiv"><pre>260
261
262
263
264
265
266
267
268
269
270</pre></div></td><td class="code"><div class="source blob-code sos-directive"><pre><span></span><span class="kn">parameter:</span> <span class="kp">workdir</span> <span class="o">=</span> <span class="bp">None</span>
<span class="kn">parameter:</span> <span class="n">src</span> <span class="o">=</span> <span class="bp">None</span>
<span class="kn">parameter:</span> <span class="n">pi0</span> <span class="o">=</span> <span class="bp">None</span> 
<span class="kn">parameter:</span> <span class="n">shape</span> <span class="o">=</span> <span class="bp">None</span>
<span class="kn">parameter:</span> <span class="n">n_rep</span> <span class="o">=</span> <span class="bp">None</span>
<span class="kn">parameter:</span> <span class="n">replicate</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span> <span class="o">+</span> <span class="mi">1</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">n_rep</span><span class="p">)]</span>
<span class="kn">parameter:</span> <span class="n">genotype_data</span> <span class="o">=</span> <span class="bp">None</span>
<span class="kn">parameter:</span> <span class="n">permuted_genotype</span> <span class="o">=</span> <span class="s2">&quot;False&quot;</span>
<span class="kn">input:</span> <span class="n">genotype_data</span> <span class="k">if</span> <span class="n">permuted_genotype</span> <span class="o">==</span> <span class="s1">&#39;False&#39;</span> <span class="k">else</span> <span class="p">[</span><span class="n">x</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;.h5&quot;</span><span class="p">,</span> <span class="s2">&quot;.permuted.h5&quot;</span><span class="p">)</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">genotype_data</span><span class="p">],</span> <span class="kp">group_by</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="kp">for_each</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;pi0&#39;</span><span class="p">,</span> <span class="s1">&#39;shape&#39;</span><span class="p">,</span> <span class="s1">&#39;replicate&#39;</span><span class="p">]</span>
<span class="kn">output:</span> <span class="s2">&quot;${workdir}/${_input!bn}.&quot;</span> <span class="o">+</span> <span class="s2">&quot;${_pi0}_${_shape}_${_replicate}&quot;</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s1">&#39;.&#39;</span><span class="p">,</span> <span class="s1">&#39;p&#39;</span><span class="p">)</span> <span class="o">+</span> <span class="s1">&#39;.expr.h5&#39;</span>
<span class="kn">task:</span> <span class="kp">workdir</span> <span class="o">=</span> <span class="kp">workdir</span>
</pre></div>
</td></tr>
<tr><td class="blob-num js-line-number"><div class="linenodiv"><pre>271</pre></div></td><td class="code"><div class="source blob-code sos-directive"><pre><span></span><span class="kn">python:</span>
</pre></div>
</td></tr>
<tr><td class="blob-num js-line-number"><div class="linenodiv"><pre>272
273
274
275
276
277
278
279
280
281
282
283
284
285
286
287
288
289
290
291
292
293
294
295
296
297</pre></div></td><td class="code"><div class="source blob-code sos-script "><pre><span></span><span class="kn">import</span> <span class="nn">sys</span><span class="o">,</span> <span class="nn">os</span><span class="o">,</span> <span class="nn">json</span><span class="o">,</span> <span class="nn">pandas</span> <span class="kn">as</span> <span class="nn">pd</span>
<span class="n">sys</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="err">$</span><span class="p">{</span><span class="n">src</span><span class="err">!</span><span class="n">ar</span><span class="p">})</span>
<span class="kn">from</span> <span class="nn">SimUtils</span> <span class="kn">import</span> <span class="n">PhenotypeSimulator</span><span class="p">,</span> <span class="n">BetaDist</span>
<span class="n">ms</span> <span class="o">=</span> <span class="n">PhenotypeSimulator</span><span class="p">(</span><span class="err">$</span><span class="p">{</span><span class="n">_input</span><span class="err">!</span><span class="n">ar</span><span class="p">})</span>
<span class="n">tables</span> <span class="o">=</span> <span class="n">ms</span><span class="o">.</span><span class="n">get_genes</span><span class="p">()</span>
<span class="n">ld</span> <span class="o">=</span> <span class="n">ms</span><span class="o">.</span><span class="n">load_ld</span><span class="p">(</span><span class="n">tables</span><span class="p">,</span> <span class="s2">&quot;${_input!an}.ld.h5&quot;</span><span class="p">)</span>
<span class="k">if</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">isfile</span><span class="p">(</span><span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="err">!</span><span class="n">r</span><span class="p">}):</span>
   <span class="n">os</span><span class="o">.</span><span class="n">remove</span><span class="p">(</span><span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="err">!</span><span class="n">r</span><span class="p">})</span>
<span class="n">param</span> <span class="o">=</span> <span class="n">BetaDist</span><span class="p">()</span>  
<span class="n">param</span><span class="o">.</span><span class="n">set_pi0</span><span class="p">(</span><span class="err">$</span><span class="p">{</span><span class="n">_pi0</span><span class="p">})</span>
<span class="n">param</span><span class="o">.</span><span class="n">set_</span><span class="err">$</span><span class="p">{</span><span class="n">_shape</span><span class="p">}()</span>
<span class="k">print</span><span class="p">(</span><span class="n">param</span><span class="p">)</span>
<span class="k">for</span> <span class="n">table</span> <span class="ow">in</span> <span class="n">tables</span><span class="p">:</span>
  <span class="n">ms</span><span class="o">.</span><span class="n">set_id</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">basename</span><span class="p">(</span><span class="n">table</span><span class="p">))</span>
  <span class="n">nbeta</span> <span class="o">=</span> <span class="n">ld</span><span class="p">[</span><span class="n">table</span><span class="p">]</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
  <span class="n">beta</span> <span class="o">=</span> <span class="n">ms</span><span class="o">.</span><span class="n">generate_betamix</span><span class="p">(</span><span class="n">nbeta</span><span class="o">=</span><span class="n">nbeta</span><span class="p">,</span><span class="n">pi0</span><span class="o">=</span><span class="n">param</span><span class="o">.</span><span class="n">pi0</span><span class="p">,</span> <span class="n">pis</span><span class="o">=</span><span class="n">param</span><span class="o">.</span><span class="n">pis</span><span class="p">,</span> <span class="n">mus</span> <span class="o">=</span> <span class="n">param</span><span class="o">.</span><span class="n">mus</span><span class="p">,</span> <span class="n">sigmas</span><span class="o">=</span><span class="n">param</span><span class="o">.</span><span class="n">sigmas</span><span class="p">)</span>
  <span class="k">if</span> <span class="ow">not</span> <span class="err">$</span><span class="p">{</span><span class="n">permuted_genotype</span><span class="p">}:</span>
     <span class="n">strong_snps_idx</span> <span class="o">=</span> <span class="n">ms</span><span class="o">.</span><span class="n">select_convoluted_snps</span><span class="p">(</span><span class="n">ld</span><span class="p">[</span><span class="n">table</span><span class="p">])</span>
     <span class="n">beta</span> <span class="o">=</span> <span class="n">ms</span><span class="o">.</span><span class="n">swap_beta</span><span class="p">(</span><span class="n">beta</span><span class="p">,</span> <span class="n">strong_snps_idx</span><span class="p">)</span>
  <span class="n">X</span> <span class="o">=</span> <span class="n">ms</span><span class="o">.</span><span class="n">get_X</span><span class="p">(</span><span class="n">table</span><span class="o">=</span><span class="n">table</span><span class="p">)</span>
  <span class="n">y</span> <span class="o">=</span> <span class="n">ms</span><span class="o">.</span><span class="n">generate_y</span><span class="p">(</span><span class="n">beta</span><span class="o">=</span><span class="n">beta</span><span class="p">,</span><span class="n">sigma</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">X</span><span class="o">=</span><span class="n">X</span><span class="p">,</span> <span class="n">force</span> <span class="o">=</span> <span class="bp">True</span><span class="p">)</span>
<span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">(</span><span class="n">ms</span><span class="o">.</span><span class="n">phenotype</span><span class="o">.</span><span class="n">values</span><span class="p">())</span><span class="o">.</span><span class="n">to_hdf</span><span class="p">(</span><span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="err">!</span><span class="n">r</span><span class="p">},</span> <span class="s1">&#39;/simulated&#39;</span><span class="p">,</span> <span class="n">mode</span> <span class="o">=</span> <span class="s1">&#39;a&#39;</span><span class="p">,</span> <span class="n">complevel</span> <span class="o">=</span> <span class="mi">9</span><span class="p">,</span> <span class="n">complib</span> <span class="o">=</span> <span class="s1">&#39;zlib&#39;</span><span class="p">)</span>
<span class="n">meta</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;pi&#39;</span><span class="p">:</span> <span class="n">param</span><span class="o">.</span><span class="n">pis</span><span class="p">,</span> <span class="s1">&#39;pi0&#39;</span><span class="p">:</span> <span class="n">param</span><span class="o">.</span><span class="n">pi0</span><span class="p">,</span> <span class="s1">&#39;sigma&#39;</span><span class="p">:</span> <span class="n">param</span><span class="o">.</span><span class="n">sigmas</span><span class="p">,</span> <span class="s1">&#39;mu&#39;</span><span class="p">:</span> <span class="n">param</span><span class="o">.</span><span class="n">mus</span><span class="p">,</span> <span class="s1">&#39;beta&#39;</span><span class="p">:</span> <span class="n">ms</span><span class="o">.</span><span class="n">beta</span><span class="p">,</span> <span class="s1">&#39;name&#39;</span><span class="p">:</span> <span class="err">$</span><span class="p">{</span><span class="n">_shape</span><span class="err">!</span><span class="n">r</span><span class="p">}}</span>
<span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="s2">&quot;${_output!n}.json&quot;</span><span class="p">,</span> <span class="s1">&#39;w&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">fp</span><span class="p">:</span>
  <span class="n">json</span><span class="o">.</span><span class="n">dump</span><span class="p">(</span><span class="n">meta</span><span class="p">,</span> <span class="n">fp</span><span class="p">)</span>
 
</pre></div>
</td></tr>
<tr><td class="blob-num js-line-number"><div class="linenodiv"><pre>298</pre></div></td><td class="code"><div class="source blob-code sos-header"><pre><span></span><span class="p">[</span><span class="n">simulation_2</span><span class="p">]</span>
</pre></div>
</td></tr>
<tr><td class="blob-num js-line-number"><div class="linenodiv"><pre>299</pre></div></td><td class="code"><div class="source blob-code sos-comment"><pre><span></span><span class="c1"># Analyze data</span>
</pre></div>
</td></tr><tr><td class="blob-num js-line-number"><div class="linenodiv"><pre>300
301
302
303
304</pre></div></td><td class="code"><div class="source blob-code sos-directive"><pre><span></span><span class="kn">parameter:</span> <span class="n">genotype_data</span> <span class="o">=</span> <span class="bp">None</span>
<span class="kn">parameter:</span> <span class="kp">workdir</span> <span class="o">=</span> <span class="bp">None</span>
<span class="kn">input:</span> <span class="kp">group_by</span> <span class="o">=</span> <span class="mi">1</span>
<span class="kn">output:</span> <span class="s2">&quot;${_input!n}.analyzed.rds&quot;</span>
<span class="kn">task:</span> <span class="kp">workdir</span> <span class="o">=</span> <span class="kp">workdir</span>
</pre></div>
</td></tr>
<tr><td class="blob-num js-line-number"><div class="linenodiv"><pre>305</pre></div></td><td class="code"><div class="source blob-code sos-directive"><pre><span></span><span class="kn">R:</span>
</pre></div>
</td></tr>
<tr><td class="blob-num js-line-number"><div class="linenodiv"><pre>306
307
308
309
310
311
312
313
314
315
316
317
318
319
320
321
322
323
324
325
326
327
328
329
330
331
332</pre></div></td><td class="code"><div class="source blob-code sos-script "><pre><span></span><span class="kn">library</span><span class="p">(</span>rhdf5<span class="p">)</span>
<span class="kn">source</span><span class="p">(</span><span class="s">&quot;src/Utils.R&quot;</span><span class="p">)</span>
genotype_file <span class="o">=</span> <span class="s">&quot;${genotype_data[0]!ad}/${_input!bnnn}.h5&quot;</span>
expr_file <span class="o">=</span> <span class="o">$</span><span class="p">{</span>_input<span class="o">!</span>r<span class="p">}</span>
expr_table <span class="o">=</span> <span class="s">&#39;/simulated&#39;</span>
tables <span class="o">=</span> get_h5_groups<span class="p">(</span>genotype_file<span class="p">)</span>
res <span class="o">=</span> <span class="kt">list</span><span class="p">()</span>
<span class="kr">for</span> <span class="p">(</span>table <span class="kr">in</span> tables<span class="p">)</span> <span class="p">{</span>
    dat <span class="o">=</span> load_data<span class="p">(</span>genotype_file<span class="p">,</span> expr_file<span class="p">,</span> <span class="kp">table</span><span class="p">,</span> expr_table<span class="p">)</span>
    X <span class="o">=</span> <span class="kp">as.matrix</span><span class="p">(</span>dat<span class="o">$</span>X<span class="p">)</span>
    <span class="kp">storage.mode</span><span class="p">(</span>X<span class="p">)</span> <span class="o">&lt;-</span> <span class="s">&quot;double&quot;</span>
    y <span class="o">=</span> <span class="kp">as.vector</span><span class="p">(</span>dat<span class="o">$</span>y<span class="p">)</span>
    <span class="c1"># Univariate analysis</span>
    res0 <span class="o">=</span> univariate_lm<span class="p">(</span>X<span class="p">,</span>y<span class="p">)</span>
    mixsd <span class="o">=</span> ashr<span class="o">:::</span>autoselect.mixsd<span class="p">(</span>res0<span class="p">,</span> <span class="kp">sqrt</span><span class="p">(</span><span class="m">2</span><span class="p">),</span> <span class="m">0</span><span class="p">,</span> <span class="kt">c</span><span class="p">(</span><span class="o">-</span><span class="kc">Inf</span><span class="p">,</span><span class="kc">Inf</span><span class="p">),</span> <span class="s">&quot;normal&quot;</span><span class="p">)</span>
    res1 <span class="o">=</span> varbvs<span class="o">::</span>varbvsmix<span class="p">(</span>X<span class="p">,</span> <span class="kc">NULL</span><span class="p">,</span> y<span class="p">,</span> sa <span class="o">=</span> <span class="kt">c</span><span class="p">(</span><span class="m">0</span><span class="p">,</span> mixsd<span class="o">^</span><span class="m">2</span><span class="p">))</span>
    res1<span class="o">$</span>pip <span class="o">=</span> res1<span class="o">$</span>alpha <span class="o">%*%</span> <span class="kt">c</span><span class="p">(</span>res1<span class="o">$</span>w<span class="p">)</span>
    res1<span class="o">$</span>beta <span class="o">=</span> <span class="kp">rowSums</span><span class="p">(</span>res1<span class="o">$</span>alpha <span class="o">*</span> res1<span class="o">$</span>mu<span class="p">)</span>
    res1<span class="o">$</span>y <span class="o">=</span> X <span class="o">%*%</span> res1<span class="o">$</span>beta <span class="o">+</span> res1<span class="o">$</span>mu.cov<span class="p">[</span><span class="m">1</span><span class="p">]</span>
    meta <span class="o">=</span> rjson<span class="o">::</span>fromJSON<span class="p">(</span>file <span class="o">=</span> <span class="s">&quot;${_input!n}.json&quot;</span><span class="p">)</span>
    meta<span class="o">$</span>y <span class="o">=</span> y
    meta<span class="o">$</span>beta <span class="o">=</span> meta<span class="o">$</span><span class="kp">beta</span><span class="p">[[</span><span class="kp">basename</span><span class="p">(</span><span class="kp">table</span><span class="p">)]]</span>
    res2 <span class="o">=</span> ashr<span class="o">::</span>ash<span class="p">(</span>res0<span class="o">$</span>x<span class="p">,</span> res0<span class="o">$</span>s<span class="p">)</span>
    res<span class="p">[[</span><span class="kp">basename</span><span class="p">(</span><span class="kp">table</span><span class="p">)]]</span> <span class="o">=</span> <span class="kt">list</span><span class="p">(</span>uni<span class="o">=</span>res0<span class="p">,</span>mr.ash<span class="o">=</span>res1<span class="p">,</span>ash<span class="o">=</span>res2<span class="p">,</span>meta<span class="o">=</span>meta<span class="p">)</span>
<span class="p">}</span>
<span class="kp">saveRDS</span><span class="p">(</span>res<span class="p">,</span> file <span class="o">=</span> <span class="o">$</span><span class="p">{</span>_output<span class="o">!</span>r<span class="p">})</span>
 
</pre></div>
</td></tr>
<tr><td class="blob-num js-line-number"><div class="linenodiv"><pre>333</pre></div></td><td class="code"><div class="source blob-code sos-header"><pre><span></span><span class="p">[</span><span class="n">simulation_3</span><span class="p">]</span>
</pre></div>
</td></tr>
<tr><td class="blob-num js-line-number"><div class="linenodiv"><pre>334</pre></div></td><td class="code"><div class="source blob-code sos-comment"><pre><span></span><span class="c1"># Make plots</span>
</pre></div>
</td></tr><tr><td class="blob-num js-line-number"><div class="linenodiv"><pre>335
336
337</pre></div></td><td class="code"><div class="source blob-code sos-directive"><pre><span></span><span class="kn">input:</span> <span class="kp">group_by</span> <span class="o">=</span> <span class="mi">1</span>
<span class="kn">output:</span> <span class="s2">&quot;${_input!n}.pdf&quot;</span>
<span class="kn">task:</span> <span class="kp">workdir</span> <span class="o">=</span> <span class="kp">workdir</span>
</pre></div>
</td></tr>
<tr><td class="blob-num js-line-number"><div class="linenodiv"><pre>338</pre></div></td><td class="code"><div class="source blob-code sos-directive"><pre><span></span><span class="kn">R:</span>
</pre></div>
</td></tr>
<tr><td class="blob-num js-line-number"><div class="linenodiv"><pre>339
340
341
342
343
344
345
346
347
348
349
350
351
352
353
354
355
356
357
358
359
360</pre></div></td><td class="code"><div class="source blob-code sos-script "><pre><span></span>dat <span class="o">=</span> <span class="kp">readRDS</span><span class="p">(</span><span class="o">$</span><span class="p">{</span>_input<span class="o">!</span>r<span class="p">})</span>
pdf<span class="p">(</span><span class="o">$</span><span class="p">{</span>_output<span class="o">!</span>r<span class="p">},</span> title <span class="o">=</span> <span class="s">&quot;${_input!bnn}&quot;</span><span class="p">)</span>
<span class="kr">for</span> <span class="p">(</span>table <span class="kr">in</span> <span class="kp">names</span><span class="p">(</span>dat<span class="p">))</span> <span class="p">{</span>
    res <span class="o">=</span> dat<span class="p">[[</span><span class="kp">table</span><span class="p">]]</span>
    <span class="c1"># Fitted mixture porportions</span>
    par<span class="p">(</span>mfrow <span class="o">=</span> <span class="kt">c</span><span class="p">(</span><span class="m">1</span><span class="p">,</span><span class="m">3</span><span class="p">),</span> oma<span class="o">=</span><span class="kt">c</span><span class="p">(</span><span class="m">0</span><span class="p">,</span><span class="m">0</span><span class="p">,</span><span class="m">2</span><span class="p">,</span><span class="m">0</span><span class="p">))</span>
    barplot<span class="p">(</span><span class="kt">c</span><span class="p">(</span>res<span class="o">$</span>meta<span class="o">$</span>pi0<span class="p">,</span> res<span class="o">$</span>meta<span class="o">$</span><span class="kc">pi</span> <span class="o">*</span> <span class="p">(</span><span class="m">1</span> <span class="o">-</span> res<span class="o">$</span>meta<span class="o">$</span>pi0<span class="p">)),</span> col<span class="o">=</span>rgb<span class="p">(</span><span class="m">0.2</span><span class="p">,</span><span class="m">0.4</span><span class="p">,</span><span class="m">0.6</span><span class="p">,</span><span class="m">0.6</span><span class="p">),</span> horiz<span class="o">=</span><span class="bp">T</span><span class="p">,</span> las<span class="o">=</span><span class="m">1</span><span class="p">,</span> main <span class="o">=</span> <span class="s">&#39;truth&#39;</span><span class="p">)</span>
    barplot<span class="p">(</span>res<span class="o">$</span>mr.ash<span class="o">$</span>w<span class="p">,</span> col<span class="o">=</span>rgb<span class="p">(</span><span class="m">0.2</span><span class="p">,</span><span class="m">0.4</span><span class="p">,</span><span class="m">0.6</span><span class="p">,</span><span class="m">0.6</span><span class="p">),</span> horiz<span class="o">=</span><span class="bp">T</span><span class="p">,</span> las<span class="o">=</span><span class="m">1</span><span class="p">,</span> main <span class="o">=</span> <span class="s">&#39;mr-ash, fitted g&#39;</span><span class="p">)</span>
    barplot<span class="p">(</span>ashr<span class="o">::</span>get_fitted_g<span class="p">(</span>res<span class="o">$</span>ash<span class="p">)</span><span class="o">$</span><span class="kc">pi</span><span class="p">,</span> col<span class="o">=</span>rgb<span class="p">(</span><span class="m">0.2</span><span class="p">,</span><span class="m">0.4</span><span class="p">,</span><span class="m">0.6</span><span class="p">,</span><span class="m">0.6</span><span class="p">),</span> horiz<span class="o">=</span><span class="bp">T</span><span class="p">,</span> las<span class="o">=</span><span class="m">1</span><span class="p">,</span> main <span class="o">=</span> <span class="s">&#39;ash, fitted g&#39;</span><span class="p">)</span>
    title<span class="p">(</span><span class="kp">paste0</span><span class="p">(</span><span class="kp">table</span><span class="p">,</span> <span class="s">&quot; &quot;</span><span class="p">,</span> <span class="kp">ifelse</span><span class="p">(</span><span class="kp">grepl</span><span class="p">(</span><span class="s">&#39;permuted&#39;</span><span class="p">,</span> <span class="o">$</span><span class="p">{</span>_input<span class="o">!</span>r<span class="p">}),</span> <span class="s">&quot;permuted&quot;</span><span class="p">,</span> <span class="s">&quot;original&quot;</span><span class="p">),</span> <span class="s">&quot; genotype, &quot;</span><span class="p">,</span> <span class="kp">round</span><span class="p">(</span>res<span class="o">$</span>meta<span class="o">$</span>pi0<span class="p">,</span><span class="m">4</span><span class="p">),</span> <span class="s">&quot; pi0 + &quot;</span><span class="p">,</span> res<span class="o">$</span>meta<span class="o">$</span>name<span class="p">,</span> <span class="s">&quot; mixture&quot;</span><span class="p">),</span> outer<span class="o">=</span><span class="kc">TRUE</span><span class="p">)</span>
    <span class="c1"># Compare fit</span>
    par<span class="p">(</span>mfrow <span class="o">=</span> <span class="kt">c</span><span class="p">(</span><span class="m">3</span><span class="p">,</span><span class="m">2</span><span class="p">),</span> mar <span class="o">=</span> <span class="kt">c</span><span class="p">(</span><span class="m">2</span><span class="p">,</span><span class="m">2</span><span class="p">,</span><span class="m">2</span><span class="p">,</span><span class="m">1</span><span class="p">))</span>
    plot<span class="p">(</span>res<span class="o">$</span>meta<span class="o">$</span><span class="kp">beta</span><span class="p">,</span> pch<span class="o">=</span><span class="m">20</span><span class="p">,</span> bg<span class="o">=</span><span class="s">&quot;white&quot;</span><span class="p">,</span> col<span class="o">=</span>rgb<span class="p">(</span><span class="m">0.3</span><span class="p">,</span><span class="m">0.6</span><span class="p">,</span><span class="m">0.9</span><span class="p">,</span><span class="m">0.9</span><span class="p">),</span> ylab <span class="o">=</span> <span class="s">&quot;beta&quot;</span><span class="p">,</span> main <span class="o">=</span> <span class="s">&#39;truth, beta&#39;</span><span class="p">)</span>
    plot<span class="p">(</span>res<span class="o">$</span>mr.ash<span class="o">$</span><span class="kp">beta</span><span class="p">,</span> pch<span class="o">=</span><span class="m">20</span><span class="p">,</span> bg<span class="o">=</span><span class="s">&quot;white&quot;</span><span class="p">,</span> col<span class="o">=</span>rgb<span class="p">(</span><span class="m">0.3</span><span class="p">,</span><span class="m">0.6</span><span class="p">,</span><span class="m">0.9</span><span class="p">,</span><span class="m">0.9</span><span class="p">),</span> ylab <span class="o">=</span> <span class="s">&quot;betahat&quot;</span><span class="p">,</span> main <span class="o">=</span> <span class="s">&#39;mr-ash, betahat&#39;</span><span class="p">)</span>
    legend<span class="p">(</span><span class="s">&quot;topleft&quot;</span><span class="p">,</span> <span class="kp">paste0</span><span class="p">(</span><span class="s">&quot;sigma = &quot;</span><span class="p">,</span> <span class="kp">round</span><span class="p">(</span>res<span class="o">$</span>mr.ash<span class="o">$</span>sigma<span class="p">,</span><span class="m">2</span><span class="p">)),</span> bty<span class="o">=</span><span class="s">&quot;n&quot;</span><span class="p">)</span> 
    plot<span class="p">(</span>ashr<span class="o">::</span>get_pm<span class="p">(</span>res<span class="o">$</span>ash<span class="p">),</span> pch<span class="o">=</span><span class="m">20</span><span class="p">,</span> bg<span class="o">=</span><span class="s">&quot;white&quot;</span><span class="p">,</span> col<span class="o">=</span>rgb<span class="p">(</span><span class="m">0.3</span><span class="p">,</span><span class="m">0.6</span><span class="p">,</span><span class="m">0.9</span><span class="p">,</span><span class="m">0.9</span><span class="p">),</span> ylab <span class="o">=</span> <span class="s">&quot;betahat&quot;</span><span class="p">,</span> main <span class="o">=</span> <span class="s">&#39;ash, betahat&#39;</span><span class="p">)</span>
    plot<span class="p">(</span>res<span class="o">$</span>uni<span class="o">$</span>x<span class="p">,</span> pch<span class="o">=</span><span class="m">20</span><span class="p">,</span> bg<span class="o">=</span><span class="s">&quot;white&quot;</span><span class="p">,</span> col<span class="o">=</span>rgb<span class="p">(</span><span class="m">0.3</span><span class="p">,</span><span class="m">0.6</span><span class="p">,</span><span class="m">0.9</span><span class="p">,</span><span class="m">0.9</span><span class="p">),</span> ylab <span class="o">=</span> <span class="s">&quot;betahat&quot;</span><span class="p">,</span> main <span class="o">=</span> <span class="s">&#39;univariate, betahat&#39;</span><span class="p">)</span>
    plot<span class="p">(</span>res<span class="o">$</span>meta<span class="o">$</span><span class="kp">beta</span><span class="p">,</span> res<span class="o">$</span>mr.ash<span class="o">$</span><span class="kp">beta</span><span class="p">,</span> pch<span class="o">=</span><span class="m">20</span><span class="p">,</span> bg<span class="o">=</span><span class="s">&quot;white&quot;</span><span class="p">,</span> col<span class="o">=</span>rgb<span class="p">(</span><span class="m">0.3</span><span class="p">,</span><span class="m">0.3</span><span class="p">,</span><span class="m">0.9</span><span class="p">,</span><span class="m">0.9</span><span class="p">),</span> ylab <span class="o">=</span> <span class="s">&quot;mr.ash&quot;</span><span class="p">,</span> xlab <span class="o">=</span> <span class="s">&quot;truth&quot;</span><span class="p">,</span> main <span class="o">=</span> <span class="s">&#39;truth vs mr.ash&#39;</span><span class="p">)</span>
    abline<span class="p">(</span><span class="m">0</span><span class="p">,</span><span class="m">1</span><span class="p">,</span> col<span class="o">=</span><span class="s">&quot;whitesmoke&quot;</span><span class="p">)</span>
    plot<span class="p">(</span>ashr<span class="o">::</span>get_pm<span class="p">(</span>res<span class="o">$</span>ash<span class="p">),</span> res<span class="o">$</span>mr.ash<span class="o">$</span><span class="kp">beta</span><span class="p">,</span> pch<span class="o">=</span><span class="m">20</span><span class="p">,</span> bg<span class="o">=</span><span class="s">&quot;white&quot;</span><span class="p">,</span> col<span class="o">=</span>rgb<span class="p">(</span><span class="m">0.3</span><span class="p">,</span><span class="m">0.3</span><span class="p">,</span><span class="m">0.9</span><span class="p">,</span><span class="m">0.9</span><span class="p">),</span> ylab <span class="o">=</span> <span class="s">&quot;mr.ash&quot;</span><span class="p">,</span> xlab <span class="o">=</span> <span class="s">&quot;ash&quot;</span><span class="p">,</span> main <span class="o">=</span> <span class="s">&#39;ash vs mr.ash&#39;</span><span class="p">)</span>
    abline<span class="p">(</span><span class="m">0</span><span class="p">,</span><span class="m">1</span><span class="p">,</span> col<span class="o">=</span><span class="s">&quot;whitesmoke&quot;</span><span class="p">)</span>
<span class="p">}</span>
</pre></div>
</td></tr>
</table>
</div>
</div>
</div>
</body></html>
