<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="Peter Carbonetto" />

<meta name="date" content="2023-11-23" />

<title>GREGOR functional enrichment analysis of SuSiE and mvSuSiE fine-mapping results for UK Biobank blood cell traits</title>

<script src="site_libs/header-attrs-2.21/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/readable.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/default.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>









<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>



<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->




</head>

<body>


<div class="container-fluid main-container">




<div id="header">



<h1 class="title toc-ignore">GREGOR functional enrichment analysis of
SuSiE and mvSuSiE fine-mapping results for UK Biobank blood cell
traits</h1>
<h4 class="author">Peter Carbonetto</h4>
<h4 class="date">November 23, 2023</h4>

</div>


<p>Here we generate some plots for the paper from the GREGOR functional
enrichment analysis performed <a
href="https://github.com/gaow/mvarbvs/blob/master/analysis/20231106_GREGOR_functional_enrichment.ipynb">here</a>.</p>
<p>Load the packages used in this analysis.</p>
<pre class="r"><code>library(ggplot2)
library(cowplot)</code></pre>
<p>These are the blood cell traits.</p>
<pre class="r"><code>traits &lt;- c(&quot;RBC&quot;,&quot;HGB&quot;,&quot;MCV&quot;,&quot;RDW&quot;,&quot;MSCV&quot;,          # mature red blood cell
            &quot;Reticulocyte_perc&quot;,&quot;HLR_perc&quot;,          # immature red blood cell
            &quot;Platelet_count&quot;,&quot;Plateletcrit&quot;,&quot;PDW&quot;,   # platelet
            &quot;WBC&quot;,&quot;Lymphocyte_perc&quot;,&quot;Monocyte_perc&quot;, # white blood cell
            &quot;Neutrophill_perc&quot;,&quot;Eosinophill_perc&quot;,&quot;Basophill_perc&quot;,
            &quot;crosstrait&quot;)</code></pre>
<p>These are the SNP annotations we will show the enrichment results
for.</p>
<pre class="r"><code>annotations &lt;- list(
  &quot;transcription factor binding sites&quot; =
    c(&quot;TFBS_ENCODE.bed&quot;,&quot;FOSL2.bed&quot;,&quot;FOXO1_DeMayo.bed&quot;,
      &quot;NR2F2.bed&quot;,&quot;PLZF.bed&quot;,&quot;POLII_DeMayo.bed&quot;),
  &quot;enhancer promoter regions&quot; =
    c(&quot;Enhancer_Andersson.bed&quot;,&quot;Enhancer_Hoffman.bed&quot;,
      &quot;PromoterFlanking_Hoffman.bed&quot;,&quot;Promoter_UCSC.bed&quot;,
      &quot;SuperEnhancer_Hnisz.bed&quot;,&quot;TSS_Hoffman.bed&quot;,
      &quot;WeakEnhancer_Hoffman.bed&quot;),
  &quot;eqtls data&quot; =
    c(&quot;eQTL_0.05_FDR.bed&quot;,&quot;eQTL_0.1_FDR.bed&quot;),
  &quot;genomic structural elements&quot; =
    c(&quot;Coding_UCSC.bed&quot;,&quot;Intron_UCSC.bed&quot;,&quot;UTR_3_UCSC.bed&quot;,&quot;UTR_5_UCSC.bed&quot;))</code></pre>
<p>Import the SuSiE-based and mvSuSiE-based enrichment results from the
files and combine into a single data frame.</p>
<pre class="r"><code>odds_ratios &lt;- NULL
for (trait in traits) {
  susie_file &lt;- paste(&quot;../output/blood_cell_traits/gregor/SuSiE&quot;,trait,
                      &quot;positive_cs.index_gregor_output_StatisticSummaryFile&quot;,
                      &quot;fisher.txt&quot;,sep = &quot;_&quot;)
  mvsusie_file &lt;- paste(&quot;../output/blood_cell_traits/gregor/mvSuSiE&quot;,trait,
                        &quot;positive_cs.index_gregor_output_StatisticSummaryFile&quot;,
                        &quot;fisher.txt&quot;,sep = &quot;_&quot;)
  susie   &lt;- read.table(susie_file,sep = &quot; &quot;,header = TRUE,
                        stringsAsFactors = FALSE)
  mvsusie &lt;- read.table(mvsusie_file,sep = &quot; &quot;,header = TRUE,
                        stringsAsFactors = FALSE)
  odds_ratios &lt;- 
    rbind(odds_ratios,
          data.frame(bed       = susie$Bed_File,
                     trait     = trait,
                     method    = &quot;susie&quot;,
                     oddsratio = susie$odds,
                     stringsAsFactors = FALSE),
          data.frame(bed       = mvsusie$Bed_File,
                     trait     = trait,
                     method    = &quot;mvsusie&quot;,
                     oddsratio = mvsusie$odds,
          stringsAsFactors = FALSE))
}</code></pre>
<p>Prepare the data frame for ggplot2.</p>
<pre class="r"><code>n &lt;- length(annotations)
for (i in 1:n) {
  x &lt;- subset(odds_ratios,
              method == &quot;mvsusie&quot; &amp;
              trait == &quot;crosstrait&quot; &amp;
              is.element(bed,annotations[[i]]))
  rows &lt;- order(x$oddsratio,decreasing = TRUE)
  annotations[[i]] &lt;- x[rows,&quot;bed&quot;]
}
annotations &lt;- unlist(annotations)
odds_ratios &lt;- subset(odds_ratios,is.element(bed,annotations))
odds_ratios &lt;- transform(odds_ratios,
                         bed    = factor(bed,rev(annotations)),
                         trait  = factor(trait,traits),
                         method = factor(method,c(&quot;susie&quot;,&quot;mvsusie&quot;)))</code></pre>
<p>This first plot shows the “cross-trait” SuSiE and mvSuSiE enrichment
results:</p>
<pre class="r"><code>p1 &lt;- ggplot(subset(odds_ratios,trait == &quot;crosstrait&quot;),
       aes(x = oddsratio,y = bed,color = method)) +
  geom_point() +
  scale_color_manual(values = c(&quot;dodgerblue&quot;,&quot;red&quot;)) +
  labs(y = &quot;&quot;) +
  theme_cowplot(font_size = 9) +
  theme(panel.grid.major.y = element_line(color = &quot;gray&quot;,size = 0.3))
p1</code></pre>
<p><img src="gregor_files/figure-html/plot-crosstrait-1.png" width="540" style="display: block; margin: auto;" /></p>
<p>Next we show the enrichment results for all traits:</p>
<pre class="r"><code>p2 &lt;- ggplot(odds_ratios,aes(x = oddsratio,y = bed,color = method)) +
  geom_point() +
  scale_color_manual(values = c(&quot;dodgerblue&quot;,&quot;red&quot;)) +
  labs(y = &quot;&quot;) +
  theme_cowplot(font_size = 7) +
  facet_wrap(vars(trait)) +
  theme(strip.background = element_blank(),
        panel.grid.major.y = element_line(color = &quot;gray&quot;,size = 0.3))
p2</code></pre>
<p><img src="gregor_files/figure-html/plot-all-traits-1.png" width="960" style="display: block; margin: auto;" /></p>




</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
