{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "kernel": "SoS"
   },
   "source": [
    "# Enrichment analysis workflow for molecular QTL results"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "kernel": "SoS"
   },
   "source": [
    "For molecular QTL analysis results we obtained we'd like to see if:\n",
    "\n",
    "1. PIP are higher on average in certain annotation groups than in the rest of genome\n",
    "2. Is there an enrichment for variables both in CS and in some annotation groups\n",
    "    - Specifically, whether or not there is an enrichment in the secondary CS that we capture\n",
    "    \n",
    "We focus only on the results that has 1 or more CS identified."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "kernel": "SoS"
   },
   "outputs": [],
   "source": [
    "%revisions -s -n 10"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "kernel": "SoS"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "usage: sos run 20180712_Enrichment_Workflow.ipynb\n",
      "               [workflow_name | -t targets] [options] [workflow_options]\n",
      "  workflow_name:        Single or combined workflows defined in this script\n",
      "  targets:              One or more targets to generate\n",
      "  options:              Single-hyphen sos parameters (see \"sos run -h\" for details)\n",
      "  workflow_options:     Double-hyphen workflow-specific parameters\n",
      "\n",
      "Workflows:\n",
      "  extract_sumstats\n",
      "  zscore2bed\n",
      "  get_variants\n",
      "  range2var_annotation\n",
      "  pip_rank_test\n",
      "  cs_fisher_test\n",
      "\n",
      "Global Workflow Options:\n",
      "  --y-data . (as path)\n",
      "                        Y data, the phenotype file paths\n",
      "  --trait VAL (required)\n",
      "                        Trait name\n",
      "  --cwd  path(f'{y_data:d}/{trait}_output')\n",
      "\n",
      "                        Specify work / output directory\n",
      "  --annotation-dir /home/gaow/Documents/GIT/LargeFiles/he_lab_annotations_bed (as path)\n",
      "                        Path to directory of annotation\n",
      "                        files\n",
      "  --single-annot . (as path)\n",
      "                        Path to list of single annotations\n",
      "                        to use\n",
      "  --max-dist 100000 (as int)\n",
      "                        Maximum distance to site of\n",
      "                        interest, set to eg. 100Kb or 1Mb\n",
      "                        up/downstream to start site of\n",
      "                        analysis unit\n",
      "\n",
      "Sections\n",
      "  extract_sumstats_1:   Extract summary stats from RDS files\n",
      "                        to plain text\n",
      "  extract_sumstats_2:   Consolidate results to one file\n",
      "  zscore2bed_1:         Auxiliary step to get variant in bed\n",
      "                        format based on variant ID in\n",
      "                        z-score file\n",
      "    Workflow Options:\n",
      "      --in-file . (as path)\n",
      "      --chr-prefix ''\n",
      "  zscore2bed_2:\n",
      "  get_variants:\n",
      "  range2var_annotation_1: Get variants in data that falls in\n",
      "                        target region\n",
      "  range2var_annotation_2: Make binary annotation file\n",
      "  pip_rank_test_1:      Test if PIP is larger in annotations\n",
      "  pip_rank_test_2, cs_fisher_test_2: Consolidate results to\n",
      "                        one file\n",
      "  cs_fisher_test_1:     Test if CS is enriched with\n",
      "                        annotations\n"
     ]
    }
   ],
   "source": [
    "! sos run 20180712_Enrichment_Workflow.ipynb -h"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "kernel": "SoS"
   },
   "source": [
    "## Annotation input\n",
    "\n",
    "\n",
    "Annotation files are in bed format (for example: `Coding_UCSC.bed`).\n",
    "\n",
    "```\n",
    "chr1    69090   70008\n",
    "chr1    367658  368597\n",
    "chr1    621095  622034\n",
    "......\n",
    "chr9    141121357       141121553\n",
    "chr9    141124188       141124276\n",
    "chr9    141134069       141134172\n",
    "```\n",
    "\n",
    "There are many annotations one can use. For this workflow one should prepare a list of annotations for input, like this:\n",
    "\n",
    "```\n",
    "# my annotation\n",
    "E8_TCM_T_48h\n",
    "E8_TCM_D_48h\n",
    "NR2F2\n",
    "PGR_Demayo\n",
    "DNaseI\n",
    "H3K27me3\n",
    "H3K4me3\n",
    "H3K27ac\n",
    "FAIRE\n",
    "```\n",
    "\n",
    "comment symbol `#` is allowed."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "kernel": "SoS"
   },
   "source": [
    "## Data input"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "kernel": "SoS"
   },
   "source": [
    "Need to format the data to:"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "kernel": "SoS"
   },
   "source": [
    "```\n",
    "Variant_ID PIP z_score CS_ID \n",
    "```\n",
    "\n",
    "Where `CS_ID` is 0 if variant is not in SuSiE CS, 1 if in CS 1, 2 in CS 2, etc. `Variant_ID` carries information of chrom and pos. eg, `rs10131831_chr14_20905250_G_A`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "kernel": "SoS"
   },
   "outputs": [],
   "source": [
    "[global]\n",
    "# Y data, the phenotype file paths\n",
    "parameter: y_data = path()\n",
    "# Trait name\n",
    "parameter: trait = None\n",
    "# Specify work / output directory\n",
    "parameter: cwd = path(f'{y_data:d}/{trait}_output')\n",
    "# Path to directory of annotation files\n",
    "parameter: annotation_dir = path('~/Documents/GIT/LargeFiles/he_lab_annotations_bed')\n",
    "# Path to list of single annotations to use\n",
    "parameter: single_annot = path() #parameter: single_annot = path(\"data/all_annotations.txt\")\n",
    "# Maximum distance to site of interest, set to eg. 100Kb or 1Mb up/downstream to start site of analysis unit\n",
    "parameter: max_dist = 100000\n",
    "fail_if(not y_data.is_file(), msg = 'Please provide valid ``--y-data``!')\n",
    "z_score = path(f'{cwd}/{y_data:bnn}_{int(max_dist/1000)}Kb/enrichment/SuSiE_loci.sumstats.gz')\n",
    "out_dir = f'{cwd}/{y_data:bnn}_{int(max_dist/1000)}Kb/enrichment/{z_score:bn}'.replace('.', '_')\n",
    "try:\n",
    "    single_anno = [f\"{annotation_dir}/{x.split()[0]}.bed\" for x in open(single_annot).readlines() if not x.startswith('#')]\n",
    "except (FileNotFoundError, IsADirectoryError):\n",
    "    single_anno = []"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "kernel": "SoS"
   },
   "source": [
    "## Prepare summary statistics file\n",
    "\n",
    "```\n",
    "sos run analysis/20180712_Enrichment_Workflow.ipynb extract_sumstats \\\n",
    "    --y-data ~/Documents/GIT/LargeFiles/JointLCL/fastqtl_qqnorm_ASintron_RNAseqGeuvadis_YangVCF.txt.gz \\\n",
    "    --trait AS -j 16 \n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "kernel": "SoS"
   },
   "outputs": [],
   "source": [
    "# Extract summary stats from RDS files to plain text\n",
    "[extract_sumstats_1]\n",
    "input: glob.glob(f'{cwd}/{y_data:bnn}_{int(max_dist/1000)}Kb/SuSiE_CS_[1-9]/*.rds'), group_by = 1, concurrent = True\n",
    "output: f'{_input:dd}/enrichment/{_input:bn}.sumstats.gz'\n",
    "R: expand = '${ }'\n",
    "    dat = readRDS(${_input:r})\n",
    "    pip = dat$pip\n",
    "    names = names(readRDS(dat$input)[[dat$idx]]$z_score)\n",
    "    zscore = readRDS(dat$input)[[dat$idx]]$z_score\n",
    "    cs_id = rep(0, length(pip))\n",
    "    for (i in 1:length(dat$sets$cs)) {\n",
    "        cs_id[dat$sets$cs[[i]]] = dat$sets$cs_index[i]\n",
    "    }\n",
    "    write.table(cbind(names,pip,zscore,cs_id), gzfile(${_output:r}), quote=FALSE, col.names=FALSE, row.names=FALSE, sep=\"\\t\")\n",
    "\n",
    "# Consolidate results to one file\n",
    "[extract_sumstats_2]\n",
    "output: f'{cwd}/{y_data:bnn}_{int(max_dist/1000)}Kb/enrichment/SuSiE_loci.sumstats.gz'\n",
    "bash: expand = True\n",
    "    zcat {_input} | gzip --best > {_output}\n",
    "_input.zap()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "kernel": "SoS"
   },
   "source": [
    "```\n",
    "[GW] zcat /home/gaow/GIT/LargeFiles/JointLCL/AS_output/fastqtl_qqnorm_ASintron_RNAseqGeuvadis_YangVCF_100Kb/SuSiE_loci.sumstats.gz | wc -l\n",
    "1667666\n",
    "[GW] zcat /home/gaow/GIT/LargeFiles/JointLCL/AS_output/fastqtl_qqnorm_ASintron_RNAseqGeuvadis_YangVCF_100Kb/SuSiE_loci.sumstats.gz | cut -f 4 | grep 1 | wc -l\n",
    "46659\n",
    "[GW] zcat /home/gaow/GIT/LargeFiles/JointLCL/AS_output/fastqtl_qqnorm_ASintron_RNAseqGeuvadis_YangVCF_100Kb/SuSiE_loci.sumstats.gz | cut -f 4 | grep 2 | wc -l\n",
    "2212\n",
    "[GW] zcat /home/gaow/GIT/LargeFiles/JointLCL/AS_output/fastqtl_qqnorm_ASintron_RNAseqGeuvadis_YangVCF_100Kb/SuSiE_loci.sumstats.gz | cut -f 4 | grep 3 | wc -l\n",
    "152\n",
    "```\n",
    "\n",
    "So we have total of 1667666 variants, 46659 in CS 1, 2212 in CS 2."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "kernel": "SoS"
   },
   "source": [
    "## Convert variants from summary statistics to bed format"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "kernel": "SoS"
   },
   "outputs": [],
   "source": [
    "# Auxiliary step to get variant in bed format based on variant ID in z-score file\n",
    "[zscore2bed_1]\n",
    "parameter: in_file = path()\n",
    "parameter: chr_prefix = \"\"\n",
    "input: in_file\n",
    "output: f'{_input:n}.bed.unsorted'\n",
    "R: expand = \"${ }\", docker_image = 'gaow/atac-gwas', workdir = cwd, stdout = f'{_output:n}.stdout'\n",
    "    library(readr)\n",
    "    library(stringr)\n",
    "    library(dplyr)\n",
    "    var_file <- ${_input:r}\n",
    "    out_file <- ${_output:r}\n",
    "\n",
    "    variants <- read_tsv(var_file)\n",
    "    colnames(variants) = c('variant', 'pip', 'zscore', 'cs')\n",
    "    var_info <- str_split(variants$variant, \"_\")\n",
    "    variants <- mutate(variants, chr = paste0(\"${chr_prefix}\", sapply(var_info, function(x){x[2]})), \n",
    "                                 pos = sapply(var_info, function(x){x[3]})) %>%\n",
    "                mutate(start = as.numeric(pos), stop=as.numeric(pos)  + 1) %>%\n",
    "                select(chr, start, stop, variant)\n",
    "    options(scipen=1000) # So that positions are always fully written out)\n",
    "    write.table(variants, file=out_file, quote=FALSE, col.names=FALSE, row.names=FALSE, sep=\"\\t\")\n",
    "\n",
    "[zscore2bed_2]\n",
    "output: f'{_input:n}'\n",
    "bash: expand = True, docker_image = 'gaow/atac-gwas', workdir = cwd\n",
    "     sort-bed {_input} > {_output}\n",
    "_input.zap()\n",
    "\n",
    "[get_variants: provides = '{data}.bed']\n",
    "output: f'{data}.bed'\n",
    "sos_run('zscore2bed', in_file = f'{_output:n}.gz')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "kernel": "SoS"
   },
   "source": [
    "## Apply ranged based annotations\n",
    "\n",
    "```\n",
    "sos run analysis/20180712_Enrichment_Workflow.ipynb range2var_annotation \\\n",
    "    --y-data ~/Documents/GIT/LargeFiles/JointLCL/fastqtl_qqnorm_ASintron_RNAseqGeuvadis_YangVCF.txt.gz \\\n",
    "    --trait AS -j 16 \\\n",
    "    --single-annot data/annotation.list \n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "kernel": "SoS"
   },
   "outputs": [],
   "source": [
    "# Get variants in data that falls in target region\n",
    "[range2var_annotation_1]\n",
    "depends: f'{z_score:n}.bed'\n",
    "input: set(paths(single_anno)), group_by = 1, concurrent = True\n",
    "output: f'{out_dir}/{_input:bn}.{z_score:bn}.bed'\n",
    "bash: expand = True, docker_image = 'gaow/atac-gwas', workdir = cwd, volumes = f'{annotation_dir}:{annotation_dir}'\n",
    "    bedops -e {z_score:n}.bed {_input} > {_output}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "kernel": "SoS"
   },
   "outputs": [],
   "source": [
    "# Make binary annotation file\n",
    "[range2var_annotation_2]\n",
    "depends: z_score\n",
    "input: group_by = 1, concurrent = True\n",
    "output: f'{_input:n}.gz'\n",
    "R: expand = \"${ }\", docker_image = 'gaow/atac-gwas', workdir = cwd, stdout = f'{_output:n}.stdout'\n",
    "    library(readr)\n",
    "    library(dplyr)\n",
    "    library(stringr)\n",
    "\n",
    "    variant_tsv <- ${z_score:r}\n",
    "    annotation_var_bed <- ${_input:r}\n",
    "    annot_name <- ${_input:bnr} %>% str_replace(paste0(\".\",${z_score:bnr}), \"\")\n",
    "    out_name <- ${_output:r}\n",
    "\n",
    "    vars <- read_tsv(variant_tsv)[,1]\n",
    "    annot_vars = read_tsv(annotation_var_bed, col_names=FALSE)\n",
    "    names(vars) <- \"SNP\"\n",
    "    vars <- vars %>%\n",
    "            mutate(annot_d = case_when(SNP %in% annot_vars$X4 ~ 1,\n",
    "                                                        TRUE ~ 0))\n",
    "    names(vars)[2] <- annot_name\n",
    "    write.table(vars, file=gzfile(out_name),\n",
    "                col.names=TRUE, row.names=FALSE, sep=\"\\t\", quote=FALSE)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "kernel": "SoS"
   },
   "source": [
    "## Enrichment analysis\n",
    "\n",
    "Test for larger PIP in annotation vs outside it.\n",
    "\n",
    "```\n",
    "sos run analysis/20180712_Enrichment_Workflow.ipynb pip_rank_test \\\n",
    "    --y-data ~/Documents/GIT/LargeFiles/JointLCL/fastqtl_qqnorm_ASintron_RNAseqGeuvadis_YangVCF.txt.gz \\\n",
    "    --trait AS -j 16 \\\n",
    "    --single-annot data/annotation.list \n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "kernel": "SoS"
   },
   "outputs": [],
   "source": [
    "# Test if PIP is larger in annotations\n",
    "[pip_rank_test_1]\n",
    "depends: z_score\n",
    "input_files = [f'{out_dir}/{value:bn}.{z_score:bn}.gz' for value in paths(single_anno)]\n",
    "input: input_files, group_by = 1, concurrent = True\n",
    "output: f'{_input:n}.{step_name}.csv'\n",
    "R: expand = '${ }', docker_image = 'gaow/atac-gwas', workdir = cwd, stdout = f'{_output:n}.stdout'\n",
    "    set.seed(1)\n",
    "    library(readr)\n",
    "    library(dplyr)\n",
    "    variants <- read_tsv(${z_score:r}, col_names=FALSE)\n",
    "    colnames(variants) = c('SNP', 'PIP', 'Z', 'CS')\n",
    "    annotation <- read_tsv(${_input:r}, col_names=TRUE)\n",
    "    name = colnames(annotation)[2]\n",
    "    colnames(annotation) = c('SNP', 'GROUP')\n",
    "    # add two random groupings\n",
    "    annotation$RAND_1 = sample(annotation$GROUP)\n",
    "    annotation$RAND_2 = sample(annotation$GROUP)\n",
    "    variants = inner_join(variants, annotation, by = \"SNP\")\n",
    "    if (length(unique(variants$GROUP)) == 1) {\n",
    "      write(paste(name, NA, NA, NA, sep=','), file = ${_output:r})\n",
    "    } else {\n",
    "    test = wilcox.test(PIP ~ GROUP, data=variants)\n",
    "    c_1 = wilcox.test(PIP ~ RAND_1, data=variants)\n",
    "    c_2 = wilcox.test(PIP ~ RAND_2, data=variants)\n",
    "    write(paste(name, test$p.value, c_1$p.value, c_2$p.value, sep=','), file = ${_output:r})\n",
    "    }\n",
    "\n",
    "# Consolidate results to one file\n",
    "[pip_rank_test_2, cs_fisher_test_2]\n",
    "output: f'{cwd}/{y_data:bnn}_{int(max_dist/1000)}Kb/enrichment/SuSiE_loci.sumstats.{step_name}.csv'\n",
    "bash: expand = True\n",
    "    cat {_input} > {_output}\n",
    "_input.zap()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "kernel": "R"
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div class=\"sos_hint\">> /home/gaow/Documents/GIT/LargeFiles/JointLCL/AS_output/fastqtl_qqnorm_ASintron_RNAseqGeuvadis_YangVCF_100Kb/enrichment/SuSiE_loci.sumstats.pip_rank_test_2.csv (2.0 KiB):</div>"
      ],
      "text/plain": [
       "\n",
       "> /home/gaow/Documents/GIT/LargeFiles/JointLCL/AS_output/fastqtl_qqnorm_ASintron_RNAseqGeuvadis_YangVCF_100Kb/enrichment/SuSiE_loci.sumstats.pip_rank_test_2.csv (2.0 KiB):"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "\n",
       "    <div class='dataframe_container' style=\"max-height:400px\">\n",
       "    <input type=\"text\" class='dataframe_input' id=\"search_5\" onkeyup=\"filterDataFrame('5')\" placeholder=\"Search for names..\">\n",
       "    <table border=\"1\" id=\"dataframe_5\" class=\"sos_dataframe dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th> &nbsp; <i class=\"fa fa-sort\" style=\"color:lightgray\" onclick=\"sortDataFrame('5', 0, 'numeric')\"></th>\n",
       "      <th>name &nbsp; <i class=\"fa fa-sort\" style=\"color:lightgray\" onclick=\"sortDataFrame('5', 1, 'alphabetic')\"></th>\n",
       "      <th>p-val &nbsp; <i class=\"fa fa-sort\" style=\"color:lightgray\" onclick=\"sortDataFrame('5', 2, 'numeric')\"></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1_GATA2-Interval-Track</td>\n",
       "      <td>1.830602e-04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CN_range_anno</td>\n",
       "      <td>8.382343e-16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Coding_UCSC</td>\n",
       "      <td>2.367650e-137</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Conserved_LindbladToh</td>\n",
       "      <td>8.528926e-200</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CTCF_Hoffman</td>\n",
       "      <td>4.655771e-04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>DGF_ENCODE</td>\n",
       "      <td>1.868320e-14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>DHS_peaks_Trynka</td>\n",
       "      <td>1.212949e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>DNaseI</td>\n",
       "      <td>3.625202e-190</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>DN_range_anno</td>\n",
       "      <td>5.706675e-61</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>E8_TCM_D_48h</td>\n",
       "      <td>2.373551e-59</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>E8_TCM_T_48h</td>\n",
       "      <td>1.539131e-37</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>Enhancer_Andersson</td>\n",
       "      <td>5.087272e-18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>Enhancer_Hoffman</td>\n",
       "      <td>4.865850e-03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>eQTL_0.05_FDR</td>\n",
       "      <td>6.681596e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>eQTL_0.1_FDR</td>\n",
       "      <td>6.619897e-02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>FAIRE</td>\n",
       "      <td>2.208255e-04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>FetalDHS_Trynka</td>\n",
       "      <td>7.464480e-06</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>FOSL2</td>\n",
       "      <td>2.293812e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>FOXO1_DeMayo</td>\n",
       "      <td>1.439729e-17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>GA_range_anno</td>\n",
       "      <td>1.684668e-43</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>H3K27ac</td>\n",
       "      <td>1.977123e-96</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>h3k27ac-final-C-48h</td>\n",
       "      <td>5.285726e-217</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>h3k27ac-final-D-48h</td>\n",
       "      <td>0.000000e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>H3K27ac_Hnisz</td>\n",
       "      <td>1.471394e-252</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>H3K27ac_PGC2</td>\n",
       "      <td>0.000000e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>H3K27me3</td>\n",
       "      <td>1.317778e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>h3k4me1-final-C-48h</td>\n",
       "      <td>1.985243e-282</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>h3k4me1-final-D-48h</td>\n",
       "      <td>0.000000e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>H3K4me1_peaks_Trynka</td>\n",
       "      <td>9.328009e-185</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>H3K4me3</td>\n",
       "      <td>3.243413e-137</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>h3k4me3-final-C-48h</td>\n",
       "      <td>5.153547e-154</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>h3k4me3-final-D-48h</td>\n",
       "      <td>6.331268e-238</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>H3K4me3_peaks_Trynka</td>\n",
       "      <td>3.150111e-88</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>H3K9ac_peaks_Trynka</td>\n",
       "      <td>1.571287e-189</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>hESC-RegElements-hg19</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>hic_all_interacting_DT1_dTL4_D_48h</td>\n",
       "      <td>3.185830e-227</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>hic_all_interacting_DT1_dTL4_D_48h_expanded</td>\n",
       "      <td>0.000000e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>hic_harmonizome_interacting_DT1_dTL4_D_48h</td>\n",
       "      <td>6.139673e-94</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>hic_human_gain_interacting_DT1_dTL4_D_48h</td>\n",
       "      <td>4.323638e-244</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>Intron_UCSC</td>\n",
       "      <td>0.000000e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>ips_range_anno</td>\n",
       "      <td>2.711120e-94</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>NR2F2</td>\n",
       "      <td>1.894557e-05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>NSC_range_anno</td>\n",
       "      <td>1.103111e-20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>PGR-A_Bagchi</td>\n",
       "      <td>8.813764e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>PGR-B_Bagchi</td>\n",
       "      <td>1.744193e-07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>PGR_Demayo</td>\n",
       "      <td>3.667659e-02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>PLZF</td>\n",
       "      <td>1.392102e-10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>POLII_DeMayo</td>\n",
       "      <td>0.000000e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>PromoterFlanking_Hoffman</td>\n",
       "      <td>1.198566e-02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49</th>\n",
       "      <td>Promoter_UCSC</td>\n",
       "      <td>1.593581e-172</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50</th>\n",
       "      <td>Repressed_Hoffman</td>\n",
       "      <td>0.000000e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51</th>\n",
       "      <td>SuperEnhancer_Hnisz</td>\n",
       "      <td>2.148822e-49</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52</th>\n",
       "      <td>TFBS_ENCODE</td>\n",
       "      <td>1.531654e-118</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53</th>\n",
       "      <td>Transcribed_Hoffman</td>\n",
       "      <td>0.000000e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54</th>\n",
       "      <td>TSS_Hoffman</td>\n",
       "      <td>1.876752e-204</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55</th>\n",
       "      <td>UTR_3_UCSC</td>\n",
       "      <td>1.342769e-43</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56</th>\n",
       "      <td>UTR_5_UCSC</td>\n",
       "      <td>3.177439e-18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57</th>\n",
       "      <td>WeakEnhancer_Hoffman</td>\n",
       "      <td>1.895438e-37</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table></div>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "dat = read.table('/home/gaow/Documents/GIT/LargeFiles/JointLCL/AS_output/fastqtl_qqnorm_ASintron_RNAseqGeuvadis_YangVCF_100Kb/enrichment/SuSiE_loci.sumstats.pip_rank_test_2.csv', head=F, sep=',')\n",
    "colnames(dat) = c('annotation', 'enrichment', 'random_1', \"random_2\")\n",
    "dat"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "kernel": "SoS"
   },
   "source": [
    "Test for enrichment of annotation in CS.\n",
    "\n",
    "```\n",
    "sos run analysis/20180712_Enrichment_Workflow.ipynb cs_fisher_test \\\n",
    "    --y-data ~/Documents/GIT/LargeFiles/JointLCL/fastqtl_qqnorm_ASintron_RNAseqGeuvadis_YangVCF.txt.gz \\\n",
    "    --trait AS -j 16 \\\n",
    "    --single-annot data/annotation.list \n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "kernel": "SoS"
   },
   "outputs": [],
   "source": [
    "# Test if CS is enriched with annotations\n",
    "[cs_fisher_test_1]\n",
    "depends: z_score\n",
    "input_files = [f'{out_dir}/{value:bn}.{z_score:bn}.gz' for value in paths(single_anno)]\n",
    "input: input_files, group_by = 1, concurrent = True\n",
    "output: f'{_input:n}.{step_name}.csv'\n",
    "R: expand = '${ }', docker_image = 'gaow/atac-gwas', workdir = cwd, stdout = f'{_output:n}.stdout'\n",
    "    run_test = function(dat) {\n",
    "        d1 = dat\n",
    "        d1$CS[which(d1$CS>0)] = 1\n",
    "        d1 = table(d1)\n",
    "        test.d1 = fisher.test(d1)\n",
    "        # test for non-first CS only\n",
    "        d2 = dat[which(dat$CS != 1),]\n",
    "        d2$CS[which(d2$CS>0)] = 1\n",
    "        d2 = table(d2)\n",
    "        test.d2 = fisher.test(d2)\n",
    "        return(c(test.d1$p.value, test.d2$p.value))\n",
    "    }\n",
    "    set.seed(1)\n",
    "    library(readr)\n",
    "    library(dplyr)\n",
    "    variants <- read_tsv(${z_score:r}, col_names=FALSE)\n",
    "    colnames(variants) = c('SNP', 'PIP', 'Z', 'CS')\n",
    "    annotation <- read_tsv(${_input:r}, col_names=TRUE)\n",
    "    name = colnames(annotation)[2]\n",
    "    colnames(annotation) = c('SNP', 'GROUP')\n",
    "    # add two random groupings\n",
    "    annotation$RAND_1 = sample(annotation$GROUP)\n",
    "    annotation$RAND_2 = sample(annotation$GROUP)\n",
    "    variants = inner_join(variants, annotation, by = \"SNP\")[,4:7]\n",
    "    if (length(unique(variants$GROUP)) == 1) {\n",
    "      write(paste(name, NA, NA, NA, NA, NA, NA, sep=','), file = ${_output:r})\n",
    "    } else {\n",
    "    # test against all CS\n",
    "    test = run_test(variants[,c('CS', 'GROUP')])\n",
    "    ctrl_1 = run_test(variants[,c('CS', 'RAND_1')])\n",
    "    ctrl_2 = run_test(variants[,c('CS', 'RAND_2')])\n",
    "    write(paste(name, test[1], test[2], ctrl_1[1], ctrl_1[2], ctrl_2[1], ctrl_2[2], sep=','), file = ${_output:r})\n",
    "    }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "kernel": "R"
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Parsed with column specification:\n",
      "cols(\n",
      "  X1 = col_character(),\n",
      "  X2 = col_double(),\n",
      "  X3 = col_double(),\n",
      "  X4 = col_double()\n",
      ")\n",
      "Warning message in rbind(names(probs), probs_f):\n",
      "“number of columns of result is not a multiple of vector length (arg 2)”\n",
      "Warning message:\n",
      "“1 parsing failure.\n",
      "row # A tibble: 1 x 5 col     row col   expected  actual    file                                          expected   <int> <chr> <chr>     <chr>     <chr>                                         actual 1    69 <NA>  4 columns 7 columns '/home/gaow/Documents/GIT/LargeFiles/JointLC… file # A tibble: 1 x 5\n",
      "”\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table>\n",
       "<thead><tr><th scope=col>X1</th><th scope=col>X2</th><th scope=col>X3</th><th scope=col>X4</th></tr></thead>\n",
       "<tbody>\n",
       "\t<tr><td>1_GATA2-Interval-Track</td><td> 8.212912e-01         </td><td>0.38092063            </td><td>0.61885253            </td></tr>\n",
       "\t<tr><td>1_GATA2-Interval-Track</td><td> 1.098493e-01         </td><td>0.77235991            </td><td>0.51011592            </td></tr>\n",
       "\t<tr><td>CN_range_anno         </td><td> 1.228953e-08         </td><td>0.51158775            </td><td>0.95074335            </td></tr>\n",
       "\t<tr><td>CN_range_anno         </td><td> 2.621717e-01         </td><td>0.50837502            </td><td>0.54243696            </td></tr>\n",
       "\t<tr><td>Coding_UCSC           </td><td> 4.079106e-86         </td><td>0.59766878            </td><td>0.36280463            </td></tr>\n",
       "\t<tr><td>Coding_UCSC           </td><td> 1.883235e-14         </td><td>0.03185144            </td><td>0.13665816            </td></tr>\n",
       "\t<tr><td>Conserved_LindbladToh </td><td> 1.502215e-71         </td><td>0.10594412            </td><td>0.24181969            </td></tr>\n",
       "\t<tr><td>Conserved_LindbladToh </td><td> 5.028735e-09         </td><td>0.29427964            </td><td>0.64958592            </td></tr>\n",
       "\t<tr><td>CTCF_Hoffman          </td><td> 3.360994e-62         </td><td>0.07425694            </td><td>0.21765072            </td></tr>\n",
       "\t<tr><td>CTCF_Hoffman          </td><td> 1.477895e-04         </td><td>0.58590582            </td><td>0.95172204            </td></tr>\n",
       "\t<tr><td>DGF_ENCODE            </td><td> 1.249363e-15         </td><td>0.52983684            </td><td>0.41062036            </td></tr>\n",
       "\t<tr><td>DGF_ENCODE            </td><td> 4.290191e-02         </td><td>0.33028711            </td><td>0.71766178            </td></tr>\n",
       "\t<tr><td>DHS_peaks_Trynka      </td><td> 4.697483e-01         </td><td>0.08847985            </td><td>0.56418635            </td></tr>\n",
       "\t<tr><td>DHS_peaks_Trynka      </td><td> 5.732683e-02         </td><td>0.18694509            </td><td>0.96103924            </td></tr>\n",
       "\t<tr><td>DNaseI                </td><td> 0.000000e+00         </td><td>0.23760666            </td><td>0.61015558            </td></tr>\n",
       "\t<tr><td>DNaseI                </td><td> 4.409109e-27         </td><td>0.41950831            </td><td>0.54231657            </td></tr>\n",
       "\t<tr><td>DN_range_anno         </td><td> 1.136175e-03         </td><td>0.76131721            </td><td>0.66269160            </td></tr>\n",
       "\t<tr><td>DN_range_anno         </td><td> 6.158265e-02         </td><td>0.19803604            </td><td>0.83541162            </td></tr>\n",
       "\t<tr><td>E8_TCM_D_48h          </td><td> 5.640626e-33         </td><td>0.50219048            </td><td>0.86787505            </td></tr>\n",
       "\t<tr><td>E8_TCM_D_48h          </td><td> 4.343954e-01         </td><td>0.86140450            </td><td>0.66296807            </td></tr>\n",
       "\t<tr><td>E8_TCM_T_48h          </td><td> 4.001675e-38         </td><td>0.33477150            </td><td>0.77902151            </td></tr>\n",
       "\t<tr><td>E8_TCM_T_48h          </td><td> 5.339539e-03         </td><td>0.64763582            </td><td>0.66467794            </td></tr>\n",
       "\t<tr><td>Enhancer_Andersson    </td><td> 2.164834e-56         </td><td>0.94735529            </td><td>0.89511744            </td></tr>\n",
       "\t<tr><td>Enhancer_Andersson    </td><td> 5.403893e-07         </td><td>0.75550629            </td><td>0.28993455            </td></tr>\n",
       "\t<tr><td>Enhancer_Hoffman      </td><td>4.861672e-193         </td><td>0.35620173            </td><td>0.13763029            </td></tr>\n",
       "\t<tr><td>Enhancer_Hoffman      </td><td> 1.690437e-07         </td><td>0.81158637            </td><td>0.57750808            </td></tr>\n",
       "\t<tr><td>eQTL_0.05_FDR         </td><td> 3.156183e-78         </td><td>0.03411059            </td><td>0.03776495            </td></tr>\n",
       "\t<tr><td>eQTL_0.05_FDR         </td><td> 1.102527e-63         </td><td>0.78966024            </td><td>0.05560667            </td></tr>\n",
       "\t<tr><td>eQTL_0.1_FDR          </td><td> 2.107911e-70         </td><td>0.39298311            </td><td>0.55061514            </td></tr>\n",
       "\t<tr><td>eQTL_0.1_FDR          </td><td> 2.809380e-53         </td><td>0.64951528            </td><td>0.23378345            </td></tr>\n",
       "\t<tr><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td></tr>\n",
       "\t<tr><td>PGR-A_Bagchi            </td><td> 8.757311e-09           </td><td>2.340265e-01            </td><td>0.9105265129            </td></tr>\n",
       "\t<tr><td>PGR-A_Bagchi            </td><td> 4.713856e-06           </td><td>1.080582e-01            </td><td>0.2892528716            </td></tr>\n",
       "\t<tr><td>PGR-B_Bagchi            </td><td> 2.789860e-02           </td><td>7.967789e-01            </td><td>0.1489434286            </td></tr>\n",
       "\t<tr><td>PGR-B_Bagchi            </td><td> 1.803904e-04           </td><td>2.243336e-01            </td><td>0.3498259219            </td></tr>\n",
       "\t<tr><td>PGR_Demayo              </td><td> 5.373188e-01           </td><td>3.502766e-01            </td><td>0.0739943853            </td></tr>\n",
       "\t<tr><td>PGR_Demayo              </td><td> 8.406633e-02           </td><td>1.000000e+00            </td><td>0.5340505193            </td></tr>\n",
       "\t<tr><td>PLZF                    </td><td> 2.060002e-36           </td><td>7.977086e-01            </td><td>0.0718702484            </td></tr>\n",
       "\t<tr><td>PLZF                    </td><td> 3.232903e-15           </td><td>4.599331e-01            </td><td>0.8063556301            </td></tr>\n",
       "\t<tr><td>POLII_DeMayo            </td><td> 0.000000e+00           </td><td>2.030460e-02            </td><td>0.1141737875            </td></tr>\n",
       "\t<tr><td>POLII_DeMayo            </td><td> 1.753827e-02           </td><td>2.986595e-01            </td><td>0.6860353557            </td></tr>\n",
       "\t<tr><td>PromoterFlanking_Hoffman</td><td> 3.538217e-05           </td><td>7.400462e-04            </td><td>0.4044164281            </td></tr>\n",
       "\t<tr><td>PromoterFlanking_Hoffman</td><td> 2.403007e-03           </td><td>9.058349e-01            </td><td>0.7530615395            </td></tr>\n",
       "\t<tr><td>Promoter_UCSC           </td><td>7.547211e-144           </td><td>2.557095e-01            </td><td>0.0123673073            </td></tr>\n",
       "\t<tr><td>Promoter_UCSC           </td><td> 1.180574e-01           </td><td>3.855754e-01            </td><td>0.9077577479            </td></tr>\n",
       "\t<tr><td>Repressed_Hoffman       </td><td> 0.000000e+00           </td><td>3.379821e-02            </td><td>0.1370881642            </td></tr>\n",
       "\t<tr><td>Repressed_Hoffman       </td><td> 9.669009e-47           </td><td>7.318258e-01            </td><td>0.2475137279            </td></tr>\n",
       "\t<tr><td>SuperEnhancer_Hnisz     </td><td> 0.000000e+00           </td><td>2.507967e-01            </td><td>0.6159840512            </td></tr>\n",
       "\t<tr><td>SuperEnhancer_Hnisz     </td><td> 2.697503e-23           </td><td>9.246390e-01            </td><td>0.0006885008            </td></tr>\n",
       "\t<tr><td>TFBS_ENCODE             </td><td> 1.394719e-19           </td><td>5.089287e-01            </td><td>0.2001090615            </td></tr>\n",
       "\t<tr><td>TFBS_ENCODE             </td><td> 9.960109e-05           </td><td>4.940091e-02            </td><td>0.5474870572            </td></tr>\n",
       "\t<tr><td>Transcribed_Hoffman     </td><td> 0.000000e+00           </td><td>2.894718e-01            </td><td>0.3107053146            </td></tr>\n",
       "\t<tr><td>Transcribed_Hoffman     </td><td> 2.532327e-53           </td><td>8.032282e-01            </td><td>0.3356351633            </td></tr>\n",
       "\t<tr><td>TSS_Hoffman             </td><td>3.628567e-116           </td><td>2.789893e-02            </td><td>0.2732594852            </td></tr>\n",
       "\t<tr><td>TSS_Hoffman             </td><td> 2.134504e-28           </td><td>7.149206e-01            </td><td>0.0745319002            </td></tr>\n",
       "\t<tr><td>UTR_3_UCSC              </td><td>1.713536e-140           </td><td>7.597747e-01            </td><td>0.9722959964            </td></tr>\n",
       "\t<tr><td>UTR_3_UCSC              </td><td> 7.073551e-01           </td><td>7.184015e-01            </td><td>0.8186212637            </td></tr>\n",
       "\t<tr><td>UTR_5_UCSC              </td><td> 1.494857e-81           </td><td>9.331046e-01            </td><td>0.9180328677            </td></tr>\n",
       "\t<tr><td>UTR_5_UCSC              </td><td> 8.872967e-03           </td><td>5.379910e-01            </td><td>0.6589439793            </td></tr>\n",
       "\t<tr><td>WeakEnhancer_Hoffman    </td><td>3.589908e-257           </td><td>4.725798e-06            </td><td>0.9176948665            </td></tr>\n",
       "\t<tr><td>WeakEnhancer_Hoffman    </td><td> 5.908310e-08           </td><td>1.003846e-01            </td><td>0.4464084510            </td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "\\begin{tabular}{r|llll}\n",
       " X1 & X2 & X3 & X4\\\\\n",
       "\\hline\n",
       "\t 1\\_GATA2-Interval-Track &  8.212912e-01            & 0.38092063               & 0.61885253              \\\\\n",
       "\t 1\\_GATA2-Interval-Track &  1.098493e-01            & 0.77235991               & 0.51011592              \\\\\n",
       "\t CN\\_range\\_anno          &  1.228953e-08              & 0.51158775                 & 0.95074335                \\\\\n",
       "\t CN\\_range\\_anno          &  2.621717e-01              & 0.50837502                 & 0.54243696                \\\\\n",
       "\t Coding\\_UCSC            &  4.079106e-86            & 0.59766878               & 0.36280463              \\\\\n",
       "\t Coding\\_UCSC            &  1.883235e-14            & 0.03185144               & 0.13665816              \\\\\n",
       "\t Conserved\\_LindbladToh  &  1.502215e-71            & 0.10594412               & 0.24181969              \\\\\n",
       "\t Conserved\\_LindbladToh  &  5.028735e-09            & 0.29427964               & 0.64958592              \\\\\n",
       "\t CTCF\\_Hoffman           &  3.360994e-62            & 0.07425694               & 0.21765072              \\\\\n",
       "\t CTCF\\_Hoffman           &  1.477895e-04            & 0.58590582               & 0.95172204              \\\\\n",
       "\t DGF\\_ENCODE             &  1.249363e-15            & 0.52983684               & 0.41062036              \\\\\n",
       "\t DGF\\_ENCODE             &  4.290191e-02            & 0.33028711               & 0.71766178              \\\\\n",
       "\t DHS\\_peaks\\_Trynka       &  4.697483e-01              & 0.08847985                 & 0.56418635                \\\\\n",
       "\t DHS\\_peaks\\_Trynka       &  5.732683e-02              & 0.18694509                 & 0.96103924                \\\\\n",
       "\t DNaseI                 &  0.000000e+00          & 0.23760666             & 0.61015558            \\\\\n",
       "\t DNaseI                 &  4.409109e-27          & 0.41950831             & 0.54231657            \\\\\n",
       "\t DN\\_range\\_anno          &  1.136175e-03              & 0.76131721                 & 0.66269160                \\\\\n",
       "\t DN\\_range\\_anno          &  6.158265e-02              & 0.19803604                 & 0.83541162                \\\\\n",
       "\t E8\\_TCM\\_D\\_48h           &  5.640626e-33                & 0.50219048                   & 0.86787505                  \\\\\n",
       "\t E8\\_TCM\\_D\\_48h           &  4.343954e-01                & 0.86140450                   & 0.66296807                  \\\\\n",
       "\t E8\\_TCM\\_T\\_48h           &  4.001675e-38                & 0.33477150                   & 0.77902151                  \\\\\n",
       "\t E8\\_TCM\\_T\\_48h           &  5.339539e-03                & 0.64763582                   & 0.66467794                  \\\\\n",
       "\t Enhancer\\_Andersson     &  2.164834e-56            & 0.94735529               & 0.89511744              \\\\\n",
       "\t Enhancer\\_Andersson     &  5.403893e-07            & 0.75550629               & 0.28993455              \\\\\n",
       "\t Enhancer\\_Hoffman       & 4.861672e-193            & 0.35620173               & 0.13763029              \\\\\n",
       "\t Enhancer\\_Hoffman       &  1.690437e-07            & 0.81158637               & 0.57750808              \\\\\n",
       "\t eQTL\\_0.05\\_FDR          &  3.156183e-78              & 0.03411059                 & 0.03776495                \\\\\n",
       "\t eQTL\\_0.05\\_FDR          &  1.102527e-63              & 0.78966024                 & 0.05560667                \\\\\n",
       "\t eQTL\\_0.1\\_FDR           &  2.107911e-70              & 0.39298311                 & 0.55061514                \\\\\n",
       "\t eQTL\\_0.1\\_FDR           &  2.809380e-53              & 0.64951528                 & 0.23378345                \\\\\n",
       "\t ⋮ & ⋮ & ⋮ & ⋮\\\\\n",
       "\t PGR-A\\_Bagchi             &  8.757311e-09              & 2.340265e-01               & 0.9105265129              \\\\\n",
       "\t PGR-A\\_Bagchi             &  4.713856e-06              & 1.080582e-01               & 0.2892528716              \\\\\n",
       "\t PGR-B\\_Bagchi             &  2.789860e-02              & 7.967789e-01               & 0.1489434286              \\\\\n",
       "\t PGR-B\\_Bagchi             &  1.803904e-04              & 2.243336e-01               & 0.3498259219              \\\\\n",
       "\t PGR\\_Demayo               &  5.373188e-01              & 3.502766e-01               & 0.0739943853              \\\\\n",
       "\t PGR\\_Demayo               &  8.406633e-02              & 1.000000e+00               & 0.5340505193              \\\\\n",
       "\t PLZF                     &  2.060002e-36            & 7.977086e-01             & 0.0718702484            \\\\\n",
       "\t PLZF                     &  3.232903e-15            & 4.599331e-01             & 0.8063556301            \\\\\n",
       "\t POLII\\_DeMayo             &  0.000000e+00              & 2.030460e-02               & 0.1141737875              \\\\\n",
       "\t POLII\\_DeMayo             &  1.753827e-02              & 2.986595e-01               & 0.6860353557              \\\\\n",
       "\t PromoterFlanking\\_Hoffman &  3.538217e-05              & 7.400462e-04               & 0.4044164281              \\\\\n",
       "\t PromoterFlanking\\_Hoffman &  2.403007e-03              & 9.058349e-01               & 0.7530615395              \\\\\n",
       "\t Promoter\\_UCSC            & 7.547211e-144              & 2.557095e-01               & 0.0123673073              \\\\\n",
       "\t Promoter\\_UCSC            &  1.180574e-01              & 3.855754e-01               & 0.9077577479              \\\\\n",
       "\t Repressed\\_Hoffman        &  0.000000e+00              & 3.379821e-02               & 0.1370881642              \\\\\n",
       "\t Repressed\\_Hoffman        &  9.669009e-47              & 7.318258e-01               & 0.2475137279              \\\\\n",
       "\t SuperEnhancer\\_Hnisz      &  0.000000e+00              & 2.507967e-01               & 0.6159840512              \\\\\n",
       "\t SuperEnhancer\\_Hnisz      &  2.697503e-23              & 9.246390e-01               & 0.0006885008              \\\\\n",
       "\t TFBS\\_ENCODE              &  1.394719e-19              & 5.089287e-01               & 0.2001090615              \\\\\n",
       "\t TFBS\\_ENCODE              &  9.960109e-05              & 4.940091e-02               & 0.5474870572              \\\\\n",
       "\t Transcribed\\_Hoffman      &  0.000000e+00              & 2.894718e-01               & 0.3107053146              \\\\\n",
       "\t Transcribed\\_Hoffman      &  2.532327e-53              & 8.032282e-01               & 0.3356351633              \\\\\n",
       "\t TSS\\_Hoffman              & 3.628567e-116              & 2.789893e-02               & 0.2732594852              \\\\\n",
       "\t TSS\\_Hoffman              &  2.134504e-28              & 7.149206e-01               & 0.0745319002              \\\\\n",
       "\t UTR\\_3\\_UCSC               & 1.713536e-140                & 7.597747e-01                 & 0.9722959964                \\\\\n",
       "\t UTR\\_3\\_UCSC               &  7.073551e-01                & 7.184015e-01                 & 0.8186212637                \\\\\n",
       "\t UTR\\_5\\_UCSC               &  1.494857e-81                & 9.331046e-01                 & 0.9180328677                \\\\\n",
       "\t UTR\\_5\\_UCSC               &  8.872967e-03                & 5.379910e-01                 & 0.6589439793                \\\\\n",
       "\t WeakEnhancer\\_Hoffman     & 3.589908e-257              & 4.725798e-06               & 0.9176948665              \\\\\n",
       "\t WeakEnhancer\\_Hoffman     &  5.908310e-08              & 1.003846e-01               & 0.4464084510              \\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "X1 | X2 | X3 | X4 | \n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| 1_GATA2-Interval-Track |  8.212912e-01          | 0.38092063             | 0.61885253             | \n",
       "| 1_GATA2-Interval-Track |  1.098493e-01          | 0.77235991             | 0.51011592             | \n",
       "| CN_range_anno          |  1.228953e-08          | 0.51158775             | 0.95074335             | \n",
       "| CN_range_anno          |  2.621717e-01          | 0.50837502             | 0.54243696             | \n",
       "| Coding_UCSC            |  4.079106e-86          | 0.59766878             | 0.36280463             | \n",
       "| Coding_UCSC            |  1.883235e-14          | 0.03185144             | 0.13665816             | \n",
       "| Conserved_LindbladToh  |  1.502215e-71          | 0.10594412             | 0.24181969             | \n",
       "| Conserved_LindbladToh  |  5.028735e-09          | 0.29427964             | 0.64958592             | \n",
       "| CTCF_Hoffman           |  3.360994e-62          | 0.07425694             | 0.21765072             | \n",
       "| CTCF_Hoffman           |  1.477895e-04          | 0.58590582             | 0.95172204             | \n",
       "| DGF_ENCODE             |  1.249363e-15          | 0.52983684             | 0.41062036             | \n",
       "| DGF_ENCODE             |  4.290191e-02          | 0.33028711             | 0.71766178             | \n",
       "| DHS_peaks_Trynka       |  4.697483e-01          | 0.08847985             | 0.56418635             | \n",
       "| DHS_peaks_Trynka       |  5.732683e-02          | 0.18694509             | 0.96103924             | \n",
       "| DNaseI                 |  0.000000e+00          | 0.23760666             | 0.61015558             | \n",
       "| DNaseI                 |  4.409109e-27          | 0.41950831             | 0.54231657             | \n",
       "| DN_range_anno          |  1.136175e-03          | 0.76131721             | 0.66269160             | \n",
       "| DN_range_anno          |  6.158265e-02          | 0.19803604             | 0.83541162             | \n",
       "| E8_TCM_D_48h           |  5.640626e-33          | 0.50219048             | 0.86787505             | \n",
       "| E8_TCM_D_48h           |  4.343954e-01          | 0.86140450             | 0.66296807             | \n",
       "| E8_TCM_T_48h           |  4.001675e-38          | 0.33477150             | 0.77902151             | \n",
       "| E8_TCM_T_48h           |  5.339539e-03          | 0.64763582             | 0.66467794             | \n",
       "| Enhancer_Andersson     |  2.164834e-56          | 0.94735529             | 0.89511744             | \n",
       "| Enhancer_Andersson     |  5.403893e-07          | 0.75550629             | 0.28993455             | \n",
       "| Enhancer_Hoffman       | 4.861672e-193          | 0.35620173             | 0.13763029             | \n",
       "| Enhancer_Hoffman       |  1.690437e-07          | 0.81158637             | 0.57750808             | \n",
       "| eQTL_0.05_FDR          |  3.156183e-78          | 0.03411059             | 0.03776495             | \n",
       "| eQTL_0.05_FDR          |  1.102527e-63          | 0.78966024             | 0.05560667             | \n",
       "| eQTL_0.1_FDR           |  2.107911e-70          | 0.39298311             | 0.55061514             | \n",
       "| eQTL_0.1_FDR           |  2.809380e-53          | 0.64951528             | 0.23378345             | \n",
       "| ⋮ | ⋮ | ⋮ | ⋮ | \n",
       "| PGR-A_Bagchi             |  8.757311e-09            | 2.340265e-01             | 0.9105265129             | \n",
       "| PGR-A_Bagchi             |  4.713856e-06            | 1.080582e-01             | 0.2892528716             | \n",
       "| PGR-B_Bagchi             |  2.789860e-02            | 7.967789e-01             | 0.1489434286             | \n",
       "| PGR-B_Bagchi             |  1.803904e-04            | 2.243336e-01             | 0.3498259219             | \n",
       "| PGR_Demayo               |  5.373188e-01            | 3.502766e-01             | 0.0739943853             | \n",
       "| PGR_Demayo               |  8.406633e-02            | 1.000000e+00             | 0.5340505193             | \n",
       "| PLZF                     |  2.060002e-36            | 7.977086e-01             | 0.0718702484             | \n",
       "| PLZF                     |  3.232903e-15            | 4.599331e-01             | 0.8063556301             | \n",
       "| POLII_DeMayo             |  0.000000e+00            | 2.030460e-02             | 0.1141737875             | \n",
       "| POLII_DeMayo             |  1.753827e-02            | 2.986595e-01             | 0.6860353557             | \n",
       "| PromoterFlanking_Hoffman |  3.538217e-05            | 7.400462e-04             | 0.4044164281             | \n",
       "| PromoterFlanking_Hoffman |  2.403007e-03            | 9.058349e-01             | 0.7530615395             | \n",
       "| Promoter_UCSC            | 7.547211e-144            | 2.557095e-01             | 0.0123673073             | \n",
       "| Promoter_UCSC            |  1.180574e-01            | 3.855754e-01             | 0.9077577479             | \n",
       "| Repressed_Hoffman        |  0.000000e+00            | 3.379821e-02             | 0.1370881642             | \n",
       "| Repressed_Hoffman        |  9.669009e-47            | 7.318258e-01             | 0.2475137279             | \n",
       "| SuperEnhancer_Hnisz      |  0.000000e+00            | 2.507967e-01             | 0.6159840512             | \n",
       "| SuperEnhancer_Hnisz      |  2.697503e-23            | 9.246390e-01             | 0.0006885008             | \n",
       "| TFBS_ENCODE              |  1.394719e-19            | 5.089287e-01             | 0.2001090615             | \n",
       "| TFBS_ENCODE              |  9.960109e-05            | 4.940091e-02             | 0.5474870572             | \n",
       "| Transcribed_Hoffman      |  0.000000e+00            | 2.894718e-01             | 0.3107053146             | \n",
       "| Transcribed_Hoffman      |  2.532327e-53            | 8.032282e-01             | 0.3356351633             | \n",
       "| TSS_Hoffman              | 3.628567e-116            | 2.789893e-02             | 0.2732594852             | \n",
       "| TSS_Hoffman              |  2.134504e-28            | 7.149206e-01             | 0.0745319002             | \n",
       "| UTR_3_UCSC               | 1.713536e-140            | 7.597747e-01             | 0.9722959964             | \n",
       "| UTR_3_UCSC               |  7.073551e-01            | 7.184015e-01             | 0.8186212637             | \n",
       "| UTR_5_UCSC               |  1.494857e-81            | 9.331046e-01             | 0.9180328677             | \n",
       "| UTR_5_UCSC               |  8.872967e-03            | 5.379910e-01             | 0.6589439793             | \n",
       "| WeakEnhancer_Hoffman     | 3.589908e-257            | 4.725798e-06             | 0.9176948665             | \n",
       "| WeakEnhancer_Hoffman     |  5.908310e-08            | 1.003846e-01             | 0.4464084510             | \n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "    X1                       X2            X3           X4          \n",
       "1   1_GATA2-Interval-Track    8.212912e-01 0.38092063   0.61885253  \n",
       "2   1_GATA2-Interval-Track    1.098493e-01 0.77235991   0.51011592  \n",
       "3   CN_range_anno             1.228953e-08 0.51158775   0.95074335  \n",
       "4   CN_range_anno             2.621717e-01 0.50837502   0.54243696  \n",
       "5   Coding_UCSC               4.079106e-86 0.59766878   0.36280463  \n",
       "6   Coding_UCSC               1.883235e-14 0.03185144   0.13665816  \n",
       "7   Conserved_LindbladToh     1.502215e-71 0.10594412   0.24181969  \n",
       "8   Conserved_LindbladToh     5.028735e-09 0.29427964   0.64958592  \n",
       "9   CTCF_Hoffman              3.360994e-62 0.07425694   0.21765072  \n",
       "10  CTCF_Hoffman              1.477895e-04 0.58590582   0.95172204  \n",
       "11  DGF_ENCODE                1.249363e-15 0.52983684   0.41062036  \n",
       "12  DGF_ENCODE                4.290191e-02 0.33028711   0.71766178  \n",
       "13  DHS_peaks_Trynka          4.697483e-01 0.08847985   0.56418635  \n",
       "14  DHS_peaks_Trynka          5.732683e-02 0.18694509   0.96103924  \n",
       "15  DNaseI                    0.000000e+00 0.23760666   0.61015558  \n",
       "16  DNaseI                    4.409109e-27 0.41950831   0.54231657  \n",
       "17  DN_range_anno             1.136175e-03 0.76131721   0.66269160  \n",
       "18  DN_range_anno             6.158265e-02 0.19803604   0.83541162  \n",
       "19  E8_TCM_D_48h              5.640626e-33 0.50219048   0.86787505  \n",
       "20  E8_TCM_D_48h              4.343954e-01 0.86140450   0.66296807  \n",
       "21  E8_TCM_T_48h              4.001675e-38 0.33477150   0.77902151  \n",
       "22  E8_TCM_T_48h              5.339539e-03 0.64763582   0.66467794  \n",
       "23  Enhancer_Andersson        2.164834e-56 0.94735529   0.89511744  \n",
       "24  Enhancer_Andersson        5.403893e-07 0.75550629   0.28993455  \n",
       "25  Enhancer_Hoffman         4.861672e-193 0.35620173   0.13763029  \n",
       "26  Enhancer_Hoffman          1.690437e-07 0.81158637   0.57750808  \n",
       "27  eQTL_0.05_FDR             3.156183e-78 0.03411059   0.03776495  \n",
       "28  eQTL_0.05_FDR             1.102527e-63 0.78966024   0.05560667  \n",
       "29  eQTL_0.1_FDR              2.107911e-70 0.39298311   0.55061514  \n",
       "30  eQTL_0.1_FDR              2.809380e-53 0.64951528   0.23378345  \n",
       "⋮   ⋮                        ⋮             ⋮            ⋮           \n",
       "86  PGR-A_Bagchi              8.757311e-09 2.340265e-01 0.9105265129\n",
       "87  PGR-A_Bagchi              4.713856e-06 1.080582e-01 0.2892528716\n",
       "88  PGR-B_Bagchi              2.789860e-02 7.967789e-01 0.1489434286\n",
       "89  PGR-B_Bagchi              1.803904e-04 2.243336e-01 0.3498259219\n",
       "90  PGR_Demayo                5.373188e-01 3.502766e-01 0.0739943853\n",
       "91  PGR_Demayo                8.406633e-02 1.000000e+00 0.5340505193\n",
       "92  PLZF                      2.060002e-36 7.977086e-01 0.0718702484\n",
       "93  PLZF                      3.232903e-15 4.599331e-01 0.8063556301\n",
       "94  POLII_DeMayo              0.000000e+00 2.030460e-02 0.1141737875\n",
       "95  POLII_DeMayo              1.753827e-02 2.986595e-01 0.6860353557\n",
       "96  PromoterFlanking_Hoffman  3.538217e-05 7.400462e-04 0.4044164281\n",
       "97  PromoterFlanking_Hoffman  2.403007e-03 9.058349e-01 0.7530615395\n",
       "98  Promoter_UCSC            7.547211e-144 2.557095e-01 0.0123673073\n",
       "99  Promoter_UCSC             1.180574e-01 3.855754e-01 0.9077577479\n",
       "100 Repressed_Hoffman         0.000000e+00 3.379821e-02 0.1370881642\n",
       "101 Repressed_Hoffman         9.669009e-47 7.318258e-01 0.2475137279\n",
       "102 SuperEnhancer_Hnisz       0.000000e+00 2.507967e-01 0.6159840512\n",
       "103 SuperEnhancer_Hnisz       2.697503e-23 9.246390e-01 0.0006885008\n",
       "104 TFBS_ENCODE               1.394719e-19 5.089287e-01 0.2001090615\n",
       "105 TFBS_ENCODE               9.960109e-05 4.940091e-02 0.5474870572\n",
       "106 Transcribed_Hoffman       0.000000e+00 2.894718e-01 0.3107053146\n",
       "107 Transcribed_Hoffman       2.532327e-53 8.032282e-01 0.3356351633\n",
       "108 TSS_Hoffman              3.628567e-116 2.789893e-02 0.2732594852\n",
       "109 TSS_Hoffman               2.134504e-28 7.149206e-01 0.0745319002\n",
       "110 UTR_3_UCSC               1.713536e-140 7.597747e-01 0.9722959964\n",
       "111 UTR_3_UCSC                7.073551e-01 7.184015e-01 0.8186212637\n",
       "112 UTR_5_UCSC                1.494857e-81 9.331046e-01 0.9180328677\n",
       "113 UTR_5_UCSC                8.872967e-03 5.379910e-01 0.6589439793\n",
       "114 WeakEnhancer_Hoffman     3.589908e-257 4.725798e-06 0.9176948665\n",
       "115 WeakEnhancer_Hoffman      5.908310e-08 1.003846e-01 0.4464084510"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "dat = read.table('/home/gaow/Documents/GIT/LargeFiles/JointLCL/AS_output/fastqtl_qqnorm_ASintron_RNAseqGeuvadis_YangVCF_100Kb/enrichment/SuSiE_loci.sumstats.cs_fisher_test_2.csv', head=F, sep=',')\n",
    "colnames(dat) = c('annotation', 'enrichment CS>0', 'enrichment CS>1', \"random_1 CS>0\", \"random_1 CS>1\", \"random_2 CS>0\", \"random_2 CS>1\")\n",
    "dat"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "kernel": "SoS"
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "SoS",
   "language": "sos",
   "name": "sos"
  },
  "language_info": {
   "codemirror_mode": "sos",
   "file_extension": ".sos",
   "mimetype": "text/x-sos",
   "name": "sos",
   "nbconvert_exporter": "sos_notebook.converter.SoS_Exporter",
   "pygments_lexer": "sos"
  },
  "sos": {
   "default_kernel": "SoS",
   "kernels": [
    [
     "R",
     "ir",
     "R",
     "#DCDCDA"
    ],
    [
     "SoS",
     "sos",
     "",
     ""
    ]
   ]
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
